<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Rui Yang">
<meta name="dcterms.date" content="2024-12-30">

<title>Cut&amp;Tag analysis pipeline â€“ NeuroBorder</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../../../">
<script src="../../../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../../../site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../../../../site_libs/quarto-html/quarto-syntax-highlighting-dark-b53751a350365c71b6c909e95f209ed1.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../../../site_libs/bootstrap/bootstrap-7656f4217d0b61e7d1e52c7a113f8883.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../../../../site_libs/bootstrap/bootstrap-dark-0e17766785cb233d275babd160f1f995.min.css" rel="prefetch" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../../../../styles.css">
</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../../../index.html">
    <span class="navbar-title">NeuroBorder</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../../../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-blogs" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Blogs</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-blogs">    
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Computer/index.html">
 <span class="dropdown-text">Computer</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Mathematics/index.html">
 <span class="dropdown-text">Mathematics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Bioinformatics/index.html">
 <span class="dropdown-text">Bioinformatics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Galaxy/index.html">
 <span class="dropdown-text">Galaxy</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../../../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-page-right">
      <h1 class="title">Cut&amp;Tag analysis pipeline</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">cut&amp;tag</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta column-page-right">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Rui Yang </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">December 30, 2024</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">March 26, 2025</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#md5sum-check-over-raw-fastq-files" id="toc-md5sum-check-over-raw-fastq-files" class="nav-link" data-scroll-target="#md5sum-check-over-raw-fastq-files"><span class="header-section-number">2</span> MD5SUM check over raw FASTQ files</a></li>
  <li><a href="#fastqc-over-raw-fastq-files" id="toc-fastqc-over-raw-fastq-files" class="nav-link" data-scroll-target="#fastqc-over-raw-fastq-files"><span class="header-section-number">3</span> FASTQC over raw FASTQ files</a></li>
  <li><a href="#quality-trimming-over-raw-fastq-files" id="toc-quality-trimming-over-raw-fastq-files" class="nav-link" data-scroll-target="#quality-trimming-over-raw-fastq-files"><span class="header-section-number">4</span> Quality trimming over raw FASTQ files</a></li>
  <li><a href="#fastqc-over-clean-fastq-files" id="toc-fastqc-over-clean-fastq-files" class="nav-link" data-scroll-target="#fastqc-over-clean-fastq-files"><span class="header-section-number">5</span> FASTQC over clean FASTQ files</a></li>
  <li><a href="#build-bowtie2-index" id="toc-build-bowtie2-index" class="nav-link" data-scroll-target="#build-bowtie2-index"><span class="header-section-number">6</span> Build Bowtie2 index</a></li>
  <li><a href="#align-reads-with-bowtie2" id="toc-align-reads-with-bowtie2" class="nav-link" data-scroll-target="#align-reads-with-bowtie2"><span class="header-section-number">7</span> Align reads with Bowtie2</a></li>
  <li><a href="#remove-reads-unmapped-and-with-low-quality" id="toc-remove-reads-unmapped-and-with-low-quality" class="nav-link" data-scroll-target="#remove-reads-unmapped-and-with-low-quality"><span class="header-section-number">8</span> Remove reads unmapped and with low quality</a></li>
  <li><a href="#remove-duplicate-reads" id="toc-remove-duplicate-reads" class="nav-link" data-scroll-target="#remove-duplicate-reads"><span class="header-section-number">9</span> Remove duplicate reads</a>
  <ul>
  <li><a href="#add-rg-line-if-it-is-not-existed" id="toc-add-rg-line-if-it-is-not-existed" class="nav-link" data-scroll-target="#add-rg-line-if-it-is-not-existed"><span class="header-section-number">9.1</span> Add <code>@RG</code> line if it is not existed</a></li>
  <li><a href="#remove-duplicate-reads-1" id="toc-remove-duplicate-reads-1" class="nav-link" data-scroll-target="#remove-duplicate-reads-1"><span class="header-section-number">9.2</span> Remove duplicate reads</a></li>
  </ul></li>
  <li><a href="#assess-fragment-length-distributions" id="toc-assess-fragment-length-distributions" class="nav-link" data-scroll-target="#assess-fragment-length-distributions"><span class="header-section-number">10</span> Assess fragment length distributions</a>
  <ul>
  <li><a href="#extract-fragment-lengths" id="toc-extract-fragment-lengths" class="nav-link" data-scroll-target="#extract-fragment-lengths"><span class="header-section-number">10.1</span> Extract fragment lengths</a></li>
  <li><a href="#visualize-fragment-length-distibutions" id="toc-visualize-fragment-length-distibutions" class="nav-link" data-scroll-target="#visualize-fragment-length-distibutions"><span class="header-section-number">10.2</span> Visualize fragment length distibutions</a></li>
  </ul></li>
  <li><a href="#assess-the-reproducibility-of-replicates" id="toc-assess-the-reproducibility-of-replicates" class="nav-link" data-scroll-target="#assess-the-reproducibility-of-replicates"><span class="header-section-number">11</span> Assess the reproducibility of replicates</a>
  <ul>
  <li><a href="#convert-bam-into-bed" id="toc-convert-bam-into-bed" class="nav-link" data-scroll-target="#convert-bam-into-bed"><span class="header-section-number">11.1</span> Convert BAM into BED</a></li>
  <li><a href="#aggregate-fragments-into-bin-counts" id="toc-aggregate-fragments-into-bin-counts" class="nav-link" data-scroll-target="#aggregate-fragments-into-bin-counts"><span class="header-section-number">11.2</span> Aggregate fragments into bin counts</a></li>
  <li><a href="#assess-the-reproducibility-of-replicates-1" id="toc-assess-the-reproducibility-of-replicates-1" class="nav-link" data-scroll-target="#assess-the-reproducibility-of-replicates-1"><span class="header-section-number">11.3</span> Assess the reproducibility of replicates</a></li>
  </ul></li>
  <li><a href="#plot-signal-enrichment-profile" id="toc-plot-signal-enrichment-profile" class="nav-link" data-scroll-target="#plot-signal-enrichment-profile"><span class="header-section-number">12</span> Plot signal enrichment profile</a></li>
  <li><a href="#convert-bam-into-bigwig" id="toc-convert-bam-into-bigwig" class="nav-link" data-scroll-target="#convert-bam-into-bigwig"><span class="header-section-number">13</span> Convert BAM into BigWig</a></li>
  <li><a href="#plot-tss-enrichment-profile" id="toc-plot-tss-enrichment-profile" class="nav-link" data-scroll-target="#plot-tss-enrichment-profile"><span class="header-section-number">14</span> Plot TSS enrichment profile</a>
  <ul>
  <li><a href="#extract-transcription-start-sites" id="toc-extract-transcription-start-sites" class="nav-link" data-scroll-target="#extract-transcription-start-sites"><span class="header-section-number">14.1</span> Extract transcription start sites</a></li>
  <li><a href="#plot-tss-enrichment-profile-1" id="toc-plot-tss-enrichment-profile-1" class="nav-link" data-scroll-target="#plot-tss-enrichment-profile-1"><span class="header-section-number">14.2</span> Plot TSS enrichment profile</a></li>
  </ul></li>
  <li><a href="#call-peaks-with-macs" id="toc-call-peaks-with-macs" class="nav-link" data-scroll-target="#call-peaks-with-macs"><span class="header-section-number">15</span> Call peaks with MACS</a>
  <ul>
  <li><a href="#pairwise-merging-of-biological-replicates" id="toc-pairwise-merging-of-biological-replicates" class="nav-link" data-scroll-target="#pairwise-merging-of-biological-replicates"><span class="header-section-number">15.1</span> Pairwise merging of biological replicates</a></li>
  <li><a href="#call-peaks-with-macs-1" id="toc-call-peaks-with-macs-1" class="nav-link" data-scroll-target="#call-peaks-with-macs-1"><span class="header-section-number">15.2</span> Call peaks with MACS</a>
  <ul>
  <li><a href="#call-peaks-with-controls-using-p-value" id="toc-call-peaks-with-controls-using-p-value" class="nav-link" data-scroll-target="#call-peaks-with-controls-using-p-value"><span class="header-section-number">15.2.1</span> Call peaks with controls (using p-value)</a></li>
  <li><a href="#call-peaks-with-controls-using-q-value" id="toc-call-peaks-with-controls-using-q-value" class="nav-link" data-scroll-target="#call-peaks-with-controls-using-q-value"><span class="header-section-number">15.2.2</span> Call peaks with controls (using q-value)</a></li>
  <li><a href="#call-peaks-without-controls-using-p-value" id="toc-call-peaks-without-controls-using-p-value" class="nav-link" data-scroll-target="#call-peaks-without-controls-using-p-value"><span class="header-section-number">15.2.3</span> Call peaks without controls (using p-value)</a></li>
  <li><a href="#call-peaks-without-controls-using-q-value" id="toc-call-peaks-without-controls-using-q-value" class="nav-link" data-scroll-target="#call-peaks-without-controls-using-q-value"><span class="header-section-number">15.2.4</span> Call peaks without controls (using q-value)</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#naive-overlapping" id="toc-naive-overlapping" class="nav-link" data-scroll-target="#naive-overlapping"><span class="header-section-number">16</span> Naive overlapping</a></li>
  <li><a href="#idr-analysis" id="toc-idr-analysis" class="nav-link" data-scroll-target="#idr-analysis"><span class="header-section-number">17</span> IDR analysis</a></li>
  <li><a href="#pairwise-overlapping-statistics" id="toc-pairwise-overlapping-statistics" class="nav-link" data-scroll-target="#pairwise-overlapping-statistics"><span class="header-section-number">18</span> Pairwise overlapping statistics</a></li>
  <li><a href="#generate-signal-tracks-with-macs" id="toc-generate-signal-tracks-with-macs" class="nav-link" data-scroll-target="#generate-signal-tracks-with-macs"><span class="header-section-number">19</span> Generate signal tracks with MACS</a></li>
  <li><a href="#extending-peaks-on-both-sides-from-summits" id="toc-extending-peaks-on-both-sides-from-summits" class="nav-link" data-scroll-target="#extending-peaks-on-both-sides-from-summits"><span class="header-section-number">20</span> Extending peaks on both sides from summits</a></li>
  <li><a href="#generate-final-peak-sets" id="toc-generate-final-peak-sets" class="nav-link" data-scroll-target="#generate-final-peak-sets"><span class="header-section-number">21</span> Generate final peak sets</a></li>
  <li><a href="#de-novo-motif-finding-with-homer" id="toc-de-novo-motif-finding-with-homer" class="nav-link" data-scroll-target="#de-novo-motif-finding-with-homer"><span class="header-section-number">22</span> De novo motif finding with Homer</a>
  <ul>
  <li><a href="#de-novo-motif-finding-with-homer-1" id="toc-de-novo-motif-finding-with-homer-1" class="nav-link" data-scroll-target="#de-novo-motif-finding-with-homer-1"><span class="header-section-number">22.1</span> De novo motif finding with Homer</a></li>
  <li><a href="#finding-instances-of-specific-motifs" id="toc-finding-instances-of-specific-motifs" class="nav-link" data-scroll-target="#finding-instances-of-specific-motifs"><span class="header-section-number">22.2</span> Finding instances of specific motifs</a></li>
  </ul></li>
  <li><a href="#peak-annotation" id="toc-peak-annotation" class="nav-link" data-scroll-target="#peak-annotation"><span class="header-section-number">23</span> Peak annotation</a>
  <ul>
  <li><a href="#make-txdb-object-from-gff3gtf-file" id="toc-make-txdb-object-from-gff3gtf-file" class="nav-link" data-scroll-target="#make-txdb-object-from-gff3gtf-file"><span class="header-section-number">23.1</span> Make TxDb object from GFF3/GTF file</a></li>
  <li><a href="#profile-of-peaks-binding-to-tssbodytts-regions" id="toc-profile-of-peaks-binding-to-tssbodytts-regions" class="nav-link" data-scroll-target="#profile-of-peaks-binding-to-tssbodytts-regions"><span class="header-section-number">23.2</span> Profile of peaks binding to TSS/body/TTS regions</a></li>
  <li><a href="#peak-annotation-1" id="toc-peak-annotation-1" class="nav-link" data-scroll-target="#peak-annotation-1"><span class="header-section-number">23.3</span> Peak annotation</a></li>
  <li><a href="#functional-enrichment-analysis" id="toc-functional-enrichment-analysis" class="nav-link" data-scroll-target="#functional-enrichment-analysis"><span class="header-section-number">23.4</span> Functional enrichment analysis</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/edit/main/Blogs/Bioinformatics/posts/Cut&amp;Tag/cuttag_analysis_pipeline/index.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/blob/main/Blogs/Bioinformatics/posts/Cut&amp;Tag/cuttag_analysis_pipeline/index.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block column-page-right" id="quarto-document-content">





<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>Before running any of the following steps, you should rename your FASTQ files according to <a href="https://www.neuroborder.com/Blogs/Galaxy/posts/Galaxy/endcode_rna_seq_pipeline_in_galaxy/#copy-your-raw-fastq-data-to-a-new-directory-and-rename-them">these rules</a>.</p>
<div id="bfae500c" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>work_dir <span class="op">=</span> <span class="st">"/data/users/yangrui/mouse/cuttag_v20250108"</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(work_dir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="{r}"><code>work_dir &lt;- "/data/users/yangrui/mouse/cuttag_v20250108"

setwd(work_dir)</code></pre>
</section>
<section id="md5sum-check-over-raw-fastq-files" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="md5sum-check-over-raw-fastq-files"><span class="header-section-number">2</span> MD5SUM check over raw FASTQ files</h2>
<div id="89da9651" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>raw_fastq_dir <span class="op">=</span> <span class="st">"raw_fastq"</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>md5_file <span class="op">=</span> <span class="st">"md5.txt"</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>md5_check_file <span class="op">=</span> <span class="st">"md5_check.txt"</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(raw_fastq_dir)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>YRUtils.BaseUtils.<span class="fu">md5_check</span>(md5_file, md5_check_file)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(work_dir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fastqc-over-raw-fastq-files" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="fastqc-over-raw-fastq-files"><span class="header-section-number">3</span> FASTQC over raw FASTQ files</h2>
<div id="b23015ab" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>raw_fastq_dir <span class="op">=</span> <span class="st">"raw_fastq"</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>raw_fastqc_dir <span class="op">=</span> <span class="st">"raw_fastqc"</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(raw_fastqc_dir)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>raw_fastq_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(raw_fastq_dir, <span class="st">r"</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>YRUtils.BioUtils.<span class="fu">fastqc</span>(raw_fastq_files, raw_fastqc_dir;</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    fastqc_options<span class="op">=</span><span class="st">"--threads 4"</span>, multiqc_options<span class="op">=</span><span class="st">"--zip-data-dir"</span>, num_jobs<span class="op">=</span><span class="fl">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="quality-trimming-over-raw-fastq-files" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="quality-trimming-over-raw-fastq-files"><span class="header-section-number">4</span> Quality trimming over raw FASTQ files</h2>
<div id="8a9c8fe4" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>raw_fastq_dir <span class="op">=</span> <span class="st">"raw_fastq"</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>clean_fastq_dir <span class="op">=</span> <span class="st">"clean_fastq"</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(clean_fastq_dir)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>raw_fastq_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(raw_fastq_dir, <span class="st">r"</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>dict <span class="op">=</span> YRUtils.BioUtils.<span class="fu">auto_detect_fastq_read_type</span>(raw_fastq_files)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>files_dict <span class="op">=</span> <span class="cf">if</span> dict[<span class="st">"paired"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>    dict[<span class="st">"paired"</span>][<span class="st">"dict"</span>]</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> dict[<span class="st">"single"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>    dict[<span class="st">"single"</span>][<span class="st">"dict"</span>]</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@error</span> <span class="st">"did not detect any paired-end or single-end files"</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>files_read_type <span class="op">=</span> <span class="cf">if</span> dict[<span class="st">"paired"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">"paired"</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> dict[<span class="st">"single"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">"single"</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@error</span> <span class="st">"did not detect any paired-end or single-end files"</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>YRUtils.BioUtils.<span class="fu">trimgalore</span>(files_dict, files_read_type, clean_fastq_dir;</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>    trimgalore_options<span class="op">=</span><span class="st">"--cores 4 --phred33 --quality 20 --length 30 --trim-n"</span>,</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    num_jobs<span class="op">=</span><span class="fl">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fastqc-over-clean-fastq-files" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="fastqc-over-clean-fastq-files"><span class="header-section-number">5</span> FASTQC over clean FASTQ files</h2>
<div id="79a3f202" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>clean_fastq_dir <span class="op">=</span> <span class="st">"clean_fastq"</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>clean_fastqc_dir <span class="op">=</span> <span class="st">"clean_fastqc"</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(clean_fastqc_dir)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>clean_fastq_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(clean_fastq_dir, <span class="st">r"</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>YRUtils.BioUtils.<span class="fu">fastqc</span>(clean_fastq_files, clean_fastqc_dir;</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    fastqc_options<span class="op">=</span><span class="st">"--threads 4"</span>, multiqc_options<span class="op">=</span><span class="st">"--zip-data-dir"</span>, num_jobs<span class="op">=</span><span class="fl">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="build-bowtie2-index" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="build-bowtie2-index"><span class="header-section-number">6</span> Build Bowtie2 index</h2>
<div id="d97dc5c7" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>ref_fa <span class="op">=</span> <span class="st">"/data/biodatabase/species/mm10/genome/genome/mm10_no_alt_analysis_set_ENCODE.fasta.gz"</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>bowtie2_index_dir <span class="op">=</span> <span class="st">"bowtie2_index"</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>bowtie2_index_prefix <span class="op">=</span> <span class="st">"mm10"</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>bowtie2_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>log_dir <span class="op">=</span> <span class="st">"log"</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"tmp"</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(bowtie2_index_dir)</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(log_dir)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(tmp_dir)</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> !<span class="fu">isnothing</span>(<span class="fu">match</span>(<span class="st">r"</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>ref_fa<span class="ch">))</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>    new_ref_fa <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(ref_fa), <span class="st">r"</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">" =&gt; ""))</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="st">    </span>YRUtils<span class="sc">.</span><span class="st">ShellUtils</span><span class="sc">.</span><span class="st">pigz</span><span class="ch">(</span><span class="st">ref_fa, new_ref_fa; decompress=true, keep=true</span><span class="ch">)</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>    new_ref_fa <span class="op">=</span> ref_fa</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>cmd <span class="op">=</span> <span class="fu">pipeline</span>(<span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"bowtie2-build"</span>, <span class="st">"--threads"</span>, bowtie2_n_threads, <span class="st">"-f"</span>, new_ref_fa, <span class="fu">joinpath</span>(bowtie2_index_dir, bowtie2_index_prefix)]));</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>    <span class="cn">stdout</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"build_bowtie2_index.log"</span>),</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>    <span class="cn">stderr</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"build_bowtie2_index.log"</span>))</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a><span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> !<span class="fu">isnothing</span>(<span class="fu">match</span>(<span class="st">r"</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>ref_fa<span class="ch">))</span></span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>(new_ref_fa)</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="align-reads-with-bowtie2" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="align-reads-with-bowtie2"><span class="header-section-number">7</span> Align reads with Bowtie2</h2>
<div id="98042360" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>clean_fastq_dir <span class="op">=</span> <span class="st">"clean_fastq"</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>bam_dir <span class="op">=</span> <span class="st">"bam"</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"tmp"</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>log_dir <span class="op">=</span> <span class="st">"log"</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>bowtie2_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>bowtie2_index <span class="op">=</span> <span class="st">"bowtie2_index/mm10"</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>samtools_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>samtools_mem <span class="op">=</span> <span class="st">"768M"</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(bam_dir)</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>clean_fastq_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(clean_fastq_dir, <span class="st">r"</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>dict <span class="op">=</span> YRUtils.BioUtils.<span class="fu">auto_detect_fastq_read_type</span>(clean_fastq_files)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>files_dict <span class="op">=</span> <span class="cf">if</span> dict[<span class="st">"paired"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>    dict[<span class="st">"paired"</span>][<span class="st">"dict"</span>]</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> dict[<span class="st">"single"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>    dict[<span class="st">"single"</span>][<span class="st">"dict"</span>]</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@error</span> <span class="st">"did not detect any paired-end or single-end files"</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>files_read_type <span class="op">=</span> <span class="cf">if</span> dict[<span class="st">"paired"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>    <span class="st">"paired"</span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> dict[<span class="st">"single"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>    <span class="st">"single"</span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@error</span> <span class="st">"did not detect any paired-end or single-end files"</span></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> files_read_type <span class="op">==</span> <span class="st">"paired"</span></span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> sample <span class="kw">in</span> <span class="fu">keys</span>(files_dict)</span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> replicate <span class="kw">in</span> <span class="fu">keys</span>(files_dict[sample])</span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>            r1_fq_files <span class="op">=</span> files_dict[sample][replicate][<span class="st">"R1"</span>]</span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>            r2_fq_files <span class="op">=</span> files_dict[sample][replicate][<span class="st">"R2"</span>]</span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>            bam_file <span class="op">=</span> <span class="fu">joinpath</span>(bam_dir, <span class="fu">string</span>(sample, <span class="st">"_"</span>, replicate, <span class="st">".chr_srt.bam"</span>))</span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="fu">length</span>(r1_fq_files) <span class="op">&gt;</span> <span class="fl">1</span></span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a>                r1_fq_file <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(sample, <span class="st">"_"</span>, replicate, <span class="st">".R1.fq.gz"</span>))</span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>                cmd <span class="op">=</span> <span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">string</span>(<span class="st">"zcat -f "</span>, <span class="fu">join</span>(r1_fq_files, <span class="st">" "</span>),</span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>                        <span class="st">" | pigz -n -c &gt; "</span>,</span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>                        r1_fq_file)]))</span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a>                <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a>                <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span></span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a>                r1_fq_file <span class="op">=</span> r1_fq_files[<span class="fl">1</span>]</span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="fu">length</span>(r2_fq_files) <span class="op">&gt;</span> <span class="fl">1</span></span>
<span id="cb8-48"><a href="#cb8-48" aria-hidden="true" tabindex="-1"></a>                r2_fq_file <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(sample, <span class="st">"_"</span>, replicate, <span class="st">".R2.fq.gz"</span>))</span>
<span id="cb8-49"><a href="#cb8-49" aria-hidden="true" tabindex="-1"></a>                cmd <span class="op">=</span> <span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb8-50"><a href="#cb8-50" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">string</span>(<span class="st">"zcat -f "</span>, <span class="fu">join</span>(r2_fq_files, <span class="st">" "</span>),</span>
<span id="cb8-51"><a href="#cb8-51" aria-hidden="true" tabindex="-1"></a>                        <span class="st">" | pigz -n -c &gt; "</span>,</span>
<span id="cb8-52"><a href="#cb8-52" aria-hidden="true" tabindex="-1"></a>                        r2_fq_file)]))</span>
<span id="cb8-53"><a href="#cb8-53" aria-hidden="true" tabindex="-1"></a>                <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb8-54"><a href="#cb8-54" aria-hidden="true" tabindex="-1"></a>                <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb8-55"><a href="#cb8-55" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span></span>
<span id="cb8-56"><a href="#cb8-56" aria-hidden="true" tabindex="-1"></a>                r2_fq_file <span class="op">=</span> r2_fq_files[<span class="fl">1</span>]</span>
<span id="cb8-57"><a href="#cb8-57" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb8-58"><a href="#cb8-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-59"><a href="#cb8-59" aria-hidden="true" tabindex="-1"></a>            cmd <span class="op">=</span> <span class="fu">pipeline</span>(</span>
<span id="cb8-60"><a href="#cb8-60" aria-hidden="true" tabindex="-1"></a>                <span class="fu">Cmd</span>(</span>
<span id="cb8-61"><a href="#cb8-61" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb8-62"><a href="#cb8-62" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">string</span>(<span class="st">"bowtie2 -X 2000 -p "</span>, bowtie2_n_threads, <span class="st">" -x "</span>, bowtie2_index, <span class="st">" -1 "</span>, r1_fq_file, <span class="st">" -2 "</span>, r2_fq_file,</span>
<span id="cb8-63"><a href="#cb8-63" aria-hidden="true" tabindex="-1"></a>                            <span class="st">" | samtools view -S -u - | samtools sort -@ "</span>, samtools_n_threads, <span class="st">" -m "</span>, samtools_mem, <span class="st">" - -o "</span>, bam_file)]),</span>
<span id="cb8-64"><a href="#cb8-64" aria-hidden="true" tabindex="-1"></a>                );</span>
<span id="cb8-65"><a href="#cb8-65" aria-hidden="true" tabindex="-1"></a>                <span class="cn">stdout</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"bowtie2_align.log"</span>),</span>
<span id="cb8-66"><a href="#cb8-66" aria-hidden="true" tabindex="-1"></a>                <span class="cn">stderr</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"bowtie2_align.log"</span>),</span>
<span id="cb8-67"><a href="#cb8-67" aria-hidden="true" tabindex="-1"></a>                append<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb8-68"><a href="#cb8-68" aria-hidden="true" tabindex="-1"></a>            <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb8-69"><a href="#cb8-69" aria-hidden="true" tabindex="-1"></a>            <span class="fu">open</span>(io <span class="op">-&gt;</span> <span class="fu">println</span>(io, <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)),</span>
<span id="cb8-70"><a href="#cb8-70" aria-hidden="true" tabindex="-1"></a>                <span class="fu">joinpath</span>(log_dir, <span class="st">"bowtie2_align.log"</span>), <span class="st">"a"</span>)</span>
<span id="cb8-71"><a href="#cb8-71" aria-hidden="true" tabindex="-1"></a>            <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb8-72"><a href="#cb8-72" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb8-73"><a href="#cb8-73" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb8-74"><a href="#cb8-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb8-75"><a href="#cb8-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-76"><a href="#cb8-76" aria-hidden="true" tabindex="-1"></a>cmd <span class="op">=</span> <span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>, <span class="fu">string</span>(<span class="st">"rm -rf "</span>, <span class="fu">joinpath</span>(tmp_dir, <span class="st">"*"</span>))]))</span>
<span id="cb8-77"><a href="#cb8-77" aria-hidden="true" tabindex="-1"></a><span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb8-78"><a href="#cb8-78" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="remove-reads-unmapped-and-with-low-quality" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="remove-reads-unmapped-and-with-low-quality"><span class="header-section-number">8</span> Remove reads unmapped and with low quality</h2>
<p>After this step, if read duplication rate is very low, you can skip the next step - removing duplicate reads.</p>
<div id="461ccd37" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>bam_dir <span class="op">=</span> <span class="st">"bam"</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>high_qual_bam_dir <span class="op">=</span> <span class="st">"high_qual_bam"</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>log_dir <span class="op">=</span> <span class="st">"log"</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"tmp"</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>samtools_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>samtools_mem <span class="op">=</span> <span class="st">"768M"</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>map_qual <span class="op">=</span> <span class="fl">30</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(high_qual_bam_dir)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>bam_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(bam_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> bam_file <span class="kw">in</span> bam_files</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>    tmp_name_srt_bam_file <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>name_srt<span class="sc">.</span><span class="st">bam"</span>))</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">pipeline</span>(<span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"samtools view -u -F 1804 -f 2 -q "</span>, map_qual, <span class="st">" "</span>, bam_file,</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | samtools sort -n -@ "</span>, samtools_n_threads, <span class="st">" -m "</span>, samtools_mem, <span class="st">" - -o "</span>, tmp_name_srt_bam_file)]));</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stdout</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>),</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stderr</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>),</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>        append<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">open</span>(io <span class="op">-&gt;</span> <span class="fu">println</span>(io, <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)),</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>        <span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>), <span class="st">"a"</span>)</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>    tmp_fixmate_bam_file <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>fixmate<span class="sc">.</span><span class="st">bam"</span>))</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">pipeline</span>(<span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"samtools fixmate -@ "</span>, samtools_n_threads, <span class="st">" -r "</span>, tmp_name_srt_bam_file, <span class="st">" "</span>, tmp_fixmate_bam_file)]));</span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stdout</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>),</span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stderr</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>),</span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a>        append<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">open</span>(io <span class="op">-&gt;</span> <span class="fu">println</span>(io, <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)),</span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a>        <span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>), <span class="st">"a"</span>)</span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a>    filtered_bam_file <span class="op">=</span> <span class="fu">joinpath</span>(high_qual_bam_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>chr_srt<span class="sc">.</span><span class="st">bam"</span>))</span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">pipeline</span>(<span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb9-39"><a href="#cb9-39" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"samtools view -u -F 1804 -f 2 "</span>, tmp_fixmate_bam_file,</span>
<span id="cb9-40"><a href="#cb9-40" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | samtools sort -@ "</span>, samtools_n_threads, <span class="st">" -m "</span>, samtools_mem, <span class="st">" - -o "</span>, filtered_bam_file)]));</span>
<span id="cb9-41"><a href="#cb9-41" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stdout</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>),</span>
<span id="cb9-42"><a href="#cb9-42" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stderr</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>),</span>
<span id="cb9-43"><a href="#cb9-43" aria-hidden="true" tabindex="-1"></a>        append<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb9-44"><a href="#cb9-44" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb9-45"><a href="#cb9-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">open</span>(io <span class="op">-&gt;</span> <span class="fu">println</span>(io, <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)),</span>
<span id="cb9-46"><a href="#cb9-46" aria-hidden="true" tabindex="-1"></a>        <span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>), <span class="st">"a"</span>)</span>
<span id="cb9-47"><a href="#cb9-47" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb9-48"><a href="#cb9-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-49"><a href="#cb9-49" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>.([tmp_name_srt_bam_file, tmp_fixmate_bam_file])</span>
<span id="cb9-50"><a href="#cb9-50" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="remove-duplicate-reads" class="level2" data-number="9">
<h2 data-number="9" class="anchored" data-anchor-id="remove-duplicate-reads"><span class="header-section-number">9</span> Remove duplicate reads</h2>
<section id="add-rg-line-if-it-is-not-existed" class="level3" data-number="9.1">
<h3 data-number="9.1" class="anchored" data-anchor-id="add-rg-line-if-it-is-not-existed"><span class="header-section-number">9.1</span> Add <code>@RG</code> line if it is not existed</h3>
<p>When you mark duplicates using <code>Picard</code>, tag <code>@RG</code> within the header section in your BAM files is mandatory.</p>
<p>You can check whether your BAM files contain tag <code>@RG</code> line with the following code (no if nothing appears):</p>
<pre class="{bash}"><code>samtools view -H &lt;your BAM file&gt; | grep '@RG'</code></pre>
<p>Check whether your BAM files are sorted:</p>
<pre class="{bash}"><code>samtools view -H &lt;your BAM file&gt; | grep '@HD'</code></pre>
<p>See <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890671-Read-groups">here</a> for how read group is defined by GATK and <a href="https://en.wikipedia.org/wiki/FASTQ_format">here</a> for how to extract some info from FASTQ files.</p>
<p>Read group defined by GATK:</p>
<pre class="{bash}"><code>@RG ID:H0164.2  PL:illumina PU:H0164ALXX140820.2    LB:Solexa-272222    PI:0    DT:2014-08-20T00:00:00-0400 SM:NA12878  CN:BI</code></pre>
<ul>
<li><p><code>ID:&lt;flow cell name&gt;.&lt;lane number&gt;</code>: read group identifier.</p></li>
<li><p><code>PU:&lt;flow cell barcode&gt;.&lt;lane number&gt;.&lt;sample barcode&gt;</code>: platform unit (optional). Flow cell barcode refers to the unique identifier for a particular flow cell. Sample barcode is a sample/library-specific identifier.</p></li>
<li><p><code>SM:&lt;sample name&gt;</code>: sample.</p></li>
<li><p><code>PL:&lt;platform name&gt;</code>: platform/technology used to produce the read, such as ILLUMINA, SOLID, LS454, HELICOS and PACBIO.</p></li>
<li><p><code>LB:&lt;library name&gt;</code>: DNA preparation library identifier. This can be used to identify duplicates derived from library preparation.</p></li>
</ul>
<div id="cae06323" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Construct tag @RG line by extracting some info from paired-end FASTQ files and their file names</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Only compatible with Casava 1.8 format</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>raw_fastq_dir <span class="op">=</span> <span class="st">"raw_fastq"</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>high_qual_bam_dir <span class="op">=</span> <span class="st">"high_qual_bam"</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>add_rg_bam_dir <span class="op">=</span> <span class="st">"add_rg_bam"</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>platform <span class="op">=</span> <span class="st">"ILLUMINA"</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>samtools_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(add_rg_bam_dir)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>raw_fastq_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(raw_fastq_dir, <span class="st">r"</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>fastq_header_names <span class="op">=</span> [<span class="st">"instrument_name"</span>, <span class="st">"run_id"</span>, <span class="st">"flowcell_id"</span>, <span class="st">"flowcell_lane"</span>, <span class="st">"lane_tile_number"</span>, <span class="st">"tile_cluster_x_coord"</span>,</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"tile_cluster_y_coord"</span>, <span class="st">"pair_member_number"</span>, <span class="st">"is_passed"</span>, <span class="st">"control_bits_on"</span>, <span class="st">"index_sequence"</span>]</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> raw_fastq_file <span class="kw">in</span> raw_fastq_files</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>    input_bam_file <span class="op">=</span> <span class="fu">joinpath</span>(high_qual_bam_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(raw_fastq_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">R</span><span class="ch">[</span><span class="st">12</span><span class="ch">]</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)(</span><span class="sc">\.</span><span class="st">gz</span><span class="ch">)</span><span class="sc">*</span><span class="st">"</span> <span class="op">=&gt;</span> <span class="st">".chr_srt.bam"</span>))</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>    output_bam_file <span class="op">=</span> <span class="fu">joinpath</span>(add_rg_bam_dir, <span class="fu">basename</span>(input_bam_file))</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>(<span class="st">"zcat -f "</span>, raw_fastq_file, <span class="st">" | grep -P '^@' | head -n 1"</span>)]))</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>    fastq_header_line <span class="op">=</span> <span class="fu">split</span>(<span class="fu">strip</span>(<span class="fu">strip</span>(<span class="fu">read</span>(cmd, <span class="dt">String</span>)), <span class="ch">'@'</span>), <span class="st">r" </span><span class="sc">+</span><span class="st">"</span>)</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="fu">length</span>(fastq_header_line) <span class="op">==</span> <span class="fl">2</span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>        fastq_header_values <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">flatten_array</span>(<span class="fu">split</span>.(fastq_header_line, <span class="st">":"</span>))</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="fu">length</span>(fastq_header_values) <span class="op">==</span> <span class="fu">length</span>(fastq_header_names)</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>            fastq_header_dict <span class="op">=</span> <span class="fu">Dict</span>(<span class="fu">zip</span>(fastq_header_names, fastq_header_values))</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>            rg_line <span class="op">=</span> <span class="fu">string</span>(<span class="st">"@RG</span><span class="sc">\\</span><span class="st">tID:"</span>, fastq_header_dict[<span class="st">"flowcell_id"</span>], <span class="st">"."</span>, fastq_header_dict[<span class="st">"flowcell_lane"</span>],</span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>                <span class="st">"</span><span class="sc">\\</span><span class="st">tPL:"</span>, platform, <span class="st">"</span><span class="sc">\\</span><span class="st">tSM:"</span>, <span class="fu">replace</span>(<span class="fu">basename</span>(raw_fastq_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">R</span><span class="ch">[</span><span class="st">12</span><span class="ch">]</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)(</span><span class="sc">\.</span><span class="st">gz</span><span class="ch">)</span><span class="sc">*</span><span class="st">"</span> <span class="op">=&gt;</span> <span class="st">""</span>),</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>                <span class="st">"</span><span class="sc">\\</span><span class="st">tLB:"</span>, <span class="fu">replace</span>(<span class="fu">basename</span>(raw_fastq_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">R</span><span class="ch">[</span><span class="st">12</span><span class="ch">]</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)(</span><span class="sc">\.</span><span class="st">gz</span><span class="ch">)</span><span class="sc">*</span><span class="st">"</span> <span class="op">=&gt;</span> <span class="st">""</span>))</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="fu">isfile</span>(input_bam_file) <span class="op">&amp;&amp;</span> !<span class="fu">isfile</span>(output_bam_file)</span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a>                cmd <span class="op">=</span> <span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">string</span>(<span class="st">"samtools addreplacerg -@ "</span>, samtools_n_threads, <span class="st">" -r '"</span>, rg_line, <span class="st">"' -o "</span>,</span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a>                        output_bam_file, <span class="st">" "</span>, input_bam_file)]))</span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a>                <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a>                <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span></span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a>                <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"input BAM file ("</span>, input_bam_file, <span class="st">") is invalid or output BAM file ("</span>, output_bam_file, <span class="st">") has already existed, you can add @RG line yourself: "</span>, rg_line)</span>
<span id="cb13-36"><a href="#cb13-36" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb13-37"><a href="#cb13-37" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span></span>
<span id="cb13-38"><a href="#cb13-38" aria-hidden="true" tabindex="-1"></a>            <span class="pp">@error</span> <span class="st">"unsupported FASTQ header format"</span></span>
<span id="cb13-39"><a href="#cb13-39" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb13-40"><a href="#cb13-40" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span></span>
<span id="cb13-41"><a href="#cb13-41" aria-hidden="true" tabindex="-1"></a>        <span class="pp">@error</span> <span class="st">"unsupported FASTQ header format"</span></span>
<span id="cb13-42"><a href="#cb13-42" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb13-43"><a href="#cb13-43" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="remove-duplicate-reads-1" class="level3" data-number="9.2">
<h3 data-number="9.2" class="anchored" data-anchor-id="remove-duplicate-reads-1"><span class="header-section-number">9.2</span> Remove duplicate reads</h3>
<div id="ea289196" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>add_rg_bam_dir <span class="op">=</span> <span class="st">"add_rg_bam"</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>mark_dup_bam_dir <span class="op">=</span> <span class="st">"mark_dup_bam"</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>rm_dup_bam_dir <span class="op">=</span> <span class="st">"rm_dup_bam"</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>log_dir <span class="op">=</span> <span class="st">"log"</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"tmp"</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>picard_path <span class="op">=</span> <span class="st">"/data/softwares/picard_v3.3.0/picard.jar"</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>samtools_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(mark_dup_bam_dir)</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(rm_dup_bam_dir)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>bam_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(add_rg_bam_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> bam_file <span class="kw">in</span> bam_files</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>    mark_dup_bam_file <span class="op">=</span> <span class="fu">joinpath</span>(mark_dup_bam_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>mark_dup<span class="sc">.</span><span class="st">bam"</span>))</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>    mark_dup_metrics_file <span class="op">=</span> <span class="fu">joinpath</span>(mark_dup_bam_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>metrics<span class="sc">.</span><span class="st">txt"</span>))</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>    rm_dup_bam_file <span class="op">=</span> <span class="fu">joinpath</span>(rm_dup_bam_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>rm_dup<span class="sc">.</span><span class="st">bam"</span>))</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">pipeline</span>(</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>        <span class="fu">Cmd</span>(</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>                <span class="fu">string</span>(<span class="st">"java -jar "</span>, picard_path, <span class="st">" MarkDuplicates --INPUT "</span>, bam_file, <span class="st">" --OUTPUT "</span>, mark_dup_bam_file,</span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" --METRICS_FILE "</span>, mark_dup_metrics_file, <span class="st">" --VALIDATION_STRINGENCY LENIENT "</span>,</span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" --USE_JDK_DEFLATER true --USE_JDK_INFLATER true --ASSUME_SORT_ORDER coordinate "</span>,</span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" --REMOVE_DUPLICATES false --TMP_DIR "</span>, tmp_dir)]),</span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a>        );</span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stdout</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"mark_rm_reads_dups.log"</span>),</span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stderr</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"mark_rm_reads_dups.log"</span>),</span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a>        append<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">open</span>(io <span class="op">-&gt;</span> <span class="fu">println</span>(io, <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)),</span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true" tabindex="-1"></a>        <span class="fu">joinpath</span>(log_dir, <span class="st">"mark_rm_reads_dups.log"</span>), <span class="st">"a"</span>)</span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-35"><a href="#cb14-35" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">pipeline</span>(<span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb14-36"><a href="#cb14-36" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"samtools view -@ "</span>, samtools_n_threads, <span class="st">" -F 1804 -f 2 -b "</span>,</span>
<span id="cb14-37"><a href="#cb14-37" aria-hidden="true" tabindex="-1"></a>                mark_dup_bam_file, <span class="st">" -o "</span>, rm_dup_bam_file)]));</span>
<span id="cb14-38"><a href="#cb14-38" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stdout</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"mark_rm_reads_dups.log"</span>),</span>
<span id="cb14-39"><a href="#cb14-39" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stderr</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"mark_rm_reads_dups.log"</span>),</span>
<span id="cb14-40"><a href="#cb14-40" aria-hidden="true" tabindex="-1"></a>        append<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb14-41"><a href="#cb14-41" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb14-42"><a href="#cb14-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">open</span>(io <span class="op">-&gt;</span> <span class="fu">println</span>(io, <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)),</span>
<span id="cb14-43"><a href="#cb14-43" aria-hidden="true" tabindex="-1"></a>        <span class="fu">joinpath</span>(log_dir, <span class="st">"mark_rm_reads_dups.log"</span>), <span class="st">"a"</span>)</span>
<span id="cb14-44"><a href="#cb14-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb14-45"><a href="#cb14-45" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="assess-fragment-length-distributions" class="level2" data-number="10">
<h2 data-number="10" class="anchored" data-anchor-id="assess-fragment-length-distributions"><span class="header-section-number">10</span> Assess fragment length distributions</h2>
<section id="extract-fragment-lengths" class="level3" data-number="10.1">
<h3 data-number="10.1" class="anchored" data-anchor-id="extract-fragment-lengths"><span class="header-section-number">10.1</span> Extract fragment lengths</h3>
<div id="0d9a1ff9" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>rm_dup_bam_dir <span class="op">=</span> <span class="st">"rm_dup_bam"</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>frag_len_dir <span class="op">=</span> <span class="st">"frag_len"</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>samtools_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(frag_len_dir)</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>bam_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(rm_dup_bam_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> bam_file <span class="kw">in</span> bam_files</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>    frag_len_stat_file <span class="op">=</span> <span class="fu">joinpath</span>(frag_len_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>frag_len_stat<span class="sc">.</span><span class="st">tsv"</span>))</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"samtools view -@ "</span>, samtools_n_threads, <span class="st">" "</span>, bam_file,</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw" | awk -v FS='\t' -v OFS='\t' 'function abs(x) {return ((x &lt; 0.0) ? -x : x)} {print abs($9)}' "</span>,</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw" | sort -n | uniq -c | awk -v OFS='\t' '{print $2,$1/2}' &gt; "</span>, frag_len_stat_file)]),</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="visualize-fragment-length-distibutions" class="level3" data-number="10.2">
<h3 data-number="10.2" class="anchored" data-anchor-id="visualize-fragment-length-distibutions"><span class="header-section-number">10.2</span> Visualize fragment length distibutions</h3>
<pre class="{r}"><code>library(tidyverse)
library(YRUtils)
library(vroom)

frag_len_dir &lt;- "frag_len"

frag_len_stat_files &lt;- list.files(frag_len_dir, pattern = "\\.tsv$", full.names = TRUE, recursive = FALSE)
df &lt;- tibble()
for (frag_len_stat_file in frag_len_stat_files) {
    tmp_df &lt;- vroom(frag_len_stat_file, col_names = c("frag_len", "count")) %&gt;%
        mutate(sample = gsub("\\.\\w+\\.tsv$", "", basename(frag_len_stat_file)))
    df &lt;- bind_rows(df, tmp_df)
}
df &lt;- df %&gt;% arrange(sample, frag_len, count)

p &lt;- ggplot(df, aes(frag_len, count, color = sample)) +
    geom_freqpoly(stat = "identity", linewidth = 0.5) +
    labs(
        x = "Fragment length",
        y = "Count", color = "Sample"
    ) +
    theme_classic(base_family = "Arial", base_size = 20)
ppreview(p, file = file.path(frag_len_dir, "frag_len_dist.freqpoly.pdf"))</code></pre>
</section>
</section>
<section id="assess-the-reproducibility-of-replicates" class="level2" data-number="11">
<h2 data-number="11" class="anchored" data-anchor-id="assess-the-reproducibility-of-replicates"><span class="header-section-number">11</span> Assess the reproducibility of replicates</h2>
<section id="convert-bam-into-bed" class="level3" data-number="11.1">
<h3 data-number="11.1" class="anchored" data-anchor-id="convert-bam-into-bed"><span class="header-section-number">11.1</span> Convert BAM into BED</h3>
<div id="bcfb2b6b" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>rm_dup_bam_dir <span class="op">=</span> <span class="st">"rm_dup_bam"</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>bam2bed_dir <span class="op">=</span> <span class="st">"bam2bed"</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"tmp"</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>samtools_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>samtools_mem <span class="op">=</span> <span class="st">"768M"</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(bam2bed_dir)</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>bam_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(rm_dup_bam_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> bam_file <span class="kw">in</span> bam_files</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>    tmp_name_srt_bam_file <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>name_srt<span class="sc">.</span><span class="st">bam"</span>))</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>    bed_file <span class="op">=</span> <span class="fu">joinpath</span>(bam2bed_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>bed<span class="st">"</span>))</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>(<span class="st">"samtools sort -n -@ "</span>, samtools_n_threads, <span class="st">" -m "</span>, samtools_mem, <span class="st">" -o "</span>, tmp_name_srt_bam_file, <span class="st">" "</span>, bam_file)]))</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>(<span class="st">"bedtools bamtobed -bedpe -i "</span>, tmp_name_srt_bam_file,</span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>            <span class="st">raw" | awk -v FS='\t' -v OFS='\t' '$1 == $4 &amp;&amp; $6 - $2 &lt; 1000 {print $0}' "</span>,</span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a>            <span class="st">raw" | cut -f 1,2,6 | sort -k1,1 -k2,2n -k3,3n &gt; "</span>, bed_file)]))</span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>(tmp_name_srt_bam_file)</span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="aggregate-fragments-into-bin-counts" class="level3" data-number="11.2">
<h3 data-number="11.2" class="anchored" data-anchor-id="aggregate-fragments-into-bin-counts"><span class="header-section-number">11.2</span> Aggregate fragments into bin counts</h3>
<div id="cd7fcf3a" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We use the middle point of each fragment to infer which bin this fragment belongs to.</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate stategy:</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co"># w - window size (e.g. 500)</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co"># $1 - seqname</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># $2 - start</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="co"># $3 - end</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co"># int() - round down</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co"># int(($2 + $3)/(2*w))*w + w/2</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="co"># e.g. all middle points belonging to the left-closed interval [1000, 1500) will have the same quotient 2 by dividing 500,</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="co"># and then, by multiplying 500, and then, by adding 500/2, and finally, the quantity is 1250.</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Finally we use the vaule 1250 to represent all middle points belonging to [1000, 1500),</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="co"># and then we just need to count the number of 1250 to know how many fragments are enriched in the interval [1000, 1500)</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>bam2bed_dir <span class="op">=</span> <span class="st">"bam2bed"</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>bin_width <span class="op">=</span> <span class="fl">500</span></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>bed_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(bam2bed_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bed</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> bed_file <span class="kw">in</span> bed_files</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>    frag_bin_count_file <span class="op">=</span> <span class="fu">joinpath</span>(bam2bed_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bed_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">bed</span><span class="sc">$</span><span class="st">" =&gt; </span>string<span class="ch">(</span><span class="st">"</span>.bin<span class="st">", bin_width, "</span>.tsv<span class="st">")))</span></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"cat "</span>, bed_file, <span class="st">" | awk -v w="</span>, bin_width,</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw" -v FS='\t' -v OFS='\t' '{print $1, int(($2 + $3)/(2*w))*w + w/2}' "</span>,</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw" | sort -k1,1V -k2,2n | uniq -c | awk -v OFS='\t' '{print $2,$3,$1}' "</span>,</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | sort -k1,1V -k2,2n &gt; "</span>, frag_bin_count_file)]),</span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="assess-the-reproducibility-of-replicates-1" class="level3" data-number="11.3">
<h3 data-number="11.3" class="anchored" data-anchor-id="assess-the-reproducibility-of-replicates-1"><span class="header-section-number">11.3</span> Assess the reproducibility of replicates</h3>
<pre class="{r}"><code># At this step, we convert raw counts to Counts Per Million (CPM) to normalize the sequencing depth
library(tidyverse)
library(vroom)
library(psych)
library(FactoMineR)
library(ggforce)
library(ggprism)
library(ggalign)
library(YRUtils)

bam2bed_dir &lt;- "bam2bed"
qc_dir &lt;- "qc"
bin_width &lt;- 500
cpm_th &lt;- 1

dir.create(qc_dir, showWarnings = FALSE, recursive = FALSE)
files &lt;- list.files(bam2bed_dir, pattern = paste0("\\.bin", bin_width, "\\.tsv$"), full.names = TRUE, recursive = FALSE)
df &lt;- tibble()
for (file in files) {
    tmp_df &lt;- vroom(file, col_names = c("seqname", "mid_point", "count")) %&gt;%
        mutate(
            id = paste0(seqname, ":", mid_point),
            sample = gsub("\\.\\w+\\.tsv$", "", basename(file))
        ) %&gt;%
        select(id, count, sample)
    df &lt;- bind_rows(df, tmp_df)
}
df &lt;- df %&gt;%
    group_by(sample) %&gt;%
    mutate(cpm = count / sum(count) * 1e6) %&gt;%
    pivot_wider(id_cols = "id", names_from = "sample", values_from = "cpm", values_fill = 0) %&gt;%
    select(-all_of("id"))
df &lt;- log2(df[rowSums(df &gt;= cpm_th) &gt;= 1, ] + 1)

# Correlation
cor_res &lt;- corr.test(df, use = "pairwise", method = "pearson", adjust = "BH")

p &lt;- ggheatmap(
    cor_res$r,
    width = ncol(cor_res$r) * unit(10, "mm"),
    height = nrow(cor_res$r) * unit(10, "mm")
) +
    scheme_align(free_spaces = "t") +
    scale_fill_gradient2(
        low = "blue", mid = "white", high = "red",
        midpoint = 0, limits = c(-1, 1),
        breaks = c(-1, -0.5, 0, 0.5, 1)
    ) +
    labs(fill = "R") +
    guides(x = guide_axis(angle = 45)) +
    theme(
        text = element_text(size = 20, family = "Arial", color = "black"),
        axis.text = element_text(size = 20, family = "Arial", color = "black")
    ) +
    anno_top() +
    ggalign(
        data = gsub("_rep\\d+$", "", colnames(cor_res$r)),
        size = unit(4, "mm")
    ) +
    geom_tile(aes(y = 1, fill = factor(value))) +
    scale_y_continuous(breaks = NULL, name = NULL, expand = expansion(0)) +
    labs(fill = "Sample") +
    theme(
        text = element_text(size = 20, family = "Arial", color = "black"),
        axis.text = element_text(size = 20, family = "Arial", color = "black")
    ) +
    with_quad(scheme_align(guides = "t"), NULL)
ppreview(p, file = file.path(qc_dir, paste0("correlation.bin", bin_width, ".pdf")))

# PCA
pca &lt;- PCA(t(df), ncp = 10, scale.unit = TRUE, graph = FALSE)

pca_coord &lt;- as.data.frame(pca$ind$coord)
pca_coord$sample &lt;- row.names(pca_coord)
pca_coord$group &lt;- factor(gsub("_rep\\d+$", "", pca_coord$sample))
pca_eig &lt;- as.data.frame(pca$eig)

p &lt;- ggplot(pca_coord, aes(Dim.1, Dim.2)) +
    geom_point(aes(color = group), size = 4) +
    xlab(paste0("PC1 (", round(pca_eig["comp 1", "percentage of variance"]), "%)")) +
    ylab(paste0("PC2 (", round(pca_eig["comp 2", "percentage of variance"]), "%)")) +
    geom_mark_ellipse(aes(fill = group), color = NA, alpha = 0.25) +
    theme_prism(base_family = "Arial", border = TRUE, base_size = 20)
ppreview(p, file = file.path(qc_dir, paste0("pca.bin", bin_width, ".pdf")))</code></pre>
</section>
</section>
<section id="plot-signal-enrichment-profile" class="level2" data-number="12">
<h2 data-number="12" class="anchored" data-anchor-id="plot-signal-enrichment-profile"><span class="header-section-number">12</span> Plot signal enrichment profile</h2>
<p>If controls are far more enriched than treatments, you may need to call peaks of treatments without using controls.</p>
<p>You can try to call peaks of controls and to explore the reason why these regions are enriched (maybe you should remove these regions from peaks of treatments if needed).</p>
<div id="b3a7da50" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>rm_dup_bam_dir <span class="op">=</span> <span class="st">"rm_dup_bam"</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>qc_dir <span class="op">=</span> <span class="st">"qc"</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>sambamba_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>deeptools_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>map_qual <span class="op">=</span> <span class="fl">30</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>bin_size <span class="op">=</span> <span class="fl">500</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>bam_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(rm_dup_bam_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>YRUtils.BioUtils.<span class="fu">bam_index</span>(bam_files, sambamba_n_threads)</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>(<span class="st">"plotFingerprint -b "</span>, <span class="fu">join</span>(bam_files, <span class="st">" "</span>),</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>            <span class="st">" --labels "</span>, <span class="fu">join</span>(<span class="fu">replace</span>.(<span class="fu">basename</span>.(bam_files), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ""), " "),</span></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a><span class="st">            " --</span>outQualityMetrics<span class="st"> "</span>, <span class="fu">joinpath</span>(qc_dir, <span class="st">"plotfingerprint_quality_metrics.log"</span>),</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>            <span class="st">" --skipZeros --minMappingQuality "</span>, map_qual,</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>            <span class="st">" --numberOfProcessors "</span>, deeptools_n_threads,</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>            <span class="st">" --binSize "</span>, bin_size,</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>            <span class="st">" --plotFile "</span>, <span class="fu">joinpath</span>(qc_dir, <span class="st">"plotfingerprint.pdf"</span>),</span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>            <span class="st">" --outRawCounts "</span>, <span class="fu">joinpath</span>(qc_dir, <span class="st">"plotfingerprint.tsv"</span>))]))</span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a><span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="convert-bam-into-bigwig" class="level2" data-number="13">
<h2 data-number="13" class="anchored" data-anchor-id="convert-bam-into-bigwig"><span class="header-section-number">13</span> Convert BAM into BigWig</h2>
<div id="5fb81ba4" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>rm_dup_bam_dir <span class="op">=</span> <span class="st">"rm_dup_bam"</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>bam2bw_dir <span class="op">=</span> <span class="st">"bam2bw"</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>deeptools_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>map_qual <span class="op">=</span> <span class="fl">30</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>bin_size <span class="op">=</span> <span class="fl">10</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>norm_method <span class="op">=</span> <span class="st">"RPKM"</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>effective_genome_size <span class="op">=</span> <span class="fl">3372855573</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(bam2bw_dir)</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>bam_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(rm_dup_bam_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> bam_file <span class="kw">in</span> bam_files</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>    bw_file <span class="op">=</span> <span class="fu">joinpath</span>(bam2bw_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>bw<span class="st">"</span>))</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> !<span class="fu">isfile</span>(bw_file)</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>        cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>                <span class="fu">string</span>(<span class="st">"bamCoverage -b "</span>, bam_file, <span class="st">" -o "</span>, bw_file,</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" --numberOfProcessors "</span>, deeptools_n_threads,</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" --binSize "</span>, bin_size,</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" --normalizeUsing "</span>, norm_method,</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" --effectiveGenomeSize "</span>, effective_genome_size,</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" --minMappingQuality "</span>, map_qual,</span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" --extendReads"</span>)]))</span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>        <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>        <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-tss-enrichment-profile" class="level2" data-number="14">
<h2 data-number="14" class="anchored" data-anchor-id="plot-tss-enrichment-profile"><span class="header-section-number">14</span> Plot TSS enrichment profile</h2>
<section id="extract-transcription-start-sites" class="level3" data-number="14.1">
<h3 data-number="14.1" class="anchored" data-anchor-id="extract-transcription-start-sites"><span class="header-section-number">14.1</span> Extract transcription start sites</h3>
<pre class="{r}"><code>library(tidyverse)
library(rtracklayer)
library(vroom)

gff_file &lt;- "/data/biodatabase/species/mm10/genome/anno/gencode.vM21.primary_assembly.annotation_UCSC_names.gtf.gz"
gff_format &lt;- "gtf"
seq_type &lt;- "transcript"
id_col &lt;- "transcript_id"

gff &lt;- import(gff_file, format = gff_format) %&gt;%
    as.data.frame() %&gt;%
    as_tibble()
df &lt;- gff %&gt;%
    filter(type == seq_type) %&gt;%
    mutate(
        tss_start = if_else(strand == "-", end, start - 1),
        tss_end = if_else(strand == "-", end + 1, start),
        score = 0
    ) %&gt;%
    select(all_of(c("seqnames", "tss_start", "tss_end", id_col, "score", "strand"))) %&gt;%
    distinct()
vroom_write(df,
    file = gsub("\\.\\w+(\\.gz)*$", ".transcription_start_sites.bed", gff_file),
    col_names = FALSE, append = FALSE
)</code></pre>
</section>
<section id="plot-tss-enrichment-profile-1" class="level3" data-number="14.2">
<h3 data-number="14.2" class="anchored" data-anchor-id="plot-tss-enrichment-profile-1"><span class="header-section-number">14.2</span> Plot TSS enrichment profile</h3>
<div id="6178d2dd" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>bam2bw_dir <span class="op">=</span> <span class="st">"bam2bw"</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>tss_file <span class="op">=</span> <span class="st">"/data/biodatabase/species/mm10/genome/anno/gencode.vM21.primary_assembly.annotation_UCSC_names.transcription_start_sites.bed"</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>qc_dir <span class="op">=</span> <span class="st">"qc"</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>deeptools_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>bin_size <span class="op">=</span> <span class="fl">10</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>before_length <span class="op">=</span> <span class="fl">2000</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>after_length <span class="op">=</span> <span class="fl">2000</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>bw_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(bam2bw_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bw</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> bw_file <span class="kw">in</span> bw_files</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>    tss_cov_mat_file <span class="op">=</span> <span class="fu">joinpath</span>(bam2bw_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bw_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">bw</span><span class="sc">$</span><span class="st">" =&gt; ".</span>mat<span class="sc">.</span><span class="st">txt"</span>))</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>    tss_cov_mat_gz <span class="op">=</span> <span class="fu">joinpath</span>(bam2bw_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bw_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">bw</span><span class="sc">$</span><span class="st">" =&gt; ".</span>mat<span class="sc">.</span><span class="st">txt</span><span class="sc">.</span><span class="st">gz"</span>))</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>    tss_cov_heatmap_pdf <span class="op">=</span> <span class="fu">joinpath</span>(qc_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bw_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">bw</span><span class="sc">$</span><span class="st">" =&gt; ".</span>heatmap<span class="sc">.</span><span class="st">pdf"</span>))</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>    tss_cov_profile_pdf <span class="op">=</span> <span class="fu">joinpath</span>(qc_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bw_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">bw</span><span class="sc">$</span><span class="st">" =&gt; ".</span>profile<span class="sc">.</span><span class="st">pdf"</span>))</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"computeMatrix reference-point --referencePoint TSS --numberOfProcessors "</span>, deeptools_n_threads,</span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -S "</span>, bw_file, <span class="st">" -R "</span>, tss_file, <span class="st">" --binSize "</span>, bin_size,</span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --beforeRegionStartLength "</span>, before_length, <span class="st">" --afterRegionStartLength "</span>, after_length,</span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --skipZeros --samplesLabel "</span>, <span class="fu">replace</span>(<span class="fu">basename</span>(bw_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">bw</span><span class="sc">$</span><span class="st">" =&gt; ""),</span></span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a><span class="st">                " --</span>outFileNameMatrix<span class="st"> "</span>, tss_cov_mat_file, <span class="st">" --outFileName "</span>, tss_cov_mat_gz)]))</span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb23-30"><a href="#cb23-30" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"plotHeatmap -m "</span>, tss_cov_mat_gz, <span class="st">" -o "</span>, tss_cov_heatmap_pdf,</span>
<span id="cb23-31"><a href="#cb23-31" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --dpi 300 --samplesLabel "</span>, <span class="fu">replace</span>(<span class="fu">basename</span>(bw_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">bw</span><span class="sc">$</span><span class="st">" =&gt; ""))]))</span></span>
<span id="cb23-32"><a href="#cb23-32" aria-hidden="true" tabindex="-1"></a><span class="st">    @</span>info<span class="st"> string</span><span class="ch">(</span><span class="st">"</span>running <span class="st">", cmd, "</span> <span class="op">...</span><span class="st">")</span></span>
<span id="cb23-33"><a href="#cb23-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb23-34"><a href="#cb23-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-35"><a href="#cb23-35" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb23-36"><a href="#cb23-36" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb23-37"><a href="#cb23-37" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"plotProfile -m "</span>, tss_cov_mat_gz, <span class="st">" -o "</span>, tss_cov_profile_pdf,</span>
<span id="cb23-38"><a href="#cb23-38" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --dpi 300 --samplesLabel "</span>, <span class="fu">replace</span>(<span class="fu">basename</span>(bw_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">bw</span><span class="sc">$</span><span class="st">" =&gt; ""))]))</span></span>
<span id="cb23-39"><a href="#cb23-39" aria-hidden="true" tabindex="-1"></a><span class="st">    @</span>info<span class="st"> string</span><span class="ch">(</span><span class="st">"</span>running <span class="st">", cmd, "</span> <span class="op">...</span><span class="st">")</span></span>
<span id="cb23-40"><a href="#cb23-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb23-41"><a href="#cb23-41" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="call-peaks-with-macs" class="level2" data-number="15">
<h2 data-number="15" class="anchored" data-anchor-id="call-peaks-with-macs"><span class="header-section-number">15</span> Call peaks with MACS</h2>
<section id="pairwise-merging-of-biological-replicates" class="level3" data-number="15.1">
<h3 data-number="15.1" class="anchored" data-anchor-id="pairwise-merging-of-biological-replicates"><span class="header-section-number">15.1</span> Pairwise merging of biological replicates</h3>
<p>e.g.&nbsp;if you have three biological replicates for a sample, you will get three pooled BAM files: <code>rep1_vs_rep2</code>, <code>rep1_vs_rep3</code>, and <code>rep2_vs_rep3</code>.</p>
<div id="83423361" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">Combinatorics</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">NaturalSort</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>rm_dup_bam_dir <span class="op">=</span> <span class="st">"rm_dup_bam"</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>samtools_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>sambamba_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>bam_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(rm_dup_bam_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>dict <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{String,Vector{String}}</span>()</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> bam_file <span class="kw">in</span> bam_files</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>    sample <span class="op">=</span> <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"_rep</span><span class="sc">\d+\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; "")</span></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="st">    </span>if<span class="st"> !haskey</span><span class="ch">(</span><span class="st">dict, sample</span><span class="ch">)</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>        dict[sample] <span class="op">=</span> [bam_file]</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>        <span class="fu">push!</span>(dict[sample], bam_file)</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> <span class="fu">keys</span>(dict)</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>    combns <span class="op">=</span> <span class="fu">collect</span>(<span class="fu">combinations</span>(dict[sample], <span class="fl">2</span>))</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> combn <span class="kw">in</span> combns</span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>        cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>                <span class="fu">string</span>(<span class="st">"samtools merge -t "</span>, samtools_n_threads,</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" -o "</span>, <span class="fu">joinpath</span>(rm_dup_bam_dir,</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">string</span>(<span class="fu">join</span>(<span class="fu">sort</span>(<span class="fu">replace</span>.(<span class="fu">basename</span>.(combn), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ""), </span>lt<span class="st">=natural</span><span class="ch">)</span><span class="st">, "</span>_vs_<span class="st">"),</span></span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a>                            <span class="st">".rm_dup.bam"</span>)),</span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" "</span>, <span class="fu">join</span>(combn, <span class="st">" "</span>))]))</span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a>        <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a>        <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a>bam_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(rm_dup_bam_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a>YRUtils.BioUtils.<span class="fu">bam_index</span>(bam_files, sambamba_n_threads)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="call-peaks-with-macs-1" class="level3" data-number="15.2">
<h3 data-number="15.2" class="anchored" data-anchor-id="call-peaks-with-macs-1"><span class="header-section-number">15.2</span> Call peaks with MACS</h3>
<p>If you want to perform IDR analysis, then use a relax threshold (e.g.&nbsp;<code>p-value = 0.01</code>) is suitable.</p>
<p>If you only want to perform naive overlapping, then use a stringent threshold (e.g.&nbsp;<code>q-value = 0.05</code>) is suitable.</p>
<section id="call-peaks-with-controls-using-p-value" class="level4" data-number="15.2.1">
<h4 data-number="15.2.1" class="anchored" data-anchor-id="call-peaks-with-controls-using-p-value"><span class="header-section-number">15.2.1</span> Call peaks with controls (using p-value)</h4>
<div id="a2441901" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Whether you use the --SPMR flag or not,</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co"># MACS will normalize the data internally to call peaks.</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co"># The --SPMR flag only affects the signal track produced.</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co"># With the flag present, the signal will be normalized to reads per million,</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="co"># this is for comparison with other samples which have been sequenced in different depths.</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="co"># The option --SPMR only affects the bedGraph output.</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>rm_dup_bam_dir <span class="op">=</span> <span class="st">"rm_dup_bam"</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"tmp"</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>peak_with_ctl_dir <span class="op">=</span> <span class="st">"peak_with_ctl_pval0.01"</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>effective_genome_size <span class="op">=</span> <span class="fl">2652684646</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>pvalue <span class="op">=</span> <span class="fl">0.01</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>cap_num <span class="op">=</span> <span class="fl">1000000</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>chrsz <span class="op">=</span> <span class="st">"/data/biodatabase/species/mm10/genome/genome/mm10_no_alt.chrom.sizes.tsv"</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="co"># [["control1", "treatment1"], ["control2", "treatment2"], ...]</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>ctl_treat_pairs <span class="op">=</span> [</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"IgG_rep1.rm_dup.bam"</span>, <span class="st">"Satb2_rep1.rm_dup.bam"</span>],</span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"IgG_rep2.rm_dup.bam"</span>, <span class="st">"Satb2_rep2.rm_dup.bam"</span>],</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"IgG_rep3.rm_dup.bam"</span>, <span class="st">"Satb2_rep3.rm_dup.bam"</span>],</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"IgG_rep1_vs_IgG_rep2.rm_dup.bam"</span>, <span class="st">"Satb2_rep1_vs_Satb2_rep2.rm_dup.bam"</span>],</span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"IgG_rep1_vs_IgG_rep3.rm_dup.bam"</span>, <span class="st">"Satb2_rep1_vs_Satb2_rep3.rm_dup.bam"</span>],</span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"IgG_rep2_vs_IgG_rep3.rm_dup.bam"</span>, <span class="st">"Satb2_rep2_vs_Satb2_rep3.rm_dup.bam"</span>],</span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(peak_with_ctl_dir)</span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ctl_treat_pair <span class="kw">in</span> ctl_treat_pairs</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>    prefix <span class="op">=</span> <span class="fu">join</span>(<span class="fu">reverse</span>(<span class="fu">replace</span>.(ctl_treat_pair, <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; "")), "</span>_vs_<span class="st">"</span>)</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a>    macs_peak_file <span class="op">=</span> <span class="fu">joinpath</span>(peak_with_ctl_dir, <span class="fu">string</span>(prefix, <span class="st">"_peaks.narrowPeak"</span>))</span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a>    final_peak_file <span class="op">=</span> <span class="fu">joinpath</span>(peak_with_ctl_dir, <span class="fu">string</span>(prefix, <span class="st">".pval"</span>, pvalue, <span class="st">".narrowPeak.gz"</span>))</span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>    tmp_peak_file1 <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(prefix, <span class="st">".tmp1"</span>))</span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a>    tmp_peak_file2 <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(prefix, <span class="st">".tmp2"</span>))</span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a>    tmp_peak_file3 <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(prefix, <span class="st">".tmp3"</span>))</span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-33"><a href="#cb25-33" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb25-34"><a href="#cb25-34" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb25-35"><a href="#cb25-35" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"macs3 callpeak "</span>,</span>
<span id="cb25-36"><a href="#cb25-36" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -c "</span>, <span class="fu">joinpath</span>(rm_dup_bam_dir, ctl_treat_pair[<span class="fl">1</span>]),</span>
<span id="cb25-37"><a href="#cb25-37" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -t "</span>, <span class="fu">joinpath</span>(rm_dup_bam_dir, ctl_treat_pair[<span class="fl">2</span>]),</span>
<span id="cb25-38"><a href="#cb25-38" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -g "</span>, effective_genome_size,</span>
<span id="cb25-39"><a href="#cb25-39" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -n "</span>, prefix,</span>
<span id="cb25-40"><a href="#cb25-40" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --outdir "</span>, peak_with_ctl_dir,</span>
<span id="cb25-41"><a href="#cb25-41" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --tempdir "</span>, tmp_dir,</span>
<span id="cb25-42"><a href="#cb25-42" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -p "</span>, pvalue,</span>
<span id="cb25-43"><a href="#cb25-43" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -f BAMPE --keep-dup all -B --SPMR --call-summits"</span>)]))</span>
<span id="cb25-44"><a href="#cb25-44" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb25-45"><a href="#cb25-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb25-46"><a href="#cb25-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-47"><a href="#cb25-47" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Sort by Column 8 in descending order and replace long peak names in Column 4 with Peak_&lt;rank&gt;</span></span>
<span id="cb25-48"><a href="#cb25-48" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb25-49"><a href="#cb25-49" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb25-50"><a href="#cb25-50" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"LC_COLLATE=C sort -k 8gr,8gr "</span>, macs_peak_file,</span>
<span id="cb25-51"><a href="#cb25-51" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw""" | awk -v FS="\t" -v OFS="\t" '{$4="Peak_"NR; if ($2&lt;0) $2=0; if ($3&lt;0) $3=0; if ($10==-1) $10=$2+int(($3-$2+1)/2.0); print $0}' &gt; """</span>, tmp_peak_file1)]))</span>
<span id="cb25-52"><a href="#cb25-52" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb25-53"><a href="#cb25-53" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb25-54"><a href="#cb25-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-55"><a href="#cb25-55" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb25-56"><a href="#cb25-56" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb25-57"><a href="#cb25-57" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"head -n "</span>, cap_num, <span class="st">" "</span>, tmp_peak_file1, <span class="st">" &gt; "</span>, tmp_peak_file2)]))</span>
<span id="cb25-58"><a href="#cb25-58" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb25-59"><a href="#cb25-59" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb25-60"><a href="#cb25-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-61"><a href="#cb25-61" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb25-62"><a href="#cb25-62" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb25-63"><a href="#cb25-63" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"bedClip "</span>, tmp_peak_file2, <span class="st">" "</span>, chrsz, <span class="st">" "</span>, tmp_peak_file3, <span class="st">" -truncate -verbose=2"</span>)]))</span>
<span id="cb25-64"><a href="#cb25-64" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb25-65"><a href="#cb25-65" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb25-66"><a href="#cb25-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-67"><a href="#cb25-67" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb25-68"><a href="#cb25-68" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb25-69"><a href="#cb25-69" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"cat "</span>, tmp_peak_file3, <span class="st">" | pigz -nc &gt; "</span>, final_peak_file)]))</span>
<span id="cb25-70"><a href="#cb25-70" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb25-71"><a href="#cb25-71" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb25-72"><a href="#cb25-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-73"><a href="#cb25-73" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>.([tmp_peak_file1, tmp_peak_file2, tmp_peak_file3])</span>
<span id="cb25-74"><a href="#cb25-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="call-peaks-with-controls-using-q-value" class="level4" data-number="15.2.2">
<h4 data-number="15.2.2" class="anchored" data-anchor-id="call-peaks-with-controls-using-q-value"><span class="header-section-number">15.2.2</span> Call peaks with controls (using q-value)</h4>
<div id="cb10febd" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Whether you use the --SPMR flag or not,</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># MACS will normalize the data internally to call peaks.</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co"># The --SPMR flag only affects the signal track produced.</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># With the flag present, the signal will be normalized to reads per million,</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="co"># this is for comparison with other samples which have been sequenced in different depths.</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="co"># The option --SPMR only affects the bedGraph output.</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>rm_dup_bam_dir <span class="op">=</span> <span class="st">"rm_dup_bam"</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"tmp"</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>peak_with_ctl_dir <span class="op">=</span> <span class="st">"peak_with_ctl_qval0.05"</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>effective_genome_size <span class="op">=</span> <span class="fl">2652684646</span></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>qvalue <span class="op">=</span> <span class="fl">0.05</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>cap_num <span class="op">=</span> <span class="fl">1000000</span></span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>chrsz <span class="op">=</span> <span class="st">"/data/biodatabase/species/mm10/genome/genome/mm10_no_alt.chrom.sizes.tsv"</span></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a><span class="co"># [["control1", "treatment1"], ["control2", "treatment2"], ...]</span></span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a>ctl_treat_pairs <span class="op">=</span> [</span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"IgG_rep1.rm_dup.bam"</span>, <span class="st">"Satb2_rep1.rm_dup.bam"</span>],</span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"IgG_rep2.rm_dup.bam"</span>, <span class="st">"Satb2_rep2.rm_dup.bam"</span>],</span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"IgG_rep3.rm_dup.bam"</span>, <span class="st">"Satb2_rep3.rm_dup.bam"</span>],</span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"IgG_rep1_vs_IgG_rep2.rm_dup.bam"</span>, <span class="st">"Satb2_rep1_vs_Satb2_rep2.rm_dup.bam"</span>],</span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"IgG_rep1_vs_IgG_rep3.rm_dup.bam"</span>, <span class="st">"Satb2_rep1_vs_Satb2_rep3.rm_dup.bam"</span>],</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"IgG_rep2_vs_IgG_rep3.rm_dup.bam"</span>, <span class="st">"Satb2_rep2_vs_Satb2_rep3.rm_dup.bam"</span>],</span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(peak_with_ctl_dir)</span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ctl_treat_pair <span class="kw">in</span> ctl_treat_pairs</span>
<span id="cb26-26"><a href="#cb26-26" aria-hidden="true" tabindex="-1"></a>    prefix <span class="op">=</span> <span class="fu">join</span>(<span class="fu">reverse</span>(<span class="fu">replace</span>.(ctl_treat_pair, <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; "")), "</span>_vs_<span class="st">"</span>)</span>
<span id="cb26-27"><a href="#cb26-27" aria-hidden="true" tabindex="-1"></a>    macs_peak_file <span class="op">=</span> <span class="fu">joinpath</span>(peak_with_ctl_dir, <span class="fu">string</span>(prefix, <span class="st">"_peaks.narrowPeak"</span>))</span>
<span id="cb26-28"><a href="#cb26-28" aria-hidden="true" tabindex="-1"></a>    final_peak_file <span class="op">=</span> <span class="fu">joinpath</span>(peak_with_ctl_dir, <span class="fu">string</span>(prefix, <span class="st">".qval"</span>, qvalue, <span class="st">".narrowPeak.gz"</span>))</span>
<span id="cb26-29"><a href="#cb26-29" aria-hidden="true" tabindex="-1"></a>    tmp_peak_file1 <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(prefix, <span class="st">".tmp1"</span>))</span>
<span id="cb26-30"><a href="#cb26-30" aria-hidden="true" tabindex="-1"></a>    tmp_peak_file2 <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(prefix, <span class="st">".tmp2"</span>))</span>
<span id="cb26-31"><a href="#cb26-31" aria-hidden="true" tabindex="-1"></a>    tmp_peak_file3 <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(prefix, <span class="st">".tmp3"</span>))</span>
<span id="cb26-32"><a href="#cb26-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-33"><a href="#cb26-33" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb26-34"><a href="#cb26-34" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb26-35"><a href="#cb26-35" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"macs3 callpeak "</span>,</span>
<span id="cb26-36"><a href="#cb26-36" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -c "</span>, <span class="fu">joinpath</span>(rm_dup_bam_dir, ctl_treat_pair[<span class="fl">1</span>]),</span>
<span id="cb26-37"><a href="#cb26-37" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -t "</span>, <span class="fu">joinpath</span>(rm_dup_bam_dir, ctl_treat_pair[<span class="fl">2</span>]),</span>
<span id="cb26-38"><a href="#cb26-38" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -g "</span>, effective_genome_size,</span>
<span id="cb26-39"><a href="#cb26-39" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -n "</span>, prefix,</span>
<span id="cb26-40"><a href="#cb26-40" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --outdir "</span>, peak_with_ctl_dir,</span>
<span id="cb26-41"><a href="#cb26-41" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --tempdir "</span>, tmp_dir,</span>
<span id="cb26-42"><a href="#cb26-42" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -q "</span>, qvalue,</span>
<span id="cb26-43"><a href="#cb26-43" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -f BAMPE --keep-dup all -B --SPMR --call-summits"</span>)]))</span>
<span id="cb26-44"><a href="#cb26-44" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb26-45"><a href="#cb26-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb26-46"><a href="#cb26-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-47"><a href="#cb26-47" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Sort by Column 8 in descending order and replace long peak names in Column 4 with Peak_&lt;rank&gt;</span></span>
<span id="cb26-48"><a href="#cb26-48" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb26-49"><a href="#cb26-49" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb26-50"><a href="#cb26-50" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"LC_COLLATE=C sort -k 8gr,8gr "</span>, macs_peak_file,</span>
<span id="cb26-51"><a href="#cb26-51" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw""" | awk -v FS="\t" -v OFS="\t" '{$4="Peak_"NR; if ($2&lt;0) $2=0; if ($3&lt;0) $3=0; if ($10==-1) $10=$2+int(($3-$2+1)/2.0); print $0}' &gt; """</span>, tmp_peak_file1)]))</span>
<span id="cb26-52"><a href="#cb26-52" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb26-53"><a href="#cb26-53" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb26-54"><a href="#cb26-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-55"><a href="#cb26-55" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb26-56"><a href="#cb26-56" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb26-57"><a href="#cb26-57" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"head -n "</span>, cap_num, <span class="st">" "</span>, tmp_peak_file1, <span class="st">" &gt; "</span>, tmp_peak_file2)]))</span>
<span id="cb26-58"><a href="#cb26-58" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb26-59"><a href="#cb26-59" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb26-60"><a href="#cb26-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-61"><a href="#cb26-61" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb26-62"><a href="#cb26-62" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb26-63"><a href="#cb26-63" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"bedClip "</span>, tmp_peak_file2, <span class="st">" "</span>, chrsz, <span class="st">" "</span>, tmp_peak_file3, <span class="st">" -truncate -verbose=2"</span>)]))</span>
<span id="cb26-64"><a href="#cb26-64" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb26-65"><a href="#cb26-65" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb26-66"><a href="#cb26-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-67"><a href="#cb26-67" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb26-68"><a href="#cb26-68" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb26-69"><a href="#cb26-69" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"cat "</span>, tmp_peak_file3, <span class="st">" | pigz -nc &gt; "</span>, final_peak_file)]))</span>
<span id="cb26-70"><a href="#cb26-70" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb26-71"><a href="#cb26-71" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb26-72"><a href="#cb26-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-73"><a href="#cb26-73" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>.([tmp_peak_file1, tmp_peak_file2, tmp_peak_file3])</span>
<span id="cb26-74"><a href="#cb26-74" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="call-peaks-without-controls-using-p-value" class="level4" data-number="15.2.3">
<h4 data-number="15.2.3" class="anchored" data-anchor-id="call-peaks-without-controls-using-p-value"><span class="header-section-number">15.2.3</span> Call peaks without controls (using p-value)</h4>
<div id="dc65df87" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>rm_dup_bam_dir <span class="op">=</span> <span class="st">"rm_dup_bam"</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"tmp"</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>peak_without_ctl_dir <span class="op">=</span> <span class="st">"peak_without_ctl_pval0.01"</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>effective_genome_size <span class="op">=</span> <span class="fl">2652684646</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>pvalue <span class="op">=</span> <span class="fl">0.01</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>cap_num <span class="op">=</span> <span class="fl">1000000</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>chrsz <span class="op">=</span> <span class="st">"/data/biodatabase/species/mm10/genome/genome/mm10_no_alt.chrom.sizes.tsv"</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>treats <span class="op">=</span> [</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Satb2_rep1.rm_dup.bam"</span>,</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Satb2_rep2.rm_dup.bam"</span>,</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Satb2_rep3.rm_dup.bam"</span>,</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Satb2_rep1_vs_Satb2_rep2.rm_dup.bam"</span>,</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Satb2_rep1_vs_Satb2_rep3.rm_dup.bam"</span>,</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Satb2_rep2_vs_Satb2_rep3.rm_dup.bam"</span>,</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(peak_without_ctl_dir)</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treat <span class="kw">in</span> treats</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>    prefix <span class="op">=</span> <span class="fu">replace</span>(treat, <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; "")</span></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a><span class="st">    </span>macs_peak_file<span class="st"> = joinpath</span><span class="ch">(</span><span class="st">peak_without_ctl_dir, string</span><span class="ch">(</span><span class="st">prefix, "</span>_peaks.narrowPeak<span class="st">"))</span></span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>    final_peak_file <span class="op">=</span> <span class="fu">joinpath</span>(peak_without_ctl_dir, <span class="fu">string</span>(prefix, <span class="st">".pval"</span>, pvalue, <span class="st">".narrowPeak.gz"</span>))</span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a>    tmp_peak_file1 <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(prefix, <span class="st">".tmp1"</span>))</span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a>    tmp_peak_file2 <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(prefix, <span class="st">".tmp2"</span>))</span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a>    tmp_peak_file3 <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(prefix, <span class="st">".tmp3"</span>))</span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-26"><a href="#cb27-26" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb27-27"><a href="#cb27-27" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb27-28"><a href="#cb27-28" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"macs3 callpeak "</span>,</span>
<span id="cb27-29"><a href="#cb27-29" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -t "</span>, <span class="fu">joinpath</span>(rm_dup_bam_dir, treat),</span>
<span id="cb27-30"><a href="#cb27-30" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -g "</span>, effective_genome_size,</span>
<span id="cb27-31"><a href="#cb27-31" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -n "</span>, prefix,</span>
<span id="cb27-32"><a href="#cb27-32" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --outdir "</span>, peak_without_ctl_dir,</span>
<span id="cb27-33"><a href="#cb27-33" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --tempdir "</span>, tmp_dir,</span>
<span id="cb27-34"><a href="#cb27-34" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -p "</span>, pvalue,</span>
<span id="cb27-35"><a href="#cb27-35" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -f BAMPE --keep-dup all -B --SPMR --call-summits"</span>)]))</span>
<span id="cb27-36"><a href="#cb27-36" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb27-37"><a href="#cb27-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb27-38"><a href="#cb27-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-39"><a href="#cb27-39" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Sort by Column 8 in descending order and replace long peak names in Column 4 with Peak_&lt;rank&gt;</span></span>
<span id="cb27-40"><a href="#cb27-40" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb27-41"><a href="#cb27-41" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb27-42"><a href="#cb27-42" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"LC_COLLATE=C sort -k 8gr,8gr "</span>, macs_peak_file,</span>
<span id="cb27-43"><a href="#cb27-43" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw""" | awk -v FS="\t" -v OFS="\t" '{$4="Peak_"NR; if ($2&lt;0) $2=0; if ($3&lt;0) $3=0; if ($10==-1) $10=$2+int(($3-$2+1)/2.0); print $0}' &gt; """</span>, tmp_peak_file1)]))</span>
<span id="cb27-44"><a href="#cb27-44" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb27-45"><a href="#cb27-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb27-46"><a href="#cb27-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-47"><a href="#cb27-47" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb27-48"><a href="#cb27-48" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb27-49"><a href="#cb27-49" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"head -n "</span>, cap_num, <span class="st">" "</span>, tmp_peak_file1, <span class="st">" &gt; "</span>, tmp_peak_file2)]))</span>
<span id="cb27-50"><a href="#cb27-50" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb27-51"><a href="#cb27-51" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb27-52"><a href="#cb27-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-53"><a href="#cb27-53" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb27-54"><a href="#cb27-54" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb27-55"><a href="#cb27-55" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"bedClip "</span>, tmp_peak_file2, <span class="st">" "</span>, chrsz, <span class="st">" "</span>, tmp_peak_file3, <span class="st">" -truncate -verbose=2"</span>)]))</span>
<span id="cb27-56"><a href="#cb27-56" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb27-57"><a href="#cb27-57" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb27-58"><a href="#cb27-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-59"><a href="#cb27-59" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb27-60"><a href="#cb27-60" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb27-61"><a href="#cb27-61" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"cat "</span>, tmp_peak_file3, <span class="st">" | pigz -nc &gt; "</span>, final_peak_file)]))</span>
<span id="cb27-62"><a href="#cb27-62" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb27-63"><a href="#cb27-63" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb27-64"><a href="#cb27-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-65"><a href="#cb27-65" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>.([tmp_peak_file1, tmp_peak_file2, tmp_peak_file3])</span>
<span id="cb27-66"><a href="#cb27-66" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="call-peaks-without-controls-using-q-value" class="level4" data-number="15.2.4">
<h4 data-number="15.2.4" class="anchored" data-anchor-id="call-peaks-without-controls-using-q-value"><span class="header-section-number">15.2.4</span> Call peaks without controls (using q-value)</h4>
<div id="43debf8c" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>rm_dup_bam_dir <span class="op">=</span> <span class="st">"rm_dup_bam"</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"tmp"</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>peak_without_ctl_dir <span class="op">=</span> <span class="st">"peak_without_ctl_qval0.05"</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>effective_genome_size <span class="op">=</span> <span class="fl">2652684646</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>qvalue <span class="op">=</span> <span class="fl">0.05</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>cap_num <span class="op">=</span> <span class="fl">1000000</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>chrsz <span class="op">=</span> <span class="st">"/data/biodatabase/species/mm10/genome/genome/mm10_no_alt.chrom.sizes.tsv"</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>treats <span class="op">=</span> [</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Satb2_rep1.rm_dup.bam"</span>,</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Satb2_rep2.rm_dup.bam"</span>,</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Satb2_rep3.rm_dup.bam"</span>,</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Satb2_rep1_vs_Satb2_rep2.rm_dup.bam"</span>,</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Satb2_rep1_vs_Satb2_rep3.rm_dup.bam"</span>,</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Satb2_rep2_vs_Satb2_rep3.rm_dup.bam"</span>,</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(peak_without_ctl_dir)</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> treat <span class="kw">in</span> treats</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>    prefix <span class="op">=</span> <span class="fu">replace</span>(treat, <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; "")</span></span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a><span class="st">    </span>macs_peak_file<span class="st"> = joinpath</span><span class="ch">(</span><span class="st">peak_without_ctl_dir, string</span><span class="ch">(</span><span class="st">prefix, "</span>_peaks.narrowPeak<span class="st">"))</span></span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>    final_peak_file <span class="op">=</span> <span class="fu">joinpath</span>(peak_without_ctl_dir, <span class="fu">string</span>(prefix, <span class="st">".qval"</span>, qvalue, <span class="st">".narrowPeak.gz"</span>))</span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a>    tmp_peak_file1 <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(prefix, <span class="st">".tmp1"</span>))</span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>    tmp_peak_file2 <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(prefix, <span class="st">".tmp2"</span>))</span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>    tmp_peak_file3 <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(prefix, <span class="st">".tmp3"</span>))</span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"macs3 callpeak "</span>,</span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -t "</span>, <span class="fu">joinpath</span>(rm_dup_bam_dir, treat),</span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -g "</span>, effective_genome_size,</span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -n "</span>, prefix,</span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --outdir "</span>, peak_without_ctl_dir,</span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --tempdir "</span>, tmp_dir,</span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -q "</span>, qvalue,</span>
<span id="cb28-35"><a href="#cb28-35" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -f BAMPE --keep-dup all -B --SPMR --call-summits"</span>)]))</span>
<span id="cb28-36"><a href="#cb28-36" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb28-37"><a href="#cb28-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb28-38"><a href="#cb28-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-39"><a href="#cb28-39" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Sort by Column 8 in descending order and replace long peak names in Column 4 with Peak_&lt;rank&gt;</span></span>
<span id="cb28-40"><a href="#cb28-40" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb28-41"><a href="#cb28-41" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb28-42"><a href="#cb28-42" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"LC_COLLATE=C sort -k 8gr,8gr "</span>, macs_peak_file,</span>
<span id="cb28-43"><a href="#cb28-43" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw""" | awk -v FS="\t" -v OFS="\t" '{$4="Peak_"NR; if ($2&lt;0) $2=0; if ($3&lt;0) $3=0; if ($10==-1) $10=$2+int(($3-$2+1)/2.0); print $0}' &gt; """</span>, tmp_peak_file1)]))</span>
<span id="cb28-44"><a href="#cb28-44" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb28-45"><a href="#cb28-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb28-46"><a href="#cb28-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-47"><a href="#cb28-47" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb28-48"><a href="#cb28-48" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb28-49"><a href="#cb28-49" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"head -n "</span>, cap_num, <span class="st">" "</span>, tmp_peak_file1, <span class="st">" &gt; "</span>, tmp_peak_file2)]))</span>
<span id="cb28-50"><a href="#cb28-50" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb28-51"><a href="#cb28-51" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb28-52"><a href="#cb28-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-53"><a href="#cb28-53" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb28-54"><a href="#cb28-54" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb28-55"><a href="#cb28-55" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"bedClip "</span>, tmp_peak_file2, <span class="st">" "</span>, chrsz, <span class="st">" "</span>, tmp_peak_file3, <span class="st">" -truncate -verbose=2"</span>)]))</span>
<span id="cb28-56"><a href="#cb28-56" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb28-57"><a href="#cb28-57" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb28-58"><a href="#cb28-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-59"><a href="#cb28-59" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb28-60"><a href="#cb28-60" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb28-61"><a href="#cb28-61" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"cat "</span>, tmp_peak_file3, <span class="st">" | pigz -nc &gt; "</span>, final_peak_file)]))</span>
<span id="cb28-62"><a href="#cb28-62" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb28-63"><a href="#cb28-63" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb28-64"><a href="#cb28-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-65"><a href="#cb28-65" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>.([tmp_peak_file1, tmp_peak_file2, tmp_peak_file3])</span>
<span id="cb28-66"><a href="#cb28-66" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
</section>
<section id="naive-overlapping" class="level2" data-number="16">
<h2 data-number="16" class="anchored" data-anchor-id="naive-overlapping"><span class="header-section-number">16</span> Naive overlapping</h2>
<div id="5a254e9d" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [["pooled_peaks", "rep1_peaks", "rep2_peaks", "output_peaks"], ...]</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>narrow_peaks <span class="op">=</span> [</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep1_vs_Satb2_rep2.qval0.05.narrowPeak.gz"</span>,</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep1.qval0.05.narrowPeak.gz"</span>,</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep2.qval0.05.narrowPeak.gz"</span>,</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep1_vs_Satb2_rep2.qval0.05.naive_overlap.narrowPeak.gz"</span>,</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep1_vs_Satb2_rep3.qval0.05.narrowPeak.gz"</span>,</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep1.qval0.05.narrowPeak.gz"</span>,</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep3.qval0.05.narrowPeak.gz"</span>,</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep1_vs_Satb2_rep3.qval0.05.naive_overlap.narrowPeak.gz"</span>,</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep2_vs_Satb2_rep3.qval0.05.narrowPeak.gz"</span>,</span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep2.qval0.05.narrowPeak.gz"</span>,</span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep3.qval0.05.narrowPeak.gz"</span>,</span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep2_vs_Satb2_rep3.qval0.05.naive_overlap.narrowPeak.gz"</span>,</span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a>input_dir <span class="op">=</span> <span class="st">"peak_without_ctl_qval0.05"</span></span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a>output_dir <span class="op">=</span> <span class="st">"naive_overlap_without_ctl_qval0.05"</span></span>
<span id="cb29-24"><a href="#cb29-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-25"><a href="#cb29-25" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(output_dir)</span>
<span id="cb29-26"><a href="#cb29-26" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> narrow_peak <span class="kw">in</span> narrow_peaks</span>
<span id="cb29-27"><a href="#cb29-27" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb29-28"><a href="#cb29-28" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb29-29"><a href="#cb29-29" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"bedtools intersect -wo "</span>,</span>
<span id="cb29-30"><a href="#cb29-30" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -a "</span>, <span class="fu">joinpath</span>(input_dir, narrow_peak[<span class="fl">1</span>]),</span>
<span id="cb29-31"><a href="#cb29-31" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -b "</span>, <span class="fu">joinpath</span>(input_dir, narrow_peak[<span class="fl">2</span>]),</span>
<span id="cb29-32"><a href="#cb29-32" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw" | awk -v FS='\t' -v OFS='\t' '{s1=$3-$2; s2=$13-$12; if (($21/s1 &gt;= 0.5) || ($21/s2 &gt;= 0.5)) {print $0}}' "</span>,</span>
<span id="cb29-33"><a href="#cb29-33" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | cut -f 1-10 | sort -k1,1 -k2,2n | uniq "</span>,</span>
<span id="cb29-34"><a href="#cb29-34" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | bedtools intersect -wo "</span>,</span>
<span id="cb29-35"><a href="#cb29-35" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -a stdin -b "</span>, <span class="fu">joinpath</span>(input_dir, narrow_peak[<span class="fl">3</span>]),</span>
<span id="cb29-36"><a href="#cb29-36" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw" | awk -v FS='\t' -v OFS='\t' '{s1=$3-$2; s2=$13-$12; if (($21/s1 &gt;= 0.5) || ($21/s2 &gt;= 0.5)) {print $0}}' "</span>,</span>
<span id="cb29-37"><a href="#cb29-37" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | cut -f 1-10 | sort -k1,1 -k2,2n | uniq | pigz -nc &gt; "</span>, <span class="fu">joinpath</span>(output_dir, narrow_peak[<span class="fl">4</span>]))]))</span>
<span id="cb29-38"><a href="#cb29-38" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb29-39"><a href="#cb29-39" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb29-40"><a href="#cb29-40" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="idr-analysis" class="level2" data-number="17">
<h2 data-number="17" class="anchored" data-anchor-id="idr-analysis"><span class="header-section-number">17</span> IDR analysis</h2>
<div id="4c988128" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [["pooled_peaks", "rep1_peaks", "rep2_peaks", "output_prefix"], ...]</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>narrow_peaks <span class="op">=</span> [</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep1_vs_Satb2_rep2.pval0.01.narrowPeak.gz"</span>,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep1.pval0.01.narrowPeak.gz"</span>,</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep2.pval0.01.narrowPeak.gz"</span>,</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep1_vs_Satb2_rep2.pval0.01"</span>,</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep1_vs_Satb2_rep3.pval0.01.narrowPeak.gz"</span>,</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep1.pval0.01.narrowPeak.gz"</span>,</span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep3.pval0.01.narrowPeak.gz"</span>,</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep1_vs_Satb2_rep3.pval0.01"</span>,</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep2_vs_Satb2_rep3.pval0.01.narrowPeak.gz"</span>,</span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep2.pval0.01.narrowPeak.gz"</span>,</span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep3.pval0.01.narrowPeak.gz"</span>,</span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Satb2_rep2_vs_Satb2_rep3.pval0.01"</span>,</span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a>input_dir <span class="op">=</span> <span class="st">"peak_without_ctl_pval0.01"</span></span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a>output_dir <span class="op">=</span> <span class="st">"idr_without_ctl_pval0.01"</span></span>
<span id="cb30-24"><a href="#cb30-24" aria-hidden="true" tabindex="-1"></a>chrsz <span class="op">=</span> <span class="st">"/data/biodatabase/species/mm10/genome/genome/mm10_no_alt.chrom.sizes.tsv"</span></span>
<span id="cb30-25"><a href="#cb30-25" aria-hidden="true" tabindex="-1"></a>idr_threshold <span class="op">=</span> <span class="fl">0.05</span></span>
<span id="cb30-26"><a href="#cb30-26" aria-hidden="true" tabindex="-1"></a>peak_type <span class="op">=</span> <span class="st">"narrowPeak"</span></span>
<span id="cb30-27"><a href="#cb30-27" aria-hidden="true" tabindex="-1"></a>rank <span class="op">=</span> <span class="st">"p.value"</span></span>
<span id="cb30-28"><a href="#cb30-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-29"><a href="#cb30-29" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(output_dir)</span>
<span id="cb30-30"><a href="#cb30-30" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> narrow_peak <span class="kw">in</span> narrow_peaks</span>
<span id="cb30-31"><a href="#cb30-31" aria-hidden="true" tabindex="-1"></a>    peak1_file <span class="op">=</span> <span class="fu">joinpath</span>(input_dir, narrow_peak[<span class="fl">2</span>])</span>
<span id="cb30-32"><a href="#cb30-32" aria-hidden="true" tabindex="-1"></a>    peak2_file <span class="op">=</span> <span class="fu">joinpath</span>(input_dir, narrow_peak[<span class="fl">3</span>])</span>
<span id="cb30-33"><a href="#cb30-33" aria-hidden="true" tabindex="-1"></a>    pooled_peak_file <span class="op">=</span> <span class="fu">joinpath</span>(input_dir, narrow_peak[<span class="fl">1</span>])</span>
<span id="cb30-34"><a href="#cb30-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-35"><a href="#cb30-35" aria-hidden="true" tabindex="-1"></a>    prefix <span class="op">=</span> <span class="fu">joinpath</span>(output_dir, <span class="fu">string</span>(narrow_peak[<span class="fl">4</span>], <span class="st">".idr"</span>, idr_threshold))</span>
<span id="cb30-36"><a href="#cb30-36" aria-hidden="true" tabindex="-1"></a>    idr_peak_file <span class="op">=</span> <span class="fu">string</span>(prefix, <span class="st">"."</span>, peak_type, <span class="st">".gz"</span>)</span>
<span id="cb30-37"><a href="#cb30-37" aria-hidden="true" tabindex="-1"></a>    idr_log_file <span class="op">=</span> <span class="fu">string</span>(prefix, <span class="st">".log"</span>)</span>
<span id="cb30-38"><a href="#cb30-38" aria-hidden="true" tabindex="-1"></a>    idr_12col_bed_file <span class="op">=</span> <span class="fu">string</span>(prefix, <span class="st">"."</span>, peak_type, <span class="st">".12-col.bed.gz"</span>)</span>
<span id="cb30-39"><a href="#cb30-39" aria-hidden="true" tabindex="-1"></a>    idr_out_file <span class="op">=</span> <span class="fu">string</span>(prefix, <span class="st">".unthresholded-peaks.txt"</span>)</span>
<span id="cb30-40"><a href="#cb30-40" aria-hidden="true" tabindex="-1"></a>    idr_tmp_file <span class="op">=</span> <span class="fu">string</span>(prefix, <span class="st">".unthresholded-peaks.txt.tmp"</span>)</span>
<span id="cb30-41"><a href="#cb30-41" aria-hidden="true" tabindex="-1"></a>    idr_out_gz_file <span class="op">=</span> <span class="fu">string</span>(prefix, <span class="st">".unthresholded-peaks.txt.gz"</span>)</span>
<span id="cb30-42"><a href="#cb30-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-43"><a href="#cb30-43" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb30-44"><a href="#cb30-44" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb30-45"><a href="#cb30-45" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"idr "</span>,</span>
<span id="cb30-46"><a href="#cb30-46" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --samples "</span>, peak1_file, <span class="st">" "</span>, peak2_file,</span>
<span id="cb30-47"><a href="#cb30-47" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --peak-list "</span>, pooled_peak_file,</span>
<span id="cb30-48"><a href="#cb30-48" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --input-file-type "</span>, peak_type,</span>
<span id="cb30-49"><a href="#cb30-49" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --output-file "</span>, idr_out_file,</span>
<span id="cb30-50"><a href="#cb30-50" aria-hidden="true" tabindex="-1"></a>                <span class="st">"  --rank "</span>, rank,</span>
<span id="cb30-51"><a href="#cb30-51" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --soft-idr-threshold "</span>, idr_threshold,</span>
<span id="cb30-52"><a href="#cb30-52" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --log-output-file "</span>, idr_log_file,</span>
<span id="cb30-53"><a href="#cb30-53" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --plot --use-best-multisummit-IDR"</span>)]))</span>
<span id="cb30-54"><a href="#cb30-54" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb30-55"><a href="#cb30-55" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb30-56"><a href="#cb30-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-57"><a href="#cb30-57" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb30-58"><a href="#cb30-58" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb30-59"><a href="#cb30-59" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"bedClip "</span>, idr_out_file, <span class="st">" "</span>, chrsz, <span class="st">" "</span>, idr_tmp_file, <span class="st">" -truncate -verbose=2"</span>)]))</span>
<span id="cb30-60"><a href="#cb30-60" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb30-61"><a href="#cb30-61" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb30-62"><a href="#cb30-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-63"><a href="#cb30-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> rank <span class="op">==</span> <span class="st">"signal.value"</span></span>
<span id="cb30-64"><a href="#cb30-64" aria-hidden="true" tabindex="-1"></a>        col <span class="op">=</span> <span class="fl">7</span></span>
<span id="cb30-65"><a href="#cb30-65" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elseif</span> rank <span class="op">==</span> <span class="st">"p.value"</span></span>
<span id="cb30-66"><a href="#cb30-66" aria-hidden="true" tabindex="-1"></a>        col <span class="op">=</span> <span class="fl">8</span></span>
<span id="cb30-67"><a href="#cb30-67" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elseif</span> rank <span class="op">==</span> <span class="st">"q.value"</span></span>
<span id="cb30-68"><a href="#cb30-68" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="fl">9</span></span>
<span id="cb30-69"><a href="#cb30-69" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span></span>
<span id="cb30-70"><a href="#cb30-70" aria-hidden="true" tabindex="-1"></a>        <span class="pp">@error</span> <span class="st">"invalid score ranking method"</span></span>
<span id="cb30-71"><a href="#cb30-71" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb30-72"><a href="#cb30-72" aria-hidden="true" tabindex="-1"></a>    minus_log10_threshold <span class="op">=</span> <span class="fu">-log10</span>(idr_threshold)</span>
<span id="cb30-73"><a href="#cb30-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-74"><a href="#cb30-74" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb30-75"><a href="#cb30-75" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb30-76"><a href="#cb30-76" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"cat "</span>, idr_tmp_file,</span>
<span id="cb30-77"><a href="#cb30-77" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw" | awk -v FS='\t' -v OFS='\t' '$12&gt;="</span>, minus_log10_threshold,</span>
<span id="cb30-78"><a href="#cb30-78" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw" {if ($2&lt;0) $2=0; print $1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12}' "</span>,</span>
<span id="cb30-79"><a href="#cb30-79" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | sort -k1,1 -k2,2n | uniq | sort -grk"</span>, col, <span class="st">","</span>, col,</span>
<span id="cb30-80"><a href="#cb30-80" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | pigz -nc &gt; "</span>, idr_12col_bed_file)]))</span>
<span id="cb30-81"><a href="#cb30-81" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb30-82"><a href="#cb30-82" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb30-83"><a href="#cb30-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-84"><a href="#cb30-84" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb30-85"><a href="#cb30-85" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb30-86"><a href="#cb30-86" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"zcat "</span>, idr_12col_bed_file,</span>
<span id="cb30-87"><a href="#cb30-87" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw" | awk -v FS='\t' -v OFS='\t' '{print $1,$2,$3,$4,$5,$6,$7,$8,$9,$10}' "</span>,</span>
<span id="cb30-88"><a href="#cb30-88" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | pigz -nc &gt; "</span>, idr_peak_file)]))</span>
<span id="cb30-89"><a href="#cb30-89" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb30-90"><a href="#cb30-90" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb30-91"><a href="#cb30-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-92"><a href="#cb30-92" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb30-93"><a href="#cb30-93" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb30-94"><a href="#cb30-94" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"cat "</span>, idr_tmp_file,</span>
<span id="cb30-95"><a href="#cb30-95" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | pigz -nc &gt; "</span>, idr_out_gz_file)]))</span>
<span id="cb30-96"><a href="#cb30-96" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb30-97"><a href="#cb30-97" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb30-98"><a href="#cb30-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-99"><a href="#cb30-99" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>.([idr_out_file, idr_tmp_file, idr_12col_bed_file, <span class="fu">string</span>(idr_out_file, <span class="st">".noalternatesummitpeaks.png"</span>)])</span>
<span id="cb30-100"><a href="#cb30-100" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="pairwise-overlapping-statistics" class="level2" data-number="18">
<h2 data-number="18" class="anchored" data-anchor-id="pairwise-overlapping-statistics"><span class="header-section-number">18</span> Pairwise overlapping statistics</h2>
<pre class="{r}"><code>library(bedtoolsr)
library(vroom)
library(tidyverse)
library(ggprism)
library(YRUtils)
library(patchwork)

# ["peaks", "peaks", ...]
narrow_peaks &lt;- c(
    "Satb2_rep1_vs_Satb2_rep2.pval0.01.naive_overlap.narrowPeak.gz",
    "Satb2_rep1_vs_Satb2_rep3.pval0.01.naive_overlap.narrowPeak.gz",
    "Satb2_rep2_vs_Satb2_rep3.pval0.01.naive_overlap.narrowPeak.gz"
)
input_dir &lt;- "naive_overlap_without_ctl_pval0.01"
overlap_stat_dir &lt;- "overlap_stat"

dir.create(overlap_stat_dir)
combns &lt;- combn(narrow_peaks, 2)
for (i in seq_len(ncol(combns))) {
    narrow_peak_pair &lt;- file.path(input_dir, combns[, i])
    raw_df1 &lt;- vroom(gzfile(narrow_peak_pair[1]), col_names = FALSE) %&gt;%
        select(X1, X2, X3) %&gt;%
        arrange(X1, X2, X3) %&gt;%
        distinct()
    raw_df2 &lt;- vroom(gzfile(narrow_peak_pair[2]), col_names = FALSE) %&gt;%
        select(X1, X2, X3) %&gt;%
        arrange(X1, X2, X3) %&gt;%
        distinct()
    overlapped_df &lt;- bt.intersect(
        a = raw_df1,
        b = raw_df2,
        wo = TRUE
    )
    overlapped_df1 &lt;- overlapped_df %&gt;%
        select(V1, V2, V3, V7) %&gt;%
        arrange(V1, V2, V3) %&gt;%
        distinct() %&gt;%
        mutate(
            percent = V7 / (V3 - V2),
            interval = cut(percent,
                breaks = seq(0, 1, 0.1),
                include.lowest = TRUE,
                right = TRUE
            )
        )
    overlapped_df2 &lt;- overlapped_df %&gt;%
        select(V4, V5, V6, V7) %&gt;%
        arrange(V4, V5, V6) %&gt;%
        distinct() %&gt;%
        mutate(
            percent = V7 / (V6 - V5),
            interval = cut(percent,
                breaks = seq(0, 1, 0.1),
                include.lowest = TRUE,
                right = TRUE
            )
        )

    qc_metrics &lt;- paste0(
        "&gt;&gt;&gt; ", paste0(gsub("\\.narrowPeak\\.gz$", "", basename(narrow_peak_pair)), collapse = " vs. "), ": \n",
        "&gt;&gt; ", gsub("\\.narrowPeak\\.gz$", "", basename(narrow_peak_pair[1])), ": \n",
        "&gt; The number of peaks in total: ", nrow(raw_df1), "\n",
        "&gt; The number of peaks overlapped: ", nrow(overlapped_df1), "\n",
        "&gt; Percentage: ", round(nrow(overlapped_df1) / nrow(raw_df1), digits = 4), "\n",
        "&gt;&gt; ", gsub("\\.narrowPeak\\.gz$", "", basename(narrow_peak_pair[2])), ": \n",
        "&gt; The number of peaks in total: ", nrow(raw_df2), "\n",
        "&gt; The number of peaks overlapped: ", nrow(overlapped_df2), "\n",
        "&gt; Percentage: ", round(nrow(overlapped_df2) / nrow(raw_df2), digits = 4)
    )

    vroom_write_lines(qc_metrics,
        file = file.path(
            overlap_stat_dir,
            paste0(paste0(gsub("\\.narrowPeak\\.gz$", "", basename(narrow_peak_pair)), collapse = "_vs_"), ".txt")
        ),
        append = FALSE
    )

    p1 &lt;- ggplot(
        overlapped_df1 %&gt;%
            group_by(interval) %&gt;%
            reframe(n = n()) %&gt;%
            mutate(percent = n / sum(n)),
        aes(interval, percent)
    ) +
        geom_bar(stat = "identity") +
        scale_y_continuous(
            expand = expansion(0),
            limits = c(0, 1)
        ) +
        guides(x = guide_axis(angle = 30)) +
        labs(title = "1 vs. 2") +
        theme_prism()

    p2 &lt;- ggplot(
        overlapped_df2 %&gt;%
            group_by(interval) %&gt;%
            reframe(n = n()) %&gt;%
            mutate(percent = n / sum(n)),
        aes(interval, percent)
    ) +
        geom_bar(stat = "identity") +
        scale_y_continuous(
            expand = expansion(0),
            limits = c(0, 1)
        ) +
        guides(x = guide_axis(angle = 30)) +
        labs(title = "2 vs. 1") +
        theme_prism()

    ppreview(p1 | p2, file = file.path(
        overlap_stat_dir,
        paste0(paste0(gsub("\\.narrowPeak\\.gz$", "", basename(narrow_peak_pair)), collapse = "_vs_"), ".pdf")
    ))
}
</code></pre>
</section>
<section id="generate-signal-tracks-with-macs" class="level2" data-number="19">
<h2 data-number="19" class="anchored" data-anchor-id="generate-signal-tracks-with-macs"><span class="header-section-number">19</span> Generate signal tracks with MACS</h2>
<div id="a0fd7213" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>bdg_dir <span class="op">=</span> <span class="st">"peak_without_ctl_qval0.05"</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>chrsz <span class="op">=</span> <span class="st">"/data/biodatabase/species/mm10/genome/genome/mm10_no_alt.chrom.sizes.tsv"</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>bdg_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(bdg_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bdg</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>samples <span class="op">=</span> <span class="fu">unique</span>(<span class="fu">replace</span>.(<span class="fu">basename</span>.(bdg_files), <span class="st">r"</span><span class="ch">(</span><span class="st">_control_lambda</span><span class="ch">|</span><span class="st">_treat_pileup</span><span class="ch">)</span><span class="sc">.</span><span class="st">bdg</span><span class="sc">$</span><span class="st">" =&gt; ""))</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>for<span class="st"> sample in samples</span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>    bdg_prefix <span class="op">=</span> <span class="fu">joinpath</span>(bdg_dir, sample)</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>    treat_bdg_file <span class="op">=</span> <span class="fu">string</span>(bdg_prefix, <span class="st">"_treat_pileup.bdg"</span>)</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>    ctl_bdg_file <span class="op">=</span> <span class="fu">string</span>(bdg_prefix, <span class="st">"_control_lambda.bdg"</span>)</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>    fc_bdg_file <span class="op">=</span> <span class="fu">string</span>(bdg_prefix, <span class="st">".fc.signal.bdg"</span>)</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>    fc_srt_bdg_file <span class="op">=</span> <span class="fu">string</span>(bdg_prefix, <span class="st">".fc.signal.srt.bdg"</span>)</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>    fc_bw_file <span class="op">=</span> <span class="fu">string</span>(bdg_prefix, <span class="st">".fc.signal.bw"</span>)</span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># For fold enrichment signal tracks</span></span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"macs3 bdgcmp "</span>,</span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -t "</span>, treat_bdg_file,</span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -c "</span>, ctl_bdg_file,</span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a>                <span class="st">" --o-prefix "</span>, bdg_prefix,</span>
<span id="cb32-23"><a href="#cb32-23" aria-hidden="true" tabindex="-1"></a>                <span class="st">" -m FE"</span>)]))</span>
<span id="cb32-24"><a href="#cb32-24" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb32-25"><a href="#cb32-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb32-26"><a href="#cb32-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-27"><a href="#cb32-27" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb32-28"><a href="#cb32-28" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb32-29"><a href="#cb32-29" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"bedtools slop -i "</span>, <span class="fu">string</span>(bdg_prefix, <span class="st">"_FE.bdg"</span>), <span class="st">" -g "</span>, chrsz, <span class="st">" -b 0 "</span>,</span>
<span id="cb32-30"><a href="#cb32-30" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | bedClip stdin "</span>, chrsz, <span class="st">" "</span>, fc_bdg_file)]))</span>
<span id="cb32-31"><a href="#cb32-31" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb32-32"><a href="#cb32-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb32-33"><a href="#cb32-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-34"><a href="#cb32-34" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Sort and remove any overlapping regions in bedgraph by comparing two lines in a row</span></span>
<span id="cb32-35"><a href="#cb32-35" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb32-36"><a href="#cb32-36" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb32-37"><a href="#cb32-37" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"LC_COLLATE=C sort -k1,1 -k2,2n "</span>, fc_bdg_file,</span>
<span id="cb32-38"><a href="#cb32-38" aria-hidden="true" tabindex="-1"></a>                <span class="st">raw" | awk -v OFS='\t' '{if (NR==1 || NR&gt;1 &amp;&amp; (prev_chr!=$1 || prev_chr==$1 &amp;&amp; prev_chr_e&lt;=$2)) {print $0}; prev_chr=$1; prev_chr_e=$3;}' &gt; "</span>, fc_srt_bdg_file)]))</span>
<span id="cb32-39"><a href="#cb32-39" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb32-40"><a href="#cb32-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb32-41"><a href="#cb32-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-42"><a href="#cb32-42" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb32-43"><a href="#cb32-43" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb32-44"><a href="#cb32-44" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"bedGraphToBigWig "</span>, fc_srt_bdg_file, <span class="st">" "</span>, chrsz, <span class="st">" "</span>, fc_bw_file)]))</span>
<span id="cb32-45"><a href="#cb32-45" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb32-46"><a href="#cb32-46" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb32-47"><a href="#cb32-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-48"><a href="#cb32-48" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>.([fc_bdg_file, fc_srt_bdg_file, treat_bdg_file, ctl_bdg_file, <span class="fu">string</span>(bdg_prefix, <span class="st">"_FE.bdg"</span>)])</span>
<span id="cb32-49"><a href="#cb32-49" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="extending-peaks-on-both-sides-from-summits" class="level2" data-number="20">
<h2 data-number="20" class="anchored" data-anchor-id="extending-peaks-on-both-sides-from-summits"><span class="header-section-number">20</span> Extending peaks on both sides from summits</h2>
<pre class="{r}"><code>library(bedtoolsr)
library(vroom)
library(tidyverse)

extend_width &lt;- 250
merge_dist &lt;- 10
chrsz_file &lt;- "/data/biodatabase/species/mm10/genome/genome/mm10_no_alt.chrom.sizes.tsv"
peak_files &lt;- c(
    "idr_without_ctl_pval0.01/Satb2_rep1_vs_Satb2_rep2.pval0.01.idr0.05.narrowPeak.gz",
    "idr_without_ctl_pval0.01/Satb2_rep1_vs_Satb2_rep3.pval0.01.idr0.05.narrowPeak.gz",
    "idr_without_ctl_pval0.01/Satb2_rep2_vs_Satb2_rep3.pval0.01.idr0.05.narrowPeak.gz"
)

chrsz &lt;- vroom(chrsz_file, col_names = FALSE)
for (peak_file in peak_files) {
    tmp_df &lt;- vroom(gzfile(peak_file), col_names = FALSE) %&gt;%
        mutate(
            summit_seqname = X1,
            summit_start = X2 + X10,
            summit_end = summit_start,
        ) %&gt;%
        select(summit_seqname, summit_start, summit_end) %&gt;%
        arrange(summit_seqname, summit_start, summit_end) %&gt;%
        distinct()
    df &lt;- bt.slop(i = tmp_df, g = chrsz, b = extend_width) %&gt;%
        distinct() %&gt;%
        arrange(V1, V2, V3)
    df &lt;- bt.merge(df, d = merge_dist) %&gt;%
        mutate(
            V4 = paste0(V1, ":", V2, "-", V3),
            V5 = 1000,
            V6 = "+"
        )
    vroom_write(df,
        file = gsub("\\.narrowPeak\\.gz$", paste0(".summits.b", 2 * extend_width, ".bed"), peak_file),
        col_names = FALSE,
        append = FALSE
    )
}</code></pre>
</section>
<section id="generate-final-peak-sets" class="level2" data-number="21">
<h2 data-number="21" class="anchored" data-anchor-id="generate-final-peak-sets"><span class="header-section-number">21</span> Generate final peak sets</h2>
<p><strong>Scheme 1: (recommended)</strong></p>
<ol type="1">
<li><p>Call peaks with MACS using <code>p-value = 0.05</code></p></li>
<li><p>Perform pairwise IDR analysis using <code>idr threshold = 0.05</code></p></li>
<li><p>Merge peaks belonging to the same sample and merge peaks extended from summits belonging to the same sample</p></li>
</ol>
<p><strong>Scheme 2:</strong></p>
<ol type="1">
<li><p>Call peaks with MACS using <code>q-value = 0.05</code></p></li>
<li><p>Perform pairwise naive overlapping</p></li>
<li><p>Merge peaks belonging to the same sample and merge peaks extended from summits belonging to the same sample</p></li>
</ol>
<pre class="{r}"><code>library(bedtoolsr)
library(vroom)
library(tidyverse)

merge_dist &lt;- 10
output_dir &lt;- "final_peak"
peak_files &lt;- c(
    "idr_without_ctl_pval0.01/Satb2_rep1_vs_Satb2_rep2.pval0.01.idr0.05.summits.b500.bed",
    "idr_without_ctl_pval0.01/Satb2_rep1_vs_Satb2_rep3.pval0.01.idr0.05.summits.b500.bed",
    "idr_without_ctl_pval0.01/Satb2_rep2_vs_Satb2_rep3.pval0.01.idr0.05.summits.b500.bed"
)
output_file &lt;- "Satb2.without_ctl.pval0.01.idr0.05.summits.b500.merged.bed"

dir.create(output_dir)
df &lt;- tibble()
for (peak_file in peak_files) {
    df &lt;- bind_rows(
        df,
        vroom(gzfile(peak_file), col_names = FALSE) %&gt;%
            select(X1, X2, X3)
    )
}
df &lt;- distinct(df) %&gt;%
    arrange(X1, X2, X3)
df &lt;- bt.merge(df, d = merge_dist) %&gt;%
    mutate(
        V4 = paste0(V1, ":", V2, "-", V3),
        V5 = 1000,
        V6 = "+"
    )
vroom_write(df, file = file.path(output_dir, output_file), col_names = FALSE, append = FALSE)</code></pre>
</section>
<section id="de-novo-motif-finding-with-homer" class="level2" data-number="22">
<h2 data-number="22" class="anchored" data-anchor-id="de-novo-motif-finding-with-homer"><span class="header-section-number">22</span> De novo motif finding with Homer</h2>
<section id="de-novo-motif-finding-with-homer-1" class="level3" data-number="22.1">
<h3 data-number="22.1" class="anchored" data-anchor-id="de-novo-motif-finding-with-homer-1"><span class="header-section-number">22.1</span> De novo motif finding with Homer</h3>
<div id="d6e1fe2d" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>output_dir <span class="op">=</span> <span class="st">"/data/users/yangrui/mouse/cuttag_v20250108/final_peak/homer/de_novo/narrow_peak"</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"/data/users/yangrui/mouse/cuttag_v20250108/tmp"</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Peak file must have at least six columns</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>peak_file <span class="op">=</span> <span class="st">"/data/users/yangrui/mouse/cuttag_v20250108/final_peak/Satb2.without_ctl.pval0.01.idr0.05.narrowPeak.merged.bed"</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>genome <span class="op">=</span> <span class="st">"/data/biodatabase/species/mm10/genome/genome/mm10_no_alt_analysis_set_ENCODE.fasta.gz"</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>known_motifs_file <span class="op">=</span> <span class="st">"/data/biodatabase/motifs/all_motifs_rmdup.from_peca2.txt"</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>scan_size <span class="op">=</span> <span class="st">"given"</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>homer_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>extra_args <span class="op">=</span> <span class="st">""</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>new_peak_file <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(peak_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">" =&gt; ""))</span></span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>cmd<span class="st"> = Cmd</span><span class="ch">(</span></span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>(<span class="st">"zcat -f "</span>, peak_file,</span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>            <span class="st">raw""" | awk -v FS="\t" -v OFS="\t" '{if ($6==".") {$6="+"}; print $0}' &gt; """</span>, new_peak_file)]))</span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a><span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> !<span class="fu">isnothing</span>(<span class="fu">match</span>(<span class="st">r"</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>genome<span class="ch">))</span></span>
<span id="cb35-22"><a href="#cb35-22" aria-hidden="true" tabindex="-1"></a>    new_genome <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(genome), <span class="st">r"</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">" =&gt; ""))</span></span>
<span id="cb35-23"><a href="#cb35-23" aria-hidden="true" tabindex="-1"></a><span class="st">    </span>YRUtils<span class="sc">.</span><span class="st">ShellUtils</span><span class="sc">.</span><span class="st">pigz</span><span class="ch">(</span><span class="st">genome, new_genome; decompress=true, keep=true</span><span class="ch">)</span></span>
<span id="cb35-24"><a href="#cb35-24" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb35-25"><a href="#cb35-25" aria-hidden="true" tabindex="-1"></a>    new_genome <span class="op">=</span> genome</span>
<span id="cb35-26"><a href="#cb35-26" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb35-27"><a href="#cb35-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-28"><a href="#cb35-28" aria-hidden="true" tabindex="-1"></a>cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb35-29"><a href="#cb35-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb35-30"><a href="#cb35-30" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>(<span class="st">"findMotifsGenome.pl "</span>, new_peak_file, <span class="st">" "</span>, new_genome, <span class="st">" "</span>, output_dir,</span>
<span id="cb35-31"><a href="#cb35-31" aria-hidden="true" tabindex="-1"></a>            <span class="st">" -size "</span>, scan_size, <span class="st">" -p "</span>, homer_n_threads, <span class="st">" -preparsedDir "</span>, tmp_dir,</span>
<span id="cb35-32"><a href="#cb35-32" aria-hidden="true" tabindex="-1"></a>            <span class="st">" "</span>, extra_args, <span class="st">" "</span>, if !<span class="fu">isempty</span>(known_motifs_file)</span>
<span id="cb35-33"><a href="#cb35-33" aria-hidden="true" tabindex="-1"></a>                <span class="fu">string</span>(<span class="st">" -mknown "</span>, known_motifs_file)</span>
<span id="cb35-34"><a href="#cb35-34" aria-hidden="true" tabindex="-1"></a>            else</span>
<span id="cb35-35"><a href="#cb35-35" aria-hidden="true" tabindex="-1"></a>                <span class="st">""</span></span>
<span id="cb35-36"><a href="#cb35-36" aria-hidden="true" tabindex="-1"></a>            <span class="kw">end</span>)]))</span>
<span id="cb35-37"><a href="#cb35-37" aria-hidden="true" tabindex="-1"></a><span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb35-38"><a href="#cb35-38" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb35-39"><a href="#cb35-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-40"><a href="#cb35-40" aria-hidden="true" tabindex="-1"></a>cmd <span class="op">=</span> <span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>, <span class="fu">string</span>(<span class="st">"rm -rf "</span>, <span class="fu">joinpath</span>(tmp_dir, <span class="st">"*"</span>))]))</span>
<span id="cb35-41"><a href="#cb35-41" aria-hidden="true" tabindex="-1"></a><span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb35-42"><a href="#cb35-42" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="finding-instances-of-specific-motifs" class="level3" data-number="22.2">
<h3 data-number="22.2" class="anchored" data-anchor-id="finding-instances-of-specific-motifs"><span class="header-section-number">22.2</span> Finding instances of specific motifs</h3>
<div id="394a6f7b" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>output_dir <span class="op">=</span> <span class="st">"/data/users/yangrui/mouse/cuttag_v20250108/tmp"</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>output_file <span class="op">=</span> <span class="st">"motif_instances.txt"</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"/data/users/yangrui/mouse/cuttag_v20250108/tmp"</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Peak file must have at least six columns</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>peak_file <span class="op">=</span> <span class="st">"/data/users/yangrui/mouse/cuttag_v20250108/final_peak/Satb2.without_ctl.pval0.01.idr0.05.summits.b500.merged.bed"</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>genome <span class="op">=</span> <span class="st">"/data/biodatabase/species/mm10/genome/genome/mm10_no_alt_analysis_set_ENCODE.fasta.gz"</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>known_motifs_file <span class="op">=</span> <span class="st">"/data/biodatabase/motifs/all_motifs_rmdup.from_peca2.txt"</span></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>scan_size <span class="op">=</span> <span class="st">"given"</span></span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>homer_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>extra_args <span class="op">=</span> <span class="st">""</span></span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>new_peak_file <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(peak_file), <span class="st">r"</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">" =&gt; ""))</span></span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>cmd<span class="st"> = Cmd</span><span class="ch">(</span></span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>(<span class="st">"zcat -f "</span>, peak_file,</span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a>            <span class="st">raw""" | awk -v FS="\t" -v OFS="\t" '{if ($6==".") {$6="+"}; print $0}' &gt; """</span>, new_peak_file)]))</span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a><span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-22"><a href="#cb36-22" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> !<span class="fu">isnothing</span>(<span class="fu">match</span>(<span class="st">r"</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>genome<span class="ch">))</span></span>
<span id="cb36-23"><a href="#cb36-23" aria-hidden="true" tabindex="-1"></a>    new_genome <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(genome), <span class="st">r"</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">" =&gt; ""))</span></span>
<span id="cb36-24"><a href="#cb36-24" aria-hidden="true" tabindex="-1"></a><span class="st">    </span>YRUtils<span class="sc">.</span><span class="st">ShellUtils</span><span class="sc">.</span><span class="st">pigz</span><span class="ch">(</span><span class="st">genome, new_genome; decompress=true, keep=true</span><span class="ch">)</span></span>
<span id="cb36-25"><a href="#cb36-25" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb36-26"><a href="#cb36-26" aria-hidden="true" tabindex="-1"></a>    new_genome <span class="op">=</span> genome</span>
<span id="cb36-27"><a href="#cb36-27" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb36-28"><a href="#cb36-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-29"><a href="#cb36-29" aria-hidden="true" tabindex="-1"></a>cmd <span class="op">=</span> <span class="fu">Cmd</span>(</span>
<span id="cb36-30"><a href="#cb36-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb36-31"><a href="#cb36-31" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>(<span class="st">"findMotifsGenome.pl "</span>, new_peak_file, <span class="st">" "</span>, new_genome, <span class="st">" "</span>, output_dir,</span>
<span id="cb36-32"><a href="#cb36-32" aria-hidden="true" tabindex="-1"></a>            <span class="st">" -size "</span>, scan_size, <span class="st">" -p "</span>, homer_n_threads, <span class="st">" -preparsedDir "</span>, tmp_dir,</span>
<span id="cb36-33"><a href="#cb36-33" aria-hidden="true" tabindex="-1"></a>            <span class="st">" -find "</span>, known_motifs_file, <span class="st">" "</span>, extra_args, <span class="st">" &gt; "</span>, <span class="fu">joinpath</span>(output_dir, output_file))]))</span>
<span id="cb36-34"><a href="#cb36-34" aria-hidden="true" tabindex="-1"></a><span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb36-35"><a href="#cb36-35" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb36-36"><a href="#cb36-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-37"><a href="#cb36-37" aria-hidden="true" tabindex="-1"></a>cmd <span class="op">=</span> <span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>, <span class="fu">string</span>(<span class="st">"rm -rf "</span>, <span class="fu">joinpath</span>(tmp_dir, <span class="st">"*"</span>))]))</span>
<span id="cb36-38"><a href="#cb36-38" aria-hidden="true" tabindex="-1"></a><span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb36-39"><a href="#cb36-39" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="peak-annotation" class="level2" data-number="23">
<h2 data-number="23" class="anchored" data-anchor-id="peak-annotation"><span class="header-section-number">23</span> Peak annotation</h2>
<section id="make-txdb-object-from-gff3gtf-file" class="level3" data-number="23.1">
<h3 data-number="23.1" class="anchored" data-anchor-id="make-txdb-object-from-gff3gtf-file"><span class="header-section-number">23.1</span> Make TxDb object from GFF3/GTF file</h3>
<pre class="{r}"><code>library(txdbmaker)
library(vroom)
library(tidyverse)
library(GenomeInfoDb)

anno_file &lt;- "/data/biodatabase/species/mm10/genome/anno/gencode.vM21.primary_assembly.annotation_UCSC_names.gtf.gz"
chrsz_file &lt;- "/data/biodatabase/species/mm10/genome/genome/mm10_no_alt.chrom.sizes.tsv"
anno_format &lt;- "gtf"
organism &lt;- "Mus musculus"
# Sugar glider (Petaurus breviceps): 34899
# Mouse (Mus musculus): 10090
taxonomy_id &lt;- 10090
genome &lt;- "mm10"
circular_chrs &lt;- c("chrM")

chrsz &lt;- vroom(chrsz_file, col_names = FALSE)
seqinfo &lt;- Seqinfo(
    seqnames = chrsz$X1,
    seqlengths = chrsz$X2,
    isCircular = if_else(chrsz$X1 %in% circular_chrs, TRUE, FALSE),
    genome = genome
)
txdb &lt;- makeTxDbFromGFF(
    file = anno_file,
    format = anno_format,
    organism = organism,
    taxonomyId = taxonomy_id,
    chrominfo = seqinfo
)
# loadDB()
saveDb(txdb, file = gsub("\\.\\w+(\\.gz)*$", ".TxDb.sqlite", anno_file))</code></pre>
</section>
<section id="profile-of-peaks-binding-to-tssbodytts-regions" class="level3" data-number="23.2">
<h3 data-number="23.2" class="anchored" data-anchor-id="profile-of-peaks-binding-to-tssbodytts-regions"><span class="header-section-number">23.2</span> Profile of peaks binding to TSS/body/TTS regions</h3>
<pre class="{r}"><code>library(ChIPseeker)
library(YRUtils)
library(AnnotationDbi)
library(ggprism)
library(ggplot2)

peak_file &lt;- "/data/users/yangrui/mouse/cuttag_v20250108/final_peak/Satb2.without_ctl.pval0.01.idr0.05.narrowPeak.merged.bed"
txdb_file &lt;- "/data/biodatabase/species/mm10/genome/anno/gencode.vM21.primary_assembly.annotation_UCSC_names.TxDb.sqlite"
peak_anno_dir &lt;- "final_peak/peak_anno/narrow_peak"
upstream_dist &lt;- 2000
downstream_dist &lt;- 2000
by_what &lt;- "gene"
types &lt;- c("start_site", "body", "end_site")
nbin &lt;- 400

peak &lt;- readPeakFile(peak_file)
txdb &lt;- loadDb(txdb_file)
for (type in types) {
    profile_p &lt;- plotPeakProf2(
        peak = peak,
        upstream = upstream_dist,
        downstream = downstream_dist,
        by = by_what,
        type = type,
        nbin = if (type == "body") nbin else NULL,
        TxDb = txdb
    ) + theme_prism(
        base_size = 14,
        base_family = "Arial",
        border = TRUE
    )

    profile_p_file &lt;- file.path(
        peak_anno_dir,
        gsub(
            "\\.[a-zA-Z0-9]+$",
            paste0(".", by_what, ".", type, ".profile.pdf"),
            basename(peak_file)
        )
    )
    ppreview(profile_p, file = profile_p_file)

    heatmap_p &lt;- peakHeatmap(
        peak = peak,
        upstream = upstream_dist,
        downstream = downstream_dist,
        by = by_what,
        type = type,
        nbin = if (type == "body") nbin else NULL,
        TxDb = txdb
    ) + theme(
        text = element_text(family = "Arial", size = 14)
    )

    heatmap_p_file &lt;- file.path(
        peak_anno_dir,
        gsub(
            "\\.[a-zA-Z0-9]+$",
            paste0(".", by_what, ".", type, ".heatmap.pdf"),
            basename(peak_file)
        )
    )
    ppreview(heatmap_p, file = heatmap_p_file)
}</code></pre>
</section>
<section id="peak-annotation-1" class="level3" data-number="23.3">
<h3 data-number="23.3" class="anchored" data-anchor-id="peak-annotation-1"><span class="header-section-number">23.3</span> Peak annotation</h3>
<pre class="{r}"><code>library(ChIPseeker)
library(YRUtils)
library(AnnotationDbi)
library(ggplot2)
library(ggprism)
library(vroom)

peak_file &lt;- "/data/users/yangrui/mouse/cuttag_v20250108/final_peak/Satb2.without_ctl.pval0.01.idr0.05.narrowPeak.merged.bed"
txdb_file &lt;- "/data/biodatabase/species/mm10/genome/anno/gencode.vM21.primary_assembly.annotation_UCSC_names.TxDb.sqlite"
peak_anno_dir &lt;- "final_peak/peak_anno/narrow_peak"
tss_region &lt;- c(-2000, 2000)
by_what &lt;- "transcript"

peak &lt;- readPeakFile(peak_file)
txdb &lt;- loadDb(txdb_file)
peak_anno &lt;- annotatePeak(
    peak = peak,
    tssRegion = tss_region,
    TxDb = txdb,
    level = by_what
)

peak_anno_file &lt;- file.path(
    peak_anno_dir,
    gsub(
        "\\.[a-zA-Z0-9]+$",
        paste0(".", by_what, ".anno.tsv"),
        basename(peak_file)
    )
)
vroom_write(as.data.frame(peak_anno),
    file = peak_anno_file,
    col_names = TRUE, append = FALSE
)

peak_anno_stat_df &lt;- getAnnoStat(peak_anno)

pie_p &lt;- ggplot(peak_anno_stat_df, aes(x = "", y = Frequency, fill = Feature)) +
    geom_col() +
    coord_polar(theta = "y") +
    theme_void(base_size = 20, base_family = "Arial")

pie_p_file &lt;- file.path(
    peak_anno_dir,
    gsub(
        "\\.[a-zA-Z0-9]+$",
        paste0(".", by_what, ".anno_pie.pdf"),
        basename(peak_file)
    )
)
ppreview(pie_p, file = pie_p_file)

tss_distribution_p &lt;- plotDistToTSS(peak_anno) +
    labs(title = NULL) +
    theme_prism(border = TRUE, base_size = 20, base_family = "Arial") +
    theme(legend.title = element_text())

tss_distribution_p_file &lt;- file.path(
    peak_anno_dir,
    gsub(
        "\\.[a-zA-Z0-9]+$",
        paste0(".", by_what, ".anno_tss_distribution.pdf"),
        basename(peak_file)
    )
)
ppreview(tss_distribution_p, file = tss_distribution_p_file)</code></pre>
</section>
<section id="functional-enrichment-analysis" class="level3" data-number="23.4">
<h3 data-number="23.4" class="anchored" data-anchor-id="functional-enrichment-analysis"><span class="header-section-number">23.4</span> Functional enrichment analysis</h3>
<pre class="{r}"><code>library(vroom)
library(magrittr)
library(clusterProfiler)
library(AnnotationDbi)
library(tidyverse)

peak_anno_file &lt;- "/data/users/yangrui/mouse/cuttag_v20250108/final_peak/peak_anno/narrow_peak/Satb2.without_ctl.pval0.01.idr0.05.narrowPeak.merged.transcript.anno.tsv"
mpt_file &lt;- "/data/biodatabase/species/mm10/genome/anno/gencode.vM21.trna.ercc.phix.gtf.gz.gene_id_name_mapping_table.tsv"
orgdb_file &lt;- "/data/biodatabase/species/mm10/genome/anno/org.Mm.eg.db.sqlite"

mpt_df &lt;- vroom(mpt_file) %&gt;%
    select(gene_id, gene_name) %&gt;%
    distinct()
peak_anno_df &lt;- vroom(peak_anno_file) %&gt;%
    inner_join(mpt_df, by = c("geneId" = "gene_id")) %&gt;%
    select(seqnames, start, end, V4, V5, V6, gene_name, distanceToTSS) %&gt;%
    set_colnames(c("seqname", "start", "end", "name", "score", "strand", "gene_name", "dist_to_tss")) %&gt;%
    distinct()
vroom_write(peak_anno_df,
    file = gsub("\\.tsv$", "\\.with_names.tsv", peak_anno_file),
    col_names = TRUE, append = FALSE
)

orgdb &lt;- loadDb(orgdb_file)
ego &lt;- enrichGO(
    unique(na.omit(peak_anno_df[["gene_name"]])),
    OrgDb = orgdb,
    keyType = "SYMBOL",
    ont = "ALL",
    pvalueCutoff = 0.05,
    qvalueCutoff = 0.05,
    pAdjustMethod = "BH",
    minGSSize = 10,
    maxGSSize = 1000,
    readable = FALSE,
    pool = FALSE
)
ego
nrow(ego@result)
saveRDS(ego, file = gsub("\\.tsv$", "\\.GO_ALL.rds", peak_anno_file))
vroom_write(ego@result,
    file = gsub("\\.tsv$", "\\.GO_ALL.tsv", peak_anno_file),
    col_names = TRUE, append = FALSE
)</code></pre>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Powered by <a href="https://quarto.org">Quarto</a></p>
<div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/edit/main/Blogs/Bioinformatics/posts/Cut&amp;Tag/cuttag_analysis_pipeline/index.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/blob/main/Blogs/Bioinformatics/posts/Cut&amp;Tag/cuttag_analysis_pipeline/index.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>