<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Rui Yang">
<meta name="dcterms.date" content="2025-01-16">

<title>DE-related analysis for bulk proteomics â€“ NeuroBorder</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../../../">
<script src="../../../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../../../site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../../../../site_libs/quarto-html/quarto-syntax-highlighting-dark-b53751a350365c71b6c909e95f209ed1.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../../../site_libs/bootstrap/bootstrap-7656f4217d0b61e7d1e52c7a113f8883.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../../../../site_libs/bootstrap/bootstrap-dark-0e17766785cb233d275babd160f1f995.min.css" rel="prefetch" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../../../../styles.css">
</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../../../index.html">
    <span class="navbar-title">NeuroBorder</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../../../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-blogs" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Blogs</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-blogs">    
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Computer/index.html">
 <span class="dropdown-text">Computer</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Mathematics/index.html">
 <span class="dropdown-text">Mathematics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Bioinformatics/index.html">
 <span class="dropdown-text">Bioinformatics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Galaxy/index.html">
 <span class="dropdown-text">Galaxy</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../../../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-page-right">
      <h1 class="title">DE-related analysis for bulk proteomics</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">bulk proteomics</div>
                <div class="quarto-category">differential expression</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta column-page-right">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Rui Yang </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">January 16, 2025</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">March 30, 2025</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#quality-check" id="toc-quality-check" class="nav-link" data-scroll-target="#quality-check"><span class="header-section-number">2</span> Quality check</a></li>
  <li><a href="#differential-abundance-analysis" id="toc-differential-abundance-analysis" class="nav-link" data-scroll-target="#differential-abundance-analysis"><span class="header-section-number">3</span> Differential abundance analysis</a>
  <ul>
  <li><a href="#dep2-limma-based-method" id="toc-dep2-limma-based-method" class="nav-link" data-scroll-target="#dep2-limma-based-method"><span class="header-section-number">3.1</span> DEP2 (<code>limma</code>-based method)</a></li>
  <li><a href="#only-using-limma-package" id="toc-only-using-limma-package" class="nav-link" data-scroll-target="#only-using-limma-package"><span class="header-section-number">3.2</span> Only using <code>limma</code> package</a></li>
  </ul></li>
  <li><a href="#merge-de-files" id="toc-merge-de-files" class="nav-link" data-scroll-target="#merge-de-files"><span class="header-section-number">4</span> Merge DE files</a></li>
  <li><a href="#retrieve-sub-cellular-location-info-from-uniprot-and-synaptome-database" id="toc-retrieve-sub-cellular-location-info-from-uniprot-and-synaptome-database" class="nav-link" data-scroll-target="#retrieve-sub-cellular-location-info-from-uniprot-and-synaptome-database"><span class="header-section-number">5</span> Retrieve sub-cellular location info from UniProt and Synaptome database</a></li>
  <li><a href="#go-enrichment-analysis" id="toc-go-enrichment-analysis" class="nav-link" data-scroll-target="#go-enrichment-analysis"><span class="header-section-number">6</span> GO enrichment analysis</a></li>
  <li><a href="#plot-expression-heatmap" id="toc-plot-expression-heatmap" class="nav-link" data-scroll-target="#plot-expression-heatmap"><span class="header-section-number">7</span> Plot expression heatmap</a></li>
  <li><a href="#some-examples-about-usages-of-uniprot-and-string-apis" id="toc-some-examples-about-usages-of-uniprot-and-string-apis" class="nav-link" data-scroll-target="#some-examples-about-usages-of-uniprot-and-string-apis"><span class="header-section-number">8</span> Some examples about usages of UniProt and STRING APIs</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/edit/main/Blogs/Bioinformatics/posts/bulk proteomics/de_related_analysis_for_bulk_proteomics/index.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/blob/main/Blogs/Bioinformatics/posts/bulk proteomics/de_related_analysis_for_bulk_proteomics/index.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block column-page-right" id="quarto-document-content">





<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>At present, this pipeline only allows two sample groups contained in your input files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>work_dir <span class="ot">&lt;-</span> <span class="st">"/data/users/yangrui/lab_projs/lym_new"</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(work_dir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="quality-check" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="quality-check"><span class="header-section-number">2</span> Quality check</h2>
<ol type="1">
<li><code>expr_file</code> must include two ID columns:</li>
</ol>
<ul>
<li><p><code>protein_groups</code>: in which each row must be unique;</p></li>
<li><p><code>gene_name</code>: in which each gene name may not be unique.</p></li>
</ul>
<p>In addition, it should also contain a number of sample columns containing protein abundance levels.</p>
<p>Each sample name must be in the form of <code>ID_N</code>, where <code>N</code> is the replicate number.</p>
<ol start="2" type="1">
<li><code>sample_file</code> must include the following three columns:</li>
</ol>
<ul>
<li><p><code>sample</code>: with the same sample names as those in the abundance file above in the form of <code>ID_N</code>;</p></li>
<li><p><code>group</code>: in the form of <code>ID</code>;</p></li>
<li><p><code>replicate</code>: in the form of <code>N</code>.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vroom)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(psych)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(FactoMineR)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggforce)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggprism)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggalign)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(YRUtils)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>expr_file <span class="ot">&lt;-</span> <span class="st">"input/BTN3A2_proteomic_quantification_data.tsv"</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>sample_file <span class="ot">&lt;-</span> <span class="st">"input/sample_sheet.tsv"</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>qc_dir <span class="ot">&lt;-</span> <span class="st">"qc"</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep those rows detected in at least N replicates in at least one group</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>expr_sample_num <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(qc_dir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">FALSE</span>)</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>sample_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(sample_file) <span class="sc">%&gt;%</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">distinct</span>()</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>expr_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(expr_file) <span class="sc">%&gt;%</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">distinct</span>()</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">duplicated</span>(expr_df<span class="sc">$</span>protein_groups))</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (s <span class="cf">in</span> sample_df<span class="sc">$</span>sample) {</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>    expr_df[[s]][<span class="fu">is.na</span>(expr_df[[s]])] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>filter_flag <span class="ot">&lt;-</span> <span class="fu">rep</span>(F, <span class="fu">nrow</span>(expr_df))</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (g <span class="cf">in</span> <span class="fu">unique</span>(sample_df<span class="sc">$</span>group)) {</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>    tmp_df <span class="ot">&lt;-</span> expr_df[, <span class="fu">filter</span>(sample_df, group <span class="sc">==</span> g) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(sample) <span class="sc">%&gt;%</span> <span class="fu">unique</span>(), drop <span class="ot">=</span> F]</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>    filter_flag <span class="ot">&lt;-</span> filter_flag <span class="sc">|</span> (<span class="fu">rowSums</span>(tmp_df <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> expr_sample_num)</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>expr_df <span class="ot">&lt;-</span> expr_df[filter_flag, ]</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>all_vs_none_expr_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>()</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>sample_groups <span class="ot">&lt;-</span> <span class="fu">unique</span>(sample_df<span class="sc">$</span>group)</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (g <span class="cf">in</span> sample_groups) {</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>    tmp_df <span class="ot">&lt;-</span> expr_df[<span class="fu">rowSums</span>(expr_df[, <span class="fu">filter</span>(sample_df, group <span class="sc">==</span> g) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(sample) <span class="sc">%&gt;%</span> <span class="fu">unique</span>(), <span class="at">drop =</span> F] <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">==</span> <span class="dv">0</span>, ] <span class="sc">%&gt;%</span></span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">diff_flag =</span> <span class="fu">paste0</span>(sample_groups[sample_groups <span class="sc">!=</span> g], <span class="st">" Up"</span>))</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a>    all_vs_none_expr_df <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(all_vs_none_expr_df, tmp_df)</span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a>both_expr_df <span class="ot">&lt;-</span> <span class="fu">filter</span>(expr_df, <span class="sc">!</span>(protein_groups <span class="sc">%in%</span> all_vs_none_expr_df<span class="sc">$</span>protein_groups))</span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a><span class="fu">vroom_write</span>(all_vs_none_expr_df,</span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.</span><span class="sc">\\</span><span class="st">w+$"</span>, <span class="st">".all_vs_none.tsv"</span>, expr_file),</span>
<span id="cb2-44"><a href="#cb2-44" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_names =</span> <span class="cn">TRUE</span>, <span class="at">append =</span> <span class="cn">FALSE</span></span>
<span id="cb2-45"><a href="#cb2-45" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-46"><a href="#cb2-46" aria-hidden="true" tabindex="-1"></a><span class="fu">vroom_write</span>(both_expr_df,</span>
<span id="cb2-47"><a href="#cb2-47" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.</span><span class="sc">\\</span><span class="st">w+$"</span>, <span class="st">".both.tsv"</span>, expr_file),</span>
<span id="cb2-48"><a href="#cb2-48" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_names =</span> <span class="cn">TRUE</span>, <span class="at">append =</span> <span class="cn">FALSE</span></span>
<span id="cb2-49"><a href="#cb2-49" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-50"><a href="#cb2-50" aria-hidden="true" tabindex="-1"></a>log_expr_df <span class="ot">&lt;-</span> <span class="fu">log2</span>(expr_df[, <span class="fu">unique</span>(sample_df<span class="sc">$</span>sample)] <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb2-51"><a href="#cb2-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-52"><a href="#cb2-52" aria-hidden="true" tabindex="-1"></a><span class="co"># Correlation</span></span>
<span id="cb2-53"><a href="#cb2-53" aria-hidden="true" tabindex="-1"></a>cor_res <span class="ot">&lt;-</span> <span class="fu">corr.test</span>(log_expr_df, <span class="at">use =</span> <span class="st">"pairwise"</span>, <span class="at">method =</span> <span class="st">"pearson"</span>, <span class="at">adjust =</span> <span class="st">"BH"</span>)</span>
<span id="cb2-54"><a href="#cb2-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-55"><a href="#cb2-55" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggheatmap</span>(</span>
<span id="cb2-56"><a href="#cb2-56" aria-hidden="true" tabindex="-1"></a>    cor_res<span class="sc">$</span>r,</span>
<span id="cb2-57"><a href="#cb2-57" aria-hidden="true" tabindex="-1"></a>    <span class="at">width =</span> <span class="fu">ncol</span>(cor_res<span class="sc">$</span>r) <span class="sc">*</span> <span class="fu">unit</span>(<span class="dv">10</span>, <span class="st">"mm"</span>),</span>
<span id="cb2-58"><a href="#cb2-58" aria-hidden="true" tabindex="-1"></a>    <span class="at">height =</span> <span class="fu">nrow</span>(cor_res<span class="sc">$</span>r) <span class="sc">*</span> <span class="fu">unit</span>(<span class="dv">10</span>, <span class="st">"mm"</span>)</span>
<span id="cb2-59"><a href="#cb2-59" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb2-60"><a href="#cb2-60" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scheme_align</span>(<span class="at">free_spaces =</span> <span class="st">"t"</span>) <span class="sc">+</span></span>
<span id="cb2-61"><a href="#cb2-61" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_gradient2</span>(</span>
<span id="cb2-62"><a href="#cb2-62" aria-hidden="true" tabindex="-1"></a>        <span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"white"</span>, <span class="at">high =</span> <span class="st">"red"</span>,</span>
<span id="cb2-63"><a href="#cb2-63" aria-hidden="true" tabindex="-1"></a>        <span class="at">midpoint =</span> <span class="dv">0</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>),</span>
<span id="cb2-64"><a href="#cb2-64" aria-hidden="true" tabindex="-1"></a>        <span class="at">breaks =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="sc">-</span><span class="fl">0.5</span>, <span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>)</span>
<span id="cb2-65"><a href="#cb2-65" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb2-66"><a href="#cb2-66" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"R"</span>) <span class="sc">+</span></span>
<span id="cb2-67"><a href="#cb2-67" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">x =</span> <span class="fu">guide_axis</span>(<span class="at">angle =</span> <span class="dv">45</span>)) <span class="sc">+</span></span>
<span id="cb2-68"><a href="#cb2-68" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb2-69"><a href="#cb2-69" aria-hidden="true" tabindex="-1"></a>        <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">20</span>, <span class="at">family =</span> <span class="st">"Arial"</span>, <span class="at">color =</span> <span class="st">"black"</span>),</span>
<span id="cb2-70"><a href="#cb2-70" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">20</span>, <span class="at">family =</span> <span class="st">"Arial"</span>, <span class="at">color =</span> <span class="st">"black"</span>)</span>
<span id="cb2-71"><a href="#cb2-71" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb2-72"><a href="#cb2-72" aria-hidden="true" tabindex="-1"></a>    <span class="fu">anno_top</span>() <span class="sc">+</span></span>
<span id="cb2-73"><a href="#cb2-73" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggalign</span>(</span>
<span id="cb2-74"><a href="#cb2-74" aria-hidden="true" tabindex="-1"></a>        <span class="at">data =</span> <span class="fu">gsub</span>(<span class="st">"_</span><span class="sc">\\</span><span class="st">d+$"</span>, <span class="st">""</span>, <span class="fu">colnames</span>(cor_res<span class="sc">$</span>r)),</span>
<span id="cb2-75"><a href="#cb2-75" aria-hidden="true" tabindex="-1"></a>        <span class="at">size =</span> <span class="fu">unit</span>(<span class="dv">4</span>, <span class="st">"mm"</span>)</span>
<span id="cb2-76"><a href="#cb2-76" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb2-77"><a href="#cb2-77" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="dv">1</span>, <span class="at">fill =</span> <span class="fu">factor</span>(value))) <span class="sc">+</span></span>
<span id="cb2-78"><a href="#cb2-78" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> <span class="cn">NULL</span>, <span class="at">name =</span> <span class="cn">NULL</span>, <span class="at">expand =</span> <span class="fu">expansion</span>(<span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb2-79"><a href="#cb2-79" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"Sample"</span>) <span class="sc">+</span></span>
<span id="cb2-80"><a href="#cb2-80" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb2-81"><a href="#cb2-81" aria-hidden="true" tabindex="-1"></a>        <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">20</span>, <span class="at">family =</span> <span class="st">"Arial"</span>, <span class="at">color =</span> <span class="st">"black"</span>),</span>
<span id="cb2-82"><a href="#cb2-82" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">20</span>, <span class="at">family =</span> <span class="st">"Arial"</span>, <span class="at">color =</span> <span class="st">"black"</span>)</span>
<span id="cb2-83"><a href="#cb2-83" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb2-84"><a href="#cb2-84" aria-hidden="true" tabindex="-1"></a>    <span class="fu">with_quad</span>(<span class="fu">scheme_align</span>(<span class="at">guides =</span> <span class="st">"t"</span>), <span class="cn">NULL</span>)</span>
<span id="cb2-85"><a href="#cb2-85" aria-hidden="true" tabindex="-1"></a><span class="fu">ppreview</span>(p, <span class="at">file =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"correlation.pdf"</span>))</span>
<span id="cb2-86"><a href="#cb2-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-87"><a href="#cb2-87" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA</span></span>
<span id="cb2-88"><a href="#cb2-88" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">PCA</span>(<span class="fu">t</span>(log_expr_df), <span class="at">ncp =</span> <span class="dv">10</span>, <span class="at">scale.unit =</span> <span class="cn">TRUE</span>, <span class="at">graph =</span> <span class="cn">FALSE</span>)</span>
<span id="cb2-89"><a href="#cb2-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-90"><a href="#cb2-90" aria-hidden="true" tabindex="-1"></a>pca_coord <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(pca<span class="sc">$</span>ind<span class="sc">$</span>coord)</span>
<span id="cb2-91"><a href="#cb2-91" aria-hidden="true" tabindex="-1"></a>pca_coord<span class="sc">$</span>sample <span class="ot">&lt;-</span> <span class="fu">row.names</span>(pca_coord)</span>
<span id="cb2-92"><a href="#cb2-92" aria-hidden="true" tabindex="-1"></a>pca_coord<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">gsub</span>(<span class="st">"_</span><span class="sc">\\</span><span class="st">d+$"</span>, <span class="st">""</span>, pca_coord<span class="sc">$</span>sample))</span>
<span id="cb2-93"><a href="#cb2-93" aria-hidden="true" tabindex="-1"></a>pca_eig <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(pca<span class="sc">$</span>eig)</span>
<span id="cb2-94"><a href="#cb2-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-95"><a href="#cb2-95" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pca_coord, <span class="fu">aes</span>(Dim<span class="fl">.1</span>, Dim<span class="fl">.2</span>)) <span class="sc">+</span></span>
<span id="cb2-96"><a href="#cb2-96" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> group), <span class="at">size =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb2-97"><a href="#cb2-97" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="fu">paste0</span>(<span class="st">"PC1 ("</span>, <span class="fu">round</span>(pca_eig[<span class="st">"comp 1"</span>, <span class="st">"percentage of variance"</span>]), <span class="st">"%)"</span>)) <span class="sc">+</span></span>
<span id="cb2-98"><a href="#cb2-98" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="fu">paste0</span>(<span class="st">"PC2 ("</span>, <span class="fu">round</span>(pca_eig[<span class="st">"comp 2"</span>, <span class="st">"percentage of variance"</span>]), <span class="st">"%)"</span>)) <span class="sc">+</span></span>
<span id="cb2-99"><a href="#cb2-99" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_mark_ellipse</span>(<span class="fu">aes</span>(<span class="at">fill =</span> group), <span class="at">color =</span> <span class="cn">NA</span>, <span class="at">alpha =</span> <span class="fl">0.25</span>) <span class="sc">+</span></span>
<span id="cb2-100"><a href="#cb2-100" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_prism</span>(<span class="at">base_family =</span> <span class="st">"Arial"</span>, <span class="at">border =</span> <span class="cn">TRUE</span>, <span class="at">base_size =</span> <span class="dv">20</span>)</span>
<span id="cb2-101"><a href="#cb2-101" aria-hidden="true" tabindex="-1"></a><span class="fu">ppreview</span>(p, <span class="at">file =</span> <span class="fu">file.path</span>(qc_dir, <span class="st">"pca.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="differential-abundance-analysis" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="differential-abundance-analysis"><span class="header-section-number">3</span> Differential abundance analysis</h2>
<p>For <strong>spectra count</strong> (MS2), the distribution of which can be approximated by a Poisson distribution. In this case, <code>DESeq2</code>, <code>edgeR</code>, etc. may be used.</p>
<p>For <strong>signal intensity</strong> (MS1), the distribution of which can be approximated by a Normal distribution. In this case, <code>limma</code> may be used.</p>
<section id="dep2-limma-based-method" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="dep2-limma-based-method"><span class="header-section-number">3.1</span> DEP2 (<code>limma</code>-based method)</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DEP2)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vroom)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggridges)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(YRUtils)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>expr_file <span class="ot">&lt;-</span> <span class="st">"input/BTN3A2_proteomic_quantification_data.both.tsv"</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>sample_file <span class="ot">&lt;-</span> <span class="st">"input/sample_sheet.tsv"</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="st">"de"</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(output_dir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>sample_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(sample_file) <span class="sc">%&gt;%</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">distinct</span>()</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>expr_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(expr_file) <span class="sc">%&gt;%</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">distinct</span>()</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="do">### 1. Perform differential expression analysis over genes detected in both treatment and control</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="co"># For duplicated IDs, add tailed numbers, delimiter is "."</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>expr_df <span class="ot">&lt;-</span> <span class="fu">make_unique</span>(expr_df, <span class="at">names =</span> <span class="st">"gene_name"</span>, <span class="at">ids =</span> <span class="st">"protein_groups"</span>, <span class="at">delim =</span> <span class="st">";"</span>)</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Take expression columns</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>expr_cols <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">names</span>(expr_df) <span class="sc">%in%</span> sample_df<span class="sc">$</span>sample)</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a><span class="do">## 1. Construct SummarizedExperiment object</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">make_se_parse</span>(expr_df,</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">columns =</span> expr_cols,</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">mode =</span> <span class="st">"delim"</span>, <span class="at">sep =</span> <span class="st">"_"</span>,</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">log2transform =</span> T</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a><span class="do">## 2. Filtering</span></span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Filtering out reverse, contaminant, and low-quality features with many missing values</span></span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Allow up to N missing values in each condition</span></span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>mv_num <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a><span class="co"># filter_se() also provides filtering options based on filter_formula,</span></span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a><span class="co"># which can be used to filter features based on given columns</span></span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>filter_se <span class="ot">&lt;-</span> <span class="fu">filter_se</span>(se, <span class="at">thr =</span> mv_num)</span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> (<span class="fu">plot_frequency</span>(se) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Protein identifications overlap before filter"</span>)) <span class="sc">/</span></span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>    (<span class="fu">plot_frequency</span>(filter_se) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Protein identifications overlap after filter"</span>))</span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a><span class="fu">ppreview</span>(p, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="st">"filter_freq_hists.pdf"</span>))</span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Even after filtering, a considerable proportion of missing values may still remain in the assay</span></span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_missval</span>(filter_se)</span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a><span class="do">## 3. Normalization</span></span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a><span class="co"># log1p --&gt; variance stabilizing transformation (vst)</span></span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a>norm_se <span class="ot">&lt;-</span> <span class="fu">normalize_vsn</span>(filter_se)</span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_normalization</span>(norm_se)</span>
<span id="cb3-57"><a href="#cb3-57" aria-hidden="true" tabindex="-1"></a><span class="fu">ppreview</span>(p, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="st">"norm_boxplot.pdf"</span>))</span>
<span id="cb3-58"><a href="#cb3-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-59"><a href="#cb3-59" aria-hidden="true" tabindex="-1"></a><span class="co"># Density and CumSum plots of intensities of proteins with and without missing values</span></span>
<span id="cb3-60"><a href="#cb3-60" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_detect</span>(norm_se)</span>
<span id="cb3-61"><a href="#cb3-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-62"><a href="#cb3-62" aria-hidden="true" tabindex="-1"></a><span class="do">## 4. Imputation</span></span>
<span id="cb3-63"><a href="#cb3-63" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. Some may be suitable for missing at random</span></span>
<span id="cb3-64"><a href="#cb3-64" aria-hidden="true" tabindex="-1"></a><span class="co"># In this case, both small and large values may be missing,</span></span>
<span id="cb3-65"><a href="#cb3-65" aria-hidden="true" tabindex="-1"></a><span class="co"># so imputing relatively larger values is reasonable</span></span>
<span id="cb3-66"><a href="#cb3-66" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. Some may be suitable for missing not at random</span></span>
<span id="cb3-67"><a href="#cb3-67" aria-hidden="true" tabindex="-1"></a><span class="co"># In this case, I think a large proportion of missing values are so small that they cannot be detected,</span></span>
<span id="cb3-68"><a href="#cb3-68" aria-hidden="true" tabindex="-1"></a><span class="co"># so using imputation with cautions, it may introduce unwanted bias</span></span>
<span id="cb3-69"><a href="#cb3-69" aria-hidden="true" tabindex="-1"></a><span class="co"># You should have a deep comprehension for the sources of missing values</span></span>
<span id="cb3-70"><a href="#cb3-70" aria-hidden="true" tabindex="-1"></a><span class="co"># c("QRILC", "bpca", "knn", "MLE", "MinDet", "MinProb", "man", "min", "zero", "mixed", "nbavg", "RF", "GSimp")</span></span>
<span id="cb3-71"><a href="#cb3-71" aria-hidden="true" tabindex="-1"></a><span class="co"># It seems that "MLE" is really time-consuming</span></span>
<span id="cb3-72"><a href="#cb3-72" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove method "mixed" because it needs specifying which rows are missing at random (the remaining rows are missing not at random), and specifying two methods for dealing with missing at random and missing not at random respectively</span></span>
<span id="cb3-73"><a href="#cb3-73" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove method "MLE" because it seems that we need to adjust parameters more carefully otherwise the imputation values of NAs are much larger than the normal</span></span>
<span id="cb3-74"><a href="#cb3-74" aria-hidden="true" tabindex="-1"></a>impute_methods <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"QRILC"</span>, <span class="st">"bpca"</span>, <span class="st">"knn"</span>, <span class="st">"MinDet"</span>, <span class="st">"MinProb"</span>, <span class="st">"man"</span>, <span class="st">"min"</span>, <span class="st">"zero"</span>, <span class="st">"nbavg"</span>, <span class="st">"RF"</span>, <span class="st">"GSimp"</span>)</span>
<span id="cb3-75"><a href="#cb3-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-76"><a href="#cb3-76" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(impute_methods) <span class="ot">&lt;-</span> impute_methods</span>
<span id="cb3-77"><a href="#cb3-77" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep only those sample-feature pairs having NAs</span></span>
<span id="cb3-78"><a href="#cb3-78" aria-hidden="true" tabindex="-1"></a>nas_df <span class="ot">&lt;-</span> <span class="fu">assay</span>(norm_se) <span class="sc">%&gt;%</span></span>
<span id="cb3-79"><a href="#cb3-79" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb3-80"><a href="#cb3-80" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">gene_name =</span> <span class="fu">row.names</span>(.)) <span class="sc">%&gt;%</span></span>
<span id="cb3-81"><a href="#cb3-81" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">!</span>gene_name, <span class="at">names_to =</span> <span class="st">"label"</span>, <span class="at">values_to =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-82"><a href="#cb3-82" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">is_na =</span> <span class="fu">is.na</span>(value)) <span class="sc">%&gt;%</span></span>
<span id="cb3-83"><a href="#cb3-83" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(is_na) <span class="sc">%&gt;%</span></span>
<span id="cb3-84"><a href="#cb3-84" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(gene_name, label)</span>
<span id="cb3-85"><a href="#cb3-85" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep only those sample-feature pairs not having NAs</span></span>
<span id="cb3-86"><a href="#cb3-86" aria-hidden="true" tabindex="-1"></a>non_nas <span class="ot">&lt;-</span> nas_df <span class="sc">%&gt;%</span></span>
<span id="cb3-87"><a href="#cb3-87" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">id =</span> <span class="fu">paste0</span>(label, <span class="st">":"</span>, gene_name)) <span class="sc">%&gt;%</span></span>
<span id="cb3-88"><a href="#cb3-88" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(id) <span class="sc">%&gt;%</span></span>
<span id="cb3-89"><a href="#cb3-89" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>()</span>
<span id="cb3-90"><a href="#cb3-90" aria-hidden="true" tabindex="-1"></a><span class="co"># Try each imputation method</span></span>
<span id="cb3-91"><a href="#cb3-91" aria-hidden="true" tabindex="-1"></a>impute_se_ls <span class="ot">&lt;-</span> <span class="fu">lapply</span>(impute_methods, <span class="cf">function</span>(x) {</span>
<span id="cb3-92"><a href="#cb3-92" aria-hidden="true" tabindex="-1"></a>    <span class="fu">message</span>(<span class="fu">paste0</span>(<span class="st">"</span><span class="sc">\n\n</span><span class="st">running "</span>, x, <span class="st">" ..."</span>))</span>
<span id="cb3-93"><a href="#cb3-93" aria-hidden="true" tabindex="-1"></a>    <span class="fu">impute</span>(norm_se, <span class="at">fun =</span> x)</span>
<span id="cb3-94"><a href="#cb3-94" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb3-95"><a href="#cb3-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-96"><a href="#cb3-96" aria-hidden="true" tabindex="-1"></a>imps <span class="ot">&lt;-</span> <span class="fu">tibble</span>()</span>
<span id="cb3-97"><a href="#cb3-97" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (m <span class="cf">in</span> <span class="fu">names</span>(impute_se_ls)) {</span>
<span id="cb3-98"><a href="#cb3-98" aria-hidden="true" tabindex="-1"></a>    tmp_df <span class="ot">&lt;-</span> <span class="fu">assay</span>(impute_se_ls[[m]]) <span class="sc">%&gt;%</span></span>
<span id="cb3-99"><a href="#cb3-99" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb3-100"><a href="#cb3-100" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">gene_name =</span> <span class="fu">row.names</span>(.)) <span class="sc">%&gt;%</span></span>
<span id="cb3-101"><a href="#cb3-101" aria-hidden="true" tabindex="-1"></a>        <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">!</span>gene_name, <span class="at">names_to =</span> <span class="st">"label"</span>, <span class="at">values_to =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-102"><a href="#cb3-102" aria-hidden="true" tabindex="-1"></a>        <span class="fu">left_join</span>(<span class="fu">as.data.frame</span>(<span class="fu">colData</span>(impute_se_ls[[m]])[, <span class="fu">c</span>(<span class="st">"label"</span>, <span class="st">"condition"</span>)]), <span class="at">by =</span> <span class="st">"label"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-103"><a href="#cb3-103" aria-hidden="true" tabindex="-1"></a>        <span class="fu">right_join</span>(nas_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"label"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb3-104"><a href="#cb3-104" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">method =</span> m)</span>
<span id="cb3-105"><a href="#cb3-105" aria-hidden="true" tabindex="-1"></a>    imps <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(imps, tmp_df)</span>
<span id="cb3-106"><a href="#cb3-106" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-107"><a href="#cb3-107" aria-hidden="true" tabindex="-1"></a>non_imps <span class="ot">&lt;-</span> <span class="fu">assay</span>(norm_se) <span class="sc">%&gt;%</span></span>
<span id="cb3-108"><a href="#cb3-108" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb3-109"><a href="#cb3-109" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">gene_name =</span> <span class="fu">row.names</span>(.)) <span class="sc">%&gt;%</span></span>
<span id="cb3-110"><a href="#cb3-110" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">!</span>gene_name, <span class="at">names_to =</span> <span class="st">"label"</span>, <span class="at">values_to =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-111"><a href="#cb3-111" aria-hidden="true" tabindex="-1"></a>    <span class="fu">left_join</span>(<span class="fu">as.data.frame</span>(<span class="fu">colData</span>(norm_se)[, <span class="fu">c</span>(<span class="st">"label"</span>, <span class="st">"condition"</span>)]), <span class="at">by =</span> <span class="st">"label"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-112"><a href="#cb3-112" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">id =</span> <span class="fu">paste0</span>(label, <span class="st">":"</span>, gene_name), <span class="at">method =</span> <span class="st">"non_impute"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-113"><a href="#cb3-113" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span>(id <span class="sc">%in%</span> non_nas)) <span class="sc">%&gt;%</span></span>
<span id="cb3-114"><a href="#cb3-114" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">!</span>id)</span>
<span id="cb3-115"><a href="#cb3-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-116"><a href="#cb3-116" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the distribution of imputation values of NAs under each method</span></span>
<span id="cb3-117"><a href="#cb3-117" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">bind_rows</span>(imps, non_imps), <span class="fu">aes</span>(<span class="at">x =</span> value, <span class="at">y =</span> <span class="fu">factor</span>(method, <span class="at">level =</span> <span class="fu">unique</span>(method)))) <span class="sc">+</span></span>
<span id="cb3-118"><a href="#cb3-118" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density_ridges</span>(</span>
<span id="cb3-119"><a href="#cb3-119" aria-hidden="true" tabindex="-1"></a>        <span class="at">fill =</span> <span class="st">"#027AD450"</span>, <span class="at">scale =</span> <span class="fl">1.2</span>,</span>
<span id="cb3-120"><a href="#cb3-120" aria-hidden="true" tabindex="-1"></a>        <span class="at">jittered_points =</span> <span class="cn">TRUE</span>, <span class="at">position =</span> <span class="fu">position_points_jitter</span>(<span class="at">height =</span> <span class="dv">0</span>),</span>
<span id="cb3-121"><a href="#cb3-121" aria-hidden="true" tabindex="-1"></a>        <span class="at">point_shape =</span> <span class="st">"|"</span>, <span class="at">point_size =</span> <span class="dv">2</span>, <span class="at">point_alpha =</span> <span class="dv">1</span>, <span class="at">alpha =</span> <span class="fl">0.7</span></span>
<span id="cb3-122"><a href="#cb3-122" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb3-123"><a href="#cb3-123" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Impute method"</span>) <span class="sc">+</span></span>
<span id="cb3-124"><a href="#cb3-124" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">"log2 Intensity"</span>) <span class="sc">+</span></span>
<span id="cb3-125"><a href="#cb3-125" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>()</span>
<span id="cb3-126"><a href="#cb3-126" aria-hidden="true" tabindex="-1"></a><span class="fu">ppreview</span>(p, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="st">"impute_density.pdf"</span>))</span>
<span id="cb3-127"><a href="#cb3-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-128"><a href="#cb3-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-129"><a href="#cb3-129" aria-hidden="true" tabindex="-1"></a><span class="do">## 5. Hypothesis testing</span></span>
<span id="cb3-130"><a href="#cb3-130" aria-hidden="true" tabindex="-1"></a><span class="co"># T-test using limma</span></span>
<span id="cb3-131"><a href="#cb3-131" aria-hidden="true" tabindex="-1"></a><span class="co"># Test all the other samples vs. control</span></span>
<span id="cb3-132"><a href="#cb3-132" aria-hidden="true" tabindex="-1"></a>control_sample <span class="ot">&lt;-</span> <span class="st">"Control"</span></span>
<span id="cb3-133"><a href="#cb3-133" aria-hidden="true" tabindex="-1"></a>padj_th <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb3-134"><a href="#cb3-134" aria-hidden="true" tabindex="-1"></a>lfc_th <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb3-135"><a href="#cb3-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-136"><a href="#cb3-136" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="st">"BTN3A2_vs_Control"</span></span>
<span id="cb3-137"><a href="#cb3-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-138"><a href="#cb3-138" aria-hidden="true" tabindex="-1"></a>pair <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(contrast, <span class="st">"_vs_"</span>)[[<span class="dv">1</span>]]</span>
<span id="cb3-139"><a href="#cb3-139" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (m <span class="cf">in</span> <span class="fu">names</span>(impute_se_ls)) {</span>
<span id="cb3-140"><a href="#cb3-140" aria-hidden="true" tabindex="-1"></a>    impute_se <span class="ot">&lt;-</span> impute_se_ls[[m]]</span>
<span id="cb3-141"><a href="#cb3-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-142"><a href="#cb3-142" aria-hidden="true" tabindex="-1"></a>    diff_se <span class="ot">&lt;-</span> <span class="fu">test_diff</span>(impute_se, <span class="at">type =</span> <span class="st">"control"</span>, <span class="at">control =</span> control_sample, <span class="at">fdr.type =</span> <span class="st">"BH"</span>)</span>
<span id="cb3-143"><a href="#cb3-143" aria-hidden="true" tabindex="-1"></a>    sig_se <span class="ot">&lt;-</span> <span class="fu">add_rejections</span>(diff_se, <span class="at">alpha =</span> padj_th, <span class="at">lfc =</span> lfc_th)</span>
<span id="cb3-144"><a href="#cb3-144" aria-hidden="true" tabindex="-1"></a>    <span class="fu">saveRDS</span>(sig_se, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="fu">paste0</span>(contrast, <span class="st">"_"</span>, m, <span class="st">"_de.rds"</span>)))</span>
<span id="cb3-145"><a href="#cb3-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-146"><a href="#cb3-146" aria-hidden="true" tabindex="-1"></a>    clean_degs <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(sig_se<span class="sc">@</span>elementMetadata)</span>
<span id="cb3-147"><a href="#cb3-147" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(clean_degs)[<span class="fu">grep</span>(<span class="st">"_diff$"</span>, <span class="fu">names</span>(clean_degs))] <span class="ot">&lt;-</span> <span class="st">"log2FoldChange"</span></span>
<span id="cb3-148"><a href="#cb3-148" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(clean_degs)[<span class="fu">grep</span>(<span class="st">"_p.adj$"</span>, <span class="fu">names</span>(clean_degs))] <span class="ot">&lt;-</span> <span class="st">"padj"</span></span>
<span id="cb3-149"><a href="#cb3-149" aria-hidden="true" tabindex="-1"></a>    clean_degs <span class="ot">&lt;-</span> clean_degs <span class="sc">%&gt;%</span></span>
<span id="cb3-150"><a href="#cb3-150" aria-hidden="true" tabindex="-1"></a>        <span class="fu">group_by</span>(gene_name) <span class="sc">%&gt;%</span></span>
<span id="cb3-151"><a href="#cb3-151" aria-hidden="true" tabindex="-1"></a>        <span class="fu">slice_min</span>(num_NAs) <span class="sc">%&gt;%</span></span>
<span id="cb3-152"><a href="#cb3-152" aria-hidden="true" tabindex="-1"></a>        <span class="fu">slice_max</span>(sequence_number) <span class="sc">%&gt;%</span></span>
<span id="cb3-153"><a href="#cb3-153" aria-hidden="true" tabindex="-1"></a>        <span class="fu">slice_min</span>(padj) <span class="sc">%&gt;%</span></span>
<span id="cb3-154"><a href="#cb3-154" aria-hidden="true" tabindex="-1"></a>        <span class="fu">slice_max</span>(log2FoldChange) <span class="sc">%&gt;%</span></span>
<span id="cb3-155"><a href="#cb3-155" aria-hidden="true" tabindex="-1"></a>        <span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-156"><a href="#cb3-156" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb3-157"><a href="#cb3-157" aria-hidden="true" tabindex="-1"></a>        <span class="fu">arrange</span>(<span class="fu">desc</span>(log2FoldChange)) <span class="sc">%&gt;%</span></span>
<span id="cb3-158"><a href="#cb3-158" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">diff_flag =</span> <span class="fu">if_else</span>(padj <span class="sc">&lt;</span> padj_th,</span>
<span id="cb3-159"><a href="#cb3-159" aria-hidden="true" tabindex="-1"></a>            <span class="fu">if_else</span>(<span class="fu">abs</span>(log2FoldChange) <span class="sc">&gt;</span> lfc_th,</span>
<span id="cb3-160"><a href="#cb3-160" aria-hidden="true" tabindex="-1"></a>                <span class="fu">if_else</span>(log2FoldChange <span class="sc">&gt;</span> lfc_th,</span>
<span id="cb3-161"><a href="#cb3-161" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">paste0</span>(pair[<span class="dv">1</span>], <span class="st">" Up"</span>),</span>
<span id="cb3-162"><a href="#cb3-162" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">paste0</span>(pair[<span class="dv">2</span>], <span class="st">" Up"</span>)</span>
<span id="cb3-163"><a href="#cb3-163" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="cb3-164"><a href="#cb3-164" aria-hidden="true" tabindex="-1"></a>                <span class="st">"NO"</span></span>
<span id="cb3-165"><a href="#cb3-165" aria-hidden="true" tabindex="-1"></a>            ),</span>
<span id="cb3-166"><a href="#cb3-166" aria-hidden="true" tabindex="-1"></a>            <span class="st">"NO"</span></span>
<span id="cb3-167"><a href="#cb3-167" aria-hidden="true" tabindex="-1"></a>        ))</span>
<span id="cb3-168"><a href="#cb3-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-169"><a href="#cb3-169" aria-hidden="true" tabindex="-1"></a>    <span class="fu">vroom_write</span>(clean_degs, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="fu">paste0</span>(contrast, <span class="st">"_"</span>, m, <span class="st">"_de.tsv"</span>)))</span>
<span id="cb3-170"><a href="#cb3-170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-171"><a href="#cb3-171" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> <span class="fu">plot_volcano</span>(sig_se, <span class="at">contrast =</span> contrast, <span class="at">adjusted =</span> T, <span class="at">add_threshold_line =</span> <span class="st">"intersect"</span>, <span class="at">pCutoff =</span> padj_th, <span class="at">fcCutoff =</span> lfc_th, <span class="at">add_names =</span> F)</span>
<span id="cb3-172"><a href="#cb3-172" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ppreview</span>(p, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="fu">paste0</span>(contrast, <span class="st">"_"</span>, m, <span class="st">"_volcano.pdf"</span>)))</span>
<span id="cb3-173"><a href="#cb3-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-174"><a href="#cb3-174" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> <span class="fu">plot_heatmap</span>(sig_se, <span class="at">manual_contrast =</span> contrast, <span class="at">kmeans =</span> T, <span class="at">k =</span> <span class="dv">2</span>, <span class="at">split_order =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>), <span class="at">show_row_names =</span> F, <span class="at">show_row_dend =</span> F, <span class="at">heatmap_width =</span> <span class="fu">unit</span>(<span class="dv">6</span>, <span class="st">"cm"</span>), <span class="at">heatmap_height =</span> <span class="fu">unit</span>(<span class="fl">0.04</span>, <span class="st">"mm"</span>) <span class="sc">*</span> <span class="fu">nrow</span>(sig_se<span class="sc">@</span>elementMetadata))</span>
<span id="cb3-175"><a href="#cb3-175" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ppreview</span>(p, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="fu">paste0</span>(contrast, <span class="st">"_"</span>, m, <span class="st">"_heatmap.pdf"</span>)))</span>
<span id="cb3-176"><a href="#cb3-176" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="only-using-limma-package" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="only-using-limma-package"><span class="header-section-number">3.2</span> Only using <code>limma</code> package</h3>
<p>For the following two methods, there must be two columns with names <code>protein_name</code> (which must be unique) and <code>gene_name</code> in the expression table file, in which all sample column names must be in the form <code>ID_repN</code>, which will be matched using the regular expression <code>[a-zA-Z0-9]+_rep[0-9]+</code>.</p>
<ol type="1">
<li>Compare two samples at a time</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vroom)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(limma)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>expr_file <span class="ot">&lt;-</span> <span class="st">"expression_sheet.tsv"</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co"># keep those proteins detected in at least N replicates in at least one sample</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>min_expr_num <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>padj_th <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>logfc_th <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="co"># the first is the control sample</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>group_levels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"WT"</span>, <span class="st">"S541N"</span>)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>expr_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(expr_file) <span class="sc">%&gt;%</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set_rownames</span>(.[[<span class="st">"protein_name"</span>]])</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">names</span>(expr_df)[<span class="fu">str_detect</span>(<span class="fu">names</span>(expr_df), <span class="st">"[a-zA-Z0-9]+_rep[0-9]+"</span>)]</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(samples) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_rep[0-9]+$"</span>, <span class="st">""</span>, samples)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>expr_mat <span class="ot">&lt;-</span> expr_df[, samples]</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>expr_mat[<span class="fu">is.na</span>(expr_mat)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>flag <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">FALSE</span>, <span class="fu">nrow</span>(expr_mat))</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (group <span class="cf">in</span> <span class="fu">unique</span>(<span class="fu">names</span>(samples))) {</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>    flag <span class="ot">&lt;-</span> flag <span class="sc">|</span> (<span class="fu">rowSums</span>(expr_mat[, samples[<span class="fu">names</span>(samples) <span class="sc">==</span> group]] <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> min_expr_num)</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>expr_mat <span class="ot">&lt;-</span> expr_mat[flag, ]</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>expr_mat_long_df <span class="ot">&lt;-</span> expr_mat <span class="sc">%&gt;%</span></span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">protein_name =</span> <span class="fu">row.names</span>(.)) <span class="sc">%&gt;%</span></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>protein_name, <span class="at">names_to =</span> <span class="st">"sample"</span>, <span class="at">values_to =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">group =</span> <span class="fu">gsub</span>(<span class="st">"_rep[0-9]+$"</span>, <span class="st">""</span>, sample))</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a>mean_df <span class="ot">&lt;-</span> expr_mat_long_df <span class="sc">%&gt;%</span></span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(value <span class="sc">!=</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(protein_name, group) <span class="sc">%&gt;%</span></span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a>    <span class="fu">reframe</span>(<span class="at">mean =</span> <span class="fu">mean</span>(value))</span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a>expr_mat_filled_df <span class="ot">&lt;-</span> <span class="fu">left_join</span>(expr_mat_long_df, mean_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"protein_name"</span>, <span class="st">"group"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">value =</span> <span class="fu">if_else</span>(<span class="sc">!</span><span class="fu">is.na</span>(mean) <span class="sc">&amp;</span> (value <span class="sc">==</span> <span class="dv">0</span>), mean, value))</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a>expr_mat <span class="ot">&lt;-</span> expr_mat_filled_df <span class="sc">%&gt;%</span></span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span><span class="fu">all_of</span>(<span class="fu">c</span>(<span class="st">"group"</span>, <span class="st">"mean"</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> <span class="st">"sample"</span>, <span class="at">values_from =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set_rownames</span>(.[[<span class="st">"protein_name"</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span>protein_name)</span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a>expr_mat <span class="ot">&lt;-</span> expr_mat[, samples]</span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>expr_df <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(expr_df, expr_mat <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">protein_name =</span> <span class="fu">row.names</span>(.)), <span class="at">suffix =</span> <span class="fu">c</span>(<span class="st">".raw"</span>, <span class="st">".mean_filled"</span>), <span class="at">by =</span> <span class="st">"protein_name"</span>)</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a>expr_mat <span class="ot">&lt;-</span> <span class="fu">log2</span>(expr_mat <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a>group_list <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">names</span>(samples), <span class="at">levels =</span> group_levels)</span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> group_list)</span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(design) <span class="ot">&lt;-</span> <span class="fu">levels</span>(group_list)</span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(design) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(expr_mat)</span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a>contrasts <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">combn</span>(group_levels, <span class="dv">2</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">contrast =</span> <span class="fu">paste0</span>(V2, <span class="st">"-"</span>, V1)) <span class="sc">%&gt;%</span></span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(contrast)</span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(expr_mat, design)</span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a>contrast_matrix <span class="ot">&lt;-</span> <span class="fu">makeContrasts</span>(<span class="at">contrasts =</span> contrasts, <span class="at">levels =</span> design)</span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a>fit2 <span class="ot">&lt;-</span> <span class="fu">contrasts.fit</span>(fit, contrast_matrix)</span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a>fit2 <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(fit2, <span class="at">trend =</span> <span class="cn">TRUE</span>, <span class="at">robust =</span> <span class="cn">TRUE</span>)</span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a><span class="co"># check valid coefficients</span></span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(fit2<span class="sc">$</span>coefficients)</span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a>coef <span class="ot">&lt;-</span> <span class="st">"S541N-WT"</span></span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">topTable</span>(fit2, <span class="at">coef =</span> coef, <span class="at">number =</span> <span class="cn">Inf</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-76"><a href="#cb4-76" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">protein_name =</span> <span class="fu">row.names</span>(.))</span>
<span id="cb4-77"><a href="#cb4-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-78"><a href="#cb4-78" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(expr_df, res, <span class="at">by =</span> <span class="st">"protein_name"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-79"><a href="#cb4-79" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb4-80"><a href="#cb4-80" aria-hidden="true" tabindex="-1"></a>        <span class="at">P.Value =</span> <span class="fu">if_else</span>(P.Value <span class="sc">==</span> <span class="dv">0</span>, .Machine<span class="sc">$</span>double.xmin, P.Value),</span>
<span id="cb4-81"><a href="#cb4-81" aria-hidden="true" tabindex="-1"></a>        <span class="at">adj.P.Val =</span> <span class="fu">if_else</span>(adj.P.Val <span class="sc">==</span> <span class="dv">0</span>, .Machine<span class="sc">$</span>double.xmin, adj.P.Val),</span>
<span id="cb4-82"><a href="#cb4-82" aria-hidden="true" tabindex="-1"></a>        <span class="at">diff_flag =</span> <span class="fu">if_else</span>(adj.P.Val <span class="sc">&lt;</span> <span class="fl">0.05</span>,</span>
<span id="cb4-83"><a href="#cb4-83" aria-hidden="true" tabindex="-1"></a>            <span class="fu">if_else</span>(<span class="fu">abs</span>(logFC) <span class="sc">&gt;</span> logfc_th,</span>
<span id="cb4-84"><a href="#cb4-84" aria-hidden="true" tabindex="-1"></a>                <span class="fu">if_else</span>(logFC <span class="sc">&gt;</span> logfc_th, <span class="fu">paste0</span>(<span class="fu">strsplit</span>(coef, <span class="st">"-"</span>)[[<span class="dv">1</span>]][<span class="dv">1</span>], <span class="st">" Up"</span>), <span class="fu">paste0</span>(<span class="fu">strsplit</span>(coef, <span class="st">"-"</span>)[[<span class="dv">1</span>]][<span class="dv">2</span>], <span class="st">" Up"</span>)),</span>
<span id="cb4-85"><a href="#cb4-85" aria-hidden="true" tabindex="-1"></a>                <span class="st">"NO"</span></span>
<span id="cb4-86"><a href="#cb4-86" aria-hidden="true" tabindex="-1"></a>            ),</span>
<span id="cb4-87"><a href="#cb4-87" aria-hidden="true" tabindex="-1"></a>            <span class="st">"NO"</span></span>
<span id="cb4-88"><a href="#cb4-88" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb4-89"><a href="#cb4-89" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb4-90"><a href="#cb4-90" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(diff_flag)</span>
<span id="cb4-91"><a href="#cb4-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-92"><a href="#cb4-92" aria-hidden="true" tabindex="-1"></a><span class="fu">vroom_write</span>(res, <span class="at">file =</span> <span class="fu">paste0</span>(coef, <span class="st">".tsv"</span>), <span class="at">col_names =</span> <span class="cn">TRUE</span>, <span class="at">append =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Compare all possible pairs at a time</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vroom)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(limma)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gtools)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>work_dir <span class="ot">&lt;-</span> <span class="st">"test"</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>expr_file <span class="ot">&lt;-</span> <span class="st">"xbf_expression_sheet.tsv"</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="co"># keep those proteins detected in at least N replicates in at least one sample</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>min_expr_num <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>padj_th <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>logfc_th <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(work_dir)</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>expr_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(expr_file) <span class="sc">%&gt;%</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set_rownames</span>(.[[<span class="st">"protein_name"</span>]])</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">names</span>(expr_df)[<span class="fu">str_detect</span>(<span class="fu">names</span>(expr_df), <span class="st">"[a-zA-Z0-9]+_rep[0-9]+"</span>)]</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(samples) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_rep[0-9]+$"</span>, <span class="st">""</span>, samples)</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>sample_pair_combns <span class="ot">&lt;-</span> <span class="fu">permutations</span>(<span class="fu">length</span>(<span class="fu">unique</span>(<span class="fu">names</span>(samples))), <span class="dv">2</span>, <span class="at">v =</span> <span class="fu">unique</span>(<span class="fu">names</span>(samples)))</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_len</span>(<span class="fu">nrow</span>(sample_pair_combns))) {</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    group_levels <span class="ot">&lt;-</span> sample_pair_combns[i, ]</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>    pair_samples <span class="ot">&lt;-</span> samples[<span class="fu">names</span>(samples) <span class="sc">%in%</span> group_levels]</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>    expr_mat <span class="ot">&lt;-</span> expr_df[, pair_samples]</span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>    expr_mat[<span class="fu">is.na</span>(expr_mat)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>    flag <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">FALSE</span>, <span class="fu">nrow</span>(expr_mat))</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (group <span class="cf">in</span> group_levels) {</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>        flag <span class="ot">&lt;-</span> flag <span class="sc">|</span> (<span class="fu">rowSums</span>(expr_mat[, samples[<span class="fu">names</span>(samples) <span class="sc">==</span> group]] <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> min_expr_num)</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>    expr_mat <span class="ot">&lt;-</span> expr_mat[flag, ]</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>    expr_mat_long_df <span class="ot">&lt;-</span> expr_mat <span class="sc">%&gt;%</span></span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">protein_name =</span> <span class="fu">row.names</span>(.)) <span class="sc">%&gt;%</span></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a>        <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>protein_name, <span class="at">names_to =</span> <span class="st">"sample"</span>, <span class="at">values_to =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">group =</span> <span class="fu">gsub</span>(<span class="st">"_rep[0-9]+$"</span>, <span class="st">""</span>, sample))</span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a>    mean_df <span class="ot">&lt;-</span> expr_mat_long_df <span class="sc">%&gt;%</span></span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a>        <span class="fu">filter</span>(value <span class="sc">!=</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a>        <span class="fu">group_by</span>(protein_name, group) <span class="sc">%&gt;%</span></span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a>        <span class="fu">reframe</span>(<span class="at">mean =</span> <span class="fu">mean</span>(value))</span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a>    expr_mat_filled_df <span class="ot">&lt;-</span> <span class="fu">left_join</span>(expr_mat_long_df, mean_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"protein_name"</span>, <span class="st">"group"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">value =</span> <span class="fu">if_else</span>(<span class="sc">!</span><span class="fu">is.na</span>(mean) <span class="sc">&amp;</span> (value <span class="sc">==</span> <span class="dv">0</span>), mean, value))</span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a>    expr_mat <span class="ot">&lt;-</span> expr_mat_filled_df <span class="sc">%&gt;%</span></span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a>        <span class="fu">select</span>(<span class="sc">-</span><span class="fu">all_of</span>(<span class="fu">c</span>(<span class="st">"group"</span>, <span class="st">"mean"</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a>        <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> <span class="st">"sample"</span>, <span class="at">values_from =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a>        <span class="fu">set_rownames</span>(.[[<span class="st">"protein_name"</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a>        <span class="fu">select</span>(<span class="sc">-</span>protein_name)</span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a>    expr_mat <span class="ot">&lt;-</span> expr_mat[, pair_samples]</span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a>    expr_mat <span class="ot">&lt;-</span> <span class="fu">log2</span>(expr_mat <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb5-59"><a href="#cb5-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-60"><a href="#cb5-60" aria-hidden="true" tabindex="-1"></a>    group_list <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">names</span>(pair_samples), <span class="at">levels =</span> group_levels)</span>
<span id="cb5-61"><a href="#cb5-61" aria-hidden="true" tabindex="-1"></a>    design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> <span class="dv">0</span> <span class="sc">+</span> group_list)</span>
<span id="cb5-62"><a href="#cb5-62" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(design) <span class="ot">&lt;-</span> <span class="fu">levels</span>(group_list)</span>
<span id="cb5-63"><a href="#cb5-63" aria-hidden="true" tabindex="-1"></a>    <span class="fu">row.names</span>(design) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(expr_mat)</span>
<span id="cb5-64"><a href="#cb5-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-65"><a href="#cb5-65" aria-hidden="true" tabindex="-1"></a>    contrasts <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">combn</span>(group_levels, <span class="dv">2</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb5-66"><a href="#cb5-66" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb5-67"><a href="#cb5-67" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">contrast =</span> <span class="fu">paste0</span>(V2, <span class="st">"-"</span>, V1)) <span class="sc">%&gt;%</span></span>
<span id="cb5-68"><a href="#cb5-68" aria-hidden="true" tabindex="-1"></a>        <span class="fu">pull</span>(contrast)</span>
<span id="cb5-69"><a href="#cb5-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-70"><a href="#cb5-70" aria-hidden="true" tabindex="-1"></a>    fit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(expr_mat, design)</span>
<span id="cb5-71"><a href="#cb5-71" aria-hidden="true" tabindex="-1"></a>    contrast_matrix <span class="ot">&lt;-</span> <span class="fu">makeContrasts</span>(<span class="at">contrasts =</span> contrasts, <span class="at">levels =</span> design)</span>
<span id="cb5-72"><a href="#cb5-72" aria-hidden="true" tabindex="-1"></a>    fit2 <span class="ot">&lt;-</span> <span class="fu">contrasts.fit</span>(fit, contrast_matrix)</span>
<span id="cb5-73"><a href="#cb5-73" aria-hidden="true" tabindex="-1"></a>    fit2 <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(fit2, <span class="at">trend =</span> <span class="cn">TRUE</span>, <span class="at">robust =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-74"><a href="#cb5-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-75"><a href="#cb5-75" aria-hidden="true" tabindex="-1"></a>    coef <span class="ot">&lt;-</span> <span class="fu">colnames</span>(fit2<span class="sc">$</span>coefficients)</span>
<span id="cb5-76"><a href="#cb5-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-77"><a href="#cb5-77" aria-hidden="true" tabindex="-1"></a>    res <span class="ot">&lt;-</span> <span class="fu">topTable</span>(fit2, <span class="at">coef =</span> coef, <span class="at">number =</span> <span class="cn">Inf</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-78"><a href="#cb5-78" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">protein_name =</span> <span class="fu">row.names</span>(.))</span>
<span id="cb5-79"><a href="#cb5-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-80"><a href="#cb5-80" aria-hidden="true" tabindex="-1"></a>    res <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(expr_df[, <span class="fu">c</span>(<span class="st">"protein_name"</span>, <span class="st">"gene_name"</span>)], res, <span class="at">by =</span> <span class="st">"protein_name"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-81"><a href="#cb5-81" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(</span>
<span id="cb5-82"><a href="#cb5-82" aria-hidden="true" tabindex="-1"></a>            <span class="at">P.Value =</span> <span class="fu">if_else</span>(P.Value <span class="sc">==</span> <span class="dv">0</span>, .Machine<span class="sc">$</span>double.xmin, P.Value),</span>
<span id="cb5-83"><a href="#cb5-83" aria-hidden="true" tabindex="-1"></a>            <span class="at">adj.P.Val =</span> <span class="fu">if_else</span>(adj.P.Val <span class="sc">==</span> <span class="dv">0</span>, .Machine<span class="sc">$</span>double.xmin, adj.P.Val),</span>
<span id="cb5-84"><a href="#cb5-84" aria-hidden="true" tabindex="-1"></a>            <span class="at">diff_flag =</span> <span class="fu">if_else</span>(adj.P.Val <span class="sc">&lt;</span> <span class="fl">0.05</span>,</span>
<span id="cb5-85"><a href="#cb5-85" aria-hidden="true" tabindex="-1"></a>                <span class="fu">if_else</span>(<span class="fu">abs</span>(logFC) <span class="sc">&gt;</span> logfc_th,</span>
<span id="cb5-86"><a href="#cb5-86" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">if_else</span>(logFC <span class="sc">&gt;</span> logfc_th, <span class="fu">paste0</span>(<span class="fu">strsplit</span>(coef, <span class="st">"-"</span>)[[<span class="dv">1</span>]][<span class="dv">1</span>], <span class="st">" Up"</span>), <span class="fu">paste0</span>(<span class="fu">strsplit</span>(coef, <span class="st">"-"</span>)[[<span class="dv">1</span>]][<span class="dv">2</span>], <span class="st">" Up"</span>)),</span>
<span id="cb5-87"><a href="#cb5-87" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"NO"</span></span>
<span id="cb5-88"><a href="#cb5-88" aria-hidden="true" tabindex="-1"></a>                ),</span>
<span id="cb5-89"><a href="#cb5-89" aria-hidden="true" tabindex="-1"></a>                <span class="st">"NO"</span></span>
<span id="cb5-90"><a href="#cb5-90" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb5-91"><a href="#cb5-91" aria-hidden="true" tabindex="-1"></a>        ) <span class="sc">%&gt;%</span></span>
<span id="cb5-92"><a href="#cb5-92" aria-hidden="true" tabindex="-1"></a>        <span class="fu">arrange</span>(diff_flag)</span>
<span id="cb5-93"><a href="#cb5-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-94"><a href="#cb5-94" aria-hidden="true" tabindex="-1"></a>    <span class="fu">vroom_write</span>(res, <span class="at">file =</span> <span class="fu">paste0</span>(<span class="fu">gsub</span>(<span class="st">"-"</span>, <span class="st">"_vs_"</span>, coef), <span class="st">".tsv"</span>), <span class="at">col_names =</span> <span class="cn">TRUE</span>, <span class="at">append =</span> <span class="cn">FALSE</span>)</span>
<span id="cb5-95"><a href="#cb5-95" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="merge-de-files" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="merge-de-files"><span class="header-section-number">4</span> Merge DE files</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vroom)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"input/BTN3A2_proteomic_quantification_data.all_vs_none.tsv"</span>,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"de/BTN3A2_vs_Control_bpca_de.tsv"</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>output_file <span class="ot">&lt;-</span> <span class="st">"de/BTN3A2_vs_Control_bpca_de.all_vs_none.tsv"</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">tibble</span>()</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (file <span class="cf">in</span> files) {</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>        df,</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>        <span class="fu">vroom</span>(file) <span class="sc">%&gt;%</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>            <span class="fu">select</span>(protein_name, gene_name, diff_flag) <span class="sc">%&gt;%</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>            <span class="fu">filter</span>(diff_flag <span class="sc">!=</span> <span class="st">"NO"</span>)</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">distinct</span>(df) <span class="sc">%&gt;%</span></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(diff_flag)</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a><span class="fu">vroom_write</span>(df, <span class="at">file =</span> output_file, <span class="at">col_names =</span> <span class="cn">TRUE</span>, <span class="at">append =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="retrieve-sub-cellular-location-info-from-uniprot-and-synaptome-database" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="retrieve-sub-cellular-location-info-from-uniprot-and-synaptome-database"><span class="header-section-number">5</span> Retrieve sub-cellular location info from UniProt and Synaptome database</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vroom)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glue)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rvest)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(synaptome.db)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>organism_id <span class="ot">&lt;-</span> <span class="dv">10116</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>url_template <span class="ot">&lt;-</span> <span class="st">"https://rest.uniprot.org/uniprotkb/search?query=(accession:{accession})%20AND%20(active:true)%20AND%20(organism_id:{organism_id})&amp;fields=accession,gene_primary,organism_name,organism_id,protein_name,annotation_score,protein_existence,reviewed,ft_intramem,cc_subcellular_location,ft_topo_dom,ft_transmem,go,go_p,go_c,go_f&amp;format=tsv"</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>de_file <span class="ot">&lt;-</span> <span class="st">"de/BTN3A2_vs_Control_bpca_de.all_vs_none.tsv"</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="st">"subcellular_location"</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>keep_cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"protein_name"</span>, <span class="st">"gene_name"</span>, <span class="st">"diff_flag"</span>)</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>keep_de_types <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"BTN3A2 Up"</span>, <span class="st">"Control Up"</span>)</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(output_dir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">FALSE</span>)</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(de_file) <span class="sc">%&gt;%</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="fu">all_of</span>(keep_cols)) <span class="sc">%&gt;%</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(diff_flag <span class="sc">%in%</span> keep_de_types)</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>()</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (accession <span class="cf">in</span> <span class="fu">unique</span>(df<span class="sc">$</span>protein_name)) {</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>    url_instance <span class="ot">&lt;-</span> <span class="fu">glue</span>(url_template)</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>    e <span class="ot">&lt;-</span> <span class="fu">try</span>(item_raw_text <span class="ot">&lt;-</span> <span class="fu">read_html</span>(url_instance) <span class="sc">%&gt;%</span> <span class="fu">html_text</span>(), <span class="at">silent =</span> T)</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="st">"try-error"</span> <span class="sc">%in%</span> <span class="fu">class</span>(e)) {</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>        <span class="fu">message</span>(<span class="st">"invalid accession: "</span>, accession)</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>        item_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(<span class="fu">I</span>(item_raw_text), <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>        res_df <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(res_df, item_df)</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> <span class="fu">distinct</span>(res_df)</span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a>res_df[[<span class="st">"scl_secret"</span>]] <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"secret"</span>, res_df[[<span class="st">"Subcellular location [CC]"</span>]], <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a>res_df[[<span class="st">"scl_membrane"</span>]] <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"membrane"</span>, res_df[[<span class="st">"Subcellular location [CC]"</span>]], <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a>res_df[[<span class="st">"scl_secret_membrane"</span>]] <span class="ot">&lt;-</span> res_df[[<span class="st">"scl_secret"</span>]] <span class="sc">|</span> res_df[[<span class="st">"scl_membrane"</span>]]</span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a>res_df[[<span class="st">"uniprot_scl_synapse"</span>]] <span class="ot">&lt;-</span> <span class="fu">str_extract_all</span>(res_df[[<span class="st">"Subcellular location [CC]"</span>]], <span class="fu">regex</span>(<span class="st">"[</span><span class="sc">\\</span><span class="st">w-]*synap[</span><span class="sc">\\</span><span class="st">w-]*"</span>, <span class="at">ignore_case =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sapply</span>(<span class="cf">function</span>(x) {</span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a>        <span class="fu">paste0</span>(<span class="fu">unique</span>(<span class="fu">na.omit</span>(x)), <span class="at">collapse =</span> <span class="st">";"</span>)</span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>res_df[[<span class="st">"uniprot_go_synapse"</span>]] <span class="ot">&lt;-</span> <span class="fu">str_extract_all</span>(res_df[[<span class="st">"Gene Ontology (GO)"</span>]], <span class="fu">regex</span>(<span class="st">"[</span><span class="sc">\\</span><span class="st">w-]*synap[</span><span class="sc">\\</span><span class="st">w-]*"</span>, <span class="at">ignore_case =</span> T)) <span class="sc">%&gt;%</span></span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sapply</span>(<span class="cf">function</span>(x) {</span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a>        <span class="fu">paste0</span>(<span class="fu">unique</span>(<span class="fu">na.omit</span>(x)), <span class="at">collapse =</span> <span class="st">";"</span>)</span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> <span class="fu">left_join</span>(res_df,</span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a>    <span class="fu">findGeneByCompartmentPaperCnt</span>(<span class="at">cnt =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>        <span class="fu">filter</span>(RatName <span class="sc">%in%</span> res_df[[<span class="st">"Gene Names (primary)"</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a>        <span class="fu">select</span>(RatName, Localisation) <span class="sc">%&gt;%</span></span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a>        <span class="fu">distinct</span>() <span class="sc">%&gt;%</span></span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true" tabindex="-1"></a>        <span class="fu">group_by</span>(RatName) <span class="sc">%&gt;%</span></span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true" tabindex="-1"></a>        <span class="fu">reframe</span>(<span class="at">synaptome_db_synapse =</span> <span class="fu">paste0</span>(<span class="fu">unique</span>(Localisation), <span class="at">collapse =</span> <span class="st">";"</span>)),</span>
<span id="cb7-52"><a href="#cb7-52" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"Gene Names (primary)"</span> <span class="ot">=</span> <span class="st">"RatName"</span>)</span>
<span id="cb7-53"><a href="#cb7-53" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-54"><a href="#cb7-54" aria-hidden="true" tabindex="-1"></a>res_df[[<span class="st">"final_synapse"</span>]] <span class="ot">&lt;-</span> <span class="fu">apply</span>(res_df, <span class="dv">1</span>, <span class="cf">function</span>(x) {</span>
<span id="cb7-55"><a href="#cb7-55" aria-hidden="true" tabindex="-1"></a>    vec <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(<span class="fu">c</span>(x[<span class="st">"uniprot_scl_synapse"</span>], x[<span class="st">"uniprot_go_synapse"</span>], x[<span class="st">"synaptome_db_synapse"</span>]))</span>
<span id="cb7-56"><a href="#cb7-56" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(vec[vec <span class="sc">!=</span> <span class="st">""</span>], <span class="at">collapse =</span> <span class="st">";"</span>)</span>
<span id="cb7-57"><a href="#cb7-57" aria-hidden="true" tabindex="-1"></a>}, <span class="at">simplify =</span> T)</span>
<span id="cb7-58"><a href="#cb7-58" aria-hidden="true" tabindex="-1"></a>res_df[[<span class="st">"scl_secret_membrane_final_synapse"</span>]] <span class="ot">&lt;-</span> res_df[[<span class="st">"scl_secret_membrane"</span>]] <span class="sc">&amp;</span> (res_df[[<span class="st">"final_synapse"</span>]] <span class="sc">!=</span> <span class="st">""</span>)</span>
<span id="cb7-59"><a href="#cb7-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-60"><a href="#cb7-60" aria-hidden="true" tabindex="-1"></a>target_res_df <span class="ot">&lt;-</span> <span class="fu">filter</span>(res_df, scl_secret_membrane_final_synapse)</span>
<span id="cb7-61"><a href="#cb7-61" aria-hidden="true" tabindex="-1"></a>target_res_df[[<span class="st">"final_synapse_label"</span>]] <span class="ot">&lt;-</span> <span class="fu">sapply</span>(target_res_df[[<span class="st">"final_synapse"</span>]], <span class="cf">function</span>(x) {</span>
<span id="cb7-62"><a href="#cb7-62" aria-hidden="true" tabindex="-1"></a>    vec <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb7-63"><a href="#cb7-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">grepl</span>(<span class="st">"postsynap|post-synap"</span>, x, <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb7-64"><a href="#cb7-64" aria-hidden="true" tabindex="-1"></a>        vec <span class="ot">&lt;-</span> <span class="fu">c</span>(vec, <span class="st">"Postsynapse"</span>)</span>
<span id="cb7-65"><a href="#cb7-65" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb7-66"><a href="#cb7-66" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">grepl</span>(<span class="st">"presynap|pre-synap"</span>, x, <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb7-67"><a href="#cb7-67" aria-hidden="true" tabindex="-1"></a>        vec <span class="ot">&lt;-</span> <span class="fu">c</span>(vec, <span class="st">"Presynapse"</span>)</span>
<span id="cb7-68"><a href="#cb7-68" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb7-69"><a href="#cb7-69" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">grepl</span>(<span class="st">"synap"</span>, x, <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb7-70"><a href="#cb7-70" aria-hidden="true" tabindex="-1"></a>        vec <span class="ot">&lt;-</span> <span class="fu">c</span>(vec, <span class="st">"Synapse"</span>)</span>
<span id="cb7-71"><a href="#cb7-71" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb7-72"><a href="#cb7-72" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">grepl</span>(<span class="st">"Synaptic_Vesicle"</span>, x, <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb7-73"><a href="#cb7-73" aria-hidden="true" tabindex="-1"></a>        vec <span class="ot">&lt;-</span> <span class="fu">c</span>(vec, <span class="st">"Synaptic vesicle"</span>)</span>
<span id="cb7-74"><a href="#cb7-74" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb7-75"><a href="#cb7-75" aria-hidden="true" tabindex="-1"></a>    vec <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(vec)</span>
<span id="cb7-76"><a href="#cb7-76" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(vec) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb7-77"><a href="#cb7-77" aria-hidden="true" tabindex="-1"></a>        <span class="fu">stop</span>(<span class="st">"no matching"</span>)</span>
<span id="cb7-78"><a href="#cb7-78" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb7-79"><a href="#cb7-79" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(vec) <span class="sc">&gt;</span> <span class="dv">1</span>) {</span>
<span id="cb7-80"><a href="#cb7-80" aria-hidden="true" tabindex="-1"></a>        vec <span class="ot">&lt;-</span> vec[vec <span class="sc">!=</span> <span class="st">"Synapse"</span>]</span>
<span id="cb7-81"><a href="#cb7-81" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb7-82"><a href="#cb7-82" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="fu">sort</span>(<span class="fu">unique</span>(vec)), <span class="at">collapse =</span> <span class="st">"/"</span>)</span>
<span id="cb7-83"><a href="#cb7-83" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb7-84"><a href="#cb7-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-85"><a href="#cb7-85" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(df, res_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"protein_name"</span> <span class="ot">=</span> <span class="st">"Entry"</span>))</span>
<span id="cb7-86"><a href="#cb7-86" aria-hidden="true" tabindex="-1"></a>target_res_df <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(df, target_res_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"protein_name"</span> <span class="ot">=</span> <span class="st">"Entry"</span>))</span>
<span id="cb7-87"><a href="#cb7-87" aria-hidden="true" tabindex="-1"></a><span class="fu">vroom_write</span>(res_df, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.</span><span class="sc">\\</span><span class="st">w+$"</span>, <span class="st">".scl.raw.tsv"</span>, <span class="fu">basename</span>(de_file))))</span>
<span id="cb7-88"><a href="#cb7-88" aria-hidden="true" tabindex="-1"></a><span class="fu">vroom_write</span>(target_res_df, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.</span><span class="sc">\\</span><span class="st">w+$"</span>, <span class="st">".scl.target.tsv"</span>, <span class="fu">basename</span>(de_file))))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="go-enrichment-analysis" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="go-enrichment-analysis"><span class="header-section-number">6</span> GO enrichment analysis</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vroom)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clusterProfiler)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(AnnotationDbi)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>go_db_file <span class="ot">&lt;-</span> <span class="st">"/data/biodatabase/species/mRatBN7/genome/anno/org.Rn.eg.db.sqlite"</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>degs_file <span class="ot">&lt;-</span> <span class="st">"de/BTN3A2_vs_Control_bpca_de.all_vs_none.tsv"</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>root_dir <span class="ot">&lt;-</span> <span class="st">"go"</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(root_dir, <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.</span><span class="sc">\\</span><span class="st">w+$"</span>, <span class="st">""</span>, <span class="fu">basename</span>(degs_file)))</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(output_dir, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>degs <span class="ot">&lt;-</span> <span class="fu">vroom</span>(degs_file)</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>gene_set_ls <span class="ot">&lt;-</span> degs[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"diff_flag"</span>)] <span class="sc">%&gt;%</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(diff_flag <span class="sc">!=</span> <span class="st">"NO"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">split</span>(.[[<span class="st">"diff_flag"</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">lapply</span>(<span class="cf">function</span>(x) {</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>        <span class="fu">unique</span>(<span class="fu">na.omit</span>(x[[<span class="st">"gene_name"</span>]]))</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>orgdb <span class="ot">&lt;-</span> <span class="fu">loadDb</span>(go_db_file)</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>ego <span class="ot">&lt;-</span> <span class="fu">compareCluster</span>(gene_set_ls,</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">fun =</span> <span class="st">"enrichGO"</span>,</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">keyType =</span> <span class="st">"SYMBOL"</span>,</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">OrgDb =</span> orgdb,</span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>    <span class="at">ont =</span> <span class="st">"ALL"</span>,</span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>    <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span>,</span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>    <span class="at">minGSSize =</span> <span class="dv">10</span>,</span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">maxGSSize =</span> <span class="dv">1000</span>,</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>,</span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">qvalueCutoff =</span> <span class="fl">0.05</span>,</span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">readable =</span> <span class="cn">FALSE</span>,</span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">pool =</span> <span class="cn">FALSE</span></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(ego<span class="sc">@</span>compareClusterResult)</span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(ego, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="st">"GO_ALL.rds"</span>))</span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a><span class="fu">vroom_write</span>(ego<span class="sc">@</span>compareClusterResult,</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="st">"GO_ALL.tsv"</span>),</span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_names =</span> <span class="cn">TRUE</span>, <span class="at">append =</span> <span class="cn">FALSE</span></span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-expression-heatmap" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="plot-expression-heatmap"><span class="header-section-number">7</span> Plot expression heatmap</h2>
<p>The followings are just various examples:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vroom)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(YRUtils)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>sample_file <span class="ot">&lt;-</span> <span class="st">"/data/users/yangrui/lab_projs/lym_new/input/sample_sheet.tsv"</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>gene_file <span class="ot">&lt;-</span> <span class="st">"/data/users/yangrui/lab_projs/lym_new/de/BTN3A2_vs_Control_bpca_de.all_vs_none.tsv"</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>quant_file <span class="ot">&lt;-</span> <span class="st">"/data/users/yangrui/lab_projs/lym_new/input/BTN3A2_proteomic_quantification_data.tsv"</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>label_file <span class="ot">&lt;-</span> <span class="st">"/data/users/yangrui/lab_projs/lym_new/subcellular_location/BTN3A2_vs_Control_bpca_de.all_vs_none.scl.target.cys.tsv"</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="st">"/home/yangrui/temp"</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>sample_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(sample_file)</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>gene_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(gene_file) <span class="sc">%&gt;%</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(diff_flag <span class="sc">==</span> <span class="st">"BTN3A2 Up"</span>)</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>label_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(label_file)</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>quant_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(quant_file) <span class="sc">%&gt;%</span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="fu">all_of</span>(<span class="fu">c</span>(<span class="st">"protein_name"</span>, sample_df<span class="sc">$</span>sample))) <span class="sc">%&gt;%</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(protein_name <span class="sc">%in%</span> gene_df<span class="sc">$</span>protein_name) <span class="sc">%&gt;%</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set_rownames</span>(.[[<span class="st">"protein_name"</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span>protein_name)</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>quant_df[<span class="fu">is.na</span>(quant_df)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>quant_mat <span class="ot">&lt;-</span> <span class="fu">scale</span>(<span class="fu">t</span>(<span class="fu">log2</span>(quant_df <span class="sc">+</span> <span class="dv">1</span>)))</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>col_fun <span class="ot">&lt;-</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">floor</span>(<span class="fu">min</span>(quant_mat)), <span class="dv">0</span>, <span class="fu">ceiling</span>(<span class="fu">max</span>(quant_mat))), <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"white"</span>, <span class="st">"orange"</span>))</span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>row_order <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"BTN3A2_1"</span>, <span class="st">"BTN3A2_2"</span>, <span class="st">"BTN3A2_3"</span>, <span class="st">"Control_1"</span>, <span class="st">"Control_2"</span>, <span class="st">"Control_3"</span>)</span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>row_split <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_</span><span class="sc">\\</span><span class="st">d+$"</span>, <span class="st">""</span>, <span class="fu">row.names</span>(quant_mat))</span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a>row_anno <span class="ot">&lt;-</span> <span class="fu">rowAnnotation</span>(</span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">Sample =</span> <span class="fu">gsub</span>(<span class="st">"_</span><span class="sc">\\</span><span class="st">d+$"</span>, <span class="st">""</span>, <span class="fu">row.names</span>(quant_mat)),</span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">simple_anno_size =</span> <span class="fu">unit</span>(<span class="dv">3</span>, <span class="st">"mm"</span>),</span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">Sample =</span> <span class="fu">c</span>(<span class="st">"BTN3A2"</span> <span class="ot">=</span> <span class="st">"orange"</span>, <span class="st">"Control"</span> <span class="ot">=</span> <span class="st">"skyblue"</span>))</span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a>col_anno <span class="ot">&lt;-</span> <span class="fu">columnAnnotation</span>(</span>
<span id="cb9-39"><a href="#cb9-39" aria-hidden="true" tabindex="-1"></a>    <span class="at">Gene =</span> <span class="fu">anno_mark</span>(</span>
<span id="cb9-40"><a href="#cb9-40" aria-hidden="true" tabindex="-1"></a>        <span class="at">at =</span> <span class="fu">sapply</span>(label_df<span class="sc">$</span>protein_name, <span class="cf">function</span>(x) {</span>
<span id="cb9-41"><a href="#cb9-41" aria-hidden="true" tabindex="-1"></a>            <span class="fu">which</span>(<span class="fu">colnames</span>(quant_mat) <span class="sc">==</span> x)</span>
<span id="cb9-42"><a href="#cb9-42" aria-hidden="true" tabindex="-1"></a>        },</span>
<span id="cb9-43"><a href="#cb9-43" aria-hidden="true" tabindex="-1"></a>        <span class="at">simplify =</span> <span class="cn">TRUE</span>, <span class="at">USE.NAMES =</span> <span class="cn">FALSE</span></span>
<span id="cb9-44"><a href="#cb9-44" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb9-45"><a href="#cb9-45" aria-hidden="true" tabindex="-1"></a>        <span class="at">labels =</span> label_df<span class="sc">$</span>gene_name</span>
<span id="cb9-46"><a href="#cb9-46" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb9-47"><a href="#cb9-47" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-48"><a href="#cb9-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-49"><a href="#cb9-49" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(</span>
<span id="cb9-50"><a href="#cb9-50" aria-hidden="true" tabindex="-1"></a>    quant_mat,</span>
<span id="cb9-51"><a href="#cb9-51" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> col_fun,</span>
<span id="cb9-52"><a href="#cb9-52" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_columns =</span> <span class="cn">TRUE</span>,</span>
<span id="cb9-53"><a href="#cb9-53" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_column_slices =</span> <span class="cn">TRUE</span>,</span>
<span id="cb9-54"><a href="#cb9-54" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>,</span>
<span id="cb9-55"><a href="#cb9-55" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_row_slices =</span> <span class="cn">FALSE</span>,</span>
<span id="cb9-56"><a href="#cb9-56" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_column_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb9-57"><a href="#cb9-57" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_column_dend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb9-58"><a href="#cb9-58" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_row_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb9-59"><a href="#cb9-59" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_order =</span> row_order,</span>
<span id="cb9-60"><a href="#cb9-60" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_names_side =</span> <span class="st">"left"</span>,</span>
<span id="cb9-61"><a href="#cb9-61" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_split =</span> row_split,</span>
<span id="cb9-62"><a href="#cb9-62" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_gap =</span> <span class="fu">unit</span>(<span class="dv">0</span>, <span class="st">"mm"</span>),</span>
<span id="cb9-63"><a href="#cb9-63" aria-hidden="true" tabindex="-1"></a>    <span class="at">left_annotation =</span> row_anno,</span>
<span id="cb9-64"><a href="#cb9-64" aria-hidden="true" tabindex="-1"></a>    <span class="at">top_annotation =</span> col_anno,</span>
<span id="cb9-65"><a href="#cb9-65" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(<span class="at">title =</span> <span class="st">"Expression"</span>)</span>
<span id="cb9-66"><a href="#cb9-66" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-67"><a href="#cb9-67" aria-hidden="true" tabindex="-1"></a><span class="fu">ppreview</span>(p, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="st">"heatmap.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vroom)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(YRUtils)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>quant_file <span class="ot">&lt;-</span> <span class="st">"temp/S541N-WT.tsv"</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>labeled_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Ddx5"</span>, <span class="st">"Gapdh"</span>, <span class="st">"Actb"</span>, <span class="st">"Actg1"</span>, <span class="st">"Actn1"</span>, <span class="st">"Dhx9"</span>, <span class="st">"Draxin"</span>, <span class="st">"Pelp1"</span>, <span class="st">"Hnrnpa3"</span>)</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="st">"temp"</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>sample_df <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="fu">c</span>(<span class="st">"S541N"</span>, <span class="st">"WT"</span>), <span class="st">"_rep"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="st">".raw"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">sample =</span> <span class="fu">paste0</span>(Var1, Var2, Var3, Var4)) <span class="sc">%&gt;%</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rename</span>(<span class="at">group =</span> Var1) <span class="sc">%&gt;%</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(group, sample)</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>quant_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(quant_file) <span class="sc">%&gt;%</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(diff_flag <span class="sc">==</span> <span class="st">"S541N Up"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="fu">all_of</span>(<span class="fu">c</span>(<span class="st">"protein_name"</span>, <span class="st">"gene_name"</span>, sample_df<span class="sc">$</span>sample))) <span class="sc">%&gt;%</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set_rownames</span>(.[[<span class="st">"protein_name"</span>]])</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>quant_mat <span class="ot">&lt;-</span> quant_df <span class="sc">%&gt;%</span></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span>protein_name, <span class="sc">-</span>gene_name)</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>quant_mat[<span class="fu">is.na</span>(quant_mat)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>quant_mat <span class="ot">&lt;-</span> <span class="fu">scale</span>(<span class="fu">t</span>(<span class="fu">log2</span>(quant_mat <span class="sc">+</span> <span class="dv">1</span>)))</span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>col_fun <span class="ot">&lt;-</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">floor</span>(<span class="fu">min</span>(quant_mat)), <span class="dv">0</span>, <span class="fu">ceiling</span>(<span class="fu">max</span>(quant_mat))), <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"white"</span>, <span class="st">"orange"</span>))</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>row_order <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"S541N_rep1.raw"</span>, <span class="st">"S541N_rep2.raw"</span>, <span class="st">"S541N_rep3.raw"</span>, <span class="st">"WT_rep1.raw"</span>, <span class="st">"WT_rep2.raw"</span>, <span class="st">"WT_rep3.raw"</span>)</span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>row_split <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_rep</span><span class="sc">\\</span><span class="st">d+.raw$"</span>, <span class="st">""</span>, <span class="fu">row.names</span>(quant_mat))</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>left_anno <span class="ot">&lt;-</span> <span class="fu">rowAnnotation</span>(</span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">Sample =</span> <span class="fu">gsub</span>(<span class="st">"_rep</span><span class="sc">\\</span><span class="st">d+.raw$"</span>, <span class="st">""</span>, <span class="fu">row.names</span>(quant_mat)),</span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">simple_anno_size =</span> <span class="fu">unit</span>(<span class="dv">3</span>, <span class="st">"mm"</span>),</span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">Sample =</span> <span class="fu">c</span>(<span class="st">"S541N"</span> <span class="ot">=</span> <span class="st">"#EE6A50"</span>, <span class="st">"WT"</span> <span class="ot">=</span> <span class="st">"#76EEC6"</span>))</span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a>quant_df <span class="ot">&lt;-</span> quant_df <span class="sc">%&gt;%</span></span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a>        <span class="at">group =</span> <span class="fu">if_else</span>(<span class="fu">is.na</span>(gene_name), <span class="st">"Other"</span>,</span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a>            <span class="fu">if_else</span>(<span class="fu">str_detect</span>(gene_name, <span class="st">"^Rpl"</span>), <span class="st">"Rpl"</span>,</span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a>                <span class="fu">if_else</span>(<span class="fu">str_detect</span>(gene_name, <span class="st">"^Rps"</span>), <span class="st">"Rps"</span>,</span>
<span id="cb10-43"><a href="#cb10-43" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">if_else</span>(<span class="fu">str_detect</span>(gene_name, <span class="st">"^Tuba"</span>), <span class="st">"Tuba"</span>,</span>
<span id="cb10-44"><a href="#cb10-44" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">if_else</span>(<span class="fu">str_detect</span>(gene_name, <span class="st">"^Tubb"</span>), <span class="st">"Tubb"</span>, <span class="st">"Other"</span>)</span>
<span id="cb10-45"><a href="#cb10-45" aria-hidden="true" tabindex="-1"></a>                    )</span>
<span id="cb10-46"><a href="#cb10-46" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb10-47"><a href="#cb10-47" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb10-48"><a href="#cb10-48" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb10-49"><a href="#cb10-49" aria-hidden="true" tabindex="-1"></a>        <span class="at">group =</span> <span class="fu">factor</span>(group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Rpl"</span>, <span class="st">"Rps"</span>, <span class="st">"Tuba"</span>, <span class="st">"Tubb"</span>, <span class="st">"Other"</span>))</span>
<span id="cb10-50"><a href="#cb10-50" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb10-51"><a href="#cb10-51" aria-hidden="true" tabindex="-1"></a>column_split <span class="ot">&lt;-</span> quant_df<span class="sc">$</span>group</span>
<span id="cb10-52"><a href="#cb10-52" aria-hidden="true" tabindex="-1"></a>bottom_anno <span class="ot">&lt;-</span> <span class="fu">columnAnnotation</span>(</span>
<span id="cb10-53"><a href="#cb10-53" aria-hidden="true" tabindex="-1"></a>    <span class="at">Group =</span> quant_df<span class="sc">$</span>group,</span>
<span id="cb10-54"><a href="#cb10-54" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_legend =</span> <span class="cn">TRUE</span>,</span>
<span id="cb10-55"><a href="#cb10-55" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-56"><a href="#cb10-56" aria-hidden="true" tabindex="-1"></a>    <span class="at">simple_anno_size =</span> <span class="fu">unit</span>(<span class="dv">3</span>, <span class="st">"mm"</span>),</span>
<span id="cb10-57"><a href="#cb10-57" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">Group =</span> <span class="fu">c</span>(<span class="st">"Rpl"</span> <span class="ot">=</span> <span class="st">"#B23AEE"</span>, <span class="st">"Rps"</span> <span class="ot">=</span> <span class="st">"#B4EEB4"</span>, <span class="st">"Tuba"</span> <span class="ot">=</span> <span class="st">"#EE7600"</span>, <span class="st">"Tubb"</span> <span class="ot">=</span> <span class="st">"#00B2EE"</span>, <span class="st">"Other"</span> <span class="ot">=</span> <span class="st">"#EEA2AD"</span>))</span>
<span id="cb10-58"><a href="#cb10-58" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-59"><a href="#cb10-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-60"><a href="#cb10-60" aria-hidden="true" tabindex="-1"></a>label_df <span class="ot">&lt;-</span> <span class="fu">filter</span>(quant_df, gene_name <span class="sc">%in%</span> labeled_genes) <span class="sc">%&gt;%</span></span>
<span id="cb10-61"><a href="#cb10-61" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(protein_name, gene_name)</span>
<span id="cb10-62"><a href="#cb10-62" aria-hidden="true" tabindex="-1"></a>top_anno <span class="ot">&lt;-</span> <span class="fu">columnAnnotation</span>(</span>
<span id="cb10-63"><a href="#cb10-63" aria-hidden="true" tabindex="-1"></a>    <span class="at">Gene =</span> <span class="fu">anno_mark</span>(</span>
<span id="cb10-64"><a href="#cb10-64" aria-hidden="true" tabindex="-1"></a>        <span class="at">at =</span> <span class="fu">sapply</span>(label_df<span class="sc">$</span>protein_name, <span class="cf">function</span>(x) {</span>
<span id="cb10-65"><a href="#cb10-65" aria-hidden="true" tabindex="-1"></a>            <span class="fu">which</span>(<span class="fu">colnames</span>(quant_mat) <span class="sc">==</span> x)</span>
<span id="cb10-66"><a href="#cb10-66" aria-hidden="true" tabindex="-1"></a>        },</span>
<span id="cb10-67"><a href="#cb10-67" aria-hidden="true" tabindex="-1"></a>        <span class="at">simplify =</span> <span class="cn">TRUE</span>, <span class="at">USE.NAMES =</span> <span class="cn">FALSE</span></span>
<span id="cb10-68"><a href="#cb10-68" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb10-69"><a href="#cb10-69" aria-hidden="true" tabindex="-1"></a>        <span class="at">labels =</span> label_df<span class="sc">$</span>gene_name</span>
<span id="cb10-70"><a href="#cb10-70" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb10-71"><a href="#cb10-71" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-72"><a href="#cb10-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-73"><a href="#cb10-73" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(</span>
<span id="cb10-74"><a href="#cb10-74" aria-hidden="true" tabindex="-1"></a>    quant_mat,</span>
<span id="cb10-75"><a href="#cb10-75" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> col_fun,</span>
<span id="cb10-76"><a href="#cb10-76" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_columns =</span> <span class="cn">TRUE</span>,</span>
<span id="cb10-77"><a href="#cb10-77" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_column_slices =</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-78"><a href="#cb10-78" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-79"><a href="#cb10-79" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_row_slices =</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-80"><a href="#cb10-80" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_column_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-81"><a href="#cb10-81" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_column_dend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-82"><a href="#cb10-82" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_row_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-83"><a href="#cb10-83" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_order =</span> row_order,</span>
<span id="cb10-84"><a href="#cb10-84" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_names_side =</span> <span class="st">"left"</span>,</span>
<span id="cb10-85"><a href="#cb10-85" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_split =</span> row_split,</span>
<span id="cb10-86"><a href="#cb10-86" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_gap =</span> <span class="fu">unit</span>(<span class="dv">0</span>, <span class="st">"mm"</span>),</span>
<span id="cb10-87"><a href="#cb10-87" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_gap =</span> <span class="fu">unit</span>(<span class="dv">0</span>, <span class="st">"mm"</span>),</span>
<span id="cb10-88"><a href="#cb10-88" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_split =</span> column_split,</span>
<span id="cb10-89"><a href="#cb10-89" aria-hidden="true" tabindex="-1"></a>    <span class="at">left_annotation =</span> left_anno,</span>
<span id="cb10-90"><a href="#cb10-90" aria-hidden="true" tabindex="-1"></a>    <span class="at">top_annotation =</span> top_anno,</span>
<span id="cb10-91"><a href="#cb10-91" aria-hidden="true" tabindex="-1"></a>    <span class="at">bottom_annotation =</span> bottom_anno,</span>
<span id="cb10-92"><a href="#cb10-92" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_title =</span> <span class="cn">NULL</span>,</span>
<span id="cb10-93"><a href="#cb10-93" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_title_rot =</span> <span class="dv">90</span>,</span>
<span id="cb10-94"><a href="#cb10-94" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_title_side =</span> <span class="st">"bottom"</span>,</span>
<span id="cb10-95"><a href="#cb10-95" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(<span class="at">title =</span> <span class="st">"scaled LFQ intensity"</span>)</span>
<span id="cb10-96"><a href="#cb10-96" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-97"><a href="#cb10-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-98"><a href="#cb10-98" aria-hidden="true" tabindex="-1"></a><span class="fu">ppreview</span>(p, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="st">"heatmap.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vroom)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(YRUtils)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>quant_file <span class="ot">&lt;-</span> <span class="st">"temp/data.tsv"</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="st">"temp"</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>quant_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(quant_file) <span class="sc">%&gt;%</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set_rownames</span>(.[[<span class="st">"protein"</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span>protein)</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>col_fun <span class="ot">&lt;-</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">floor</span>(<span class="fu">min</span>(quant_df) <span class="sc">*</span> <span class="dv">10</span>) <span class="sc">/</span> <span class="dv">10</span>, <span class="fl">1.0</span>, <span class="fu">ceiling</span>(<span class="fu">max</span>(quant_df) <span class="sc">*</span> <span class="dv">10</span>) <span class="sc">/</span> <span class="dv">10</span>), <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"white"</span>, <span class="st">"orange"</span>))</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>column_order <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>    <span class="st">"WT"</span>,</span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">"P152A"</span>, <span class="st">"H70Q"</span>, <span class="st">"A200V"</span>, <span class="st">"Y48N"</span>, <span class="st">"A200T"</span>,</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>    <span class="st">"D99G"</span>, <span class="st">"L04P"</span>, <span class="st">"N199I"</span>, <span class="st">"N199H"</span>, <span class="st">"L54P"</span>, <span class="st">"K217R"</span>, <span class="st">"R18K"</span>, <span class="st">"R198Q"</span>, <span class="st">"R213W"</span>,</span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>    <span class="st">"C203S"</span>, <span class="st">"C206F"</span>, <span class="st">"C206Y"</span>, <span class="st">"V63L"</span>, <span class="st">"L66H"</span>, <span class="st">"P201S"</span>, <span class="st">"K209N"</span>, <span class="st">"R112P"</span>, <span class="st">"Q82H"</span></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>column_order_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster =</span> <span class="fu">c</span>(<span class="st">"WT"</span>, <span class="fu">rep</span>(<span class="st">"NO"</span>, <span class="dv">5</span>), <span class="fu">rep</span>(<span class="st">"LOF"</span>, <span class="dv">9</span>), <span class="fu">rep</span>(<span class="st">"GOF"</span>, <span class="dv">9</span>)),</span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">protein =</span> column_order</span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a>        <span class="at">cluster =</span> <span class="fu">factor</span>(cluster, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"WT"</span>, <span class="st">"NO"</span>, <span class="st">"LOF"</span>, <span class="st">"GOF"</span>)),</span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true" tabindex="-1"></a>        <span class="at">protein =</span> <span class="fu">factor</span>(protein, <span class="at">levels =</span> <span class="fu">colnames</span>(quant_df))</span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(protein)</span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true" tabindex="-1"></a>row_order <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GluA1_rep1"</span>, <span class="st">"GluA1_rep2"</span>, <span class="st">"GluA1_rep3"</span>, <span class="st">"GluA2_rep1"</span>, <span class="st">"GluA2_rep2"</span>, <span class="st">"GluA2_rep3"</span>, <span class="st">"GABAARAlpha1_rep1"</span>, <span class="st">"GABAARAlpha1_rep2"</span>, <span class="st">"GABAARAlpha1_rep3"</span>, <span class="st">"GABAARBeta1_rep1"</span>, <span class="st">"GABAARBeta1_rep2"</span>, <span class="st">"GABAARBeta1_rep3"</span>)</span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true" tabindex="-1"></a>row_split <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_rep</span><span class="sc">\\</span><span class="st">d+$"</span>, <span class="st">""</span>, <span class="fu">row.names</span>(quant_df))</span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true" tabindex="-1"></a>row_split <span class="ot">&lt;-</span> <span class="fu">factor</span>(row_split, <span class="at">levels =</span> <span class="fu">unique</span>(row_split))</span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true" tabindex="-1"></a>left_anno <span class="ot">&lt;-</span> <span class="fu">rowAnnotation</span>(</span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">Sample =</span> <span class="fu">gsub</span>(<span class="st">"_rep</span><span class="sc">\\</span><span class="st">d+$"</span>, <span class="st">""</span>, <span class="fu">row.names</span>(quant_df)),</span>
<span id="cb11-37"><a href="#cb11-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-38"><a href="#cb11-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-39"><a href="#cb11-39" aria-hidden="true" tabindex="-1"></a>    <span class="at">simple_anno_size =</span> <span class="fu">unit</span>(<span class="dv">3</span>, <span class="st">"mm"</span>),</span>
<span id="cb11-40"><a href="#cb11-40" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">Sample =</span> <span class="fu">c</span>(<span class="st">"GluA1"</span> <span class="ot">=</span> <span class="st">"#76EEC6"</span>, <span class="st">"GluA2"</span> <span class="ot">=</span> <span class="st">"#EEAEEE"</span>, <span class="st">"GABAARAlpha1"</span> <span class="ot">=</span> <span class="st">"#9F79EE"</span>, <span class="st">"GABAARBeta1"</span> <span class="ot">=</span> <span class="st">"#EE799F"</span>))</span>
<span id="cb11-41"><a href="#cb11-41" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-42"><a href="#cb11-42" aria-hidden="true" tabindex="-1"></a>top_anno <span class="ot">&lt;-</span> <span class="fu">columnAnnotation</span>(</span>
<span id="cb11-43"><a href="#cb11-43" aria-hidden="true" tabindex="-1"></a>    <span class="at">Type =</span> column_order_df<span class="sc">$</span>cluster,</span>
<span id="cb11-44"><a href="#cb11-44" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-45"><a href="#cb11-45" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-46"><a href="#cb11-46" aria-hidden="true" tabindex="-1"></a>    <span class="at">simple_anno_size =</span> <span class="fu">unit</span>(<span class="dv">3</span>, <span class="st">"mm"</span>),</span>
<span id="cb11-47"><a href="#cb11-47" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">Type =</span> <span class="fu">c</span>(<span class="st">"WT"</span> <span class="ot">=</span> <span class="st">"#BEBEBE"</span>, <span class="st">"NO"</span> <span class="ot">=</span> <span class="st">"#EED8AE"</span>, <span class="st">"LOF"</span> <span class="ot">=</span> <span class="st">"#EE8262"</span>, <span class="st">"GOF"</span> <span class="ot">=</span> <span class="st">"#BCD2EE"</span>))</span>
<span id="cb11-48"><a href="#cb11-48" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-49"><a href="#cb11-49" aria-hidden="true" tabindex="-1"></a>font_family <span class="ot">&lt;-</span> <span class="st">"Arial"</span></span>
<span id="cb11-50"><a href="#cb11-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-51"><a href="#cb11-51" aria-hidden="true" tabindex="-1"></a><span class="co"># `gpar()` is used to handle grid graphical parameters</span></span>
<span id="cb11-52"><a href="#cb11-52" aria-hidden="true" tabindex="-1"></a><span class="co"># for how to use `gpar()`, see `?gpar`</span></span>
<span id="cb11-53"><a href="#cb11-53" aria-hidden="true" tabindex="-1"></a><span class="co"># for ComplexHeatmap, you can set some parameters globally using `ht_opt`</span></span>
<span id="cb11-54"><a href="#cb11-54" aria-hidden="true" tabindex="-1"></a><span class="co"># e.g. `ht_opt$legend_title_gp &lt;- gpar(fontfamily = "Times New Roman")`</span></span>
<span id="cb11-55"><a href="#cb11-55" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(</span>
<span id="cb11-56"><a href="#cb11-56" aria-hidden="true" tabindex="-1"></a>    quant_df,</span>
<span id="cb11-57"><a href="#cb11-57" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> col_fun,</span>
<span id="cb11-58"><a href="#cb11-58" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-59"><a href="#cb11-59" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_column_slices =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-60"><a href="#cb11-60" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-61"><a href="#cb11-61" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster_row_slices =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-62"><a href="#cb11-62" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_column_names =</span> <span class="cn">TRUE</span>,</span>
<span id="cb11-63"><a href="#cb11-63" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_column_dend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-64"><a href="#cb11-64" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_row_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-65"><a href="#cb11-65" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_row_dend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-66"><a href="#cb11-66" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_order =</span> column_order,</span>
<span id="cb11-67"><a href="#cb11-67" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_split =</span> column_order_df<span class="sc">$</span>cluster,</span>
<span id="cb11-68"><a href="#cb11-68" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_order =</span> row_order,</span>
<span id="cb11-69"><a href="#cb11-69" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_split =</span> row_split,</span>
<span id="cb11-70"><a href="#cb11-70" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_gap =</span> <span class="fu">unit</span>(<span class="fl">0.8</span>, <span class="st">"mm"</span>),</span>
<span id="cb11-71"><a href="#cb11-71" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_gap =</span> <span class="fu">unit</span>(<span class="fl">0.8</span>, <span class="st">"mm"</span>),</span>
<span id="cb11-72"><a href="#cb11-72" aria-hidden="true" tabindex="-1"></a>    <span class="at">left_annotation =</span> left_anno,</span>
<span id="cb11-73"><a href="#cb11-73" aria-hidden="true" tabindex="-1"></a>    <span class="at">top_anno =</span> top_anno,</span>
<span id="cb11-74"><a href="#cb11-74" aria-hidden="true" tabindex="-1"></a>    <span class="at">height =</span> <span class="fu">nrow</span>(quant_df) <span class="sc">*</span> <span class="fu">unit</span>(<span class="dv">5</span>, <span class="st">"mm"</span>),</span>
<span id="cb11-75"><a href="#cb11-75" aria-hidden="true" tabindex="-1"></a>    <span class="at">width =</span> <span class="fu">ncol</span>(quant_df) <span class="sc">*</span> <span class="fu">unit</span>(<span class="dv">5</span>, <span class="st">"mm"</span>),</span>
<span id="cb11-76"><a href="#cb11-76" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_title_rot =</span> <span class="dv">90</span>,</span>
<span id="cb11-77"><a href="#cb11-77" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_title_rot =</span> <span class="dv">0</span>,</span>
<span id="cb11-78"><a href="#cb11-78" aria-hidden="true" tabindex="-1"></a>    <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(</span>
<span id="cb11-79"><a href="#cb11-79" aria-hidden="true" tabindex="-1"></a>        <span class="at">title =</span> <span class="st">"normalized expression level"</span>,</span>
<span id="cb11-80"><a href="#cb11-80" aria-hidden="true" tabindex="-1"></a>        <span class="at">title_gp =</span> <span class="fu">gpar</span>(<span class="at">fontfamily =</span> font_family),</span>
<span id="cb11-81"><a href="#cb11-81" aria-hidden="true" tabindex="-1"></a>        <span class="at">labels_gp =</span> <span class="fu">gpar</span>(<span class="at">fontfamily =</span> font_family)</span>
<span id="cb11-82"><a href="#cb11-82" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb11-83"><a href="#cb11-83" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_names_gp =</span> <span class="fu">gpar</span>(<span class="at">fontfamily =</span> font_family),</span>
<span id="cb11-84"><a href="#cb11-84" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_names_gp =</span> <span class="fu">gpar</span>(<span class="at">fontfamily =</span> font_family),</span>
<span id="cb11-85"><a href="#cb11-85" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_title_gp =</span> <span class="fu">gpar</span>(<span class="at">fontfamily =</span> font_family),</span>
<span id="cb11-86"><a href="#cb11-86" aria-hidden="true" tabindex="-1"></a>    <span class="at">column_title_gp =</span> <span class="fu">gpar</span>(<span class="at">fontfamily =</span> font_family)</span>
<span id="cb11-87"><a href="#cb11-87" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-88"><a href="#cb11-88" aria-hidden="true" tabindex="-1"></a><span class="fu">ppreview</span>(p, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="st">"heatmap.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="some-examples-about-usages-of-uniprot-and-string-apis" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="some-examples-about-usages-of-uniprot-and-string-apis"><span class="header-section-number">8</span> Some examples about usages of UniProt and STRING APIs</h2>
<ol type="1">
<li>Retrieve subcellular locations and interactions from UniProt</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vroom)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glue)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rvest)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>work_dir <span class="ot">&lt;-</span> <span class="st">"test"</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>organism_id <span class="ot">&lt;-</span> <span class="dv">9606</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>url_template <span class="ot">&lt;-</span> <span class="st">"https://rest.uniprot.org/uniprotkb/search?query=(accession:{accession})%20AND%20(active:true)%20AND%20(organism_id:{organism_id})&amp;fields=accession,gene_primary,organism_name,organism_id,protein_name,annotation_score,protein_existence,reviewed,ft_intramem,cc_subcellular_location,ft_topo_dom,ft_transmem,go,go_p,go_c,go_f,cc_interaction&amp;format=tsv"</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>de_file <span class="ot">&lt;-</span> <span class="st">"degs/xbfGlia_vs_xbfAstro.tsv"</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="st">"subcellular_locations_and_interactions"</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(work_dir)</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(output_dir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">FALSE</span>)</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(de_file) <span class="sc">%&gt;%</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="fu">all_of</span>(<span class="fu">c</span>(<span class="st">"protein_name"</span>, <span class="st">"gene_name"</span>, <span class="st">"diff_flag"</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(diff_flag <span class="sc">!=</span> <span class="st">"NO"</span>)</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>()</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (accession <span class="cf">in</span> <span class="fu">unique</span>(df<span class="sc">$</span>protein_name)) {</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>    url_instance <span class="ot">&lt;-</span> <span class="fu">glue</span>(url_template)</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>    e <span class="ot">&lt;-</span> <span class="fu">try</span>(item_raw_text <span class="ot">&lt;-</span> <span class="fu">read_html</span>(url_instance) <span class="sc">%&gt;%</span> <span class="fu">html_text</span>(), <span class="at">silent =</span> T)</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="st">"try-error"</span> <span class="sc">%in%</span> <span class="fu">class</span>(e)) {</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>        <span class="fu">message</span>(<span class="st">"invalid accession: "</span>, accession)</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>        item_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(<span class="fu">I</span>(item_raw_text), <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>            <span class="fu">mutate</span>(</span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>                <span class="st">`</span><span class="at">Organism (ID)</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">as.integer</span>(<span class="st">`</span><span class="at">Organism (ID)</span><span class="st">`</span>),</span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>                <span class="at">Annotation =</span> <span class="fu">as.integer</span>(Annotation)</span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a>        res_df <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(res_df, item_df)</span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> <span class="fu">distinct</span>(res_df)</span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a>res_df[[<span class="st">"scl_secret"</span>]] <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"secret"</span>, res_df[[<span class="st">"Subcellular location [CC]"</span>]], <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a>res_df[[<span class="st">"scl_membrane"</span>]] <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"membrane"</span>, res_df[[<span class="st">"Subcellular location [CC]"</span>]], <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a>res_df[[<span class="st">"scl_extracellular"</span>]] <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"extracellular"</span>, res_df[[<span class="st">"Subcellular location [CC]"</span>]], <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a>res_df[[<span class="st">"scl_pass"</span>]] <span class="ot">&lt;-</span> res_df[[<span class="st">"scl_secret"</span>]] <span class="sc">|</span> res_df[[<span class="st">"scl_membrane"</span>]] <span class="sc">|</span> res_df[[<span class="st">"scl_extracellular"</span>]]</span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> <span class="fu">filter</span>(res_df, scl_pass)</span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(res_df)</span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-45"><a href="#cb12-45" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(df, res_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"protein_name"</span> <span class="ot">=</span> <span class="st">"Entry"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb12-46"><a href="#cb12-46" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(diff_flag)</span>
<span id="cb12-47"><a href="#cb12-47" aria-hidden="true" tabindex="-1"></a><span class="fu">vroom_write</span>(res_df, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.(txt|tsv)$"</span>, <span class="st">".secreted.tsv"</span>, <span class="fu">basename</span>(de_file))), <span class="at">col_names =</span> <span class="cn">TRUE</span>, <span class="at">append =</span> <span class="cn">FALSE</span>)</span>
<span id="cb12-48"><a href="#cb12-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-49"><a href="#cb12-49" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> res_df <span class="sc">%&gt;%</span></span>
<span id="cb12-50"><a href="#cb12-50" aria-hidden="true" tabindex="-1"></a>    <span class="fu">separate_rows</span>(<span class="fu">all_of</span>(<span class="st">"Interacts with"</span>), <span class="at">sep =</span> <span class="st">"; "</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-51"><a href="#cb12-51" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb12-52"><a href="#cb12-52" aria-hidden="true" tabindex="-1"></a>        <span class="at">interacted_protein =</span> <span class="st">`</span><span class="at">Interacts with</span><span class="st">`</span>,</span>
<span id="cb12-53"><a href="#cb12-53" aria-hidden="true" tabindex="-1"></a>        <span class="at">interacted_protein =</span> <span class="fu">if_else</span>(</span>
<span id="cb12-54"><a href="#cb12-54" aria-hidden="true" tabindex="-1"></a>            <span class="fu">str_detect</span>(interacted_protein, <span class="st">".*</span><span class="sc">\\</span><span class="st">[[a-zA-Z0-9-]+</span><span class="sc">\\</span><span class="st">]$"</span>),</span>
<span id="cb12-55"><a href="#cb12-55" aria-hidden="true" tabindex="-1"></a>            <span class="fu">str_extract</span>(interacted_protein, <span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">[)[a-zA-Z0-9-]+(?=</span><span class="sc">\\</span><span class="st">])"</span>),</span>
<span id="cb12-56"><a href="#cb12-56" aria-hidden="true" tabindex="-1"></a>            interacted_protein</span>
<span id="cb12-57"><a href="#cb12-57" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb12-58"><a href="#cb12-58" aria-hidden="true" tabindex="-1"></a>        <span class="at">interacted_protein =</span> <span class="fu">gsub</span>(<span class="st">"-[0-9]+$"</span>, <span class="st">""</span>, interacted_protein)</span>
<span id="cb12-59"><a href="#cb12-59" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb12-60"><a href="#cb12-60" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(interacted_protein))</span>
<span id="cb12-61"><a href="#cb12-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-62"><a href="#cb12-62" aria-hidden="true" tabindex="-1"></a>interact_res_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>()</span>
<span id="cb12-63"><a href="#cb12-63" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (accession <span class="cf">in</span> <span class="fu">unique</span>(res_df[[<span class="st">"interacted_protein"</span>]])) {</span>
<span id="cb12-64"><a href="#cb12-64" aria-hidden="true" tabindex="-1"></a>    url_instance <span class="ot">&lt;-</span> <span class="fu">glue</span>(url_template)</span>
<span id="cb12-65"><a href="#cb12-65" aria-hidden="true" tabindex="-1"></a>    e <span class="ot">&lt;-</span> <span class="fu">try</span>(item_raw_text <span class="ot">&lt;-</span> <span class="fu">read_html</span>(url_instance) <span class="sc">%&gt;%</span> <span class="fu">html_text</span>(), <span class="at">silent =</span> T)</span>
<span id="cb12-66"><a href="#cb12-66" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="st">"try-error"</span> <span class="sc">%in%</span> <span class="fu">class</span>(e)) {</span>
<span id="cb12-67"><a href="#cb12-67" aria-hidden="true" tabindex="-1"></a>        <span class="fu">message</span>(<span class="st">"invalid accession: "</span>, accession)</span>
<span id="cb12-68"><a href="#cb12-68" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb12-69"><a href="#cb12-69" aria-hidden="true" tabindex="-1"></a>        item_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(<span class="fu">I</span>(item_raw_text), <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-70"><a href="#cb12-70" aria-hidden="true" tabindex="-1"></a>            <span class="fu">mutate</span>(</span>
<span id="cb12-71"><a href="#cb12-71" aria-hidden="true" tabindex="-1"></a>                <span class="st">`</span><span class="at">Organism (ID)</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">as.integer</span>(<span class="st">`</span><span class="at">Organism (ID)</span><span class="st">`</span>),</span>
<span id="cb12-72"><a href="#cb12-72" aria-hidden="true" tabindex="-1"></a>                <span class="at">Annotation =</span> <span class="fu">as.integer</span>(Annotation)</span>
<span id="cb12-73"><a href="#cb12-73" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb12-74"><a href="#cb12-74" aria-hidden="true" tabindex="-1"></a>        interact_res_df <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(interact_res_df, item_df)</span>
<span id="cb12-75"><a href="#cb12-75" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb12-76"><a href="#cb12-76" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-77"><a href="#cb12-77" aria-hidden="true" tabindex="-1"></a>interact_res_df <span class="ot">&lt;-</span> <span class="fu">distinct</span>(interact_res_df)</span>
<span id="cb12-78"><a href="#cb12-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-79"><a href="#cb12-79" aria-hidden="true" tabindex="-1"></a>interact_res_df[[<span class="st">"scl_secret"</span>]] <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"secret"</span>, interact_res_df[[<span class="st">"Subcellular location [CC]"</span>]], <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-80"><a href="#cb12-80" aria-hidden="true" tabindex="-1"></a>interact_res_df[[<span class="st">"scl_membrane"</span>]] <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"membrane"</span>, interact_res_df[[<span class="st">"Subcellular location [CC]"</span>]], <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-81"><a href="#cb12-81" aria-hidden="true" tabindex="-1"></a>interact_res_df[[<span class="st">"scl_extracellular"</span>]] <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"extracellular"</span>, interact_res_df[[<span class="st">"Subcellular location [CC]"</span>]], <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-82"><a href="#cb12-82" aria-hidden="true" tabindex="-1"></a>interact_res_df[[<span class="st">"scl_pass"</span>]] <span class="ot">&lt;-</span> interact_res_df[[<span class="st">"scl_secret"</span>]] <span class="sc">|</span> interact_res_df[[<span class="st">"scl_membrane"</span>]] <span class="sc">|</span> interact_res_df[[<span class="st">"scl_extracellular"</span>]]</span>
<span id="cb12-83"><a href="#cb12-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-84"><a href="#cb12-84" aria-hidden="true" tabindex="-1"></a>interact_res_df <span class="ot">&lt;-</span> <span class="fu">filter</span>(interact_res_df, scl_pass) <span class="sc">%&gt;%</span></span>
<span id="cb12-85"><a href="#cb12-85" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">interact_database =</span> <span class="st">"UniProt"</span>)</span>
<span id="cb12-86"><a href="#cb12-86" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(interact_res_df)</span>
<span id="cb12-87"><a href="#cb12-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-88"><a href="#cb12-88" aria-hidden="true" tabindex="-1"></a>final_res_df <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(res_df, interact_res_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"interacted_protein"</span> <span class="ot">=</span> <span class="st">"Entry"</span>), <span class="at">suffix =</span> <span class="fu">c</span>(<span class="st">".dep"</span>, <span class="st">".interacted"</span>))</span>
<span id="cb12-89"><a href="#cb12-89" aria-hidden="true" tabindex="-1"></a><span class="fu">vroom_write</span>(final_res_df, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.(txt|tsv)$"</span>, <span class="st">".interacted.uniprot.tsv"</span>, <span class="fu">basename</span>(de_file))), <span class="at">col_names =</span> <span class="cn">TRUE</span>, <span class="at">append =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Retrieve protein interactions from STRING</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vroom)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glue)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rvest)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(jsonlite)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>work_dir <span class="ot">&lt;-</span> <span class="st">"test"</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>organism_id <span class="ot">&lt;-</span> <span class="dv">9606</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>url_template <span class="ot">&lt;-</span> <span class="st">"https://rest.uniprot.org/uniprotkb/search?query=(accession:{accession})%20AND%20(active:true)%20AND%20(organism_id:{organism_id})&amp;fields=accession,gene_primary,organism_name,organism_id,protein_name,annotation_score,protein_existence,reviewed,ft_intramem,cc_subcellular_location,ft_topo_dom,ft_transmem,go,go_p,go_c,go_f,cc_interaction&amp;format=tsv"</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>submit_cmd <span class="ot">&lt;-</span> <span class="st">"curl --request POST 'https://rest.uniprot.org/idmapping/run' --form 'ids=</span><span class="sc">\"</span><span class="st">{ids}</span><span class="sc">\"</span><span class="st">' --form 'from=</span><span class="sc">\"</span><span class="st">UniProtKB_AC-ID</span><span class="sc">\"</span><span class="st">' --form 'to=</span><span class="sc">\"</span><span class="st">UniProtKB</span><span class="sc">\"</span><span class="st">'"</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>status_cmd <span class="ot">&lt;-</span> <span class="st">"curl -i 'https://rest.uniprot.org/idmapping/status/{submit_cmd_instance_response_job_id}'"</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>result_cmd <span class="ot">&lt;-</span> <span class="st">"curl -s 'https://rest.uniprot.org/idmapping/results/{submit_cmd_instance_response_job_id}'"</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>max_length <span class="ot">&lt;-</span> <span class="dv">20</span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>network_file <span class="ot">&lt;-</span> <span class="st">"string_db/9606.protein.physical.links.full.v12.0.txt.gz"</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>info_file <span class="ot">&lt;-</span> <span class="st">"string_db/9606.protein.info.v12.0.txt.gz"</span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>alias_file <span class="ot">&lt;-</span> <span class="st">"string_db/9606.protein.aliases.v12.0.txt.gz"</span></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>secret_file <span class="ot">&lt;-</span> <span class="st">"subcellular_locations_and_interactions/xbfGlia_vs_xbfAstro.secreted.tsv"</span></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="st">"subcellular_locations_and_interactions"</span></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(work_dir)</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(output_dir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">FALSE</span>)</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>info_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(info_file) <span class="sc">%&gt;%</span></span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="fu">all_of</span>(<span class="fu">c</span>(<span class="st">"#string_protein_id"</span>, <span class="st">"preferred_name"</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">distinct</span>()</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>secret_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(secret_file) <span class="sc">%&gt;%</span></span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">inner_join</span>(info_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene_name"</span> <span class="ot">=</span> <span class="st">"preferred_name"</span>))</span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a>network_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(network_file) <span class="sc">%&gt;%</span></span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(protein1 <span class="sc">%in%</span> secret_df[[<span class="st">"#string_protein_id"</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">inner_join</span>(info_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"protein2"</span> <span class="ot">=</span> <span class="st">"#string_protein_id"</span>))</span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a>source_levels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Ensembl_HGNC_uniprot_ids"</span>, <span class="st">"UniProt_ID"</span>, <span class="st">"UniProt_AC"</span>)</span>
<span id="cb13-36"><a href="#cb13-36" aria-hidden="true" tabindex="-1"></a>alias_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(alias_file) <span class="sc">%&gt;%</span></span>
<span id="cb13-37"><a href="#cb13-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>((<span class="st">`</span><span class="at">#string_protein_id</span><span class="st">`</span> <span class="sc">%in%</span> network_df<span class="sc">$</span>protein2) <span class="sc">&amp;</span> (source <span class="sc">%in%</span> source_levels)) <span class="sc">%&gt;%</span></span>
<span id="cb13-38"><a href="#cb13-38" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">source =</span> <span class="fu">factor</span>(source, <span class="at">levels =</span> source_levels)) <span class="sc">%&gt;%</span></span>
<span id="cb13-39"><a href="#cb13-39" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(<span class="st">`</span><span class="at">#string_protein_id</span><span class="st">`</span>) <span class="sc">%&gt;%</span></span>
<span id="cb13-40"><a href="#cb13-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice_min</span>(source) <span class="sc">%&gt;%</span></span>
<span id="cb13-41"><a href="#cb13-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb13-42"><a href="#cb13-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ungroup</span>()</span>
<span id="cb13-43"><a href="#cb13-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-44"><a href="#cb13-44" aria-hidden="true" tabindex="-1"></a>network_df <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(network_df, alias_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"protein2"</span> <span class="ot">=</span> <span class="st">"#string_protein_id"</span>))</span>
<span id="cb13-45"><a href="#cb13-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-46"><a href="#cb13-46" aria-hidden="true" tabindex="-1"></a>uniq_ids <span class="ot">&lt;-</span> <span class="fu">unique</span>(network_df<span class="sc">$</span>alias)</span>
<span id="cb13-47"><a href="#cb13-47" aria-hidden="true" tabindex="-1"></a>mpt_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>()</span>
<span id="cb13-48"><a href="#cb13-48" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (id_vec <span class="cf">in</span> <span class="fu">split</span>(uniq_ids, <span class="fu">ceiling</span>(<span class="fu">seq_along</span>(uniq_ids) <span class="sc">/</span> max_length))) {</span>
<span id="cb13-49"><a href="#cb13-49" aria-hidden="true" tabindex="-1"></a>    ids <span class="ot">&lt;-</span> <span class="fu">paste0</span>(id_vec, <span class="at">collapse =</span> <span class="st">","</span>)</span>
<span id="cb13-50"><a href="#cb13-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-51"><a href="#cb13-51" aria-hidden="true" tabindex="-1"></a>    submit_cmd_instance <span class="ot">&lt;-</span> <span class="fu">glue</span>(submit_cmd)</span>
<span id="cb13-52"><a href="#cb13-52" aria-hidden="true" tabindex="-1"></a>    submit_cmd_instance_response <span class="ot">&lt;-</span> <span class="fu">system</span>(submit_cmd_instance, <span class="at">intern =</span> <span class="cn">TRUE</span>, <span class="at">ignore.stdout =</span> <span class="cn">FALSE</span>, <span class="at">ignore.stderr =</span> <span class="cn">FALSE</span>, <span class="at">wait =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-53"><a href="#cb13-53" aria-hidden="true" tabindex="-1"></a>    submit_cmd_instance_response_ls <span class="ot">&lt;-</span> <span class="fu">fromJSON</span>(submit_cmd_instance_response)</span>
<span id="cb13-54"><a href="#cb13-54" aria-hidden="true" tabindex="-1"></a>    submit_cmd_instance_response_job_id <span class="ot">&lt;-</span> submit_cmd_instance_response_ls<span class="sc">$</span>jobId</span>
<span id="cb13-55"><a href="#cb13-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-56"><a href="#cb13-56" aria-hidden="true" tabindex="-1"></a>    status_cmd_instance <span class="ot">&lt;-</span> <span class="fu">glue</span>(status_cmd)</span>
<span id="cb13-57"><a href="#cb13-57" aria-hidden="true" tabindex="-1"></a>    <span class="cf">while</span> (<span class="cn">TRUE</span>) {</span>
<span id="cb13-58"><a href="#cb13-58" aria-hidden="true" tabindex="-1"></a>        status_cmd_instance_response <span class="ot">&lt;-</span> <span class="fu">system</span>(status_cmd_instance, <span class="at">intern =</span> <span class="cn">TRUE</span>, <span class="at">ignore.stdout =</span> <span class="cn">FALSE</span>, <span class="at">ignore.stderr =</span> <span class="cn">FALSE</span>, <span class="at">wait =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-59"><a href="#cb13-59" aria-hidden="true" tabindex="-1"></a>        status_cmd_instance_response <span class="ot">&lt;-</span> <span class="fu">trimws</span>(status_cmd_instance_response)</span>
<span id="cb13-60"><a href="#cb13-60" aria-hidden="true" tabindex="-1"></a>        status_cmd_instance_response_ls <span class="ot">&lt;-</span> <span class="fu">fromJSON</span>(status_cmd_instance_response[<span class="fu">length</span>(status_cmd_instance_response)])</span>
<span id="cb13-61"><a href="#cb13-61" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> ((status_cmd_instance_response[<span class="dv">1</span>] <span class="sc">==</span> <span class="st">"HTTP/2 303"</span>) <span class="sc">&amp;&amp;</span> (status_cmd_instance_response_ls<span class="sc">$</span>jobStatus <span class="sc">==</span> <span class="st">"FINISHED"</span>)) {</span>
<span id="cb13-62"><a href="#cb13-62" aria-hidden="true" tabindex="-1"></a>            <span class="fu">message</span>(<span class="st">"results can be retrieved!"</span>)</span>
<span id="cb13-63"><a href="#cb13-63" aria-hidden="true" tabindex="-1"></a>            <span class="cf">break</span></span>
<span id="cb13-64"><a href="#cb13-64" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb13-65"><a href="#cb13-65" aria-hidden="true" tabindex="-1"></a>        <span class="fu">Sys.sleep</span>(<span class="dv">3</span>)</span>
<span id="cb13-66"><a href="#cb13-66" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb13-67"><a href="#cb13-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-68"><a href="#cb13-68" aria-hidden="true" tabindex="-1"></a>    result_cmd_instance <span class="ot">&lt;-</span> <span class="fu">glue</span>(result_cmd)</span>
<span id="cb13-69"><a href="#cb13-69" aria-hidden="true" tabindex="-1"></a>    result_cmd_instance_response <span class="ot">&lt;-</span> <span class="fu">system</span>(result_cmd_instance, <span class="at">intern =</span> <span class="cn">TRUE</span>, <span class="at">ignore.stdout =</span> <span class="cn">FALSE</span>, <span class="at">ignore.stderr =</span> <span class="cn">FALSE</span>, <span class="at">wait =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-70"><a href="#cb13-70" aria-hidden="true" tabindex="-1"></a>    result_cmd_instance_response_ls <span class="ot">&lt;-</span> <span class="fu">fromJSON</span>(result_cmd_instance_response)</span>
<span id="cb13-71"><a href="#cb13-71" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(result_cmd_instance_response_ls<span class="sc">$</span>results) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb13-72"><a href="#cb13-72" aria-hidden="true" tabindex="-1"></a>        <span class="fu">stop</span>(<span class="st">"no ID was mapped successfully"</span>)</span>
<span id="cb13-73"><a href="#cb13-73" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb13-74"><a href="#cb13-74" aria-hidden="true" tabindex="-1"></a>    tmp_df <span class="ot">&lt;-</span> result_cmd_instance_response_ls<span class="sc">$</span>results</span>
<span id="cb13-75"><a href="#cb13-75" aria-hidden="true" tabindex="-1"></a>    mpt_df <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(mpt_df, tmp_df)</span>
<span id="cb13-76"><a href="#cb13-76" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb13-77"><a href="#cb13-77" aria-hidden="true" tabindex="-1"></a>mpt_df <span class="ot">&lt;-</span> <span class="fu">distinct</span>(mpt_df)</span>
<span id="cb13-78"><a href="#cb13-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-79"><a href="#cb13-79" aria-hidden="true" tabindex="-1"></a>network_df <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(network_df, mpt_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"alias"</span> <span class="ot">=</span> <span class="st">"from"</span>))</span>
<span id="cb13-80"><a href="#cb13-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-81"><a href="#cb13-81" aria-hidden="true" tabindex="-1"></a>interact_res_df <span class="ot">&lt;-</span> <span class="fu">tibble</span>()</span>
<span id="cb13-82"><a href="#cb13-82" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (accession <span class="cf">in</span> <span class="fu">unique</span>(network_df<span class="sc">$</span>to)) {</span>
<span id="cb13-83"><a href="#cb13-83" aria-hidden="true" tabindex="-1"></a>    url_instance <span class="ot">&lt;-</span> <span class="fu">glue</span>(url_template)</span>
<span id="cb13-84"><a href="#cb13-84" aria-hidden="true" tabindex="-1"></a>    e <span class="ot">&lt;-</span> <span class="fu">try</span>(item_raw_text <span class="ot">&lt;-</span> <span class="fu">read_html</span>(url_instance) <span class="sc">%&gt;%</span> <span class="fu">html_text</span>(), <span class="at">silent =</span> T)</span>
<span id="cb13-85"><a href="#cb13-85" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="st">"try-error"</span> <span class="sc">%in%</span> <span class="fu">class</span>(e)) {</span>
<span id="cb13-86"><a href="#cb13-86" aria-hidden="true" tabindex="-1"></a>        <span class="fu">message</span>(<span class="st">"invalid accession: "</span>, accession)</span>
<span id="cb13-87"><a href="#cb13-87" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb13-88"><a href="#cb13-88" aria-hidden="true" tabindex="-1"></a>        item_df <span class="ot">&lt;-</span> <span class="fu">vroom</span>(<span class="fu">I</span>(item_raw_text), <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb13-89"><a href="#cb13-89" aria-hidden="true" tabindex="-1"></a>            <span class="fu">mutate</span>(</span>
<span id="cb13-90"><a href="#cb13-90" aria-hidden="true" tabindex="-1"></a>                <span class="st">`</span><span class="at">Organism (ID)</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">as.integer</span>(<span class="st">`</span><span class="at">Organism (ID)</span><span class="st">`</span>),</span>
<span id="cb13-91"><a href="#cb13-91" aria-hidden="true" tabindex="-1"></a>                <span class="at">Annotation =</span> <span class="fu">as.integer</span>(Annotation)</span>
<span id="cb13-92"><a href="#cb13-92" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb13-93"><a href="#cb13-93" aria-hidden="true" tabindex="-1"></a>        interact_res_df <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(interact_res_df, item_df)</span>
<span id="cb13-94"><a href="#cb13-94" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb13-95"><a href="#cb13-95" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb13-96"><a href="#cb13-96" aria-hidden="true" tabindex="-1"></a>interact_res_df <span class="ot">&lt;-</span> <span class="fu">distinct</span>(interact_res_df)</span>
<span id="cb13-97"><a href="#cb13-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-98"><a href="#cb13-98" aria-hidden="true" tabindex="-1"></a>interact_res_df[[<span class="st">"scl_secret"</span>]] <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"secret"</span>, interact_res_df[[<span class="st">"Subcellular location [CC]"</span>]], <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-99"><a href="#cb13-99" aria-hidden="true" tabindex="-1"></a>interact_res_df[[<span class="st">"scl_membrane"</span>]] <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"membrane"</span>, interact_res_df[[<span class="st">"Subcellular location [CC]"</span>]], <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-100"><a href="#cb13-100" aria-hidden="true" tabindex="-1"></a>interact_res_df[[<span class="st">"scl_extracellular"</span>]] <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"extracellular"</span>, interact_res_df[[<span class="st">"Subcellular location [CC]"</span>]], <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-101"><a href="#cb13-101" aria-hidden="true" tabindex="-1"></a>interact_res_df[[<span class="st">"scl_pass"</span>]] <span class="ot">&lt;-</span> interact_res_df[[<span class="st">"scl_secret"</span>]] <span class="sc">|</span> interact_res_df[[<span class="st">"scl_membrane"</span>]] <span class="sc">|</span> interact_res_df[[<span class="st">"scl_extracellular"</span>]]</span>
<span id="cb13-102"><a href="#cb13-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-103"><a href="#cb13-103" aria-hidden="true" tabindex="-1"></a>interact_res_df <span class="ot">&lt;-</span> <span class="fu">filter</span>(interact_res_df, scl_pass) <span class="sc">%&gt;%</span></span>
<span id="cb13-104"><a href="#cb13-104" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">interact_database =</span> <span class="st">"STRING"</span>)</span>
<span id="cb13-105"><a href="#cb13-105" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(interact_res_df)</span>
<span id="cb13-106"><a href="#cb13-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-107"><a href="#cb13-107" aria-hidden="true" tabindex="-1"></a>final_res_df <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(secret_df, network_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"#string_protein_id"</span> <span class="ot">=</span> <span class="st">"protein1"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb13-108"><a href="#cb13-108" aria-hidden="true" tabindex="-1"></a>    <span class="fu">inner_join</span>(interact_res_df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"to"</span> <span class="ot">=</span> <span class="st">"Entry"</span>), <span class="at">suffix =</span> <span class="fu">c</span>(<span class="st">".dep"</span>, <span class="st">".interacted"</span>))</span>
<span id="cb13-109"><a href="#cb13-109" aria-hidden="true" tabindex="-1"></a><span class="fu">vroom_write</span>(final_res_df, <span class="at">file =</span> <span class="fu">file.path</span>(output_dir, <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.secreted</span><span class="sc">\\</span><span class="st">.(txt|tsv)$"</span>, <span class="st">".interacted.string.tsv"</span>, <span class="fu">basename</span>(secret_file))), <span class="at">col_names =</span> <span class="cn">TRUE</span>, <span class="at">append =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Powered by <a href="https://quarto.org">Quarto</a></p>
<div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/edit/main/Blogs/Bioinformatics/posts/bulk proteomics/de_related_analysis_for_bulk_proteomics/index.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/blob/main/Blogs/Bioinformatics/posts/bulk proteomics/de_related_analysis_for_bulk_proteomics/index.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>