<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.56">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Rui Yang">
<meta name="dcterms.date" content="2024-11-20">

<title>How to run ENCODE bulk RNA-seq pipeline in NeuroBorder Galaxy – NeuroBorder</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../../../">
<script src="../../../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../../../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../../../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../../../../styles.css">
</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../../../index.html">
    <span class="navbar-title">NeuroBorder</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../../../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-blogs" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Blogs</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-blogs">    
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Computer/index.html">
 <span class="dropdown-text">Computer</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Mathematics/index.html">
 <span class="dropdown-text">Mathematics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Bioinformatics/index.html">
 <span class="dropdown-text">Bioinformatics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Galaxy/index.html">
 <span class="dropdown-text">Galaxy</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../../../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <a href="https://github.com/yr-neurospace/NeuroBorder" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-page-right">
      <h1 class="title">How to run ENCODE bulk RNA-seq pipeline in NeuroBorder Galaxy</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">encode</div>
                <div class="quarto-category">bulk rna-seq</div>
                <div class="quarto-category">galaxy</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta column-page-right">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Rui Yang </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">November 20, 2024</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">November 20, 2024</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#steps" id="toc-steps" class="nav-link" data-scroll-target="#steps"><span class="header-section-number">2</span> Steps</a>
  <ul>
  <li><a href="#check-md5-sums-of-your-raw-fastq-data-in-your-computer" id="toc-check-md5-sums-of-your-raw-fastq-data-in-your-computer" class="nav-link" data-scroll-target="#check-md5-sums-of-your-raw-fastq-data-in-your-computer"><span class="header-section-number">2.1</span> Check MD5 sums of your raw FASTQ data in your computer</a></li>
  <li><a href="#copy-your-raw-fastq-data-to-a-new-directory-and-rename-them" id="toc-copy-your-raw-fastq-data-to-a-new-directory-and-rename-them" class="nav-link" data-scroll-target="#copy-your-raw-fastq-data-to-a-new-directory-and-rename-them"><span class="header-section-number">2.2</span> Copy your raw FASTQ data to a new directory and rename them</a></li>
  <li><a href="#upload-your-raw-fastq-files-and-md5-sums-file-to-neuroborder-galaxy" id="toc-upload-your-raw-fastq-files-and-md5-sums-file-to-neuroborder-galaxy" class="nav-link" data-scroll-target="#upload-your-raw-fastq-files-and-md5-sums-file-to-neuroborder-galaxy"><span class="header-section-number">2.3</span> Upload your raw FASTQ files and MD5 sums file to NeuroBorder Galaxy</a></li>
  <li><a href="#run-md5sum" id="toc-run-md5sum" class="nav-link" data-scroll-target="#run-md5sum"><span class="header-section-number">2.4</span> Run <code>MD5SUM</code></a></li>
  <li><a href="#run-fastqcmultiqc-over-raw-fastq-files" id="toc-run-fastqcmultiqc-over-raw-fastq-files" class="nav-link" data-scroll-target="#run-fastqcmultiqc-over-raw-fastq-files"><span class="header-section-number">2.5</span> Run <code>FASTQC/MULTIQC</code> over raw FASTQ files</a></li>
  <li><a href="#run-trim-galore" id="toc-run-trim-galore" class="nav-link" data-scroll-target="#run-trim-galore"><span class="header-section-number">2.6</span> Run <code>Trim Galore</code></a></li>
  <li><a href="#run-fastqcmultiqc-over-trimmed-fastq-files" id="toc-run-fastqcmultiqc-over-trimmed-fastq-files" class="nav-link" data-scroll-target="#run-fastqcmultiqc-over-trimmed-fastq-files"><span class="header-section-number">2.7</span> Run <code>FASTQC/MULTIQC</code> over trimmed FASTQ files</a></li>
  <li><a href="#run-encode-bulk-rna-seq-pipeline" id="toc-run-encode-bulk-rna-seq-pipeline" class="nav-link" data-scroll-target="#run-encode-bulk-rna-seq-pipeline"><span class="header-section-number">2.8</span> Run <code>ENCODE Bulk RNA-seq Pipeline</code></a></li>
  <li><a href="#run-extract-rsem-results" id="toc-run-extract-rsem-results" class="nav-link" data-scroll-target="#run-extract-rsem-results"><span class="header-section-number">2.9</span> Run <code>Extract RSEM Results</code></a></li>
  <li><a href="#run-bulk-rna-seq-de-with-deseq2" id="toc-run-bulk-rna-seq-de-with-deseq2" class="nav-link" data-scroll-target="#run-bulk-rna-seq-de-with-deseq2"><span class="header-section-number">2.10</span> Run <code>Bulk RNA-seq DE with DESeq2</code></a></li>
  <li><a href="#run-go-enrichment-analysis" id="toc-run-go-enrichment-analysis" class="nav-link" data-scroll-target="#run-go-enrichment-analysis"><span class="header-section-number">2.11</span> Run <code>GO Enrichment Analysis</code></a></li>
  <li><a href="#run-gsea-analysis" id="toc-run-gsea-analysis" class="nav-link" data-scroll-target="#run-gsea-analysis"><span class="header-section-number">2.12</span> Run <code>GSEA Analysis</code></a></li>
  <li><a href="#epilogue" id="toc-epilogue" class="nav-link" data-scroll-target="#epilogue"><span class="header-section-number">2.13</span> Epilogue</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/yr-neurospace/NeuroBorder/edit/main/Blogs/Galaxy/posts/Galaxy/endcode_rna_seq_pipeline_in_galaxy/index.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder/blob/main/Blogs/Galaxy/posts/Galaxy/endcode_rna_seq_pipeline_in_galaxy/index.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block column-page-right" id="quarto-document-content">





<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>NeuroBorder Galaxy platform is only responsible for performing analyses, not for visualizing results, which can be done by using <a href="https://github.com/yr-neurospace/NeuroBorder-ShinyWebApp">NeuroBorder Shiny App</a> in part. For some of results, you can visualize them via R, Python, GraphPad Prism, etc. yourself.</p>
</section>
<section id="steps" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="steps"><span class="header-section-number">2</span> Steps</h2>
<section id="check-md5-sums-of-your-raw-fastq-data-in-your-computer" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="check-md5-sums-of-your-raw-fastq-data-in-your-computer"><span class="header-section-number">2.1</span> Check MD5 sums of your raw FASTQ data in your computer</h3>
<p>For Windows users, I recommend using <a href="https://gitforwindows.org">git for windows</a>, which provides <code>md5sum</code> program equivalent to the Linux one. For Linux and Mac users, <code>md5sum</code> program has already been built into the system.</p>
<p>Once you have installed it, you can right-click the mouse on your target directory and choose <code>Open Git Bash here</code>. This will open Git Bash terminal and set your target directory as the current working directory. Then you can use <code>ls</code> or <code>cd &lt;target directory&gt;</code> command to list files/directories in the current working directory or change the current working directory to your specified target directory. Change your current working directory using <code>cd</code> into the directory containing MD5 file such as <code>md5.txt</code> or <code>MD5.txt</code>, and then run the following code:</p>
<pre class="{bash}"><code>md5sum -c md5.txt &gt; md5_check.txt</code></pre>
<p>Now, once the program has been run successfully, you can view the file <code>md5_check.txt</code> to see whether all files are correct.</p>
</section>
<section id="copy-your-raw-fastq-data-to-a-new-directory-and-rename-them" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="copy-your-raw-fastq-data-to-a-new-directory-and-rename-them"><span class="header-section-number">2.2</span> Copy your raw FASTQ data to a new directory and rename them</h3>
<p>Within your Git Bash, running the following code in the directory containing your raw FASTQ data needed to be analyzed:</p>
<pre class="{bash}"><code>find &lt;raw data directory&gt; -type f | grep -P ".+\.(fastq|fq)\.gz$" | xargs -I {} realpath {} | xargs -I {} cp {} &lt;target directory&gt;</code></pre>
<p>In the above code, you should replace <code>&lt;raw data directory&gt;</code> with your directory containing your raw FASTQ data needed to be analyzed (you can use <code>.</code> to denote the current working directory), and replace <code>&lt;target directory&gt;</code> with your new directory into which all raw FASTQ data will be copied.</p>
<p>After this step, rename your raw FASTQ files with the following specifications (<strong>This is absolutely necessary!</strong>):</p>
<p><strong>Note: input FASTQ file names must strictly adhere to the following rules otherwise you will get unexpected outputs!</strong></p>
<ul>
<li><p>Single-end: <code>ID_repN[_partN].(fastq|fq).gz</code></p></li>
<li><p>Paired-end: <code>ID_repN[_partN].R(1|2).(fastq|fq).gz</code></p></li>
</ul>
<p>Among them:</p>
<ul>
<li><p><code>ID</code> is the sample name, which can only contain <code>[a-zA-Z0-9]</code> and does <strong>NOT</strong> start with <code>[0-9]</code>.</p></li>
<li><p><code>repN</code> means the <code>N</code>th biological replicate.</p></li>
<li><p><code>partN</code> means the <code>N</code>th technical replicate. All technical replicates with the same <code>ID_repN</code> should be merged into a single FASTQ file before running downstream analyses. This will be done by ENCODE bulk RNA-seq pipeline before performing alignment and quantification.</p></li>
<li><p><code>N</code> can only contain <code>[0-9]</code>.</p></li>
</ul>
<p><strong>Note: Even though you have only one biological replicate, <code>rep1</code> CANNOT be omitted. If you have only one technical replicate, <code>part1</code> CAN be omitted.</strong></p>
<p>e.g.</p>
<p>For paired-end samples: <code>Brain1_rep1.fastq.R1.gz</code>, <code>Brain1_rep1.fastq.R2.gz</code>, <code>Brain1_rep2_part1.R1.fq.gz</code>, <code>Brain1_rep2_part2.R1.fastq.gz</code>, <code>Brain1_rep2_part1.R2.fq.gz</code>, <code>Brain1_rep2_part2.R2.fastq.gz</code>.</p>
<p>For single-end samples: <code>Cortex_rep1.fq.gz</code>, <code>Cortex_rep2_part1.fq.gz</code>, <code>Cortex_rep2_part2.fastq.gz</code>.</p>
<p>Once renaming is done, generate MD5 sums again for your newly renamed raw FASTQ files by running the following code in the directory containing your newly renamed FASTQ files:</p>
<pre class="{bash}"><code>md5sum.exe *.fastq.gz &gt; md5.txt</code></pre>
<p>This will generate all MD5 sums of your newly renamed FASTQ files in file <code>md5.txt</code>.</p>
</section>
<section id="upload-your-raw-fastq-files-and-md5-sums-file-to-neuroborder-galaxy" class="level3" data-number="2.3">
<h3 data-number="2.3" class="anchored" data-anchor-id="upload-your-raw-fastq-files-and-md5-sums-file-to-neuroborder-galaxy"><span class="header-section-number">2.3</span> Upload your raw FASTQ files and MD5 sums file to NeuroBorder Galaxy</h3>
<p>At present, the IP address of NeuroBorder Galaxy is an internal IP address <code>http://172.16.50.95</code>. This means that you cannot visit it outside. In addition, this IP is not permanent, which means it may be changed in the future. If you find it is invalid, please let me know. I will update it as I can.</p>
<p>Once you have typed the IP address into your browser, you can upload data and perform various analyses within NeuroBorder Galaxy.</p>
<p>For how to use some basic functionalities of NeuroBorder Galaxy, such as uploading data, managing history, etc., you can learn them from <a href="https://training.galaxyproject.org">the Galaxy Training</a> or I can give you some demos in real time if necessary.</p>
<p>At present, all tools used for ENCODE RNA-seq pipeline is under <code>Tools</code> <span class="math inline">\(\to\)</span> <code>NeuroBorder Tools</code>, in which you can choose the concrete tool for your purpose.</p>
<p>Once you have uploaded your raw FASTQ files and MD5 sums file to NeuroBorder Galaxy, they will be shown in the right <code>History</code> panel (You can give each of your histories a meaningful name and delete them when you don’t need them anymore. This is a good behavior for saving storage resource when available storage resource is inadequate.). Next, you should build your <strong>raw FASTQ files</strong> (note: not including the MD5 sums file) into a dataset list and give it a meaningful name.</p>
<p>By now, you can run each of the following step one by one.</p>
</section>
<section id="run-md5sum" class="level3" data-number="2.4">
<h3 data-number="2.4" class="anchored" data-anchor-id="run-md5sum"><span class="header-section-number">2.4</span> Run <code>MD5SUM</code></h3>
<p>This step ensures that your raw FASTQ files are not malformed during uploading.</p>
<p>In this step, you can check or generate (if <code>MD5 file</code> is not provided) MD5 sums of your files.</p>
</section>
<section id="run-fastqcmultiqc-over-raw-fastq-files" class="level3" data-number="2.5">
<h3 data-number="2.5" class="anchored" data-anchor-id="run-fastqcmultiqc-over-raw-fastq-files"><span class="header-section-number">2.5</span> Run <code>FASTQC/MULTIQC</code> over raw FASTQ files</h3>
<p>Generate quality reports for your raw FASTQ files.</p>
</section>
<section id="run-trim-galore" class="level3" data-number="2.6">
<h3 data-number="2.6" class="anchored" data-anchor-id="run-trim-galore"><span class="header-section-number">2.6</span> Run <code>Trim Galore</code></h3>
<p>This step performs quality, adapter, and Ns trimmings of your sequencing reads, discard reads too short, and ensure that you only use sequencing reads with high quality for downstream analyses.</p>
</section>
<section id="run-fastqcmultiqc-over-trimmed-fastq-files" class="level3" data-number="2.7">
<h3 data-number="2.7" class="anchored" data-anchor-id="run-fastqcmultiqc-over-trimmed-fastq-files"><span class="header-section-number">2.7</span> Run <code>FASTQC/MULTIQC</code> over trimmed FASTQ files</h3>
<p>Generate quality reports for your trimmed FASTQ files.</p>
</section>
<section id="run-encode-bulk-rna-seq-pipeline" class="level3" data-number="2.8">
<h3 data-number="2.8" class="anchored" data-anchor-id="run-encode-bulk-rna-seq-pipeline"><span class="header-section-number">2.8</span> Run <code>ENCODE Bulk RNA-seq Pipeline</code></h3>
<p>This step will run the ENCODE Bulk RNA-seq Pipeline and may take a long time from hours to days depending on your sample sizes.</p>
<p><strong>Please make sure you have selected the right species (e.g.&nbsp;<code>mm10</code> if your data is obtained from mouse) for your data.</strong></p>
<p>This step will generate some extremely large files, so you should download them with care. Once all files have been downloaded and unzipped, some files of them need to be uploaded again for performing downstream analyses. These files are mainly quantification files (with suffix such as <code>_anno_rsem.genes.results</code> or <code>_anno_rsem.isoforms.results</code>) generated by RSEM.</p>
<p>As mentioned before, all uploaded files still need to be built into a dataset list with a name when feeded into the next tool.</p>
</section>
<section id="run-extract-rsem-results" class="level3" data-number="2.9">
<h3 data-number="2.9" class="anchored" data-anchor-id="run-extract-rsem-results"><span class="header-section-number">2.9</span> Run <code>Extract RSEM Results</code></h3>
<p>This step will extract some expression metrics and store them into some files from RSEM results.</p>
</section>
<section id="run-bulk-rna-seq-de-with-deseq2" class="level3" data-number="2.10">
<h3 data-number="2.10" class="anchored" data-anchor-id="run-bulk-rna-seq-de-with-deseq2"><span class="header-section-number">2.10</span> Run <code>Bulk RNA-seq DE with DESeq2</code></h3>
<p>This step will perform differential expression analysis over all comparison pairs.</p>
</section>
<section id="run-go-enrichment-analysis" class="level3" data-number="2.11">
<h3 data-number="2.11" class="anchored" data-anchor-id="run-go-enrichment-analysis"><span class="header-section-number">2.11</span> Run <code>GO Enrichment Analysis</code></h3>
<p>This step performs GO enrichment analysis over one or more gene sets.</p>
</section>
<section id="run-gsea-analysis" class="level3" data-number="2.12">
<h3 data-number="2.12" class="anchored" data-anchor-id="run-gsea-analysis"><span class="header-section-number">2.12</span> Run <code>GSEA Analysis</code></h3>
<p>This step performs GSEA analysis.</p>
</section>
<section id="epilogue" class="level3" data-number="2.13">
<h3 data-number="2.13" class="anchored" data-anchor-id="epilogue"><span class="header-section-number">2.13</span> Epilogue</h3>
<p>To date, you have run all basic steps of bulk RNA-seq analysis. You can turn to visualize them or perform more adavnced analysis yourself.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Powered by <a href="https://quarto.org">Quarto</a></p>
<div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/yr-neurospace/NeuroBorder/edit/main/Blogs/Galaxy/posts/Galaxy/endcode_rna_seq_pipeline_in_galaxy/index.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder/blob/main/Blogs/Galaxy/posts/Galaxy/endcode_rna_seq_pipeline_in_galaxy/index.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>