<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Rui Yang">
<meta name="dcterms.date" content="2025-01-08">

<title>MPRA analysis pipeline â€“ NeuroBorder</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../../../">
<script src="../../../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../../../site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../../../../site_libs/quarto-html/quarto-syntax-highlighting-dark-b53751a350365c71b6c909e95f209ed1.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../../../site_libs/bootstrap/bootstrap-7656f4217d0b61e7d1e52c7a113f8883.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../../../../site_libs/bootstrap/bootstrap-dark-0e17766785cb233d275babd160f1f995.min.css" rel="prefetch" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../../../../styles.css">
</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../../../index.html">
    <span class="navbar-title">NeuroBorder</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../../../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-blogs" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Blogs</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-blogs">    
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Computer/index.html">
 <span class="dropdown-text">Computer</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Mathematics/index.html">
 <span class="dropdown-text">Mathematics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Bioinformatics/index.html">
 <span class="dropdown-text">Bioinformatics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Biology/index.html">
 <span class="dropdown-text">Biology</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../../../Blogs/Galaxy/index.html">
 <span class="dropdown-text">Galaxy</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../../../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-page-right">
      <h1 class="title">MPRA analysis pipeline</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">mpra</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta column-page-right">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Rui Yang </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">January 8, 2025</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">June 22, 2025</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#link-barcodes-to-cres" id="toc-link-barcodes-to-cres" class="nav-link active" data-scroll-target="#link-barcodes-to-cres"><span class="header-section-number">1</span> Link barcodes to CREs</a>
  <ul>
  <li><a href="#md5sum-check-over-raw-fastq-files" id="toc-md5sum-check-over-raw-fastq-files" class="nav-link" data-scroll-target="#md5sum-check-over-raw-fastq-files"><span class="header-section-number">1.1</span> MD5SUM check over raw FASTQ files</a></li>
  <li><a href="#fastqc-over-raw-fastq-files" id="toc-fastqc-over-raw-fastq-files" class="nav-link" data-scroll-target="#fastqc-over-raw-fastq-files"><span class="header-section-number">1.2</span> FASTQC over raw FASTQ files</a></li>
  <li><a href="#quality-trimming-over-raw-fastq-files" id="toc-quality-trimming-over-raw-fastq-files" class="nav-link" data-scroll-target="#quality-trimming-over-raw-fastq-files"><span class="header-section-number">1.3</span> Quality trimming over raw FASTQ files</a></li>
  <li><a href="#fastqc-over-clean-fastq-files" id="toc-fastqc-over-clean-fastq-files" class="nav-link" data-scroll-target="#fastqc-over-clean-fastq-files"><span class="header-section-number">1.4</span> FASTQC over clean FASTQ files</a></li>
  <li><a href="#generate-reference-sequences" id="toc-generate-reference-sequences" class="nav-link" data-scroll-target="#generate-reference-sequences"><span class="header-section-number">1.5</span> Generate reference sequences</a></li>
  <li><a href="#build-bowtie2-index" id="toc-build-bowtie2-index" class="nav-link" data-scroll-target="#build-bowtie2-index"><span class="header-section-number">1.6</span> Build Bowtie2 index</a></li>
  <li><a href="#align-reads-with-bowtie2" id="toc-align-reads-with-bowtie2" class="nav-link" data-scroll-target="#align-reads-with-bowtie2"><span class="header-section-number">1.7</span> Align reads with Bowtie2</a></li>
  <li><a href="#remove-reads-unmapped-and-with-low-quality" id="toc-remove-reads-unmapped-and-with-low-quality" class="nav-link" data-scroll-target="#remove-reads-unmapped-and-with-low-quality"><span class="header-section-number">1.8</span> Remove reads unmapped and with low quality</a></li>
  <li><a href="#extract-cre-barcode-pairs" id="toc-extract-cre-barcode-pairs" class="nav-link" data-scroll-target="#extract-cre-barcode-pairs"><span class="header-section-number">1.9</span> Extract CRE-Barcode pairs</a></li>
  <li><a href="#quality-check" id="toc-quality-check" class="nav-link" data-scroll-target="#quality-check"><span class="header-section-number">1.10</span> Quality check</a></li>
  <li><a href="#merge-all-cre-barcode-pairs" id="toc-merge-all-cre-barcode-pairs" class="nav-link" data-scroll-target="#merge-all-cre-barcode-pairs"><span class="header-section-number">1.11</span> Merge all CRE-Barcode pairs</a></li>
  </ul></li>
  <li><a href="#count-barcodes" id="toc-count-barcodes" class="nav-link" data-scroll-target="#count-barcodes"><span class="header-section-number">2</span> Count barcodes</a>
  <ul>
  <li><a href="#md5sum-check-over-raw-fastq-files-1" id="toc-md5sum-check-over-raw-fastq-files-1" class="nav-link" data-scroll-target="#md5sum-check-over-raw-fastq-files-1"><span class="header-section-number">2.1</span> MD5SUM check over raw FASTQ files</a></li>
  <li><a href="#fastqc-over-raw-fastq-files-1" id="toc-fastqc-over-raw-fastq-files-1" class="nav-link" data-scroll-target="#fastqc-over-raw-fastq-files-1"><span class="header-section-number">2.2</span> FASTQC over raw FASTQ files</a></li>
  <li><a href="#count-barcodes-1" id="toc-count-barcodes-1" class="nav-link" data-scroll-target="#count-barcodes-1"><span class="header-section-number">2.3</span> Count barcodes</a>
  <ul>
  <li><a href="#solution-1" id="toc-solution-1" class="nav-link" data-scroll-target="#solution-1"><span class="header-section-number">2.3.1</span> Solution 1</a></li>
  <li><a href="#solution-2" id="toc-solution-2" class="nav-link" data-scroll-target="#solution-2"><span class="header-section-number">2.3.2</span> Solution 2</a></li>
  </ul></li>
  <li><a href="#attach-barcodes-to-cres" id="toc-attach-barcodes-to-cres" class="nav-link" data-scroll-target="#attach-barcodes-to-cres"><span class="header-section-number">2.4</span> Attach barcodes to CREs</a>
  <ul>
  <li><a href="#solution-1-1" id="toc-solution-1-1" class="nav-link" data-scroll-target="#solution-1-1"><span class="header-section-number">2.4.1</span> Solution 1</a></li>
  <li><a href="#solution-2-1" id="toc-solution-2-1" class="nav-link" data-scroll-target="#solution-2-1"><span class="header-section-number">2.4.2</span> Solution 2</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#identify-active-cres" id="toc-identify-active-cres" class="nav-link" data-scroll-target="#identify-active-cres"><span class="header-section-number">3</span> Identify active CREs</a>
  <ul>
  <li><a href="#q95-threshold-method" id="toc-q95-threshold-method" class="nav-link" data-scroll-target="#q95-threshold-method"><span class="header-section-number">3.1</span> Q95 threshold method</a></li>
  <li><a href="#mpranalyze-method" id="toc-mpranalyze-method" class="nav-link" data-scroll-target="#mpranalyze-method"><span class="header-section-number">3.2</span> MPRAnalyze method</a></li>
  </ul></li>
  <li><a href="#peak-annotation" id="toc-peak-annotation" class="nav-link" data-scroll-target="#peak-annotation"><span class="header-section-number">4</span> Peak annotation</a></li>
  <li><a href="#appendix" id="toc-appendix" class="nav-link" data-scroll-target="#appendix"><span class="header-section-number">5</span> Appendix</a>
  <ul>
  <li><a href="#identify-active-barcodes" id="toc-identify-active-barcodes" class="nav-link" data-scroll-target="#identify-active-barcodes"><span class="header-section-number">5.1</span> Identify active barcodes</a>
  <ul>
  <li><a href="#extract-barcode-sample-matrix" id="toc-extract-barcode-sample-matrix" class="nav-link" data-scroll-target="#extract-barcode-sample-matrix"><span class="header-section-number">5.1.1</span> Extract barcode-sample matrix</a></li>
  <li><a href="#simple-threshold-method" id="toc-simple-threshold-method" class="nav-link" data-scroll-target="#simple-threshold-method"><span class="header-section-number">5.1.2</span> Simple threshold method</a></li>
  <li><a href="#mpranalyze-method-1" id="toc-mpranalyze-method-1" class="nav-link" data-scroll-target="#mpranalyze-method-1"><span class="header-section-number">5.1.3</span> MPRAnalyze method</a></li>
  </ul></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/edit/main/Blogs/Bioinformatics/posts/MPRA/mpra_analysis_pipeline/index.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/blob/main/Blogs/Bioinformatics/posts/MPRA/mpra_analysis_pipeline/index.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block column-page-right" id="quarto-document-content">





<section id="link-barcodes-to-cres" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="link-barcodes-to-cres"><span class="header-section-number">1</span> Link barcodes to CREs</h2>
<p>Before running any of the following steps, you should rename your FASTQ files according to <a href="https://www.neuroborder.com/Blogs/Galaxy/posts/Galaxy/endcode_rna_seq_pipeline_in_galaxy/#copy-your-raw-fastq-data-to-a-new-directory-and-rename-them">these rules</a>.</p>
<div id="1ad44b0e" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>work_dir <span class="op">=</span> <span class="st">"/data/users/dell/mpra/link_barcode_to_cre/enzyme_v20250430"</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(work_dir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="{r}"><code>work_dir &lt;- "/data/users/dell/mpra/link_barcode_to_cre/enzyme_v20250430"

setwd(work_dir)</code></pre>
<section id="md5sum-check-over-raw-fastq-files" class="level3" data-number="1.1">
<h3 data-number="1.1" class="anchored" data-anchor-id="md5sum-check-over-raw-fastq-files"><span class="header-section-number">1.1</span> MD5SUM check over raw FASTQ files</h3>
<div id="9cf08585" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>raw_fastq_dir <span class="op">=</span> <span class="st">"raw_fastq"</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>md5_file <span class="op">=</span> <span class="st">"md5.txt"</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>md5_check_file <span class="op">=</span> <span class="st">"md5_check.txt"</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(raw_fastq_dir)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>YRUtils.BaseUtils.<span class="fu">md5_check</span>(md5_file, md5_check_file)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(work_dir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fastqc-over-raw-fastq-files" class="level3" data-number="1.2">
<h3 data-number="1.2" class="anchored" data-anchor-id="fastqc-over-raw-fastq-files"><span class="header-section-number">1.2</span> FASTQC over raw FASTQ files</h3>
<div id="9b787e89" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>raw_fastq_dir <span class="op">=</span> <span class="st">"raw_fastq"</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>raw_fastqc_dir <span class="op">=</span> <span class="st">"raw_fastqc"</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(raw_fastqc_dir)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>raw_fastq_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(raw_fastq_dir, <span class="st">r"</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>YRUtils.BioUtils.<span class="fu">fastqc</span>(raw_fastq_files, raw_fastqc_dir;</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    fastqc_options<span class="op">=</span><span class="st">"--threads 4"</span>, multiqc_options<span class="op">=</span><span class="st">"--zip-data-dir"</span>, num_jobs<span class="op">=</span><span class="fl">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="quality-trimming-over-raw-fastq-files" class="level3" data-number="1.3">
<h3 data-number="1.3" class="anchored" data-anchor-id="quality-trimming-over-raw-fastq-files"><span class="header-section-number">1.3</span> Quality trimming over raw FASTQ files</h3>
<div id="b962e5fd" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>raw_fastq_dir <span class="op">=</span> <span class="st">"raw_fastq"</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>clean_fastq_dir <span class="op">=</span> <span class="st">"clean_fastq"</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(clean_fastq_dir)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>raw_fastq_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(raw_fastq_dir, <span class="st">r"</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>dict <span class="op">=</span> YRUtils.BioUtils.<span class="fu">auto_detect_fastq_read_type</span>(raw_fastq_files)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>files_dict <span class="op">=</span> <span class="cf">if</span> dict[<span class="st">"paired"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>    dict[<span class="st">"paired"</span>][<span class="st">"dict"</span>]</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> dict[<span class="st">"single"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>    dict[<span class="st">"single"</span>][<span class="st">"dict"</span>]</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@error</span> <span class="st">"did not detect any paired-end or single-end files"</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>files_read_type <span class="op">=</span> <span class="cf">if</span> dict[<span class="st">"paired"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">"paired"</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> dict[<span class="st">"single"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">"single"</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@error</span> <span class="st">"did not detect any paired-end or single-end files"</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>YRUtils.BioUtils.<span class="fu">trimgalore</span>(files_dict, files_read_type, clean_fastq_dir;</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>    trimgalore_options<span class="op">=</span><span class="st">"--cores 4 --phred33 --quality 20 --length 30 --trim-n"</span>,</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    num_jobs<span class="op">=</span><span class="fl">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fastqc-over-clean-fastq-files" class="level3" data-number="1.4">
<h3 data-number="1.4" class="anchored" data-anchor-id="fastqc-over-clean-fastq-files"><span class="header-section-number">1.4</span> FASTQC over clean FASTQ files</h3>
<div id="eeb28ede" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>clean_fastq_dir <span class="op">=</span> <span class="st">"clean_fastq"</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>clean_fastqc_dir <span class="op">=</span> <span class="st">"clean_fastqc"</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(clean_fastqc_dir)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>clean_fastq_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(clean_fastq_dir, <span class="st">r"</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>YRUtils.BioUtils.<span class="fu">fastqc</span>(clean_fastq_files, clean_fastqc_dir;</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    fastqc_options<span class="op">=</span><span class="st">"--threads 4"</span>, multiqc_options<span class="op">=</span><span class="st">"--zip-data-dir"</span>, num_jobs<span class="op">=</span><span class="fl">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="generate-reference-sequences" class="level3" data-number="1.5">
<h3 data-number="1.5" class="anchored" data-anchor-id="generate-reference-sequences"><span class="header-section-number">1.5</span> Generate reference sequences</h3>
<p>Write each reference sequence containing CRE as well as other necessary sequences and its ID into a FASTA file, which will be used later to build Bowtie2 reference index.</p>
<div id="6f58342d" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">CSV</span>, <span class="bu">DataFrames</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>left_seq <span class="op">=</span> <span class="st">"TTCTCTGGCCTAACTGTCTAGACCTGCAGGAGGACCGGATCAACT"</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>right_seq <span class="op">=</span> <span class="st">"CATTGCGTGAACCGACACTAGAGGGTATATAATGGAAGCTCGACTTCCAGCTTGGCAATCCGGTACTGTGCAAAGTGAACACATCGCTAAGCGAAAGCTAAGNNNNNNNNNNNNNNNACCGGTCGCCACCATGGTGAGCAAGG"</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>mpra_test_file <span class="op">=</span> <span class="st">"ref/2w_library.165bp.no_enzyme_cutting_sites.tsv"</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>mpra_ctl_file <span class="op">=</span> <span class="st">"ref/null_sequences.tsv"</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>output_mpra_test_file <span class="op">=</span> <span class="st">"ref/2w_library.165bp.no_enzyme_cutting_sites.dealed.tsv"</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>output_mpra_ctl_file <span class="op">=</span> <span class="st">"ref/null_sequences.dealed.tsv"</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>output_ref_fa_file <span class="op">=</span> <span class="st">"ref/mpra_ref.fa"</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>mpra_test <span class="op">=</span> CSV.<span class="fu">read</span>(mpra_test_file, DataFrame)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>mpra_test <span class="op">=</span> <span class="fu">unique</span>(mpra_test)</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>mpra_test <span class="op">=</span> <span class="fu">transform</span>(mpra_test, <span class="st">"PSCE"</span>, <span class="st">"extended_mm10_seq"</span> <span class="op">=&gt;</span> (x <span class="op">-&gt;</span> <span class="fu">string</span>.(left_seq, x, right_seq)) <span class="op">=&gt;</span> <span class="st">"attached_seq"</span>)</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>mpra_test <span class="op">=</span> <span class="fu">groupby</span>(mpra_test, <span class="st">"PSCE"</span>)</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>mpra_test <span class="op">=</span> <span class="fu">transform</span>(mpra_test, nrow <span class="op">=&gt;</span> <span class="st">"num_per_PSCE"</span>, eachindex <span class="op">=&gt;</span> <span class="st">"PSCE_sub_rank"</span>)</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>mpra_test <span class="op">=</span> <span class="fu">transform</span>(mpra_test, [<span class="st">"PSCE"</span>, <span class="st">"PSCE_sub_rank"</span>, <span class="st">"num_per_PSCE"</span>] <span class="op">=&gt;</span> <span class="fu">ByRow</span>((x, y, z) <span class="op">-&gt;</span> <span class="cf">begin</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> z <span class="op">==</span> <span class="fl">1</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>        x</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>        <span class="fu">string</span>.(x, <span class="st">"_"</span>, y)</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span>) <span class="op">=&gt;</span> <span class="st">"PSCE_new_id"</span>)</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(output_mpra_test_file, mpra_test; delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>mpra_ctl <span class="op">=</span> CSV.<span class="fu">read</span>(mpra_ctl_file, DataFrame)</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>mpra_ctl <span class="op">=</span> <span class="fu">unique</span>(mpra_ctl)</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>mpra_ctl <span class="op">=</span> <span class="fu">transform</span>(mpra_ctl, eachindex <span class="op">=&gt;</span> <span class="st">"rank"</span>)</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>mpra_ctl <span class="op">=</span> <span class="fu">transform</span>(mpra_ctl, <span class="st">"rank"</span> <span class="op">=&gt;</span> (x <span class="op">-&gt;</span> <span class="fu">string</span>.(<span class="st">"CTL"</span>, x)) <span class="op">=&gt;</span> <span class="st">"PSCE"</span>, <span class="st">"seq"</span> <span class="op">=&gt;</span> (x <span class="op">-&gt;</span> <span class="fu">string</span>.(left_seq, x, right_seq)) <span class="op">=&gt;</span> <span class="st">"attached_seq"</span>)</span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(output_mpra_ctl_file, mpra_ctl; delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a>ref_fa <span class="op">=</span> <span class="fu">vcat</span>(<span class="fu">string</span>.(<span class="st">"&gt;"</span>, mpra_test[!, <span class="st">"PSCE_new_id"</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>, mpra_test[!, <span class="st">"attached_seq"</span>]),</span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a>    <span class="fu">string</span>.(<span class="st">"&gt;"</span>, mpra_ctl[!, <span class="st">"PSCE"</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>, mpra_ctl[!, <span class="st">"attached_seq"</span>]))</span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a><span class="fu">open</span>(output_ref_fa_file, <span class="st">"w"</span>) <span class="cf">do</span> io</span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> ref_fa</span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a>        <span class="fu">println</span>(io, line)</span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="build-bowtie2-index" class="level3" data-number="1.6">
<h3 data-number="1.6" class="anchored" data-anchor-id="build-bowtie2-index"><span class="header-section-number">1.6</span> Build Bowtie2 index</h3>
<div id="0d5f4949" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>ref_fa <span class="op">=</span> <span class="st">"ref/mpra_ref.fa"</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>bowtie2_index_dir <span class="op">=</span> <span class="st">"bowtie2_index"</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>bowtie2_index_prefix <span class="op">=</span> <span class="st">"mpra_ref"</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>bowtie2_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>log_dir <span class="op">=</span> <span class="st">"log"</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"tmp"</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(bowtie2_index_dir)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(log_dir)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(tmp_dir)</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> !<span class="fu">isnothing</span>(<span class="fu">match</span>(<span class="st">r"</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>ref_fa<span class="ch">))</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>    new_ref_fa <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(ref_fa), <span class="st">r"</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">" =&gt; ""))</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="st">    </span>YRUtils<span class="sc">.</span><span class="st">ShellUtils</span><span class="sc">.</span><span class="st">pigz</span><span class="ch">(</span><span class="st">ref_fa, new_ref_fa; decompress=true, keep=true</span><span class="ch">)</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>    new_ref_fa <span class="op">=</span> ref_fa</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>cmd <span class="op">=</span> <span class="fu">pipeline</span>(<span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"bowtie2-build"</span>, <span class="st">"--threads"</span>, bowtie2_n_threads, <span class="st">"-f"</span>, new_ref_fa, <span class="fu">joinpath</span>(bowtie2_index_dir, bowtie2_index_prefix)]));</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>    <span class="cn">stdout</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"build_bowtie2_index.log"</span>),</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>    <span class="cn">stderr</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"build_bowtie2_index.log"</span>))</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a><span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> !<span class="fu">isnothing</span>(<span class="fu">match</span>(<span class="st">r"</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>ref_fa<span class="ch">))</span></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>(new_ref_fa)</span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="align-reads-with-bowtie2" class="level3" data-number="1.7">
<h3 data-number="1.7" class="anchored" data-anchor-id="align-reads-with-bowtie2"><span class="header-section-number">1.7</span> Align reads with Bowtie2</h3>
<div id="261cb14a" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>clean_fastq_dir <span class="op">=</span> <span class="st">"clean_fastq"</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>bam_dir <span class="op">=</span> <span class="st">"bam"</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"tmp"</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>log_dir <span class="op">=</span> <span class="st">"log"</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>bowtie2_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>bowtie2_index <span class="op">=</span> <span class="st">"bowtie2_index/mpra_ref"</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>samtools_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>samtools_mem <span class="op">=</span> <span class="st">"768M"</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(bam_dir)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(log_dir)</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(tmp_dir)</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>clean_fastq_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(clean_fastq_dir, <span class="st">r"</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>dict <span class="op">=</span> YRUtils.BioUtils.<span class="fu">auto_detect_fastq_read_type</span>(clean_fastq_files)</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>files_dict <span class="op">=</span> <span class="cf">if</span> dict[<span class="st">"paired"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>    dict[<span class="st">"paired"</span>][<span class="st">"dict"</span>]</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> dict[<span class="st">"single"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>    dict[<span class="st">"single"</span>][<span class="st">"dict"</span>]</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@error</span> <span class="st">"did not detect any paired-end or single-end files"</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>files_read_type <span class="op">=</span> <span class="cf">if</span> dict[<span class="st">"paired"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>    <span class="st">"paired"</span></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a><span class="cf">elseif</span> dict[<span class="st">"single"</span>][<span class="st">"status"</span>] <span class="op">==</span> <span class="st">"yes"</span></span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>    <span class="st">"single"</span></span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@error</span> <span class="st">"did not detect any paired-end or single-end files"</span></span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> files_read_type <span class="op">==</span> <span class="st">"paired"</span></span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> sample <span class="kw">in</span> <span class="fu">keys</span>(files_dict)</span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> replicate <span class="kw">in</span> <span class="fu">keys</span>(files_dict[sample])</span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a>            r1_fq_files <span class="op">=</span> files_dict[sample][replicate][<span class="st">"R1"</span>]</span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a>            r2_fq_files <span class="op">=</span> files_dict[sample][replicate][<span class="st">"R2"</span>]</span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a>            bam_file <span class="op">=</span> <span class="fu">joinpath</span>(bam_dir, <span class="fu">string</span>(sample, <span class="st">"_"</span>, replicate, <span class="st">".chr_srt.bam"</span>))</span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="fu">length</span>(r1_fq_files) <span class="op">&gt;</span> <span class="fl">1</span></span>
<span id="cb9-39"><a href="#cb9-39" aria-hidden="true" tabindex="-1"></a>                r1_fq_file <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(sample, <span class="st">"_"</span>, replicate, <span class="st">".R1.fq.gz"</span>))</span>
<span id="cb9-40"><a href="#cb9-40" aria-hidden="true" tabindex="-1"></a>                cmd <span class="op">=</span> <span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb9-41"><a href="#cb9-41" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">string</span>(<span class="st">"zcat -f "</span>, <span class="fu">join</span>(r1_fq_files, <span class="st">" "</span>),</span>
<span id="cb9-42"><a href="#cb9-42" aria-hidden="true" tabindex="-1"></a>                        <span class="st">" | pigz -n -c &gt; "</span>,</span>
<span id="cb9-43"><a href="#cb9-43" aria-hidden="true" tabindex="-1"></a>                        r1_fq_file)]))</span>
<span id="cb9-44"><a href="#cb9-44" aria-hidden="true" tabindex="-1"></a>                <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb9-45"><a href="#cb9-45" aria-hidden="true" tabindex="-1"></a>                <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb9-46"><a href="#cb9-46" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span></span>
<span id="cb9-47"><a href="#cb9-47" aria-hidden="true" tabindex="-1"></a>                r1_fq_file <span class="op">=</span> r1_fq_files[<span class="fl">1</span>]</span>
<span id="cb9-48"><a href="#cb9-48" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb9-49"><a href="#cb9-49" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="fu">length</span>(r2_fq_files) <span class="op">&gt;</span> <span class="fl">1</span></span>
<span id="cb9-50"><a href="#cb9-50" aria-hidden="true" tabindex="-1"></a>                r2_fq_file <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">string</span>(sample, <span class="st">"_"</span>, replicate, <span class="st">".R2.fq.gz"</span>))</span>
<span id="cb9-51"><a href="#cb9-51" aria-hidden="true" tabindex="-1"></a>                cmd <span class="op">=</span> <span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb9-52"><a href="#cb9-52" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">string</span>(<span class="st">"zcat -f "</span>, <span class="fu">join</span>(r2_fq_files, <span class="st">" "</span>),</span>
<span id="cb9-53"><a href="#cb9-53" aria-hidden="true" tabindex="-1"></a>                        <span class="st">" | pigz -n -c &gt; "</span>,</span>
<span id="cb9-54"><a href="#cb9-54" aria-hidden="true" tabindex="-1"></a>                        r2_fq_file)]))</span>
<span id="cb9-55"><a href="#cb9-55" aria-hidden="true" tabindex="-1"></a>                <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb9-56"><a href="#cb9-56" aria-hidden="true" tabindex="-1"></a>                <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb9-57"><a href="#cb9-57" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span></span>
<span id="cb9-58"><a href="#cb9-58" aria-hidden="true" tabindex="-1"></a>                r2_fq_file <span class="op">=</span> r2_fq_files[<span class="fl">1</span>]</span>
<span id="cb9-59"><a href="#cb9-59" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb9-60"><a href="#cb9-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-61"><a href="#cb9-61" aria-hidden="true" tabindex="-1"></a>            cmd <span class="op">=</span> <span class="fu">pipeline</span>(</span>
<span id="cb9-62"><a href="#cb9-62" aria-hidden="true" tabindex="-1"></a>                <span class="fu">Cmd</span>(</span>
<span id="cb9-63"><a href="#cb9-63" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb9-64"><a href="#cb9-64" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">string</span>(<span class="st">"bowtie2 --np 0 -p "</span>, bowtie2_n_threads, <span class="st">" -x "</span>, bowtie2_index, <span class="st">" -1 "</span>, r1_fq_file, <span class="st">" -2 "</span>, r2_fq_file,</span>
<span id="cb9-65"><a href="#cb9-65" aria-hidden="true" tabindex="-1"></a>                            <span class="st">" | samtools view -S -u - | samtools sort -@ "</span>, samtools_n_threads, <span class="st">" -m "</span>, samtools_mem, <span class="st">" - -o "</span>, bam_file)]),</span>
<span id="cb9-66"><a href="#cb9-66" aria-hidden="true" tabindex="-1"></a>                );</span>
<span id="cb9-67"><a href="#cb9-67" aria-hidden="true" tabindex="-1"></a>                <span class="cn">stdout</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"bowtie2_align.log"</span>),</span>
<span id="cb9-68"><a href="#cb9-68" aria-hidden="true" tabindex="-1"></a>                <span class="cn">stderr</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"bowtie2_align.log"</span>),</span>
<span id="cb9-69"><a href="#cb9-69" aria-hidden="true" tabindex="-1"></a>                append<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb9-70"><a href="#cb9-70" aria-hidden="true" tabindex="-1"></a>            <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb9-71"><a href="#cb9-71" aria-hidden="true" tabindex="-1"></a>            <span class="fu">open</span>(io <span class="op">-&gt;</span> <span class="fu">println</span>(io, <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)),</span>
<span id="cb9-72"><a href="#cb9-72" aria-hidden="true" tabindex="-1"></a>                <span class="fu">joinpath</span>(log_dir, <span class="st">"bowtie2_align.log"</span>), <span class="st">"a"</span>)</span>
<span id="cb9-73"><a href="#cb9-73" aria-hidden="true" tabindex="-1"></a>            <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb9-74"><a href="#cb9-74" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb9-75"><a href="#cb9-75" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb9-76"><a href="#cb9-76" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb9-77"><a href="#cb9-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-78"><a href="#cb9-78" aria-hidden="true" tabindex="-1"></a>cmd <span class="op">=</span> <span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>, <span class="fu">string</span>(<span class="st">"rm -rf "</span>, <span class="fu">joinpath</span>(tmp_dir, <span class="st">"*"</span>))]))</span>
<span id="cb9-79"><a href="#cb9-79" aria-hidden="true" tabindex="-1"></a><span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb9-80"><a href="#cb9-80" aria-hidden="true" tabindex="-1"></a><span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="remove-reads-unmapped-and-with-low-quality" class="level3" data-number="1.8">
<h3 data-number="1.8" class="anchored" data-anchor-id="remove-reads-unmapped-and-with-low-quality"><span class="header-section-number">1.8</span> Remove reads unmapped and with low quality</h3>
<div id="032fd35f" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>bam_dir <span class="op">=</span> <span class="st">"bam"</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>high_qual_bam_dir <span class="op">=</span> <span class="st">"high_qual_bam"</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>log_dir <span class="op">=</span> <span class="st">"log"</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>tmp_dir <span class="op">=</span> <span class="st">"tmp"</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>samtools_n_threads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>samtools_mem <span class="op">=</span> <span class="st">"768M"</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>map_qual <span class="op">=</span> <span class="fl">30</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(high_qual_bam_dir)</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>bam_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(bam_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> bam_file <span class="kw">in</span> bam_files</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>    tmp_name_srt_bam_file <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>name_srt<span class="sc">.</span><span class="st">bam"</span>))</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">pipeline</span>(<span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"samtools view -u -F 1804 -f 2 -q "</span>, map_qual, <span class="st">" "</span>, bam_file,</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | samtools sort -n -@ "</span>, samtools_n_threads, <span class="st">" -m "</span>, samtools_mem, <span class="st">" - -o "</span>, tmp_name_srt_bam_file)]));</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stdout</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>),</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stderr</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>),</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>        append<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">open</span>(io <span class="op">-&gt;</span> <span class="fu">println</span>(io, <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)),</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>        <span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>), <span class="st">"a"</span>)</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>    tmp_fixmate_bam_file <span class="op">=</span> <span class="fu">joinpath</span>(tmp_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>fixmate<span class="sc">.</span><span class="st">bam"</span>))</span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">pipeline</span>(<span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"samtools fixmate -@ "</span>, samtools_n_threads, <span class="st">" -r "</span>, tmp_name_srt_bam_file, <span class="st">" "</span>, tmp_fixmate_bam_file)]));</span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stdout</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>),</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stderr</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>),</span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>        append<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">open</span>(io <span class="op">-&gt;</span> <span class="fu">println</span>(io, <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)),</span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a>        <span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>), <span class="st">"a"</span>)</span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a>    filtered_bam_file <span class="op">=</span> <span class="fu">joinpath</span>(high_qual_bam_dir, <span class="fu">replace</span>(<span class="fu">basename</span>(bam_file), <span class="st">r"</span><span class="sc">\.\w+\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>chr_srt<span class="sc">.</span><span class="st">bam"</span>))</span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="fu">pipeline</span>(<span class="fu">Cmd</span>(<span class="fu">string</span>.([<span class="st">"/usr/bin/bash"</span>, <span class="st">"-e"</span>, <span class="st">"-c"</span>,</span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a>            <span class="fu">string</span>(<span class="st">"samtools view -u -F 1804 -f 2 "</span>, tmp_fixmate_bam_file,</span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a>                <span class="st">" | samtools sort -@ "</span>, samtools_n_threads, <span class="st">" -m "</span>, samtools_mem, <span class="st">" - -o "</span>, filtered_bam_file)]));</span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stdout</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>),</span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a>        <span class="cn">stderr</span><span class="op">=</span><span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>),</span>
<span id="cb10-43"><a href="#cb10-43" aria-hidden="true" tabindex="-1"></a>        append<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb10-44"><a href="#cb10-44" aria-hidden="true" tabindex="-1"></a>    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)</span>
<span id="cb10-45"><a href="#cb10-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">open</span>(io <span class="op">-&gt;</span> <span class="fu">println</span>(io, <span class="fu">string</span>(<span class="st">"running "</span>, cmd, <span class="st">" ..."</span>)),</span>
<span id="cb10-46"><a href="#cb10-46" aria-hidden="true" tabindex="-1"></a>        <span class="fu">joinpath</span>(log_dir, <span class="st">"reads_filter.log"</span>), <span class="st">"a"</span>)</span>
<span id="cb10-47"><a href="#cb10-47" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run</span>(cmd; wait<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb10-48"><a href="#cb10-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-49"><a href="#cb10-49" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>.([tmp_name_srt_bam_file, tmp_fixmate_bam_file])</span>
<span id="cb10-50"><a href="#cb10-50" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="extract-cre-barcode-pairs" class="level3" data-number="1.9">
<h3 data-number="1.9" class="anchored" data-anchor-id="extract-cre-barcode-pairs"><span class="header-section-number">1.9</span> Extract CRE-Barcode pairs</h3>
<div id="f049cafb" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">XAM</span>, <span class="bu">FASTX</span>, <span class="bu">CSV</span>, <span class="bu">DataFrames</span>, <span class="bu">YRUtils</span>, <span class="bu">Serialization</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="kw">function</span> <span class="fu">extract_cre_bc_pairs</span>(bam_file<span class="op">::</span><span class="dt">AbstractString</span>, ref_dict<span class="op">::</span><span class="dt">Dict{String,String}</span>;</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    barcode_length<span class="op">::</span><span class="dt">Int</span>=<span class="fl">15</span>, quality_scheme<span class="op">::</span><span class="dt">Int</span>=<span class="fl">33</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    valid_dna_bases <span class="op">=</span> (<span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    total_num_records <span class="op">=</span> <span class="fl">0</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    xn_num_records <span class="op">=</span> <span class="fl">0</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>    complete_xn_num_records <span class="op">=</span> <span class="fl">0</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    valid_xn_num_records <span class="op">=</span> <span class="fl">0</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>    typical_aln_vec_dict <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{String,Vector{String}}</span>()</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>    cre_bc_vec <span class="op">=</span> <span class="dt">Tuple</span>{<span class="dt">String</span>,<span class="dt">String</span>}[]</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>    reader <span class="op">=</span> <span class="fu">open</span>(BAM.Reader, bam_file)</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>    record <span class="op">=</span> BAM.<span class="fu">Record</span>()</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">while</span> !<span class="fu">eof</span>(reader)</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>        <span class="fu">empty!</span>(record)</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>        <span class="fu">read!</span>(reader, record)</span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>        total_num_records <span class="op">+=</span> <span class="fl">1</span></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>        <span class="co"># The optional field XN:i:&lt;N&gt; reports the number of ambiguous reference characters (e.g. N) overlapped by an alignment</span></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="fu">haskey</span>(record, <span class="st">"XN"</span>) <span class="op">&amp;&amp;</span> record[<span class="st">"XN"</span>] <span class="op">==</span> barcode_length</span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>            xn_num_records <span class="op">+=</span> <span class="fl">1</span></span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>            ref_name <span class="op">=</span> BAM.<span class="fu">refname</span>(record)</span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>            <span class="co"># The leftmost mapping position</span></span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a>            <span class="co"># BAM is 0-based, while SAM is 1-based</span></span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true" tabindex="-1"></a>            <span class="co"># BAM.position() gets the 1-based leftmost mapping position of record</span></span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true" tabindex="-1"></a>            ref_pos <span class="op">=</span> BAM.<span class="fu">position</span>(record)</span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true" tabindex="-1"></a>            ref_seq <span class="op">=</span> ref_dict[ref_name]</span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true" tabindex="-1"></a>            cigar_str <span class="op">=</span> BAM.<span class="fu">cigar</span>(record)</span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true" tabindex="-1"></a>            query_seq <span class="op">=</span> <span class="fu">string</span>(BAM.<span class="fu">sequence</span>(record))</span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true" tabindex="-1"></a>            query_qual_char_seq <span class="op">=</span> <span class="fu">join</span>(<span class="fu">Char</span>.(BAM.<span class="fu">quality</span>(record) <span class="op">.+</span> quality_scheme))</span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true" tabindex="-1"></a>            aln_vec <span class="op">=</span> <span class="fu">collect</span>(YRUtils.BioUtils.<span class="fu">parse_cigar</span>(cigar_str, ref_seq, query_seq, ref_pos; truncate_ref<span class="op">=</span><span class="cn">false</span>))</span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true" tabindex="-1"></a>            qual_aln_vec <span class="op">=</span> <span class="fu">collect</span>(YRUtils.BioUtils.<span class="fu">parse_cigar</span>(cigar_str, ref_seq, query_qual_char_seq, ref_pos; truncate_ref<span class="op">=</span><span class="cn">false</span>))</span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true" tabindex="-1"></a>            ref_m <span class="op">=</span> <span class="fu">match</span>(<span class="fu">Regex</span>(<span class="fu">string</span>(<span class="st">"N{"</span>, barcode_length, <span class="st">"}"</span>)), aln_vec[<span class="fl">1</span>])</span>
<span id="cb11-37"><a href="#cb11-37" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> !<span class="fu">isnothing</span>(ref_m)</span>
<span id="cb11-38"><a href="#cb11-38" aria-hidden="true" tabindex="-1"></a>                complete_xn_num_records <span class="op">+=</span> <span class="fl">1</span></span>
<span id="cb11-39"><a href="#cb11-39" aria-hidden="true" tabindex="-1"></a>                extract_range <span class="op">=</span> ref_m.offset<span class="op">:</span>(ref_m.offset<span class="op">+</span>barcode_length<span class="op">-</span><span class="fl">1</span>)</span>
<span id="cb11-40"><a href="#cb11-40" aria-hidden="true" tabindex="-1"></a>                barcode_seq <span class="op">=</span> aln_vec[<span class="fl">2</span>][extract_range]</span>
<span id="cb11-41"><a href="#cb11-41" aria-hidden="true" tabindex="-1"></a>                barcode_qual_char_seq <span class="op">=</span> qual_aln_vec[<span class="fl">2</span>][extract_range]</span>
<span id="cb11-42"><a href="#cb11-42" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> <span class="fu">all</span>(<span class="fu">split</span>(barcode_seq, <span class="st">""</span>) <span class="op">.âˆˆ</span> <span class="fu">Ref</span>(valid_dna_bases)) <span class="op">&amp;&amp;</span> <span class="fu">all</span>([<span class="fu">Int</span>(c) <span class="op">-</span> quality_scheme for c <span class="kw">in</span> barcode_qual_char_seq] <span class="op">.&gt;=</span> base_qual)</span>
<span id="cb11-43"><a href="#cb11-43" aria-hidden="true" tabindex="-1"></a>                    valid_xn_num_records <span class="op">+=</span> <span class="fl">1</span></span>
<span id="cb11-44"><a href="#cb11-44" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">push!</span>(cre_bc_vec, (ref_name, barcode_seq))</span>
<span id="cb11-45"><a href="#cb11-45" aria-hidden="true" tabindex="-1"></a>                    typical_aln_vec_dict[<span class="fu">string</span>(ref_name, <span class="st">":"</span>, barcode_seq)] <span class="op">=</span> aln_vec</span>
<span id="cb11-46"><a href="#cb11-46" aria-hidden="true" tabindex="-1"></a>                <span class="cf">end</span></span>
<span id="cb11-47"><a href="#cb11-47" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb11-48"><a href="#cb11-48" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb11-49"><a href="#cb11-49" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb11-50"><a href="#cb11-50" aria-hidden="true" tabindex="-1"></a>    <span class="fu">close</span>(reader)</span>
<span id="cb11-51"><a href="#cb11-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-52"><a href="#cb11-52" aria-hidden="true" tabindex="-1"></a>    cre_bc_gdf <span class="op">=</span> <span class="fu">groupby</span>(<span class="fu">DataFrame</span>(cre_bc_vec, [<span class="op">:</span>cre, <span class="op">:</span>barcode]), [<span class="op">:</span>cre, <span class="op">:</span>barcode])</span>
<span id="cb11-53"><a href="#cb11-53" aria-hidden="true" tabindex="-1"></a>    uniq_cre_bc_df <span class="op">=</span> <span class="fu">sort</span>(<span class="fu">combine</span>(cre_bc_gdf, nrow <span class="op">=&gt;</span> <span class="st">"num"</span>, proprow <span class="op">=&gt;</span> <span class="st">"prop"</span>), <span class="op">:</span>num, rev<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb11-54"><a href="#cb11-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-55"><a href="#cb11-55" aria-hidden="true" tabindex="-1"></a>    <span class="fu">open</span>(<span class="fu">replace</span>(bam_file, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>extract_cre_bc_pairs<span class="sc">.</span><span class="st">log"</span>), <span class="st">"w"</span>) <span class="cf">do</span> io</span>
<span id="cb11-56"><a href="#cb11-56" aria-hidden="true" tabindex="-1"></a>        <span class="fu">println</span>(io, <span class="fu">string</span>(</span>
<span id="cb11-57"><a href="#cb11-57" aria-hidden="true" tabindex="-1"></a>            <span class="st">"The number of records in total: "</span>, total_num_records, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb11-58"><a href="#cb11-58" aria-hidden="true" tabindex="-1"></a>            <span class="st">"The number of records with XN field: "</span>, xn_num_records, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb11-59"><a href="#cb11-59" aria-hidden="true" tabindex="-1"></a>            <span class="st">"The number of records with complete barcode: "</span>, complete_xn_num_records, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb11-60"><a href="#cb11-60" aria-hidden="true" tabindex="-1"></a>            <span class="st">"The number of records passing base and quality check: "</span>, valid_xn_num_records, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb11-61"><a href="#cb11-61" aria-hidden="true" tabindex="-1"></a>            <span class="st">"The number of records non-redundant: "</span>, <span class="fu">nrow</span>(uniq_cre_bc_df)</span>
<span id="cb11-62"><a href="#cb11-62" aria-hidden="true" tabindex="-1"></a>        ))</span>
<span id="cb11-63"><a href="#cb11-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb11-64"><a href="#cb11-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-65"><a href="#cb11-65" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> [uniq_cre_bc_df, typical_aln_vec_dict]</span>
<span id="cb11-66"><a href="#cb11-66" aria-hidden="true" tabindex="-1"></a><span class="kw">end</span></span>
<span id="cb11-67"><a href="#cb11-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-68"><a href="#cb11-68" aria-hidden="true" tabindex="-1"></a>ref_file <span class="op">=</span> <span class="st">"ref/mpra_ref.fa"</span></span>
<span id="cb11-69"><a href="#cb11-69" aria-hidden="true" tabindex="-1"></a>high_qual_bam_dir <span class="op">=</span> <span class="st">"high_qual_bam"</span></span>
<span id="cb11-70"><a href="#cb11-70" aria-hidden="true" tabindex="-1"></a>base_qual <span class="op">=</span> <span class="fl">20</span></span>
<span id="cb11-71"><a href="#cb11-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-72"><a href="#cb11-72" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in reference sequences</span></span>
<span id="cb11-73"><a href="#cb11-73" aria-hidden="true" tabindex="-1"></a>ref_dict <span class="op">=</span> <span class="fu">FASTAReader</span>(<span class="fu">open</span>(ref_file, <span class="st">"r"</span>)) <span class="cf">do</span> reader</span>
<span id="cb11-74"><a href="#cb11-74" aria-hidden="true" tabindex="-1"></a>    dict <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{String,String}</span>()</span>
<span id="cb11-75"><a href="#cb11-75" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> record <span class="kw">in</span> reader</span>
<span id="cb11-76"><a href="#cb11-76" aria-hidden="true" tabindex="-1"></a>        dict[<span class="fu">identifier</span>(record)] <span class="op">=</span> <span class="fu">sequence</span>(record)</span>
<span id="cb11-77"><a href="#cb11-77" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb11-78"><a href="#cb11-78" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> dict</span>
<span id="cb11-79"><a href="#cb11-79" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb11-80"><a href="#cb11-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-81"><a href="#cb11-81" aria-hidden="true" tabindex="-1"></a>bam_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(high_qual_bam_dir, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb11-82"><a href="#cb11-82" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> bam_file <span class="kw">in</span> bam_files</span>
<span id="cb11-83"><a href="#cb11-83" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Extract CRE-Barcode pairs</span></span>
<span id="cb11-84"><a href="#cb11-84" aria-hidden="true" tabindex="-1"></a>    cre_bc_res <span class="op">=</span> <span class="fu">extract_cre_bc_pairs</span>(bam_file, ref_dict)</span>
<span id="cb11-85"><a href="#cb11-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-86"><a href="#cb11-86" aria-hidden="true" tabindex="-1"></a>    CSV.<span class="fu">write</span>(<span class="fu">replace</span>(bam_file, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>uniq_cre_bc_pairs<span class="sc">.</span><span class="st">tsv"</span>),</span>
<span id="cb11-87"><a href="#cb11-87" aria-hidden="true" tabindex="-1"></a>        cre_bc_res[<span class="fl">1</span>]; delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>, writeheader<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb11-88"><a href="#cb11-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-89"><a href="#cb11-89" aria-hidden="true" tabindex="-1"></a>    <span class="co"># obj = open(jls_file, "r") do io</span></span>
<span id="cb11-90"><a href="#cb11-90" aria-hidden="true" tabindex="-1"></a>    <span class="co">#     deserialize(io)</span></span>
<span id="cb11-91"><a href="#cb11-91" aria-hidden="true" tabindex="-1"></a>    <span class="co"># end</span></span>
<span id="cb11-92"><a href="#cb11-92" aria-hidden="true" tabindex="-1"></a>    <span class="fu">open</span>(<span class="fu">replace</span>(bam_file, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>typical_cre_bc_aligned_sequences<span class="sc">.</span><span class="st">jls"</span>), <span class="st">"w"</span>) <span class="cf">do</span> io</span>
<span id="cb11-93"><a href="#cb11-93" aria-hidden="true" tabindex="-1"></a>        <span class="fu">serialize</span>(io, cre_bc_res[<span class="fl">2</span>])</span>
<span id="cb11-94"><a href="#cb11-94" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb11-95"><a href="#cb11-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-96"><a href="#cb11-96" aria-hidden="true" tabindex="-1"></a>    rand_keys <span class="op">=</span> <span class="fu">rand</span>(<span class="fu">keys</span>(cre_bc_res[<span class="fl">2</span>]), <span class="fl">100</span>)</span>
<span id="cb11-97"><a href="#cb11-97" aria-hidden="true" tabindex="-1"></a>    rand_dict <span class="op">=</span> <span class="fu">Dict</span>(k <span class="op">=&gt;</span> cre_bc_res[<span class="fl">2</span>][k] <span class="cf">for</span> k <span class="kw">in</span> rand_keys)</span>
<span id="cb11-98"><a href="#cb11-98" aria-hidden="true" tabindex="-1"></a>    YRUtils.BioUtils.<span class="fu">show_align</span>(rand_dict,</span>
<span id="cb11-99"><a href="#cb11-99" aria-hidden="true" tabindex="-1"></a>        <span class="fu">replace</span>(bam_file, <span class="st">r"</span><span class="sc">\.</span><span class="st">bam</span><span class="sc">$</span><span class="st">" =&gt; ".</span>typical_cre_bc_aligned_sequences<span class="sc">.</span><span class="st">100</span><span class="sc">.</span><span class="st">html"</span>);</span>
<span id="cb11-100"><a href="#cb11-100" aria-hidden="true" tabindex="-1"></a>        wrap_width<span class="op">=</span><span class="fl">120</span>)</span>
<span id="cb11-101"><a href="#cb11-101" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="quality-check" class="level3" data-number="1.10">
<h3 data-number="1.10" class="anchored" data-anchor-id="quality-check"><span class="header-section-number">1.10</span> Quality check</h3>
<pre class="{r}"><code>library(vroom)
library(tidyverse)
library(YRUtils)
library(ggprism)

input_dir &lt;- "high_qual_bam"

cre_bc_files &lt;- list.files(input_dir, pattern = "\\.tsv$", full.names = TRUE, recursive = FALSE)
for (cre_bc_file in cre_bc_files) {
    cre_bc_df &lt;- vroom(cre_bc_file) %&gt;%
        select(cre, barcode) %&gt;%
        distinct()

    cre_count_df &lt;- count(cre_bc_df, cre) %&gt;%
        mutate(type = if_else(str_detect(cre, "^CTL"), "CTL", "CRE")) %&gt;%
        rename(cre_bc = cre)
    bc_count_df &lt;- count(cre_bc_df, barcode) %&gt;%
        mutate(type = "BC") %&gt;%
        rename(cre_bc = barcode)
    count_df &lt;- bind_rows(cre_count_df, bc_count_df)

    cre_quantiles &lt;- quantile(count_df$n[count_df$type %in% c("CTL", "CRE")], probs = seq(0, 1, 0.1))
    cre_only_quantiles &lt;- quantile(count_df$n[count_df$type == "CRE"], probs = seq(0, 1, 0.1))
    ctl_only_quantiles &lt;- quantile(count_df$n[count_df$type == "CTL"], probs = seq(0, 1, 0.1))
    bc_quantiles &lt;- quantile(count_df$n[count_df$type == "BC"], probs = seq(0, 1, 0.1))
    type_nums &lt;- table(count_df$type)

    paste0(
        "1. CRE/CTL quantiles (", type_nums["CTL"] + type_nums["CRE"], "): \n",
        paste0(paste0(names(cre_quantiles), "\t", cre_quantiles), collapse = "\n"), "\n\n",
        "2. CRE only quantiles (", type_nums["CRE"], "): \n",
        paste0(paste0(names(cre_only_quantiles), "\t", cre_only_quantiles), collapse = "\n"), "\n\n",
        "3. CTL only quantiles (", type_nums["CTL"], "): \n",
        paste0(paste0(names(ctl_only_quantiles), "\t", ctl_only_quantiles), collapse = "\n"), "\n\n",
        "4. BC quantiles (", type_nums["BC"], "): \n",
        paste0(paste0(names(bc_quantiles), "\t", bc_quantiles), collapse = "\n")
    ) %&gt;% vroom_write_lines(file = gsub("\\.tsv$", "\\.quantiles.txt", cre_bc_file))

    p &lt;- ggplot(count_df, aes(type, log2(n), fill = type, color = type)) +
        geom_violin(scale = "width", alpha = 0.25, trim = TRUE) +
        geom_boxplot(width = 0.2, outliers = FALSE, alpha = 0.25) +
        scale_y_continuous(expand = expansion(mult = c(0.05, 0))) +
        labs(
            x = "Sequence Type",
            y = "log2(Count)"
        ) +
        theme_prism(base_size = 20, base_family = "Arial", border = FALSE) +
        theme(legend.position = "none")
    ppreview(p, file = gsub("\\.tsv$", "\\.violin.pdf", cre_bc_file))
}</code></pre>
</section>
<section id="merge-all-cre-barcode-pairs" class="level3" data-number="1.11">
<h3 data-number="1.11" class="anchored" data-anchor-id="merge-all-cre-barcode-pairs"><span class="header-section-number">1.11</span> Merge all CRE-Barcode pairs</h3>
<pre class="{r}"><code>library(tidyverse)
library(patchwork)
library(vroom)
library(YRUtils)

# accept only one parameter (i.e., sample file names)
# and return the sample IDs derived from the sample file names
extract_sample_id &lt;- function(x) {
    gsub(
        "^/data/users/dell/mpra/link_barcode_to_cre/|high_qual_bam/|\\.chr_srt\\.uniq_cre_bc_pairs\\.tsv$",
        "", file
    )
}

input_dirs &lt;- c(
    "/data/users/dell/mpra/link_barcode_to_cre/enzyme_v20241230/high_qual_bam",
    "/data/users/dell/mpra/link_barcode_to_cre/enzyme_v20231027/high_qual_bam",
    "/data/users/dell/mpra/link_barcode_to_cre/pcr_v20230922/high_qual_bam",
    "/data/users/dell/mpra/link_barcode_to_cre/enzyme_v20250430/high_qual_bam"
)
output_dir &lt;- "/data/users/dell/mpra/link_barcode_to_cre/final_result"
stats_columns &lt;- c("id", "cre", "barcode")

files &lt;- list.files(input_dirs, pattern = "\\.tsv$", recursive = FALSE, full.names = TRUE)
cre_bc_df &lt;- tibble()
for (file in files) {
    cre_bc_df &lt;- bind_rows(
        cre_bc_df,
        vroom(file) %&gt;%
            select(cre, barcode) %&gt;%
            mutate(
                id = paste0(cre, ":", barcode),
                sample = extract_sample_id(file)
            )
    )
}

# check overlapping ratios across samples using unique CRE-Barcode pairs/CREs/barcodes
layout &lt;- "
#A
BC
"
for (stats_column in stats_columns) {
    ls &lt;- lapply(split(cre_bc_df, cre_bc_df$sample), function(df) {
        unique(na.omit(df[[stats_column]]))
    })

    ls_copy &lt;- ls
    df &lt;- data.frame()
    for (k in length(ls_copy):2) {
        name_combn_vec &lt;- combn(names(ls_copy), k) |&gt;
            apply(2, function(name_combn) {
                paste0(name_combn, collapse = "&lt;*:*&gt;")
            })

        intersect_num_vec &lt;- c()
        for (name_combn in name_combn_vec) {
            name_vec &lt;- strsplit(name_combn, "&lt;*:*&gt;", fixed = TRUE)[[1]]
            intersect_items &lt;- Reduce(intersect, ls_copy[name_vec])
            intersect_num_vec &lt;- c(intersect_num_vec, length(intersect_items))
            for (name in name_vec) {
                ls_copy[[name]] &lt;- ls_copy[[name]][!(ls_copy[[name]] %in% intersect_items)]
            }
        }

        df &lt;- rbind(
            df,
            data.frame(
                combn_k = k,
                combn = name_combn_vec,
                intersect_num = intersect_num_vec
            )
        )
    }
    df &lt;- rbind(
        df,
        data.frame(
            combn_k = 1,
            combn = names(ls_copy),
            intersect_num = sapply(ls_copy, length)
        )
    )
    row.names(df) &lt;- NULL

    for (name in names(ls)) {
        df[[name]] &lt;- grepl(name, df$combn, fixed = TRUE)
    }

    df &lt;- pivot_longer(
        df,
        cols = -all_of(c("combn_k", "combn", "intersect_num")),
        names_to = "sample", values_to = "is_match"
    ) |&gt;
        filter((intersect_num &gt; 0) &amp; is_match)
    df &lt;- df |&gt;
        arrange(combn_k, desc(intersect_num)) |&gt;
        mutate(
            sample = factor(sample, levels = rev(unique(sample))),
            combn = factor(combn, levels = unique(combn))
        )

    set_size_df &lt;- data.frame(
        sample = names(ls),
        num = sapply(ls, length)
    ) |&gt;
        mutate(sample = factor(sample, levels = levels(df$sample)))
    row.names(set_size_df) &lt;- NULL

    mlog10 &lt;- scales::new_transform(
        name = "mlog10",
        transform = function(x) -log10(x),
        inverse = function(x) 10^-x,
        breaks = scales::breaks_log(),
        format = scales::label_scientific(),
        domain = c(.Machine$double.xmin, .Machine$double.xmax)
    )

    p_set_size_bar &lt;- ggplot(set_size_df, aes(num, sample)) +
        geom_bar(stat = "identity") +
        scale_y_discrete(position = "right") +
        scale_x_continuous(
            expand = expansion(0),
            transform = mlog10
        ) +
        labs(x = NULL, y = NULL) +
        theme_minimal() +
        theme(
            axis.text.y = element_blank(),
            axis.ticks.y = element_blank(),
            axis.ticks.x = element_line()
        )

    p_set_intersection_point &lt;- ggplot(df, aes(combn, sample)) +
        geom_point() +
        geom_line(aes(group = combn)) +
        scale_y_discrete(position = "right") +
        labs(x = NULL, y = NULL) +
        theme_bw() +
        theme(
            axis.text.x = element_blank(),
            axis.ticks.x = element_blank()
        )

    p_set_intersection_bar &lt;- ggplot(
        distinct(select(df, combn, intersect_num)),
        aes(combn, intersect_num)
    ) +
        geom_bar(stat = "identity") +
        scale_y_continuous(
            expand = expansion(0),
            transform = "log10",
            position = "right",
            labels = scales::label_scientific()
        ) +
        labs(x = NULL, y = NULL) +
        theme_bw() +
        theme(
            axis.text.x = element_blank(),
            axis.ticks.x = element_blank()
        )

    p &lt;- p_set_intersection_bar + p_set_size_bar + p_set_intersection_point +
        plot_layout(
            design = layout,
            widths = c(1, 3),
            heights = c(2, 1)
        )
    ppreview(p, file = file.path(output_dir, paste0(stats_column, "_overlapping_ratios.upset.pdf")))

    stats_df &lt;- df |&gt;
        select(combn_k, combn, intersect_num) |&gt;
        distinct() |&gt;
        reframe(
            combn_k = combn_k,
            combn = combn,
            intersect_num = intersect_num,
            intersect_num_percent = intersect_num / sum(intersect_num) * 100
        ) |&gt;
        arrange(combn_k, intersect_num)
    vroom_write(stats_df, file = file.path(output_dir, paste0(stats_column, "_overlapping_ratios.stats.txt")), col_names = TRUE, append = FALSE)
}

# output merged CRE-Barcode pairs
merged_cre_bc_df &lt;- cre_bc_df %&gt;%
    select(cre, barcode) %&gt;%
    distinct()
one_cre_barocdes &lt;- merged_cre_bc_df %&gt;%
    group_by(barcode) %&gt;%
    count() %&gt;%
    filter(n == 1) %&gt;%
    pull(barcode) %&gt;%
    unique()
one_cre_bc_df &lt;- merged_cre_bc_df %&gt;%
    filter(barcode %in% one_cre_barocdes) %&gt;%
    distinct()

vroom_write(merged_cre_bc_df, file = file.path(output_dir, "redundant_cre_bc_pairs.tsv"), col_names = TRUE, append = FALSE)
vroom_write(one_cre_bc_df, file = file.path(output_dir, "non_redundant_cre_bc_pairs.tsv"), col_names = TRUE, append = FALSE)</code></pre>
</section>
</section>
<section id="count-barcodes" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="count-barcodes"><span class="header-section-number">2</span> Count barcodes</h2>
<p>Before running any of the following steps, you should rename your FASTQ files in this form: <code>ID_(RNA|DNA)_repN[_partN].R[123].(fq|fastq).gz</code> (<code>ID</code> can only contain <code>[a-zA-Z0-9]</code>; <code>N</code> can only contain <code>[0-9]</code>).</p>
<div id="060da5df" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>work_dir <span class="op">=</span> <span class="st">"/data/users/dell/mpra/count_barcode/15bp_v20240627"</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(work_dir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="{r}"><code>work_dir &lt;- "/data/users/dell/mpra/count_barcode/15bp_v20240627"

setwd(work_dir)</code></pre>
<section id="md5sum-check-over-raw-fastq-files-1" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="md5sum-check-over-raw-fastq-files-1"><span class="header-section-number">2.1</span> MD5SUM check over raw FASTQ files</h3>
<div id="6a330950" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>raw_fastq_dir <span class="op">=</span> <span class="st">"raw_fastq"</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>md5_file <span class="op">=</span> <span class="st">"md5.txt"</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>md5_check_file <span class="op">=</span> <span class="st">"md5_check.txt"</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(raw_fastq_dir)</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>YRUtils.BaseUtils.<span class="fu">md5_check</span>(md5_file, md5_check_file)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cd</span>(work_dir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fastqc-over-raw-fastq-files-1" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="fastqc-over-raw-fastq-files-1"><span class="header-section-number">2.2</span> FASTQC over raw FASTQ files</h3>
<div id="f2c87a29" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">YRUtils</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>raw_fastq_dir <span class="op">=</span> <span class="st">"raw_fastq"</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>raw_fastqc_dir <span class="op">=</span> <span class="st">"raw_fastqc"</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(raw_fastqc_dir)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>raw_fastq_files <span class="op">=</span> YRUtils.BaseUtils.<span class="fu">list_files</span>(raw_fastq_dir, <span class="st">r"</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fastq</span><span class="ch">|</span><span class="st">fq</span><span class="ch">)</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">", </span>recursive<span class="st">=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>YRUtils.BioUtils.<span class="fu">fastqc</span>(raw_fastq_files, raw_fastqc_dir;</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>    fastqc_options<span class="op">=</span><span class="st">"--threads 4"</span>, multiqc_options<span class="op">=</span><span class="st">"--zip-data-dir"</span>, num_jobs<span class="op">=</span><span class="fl">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="count-barcodes-1" class="level3" data-number="2.3">
<h3 data-number="2.3" class="anchored" data-anchor-id="count-barcodes-1"><span class="header-section-number">2.3</span> Count barcodes</h3>
<section id="solution-1" class="level4" data-number="2.3.1">
<h4 data-number="2.3.1" class="anchored" data-anchor-id="solution-1"><span class="header-section-number">2.3.1</span> Solution 1</h4>
<div id="47a0a501" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">FASTX</span>, <span class="bu">DataFrames</span>, <span class="bu">CSV</span>, <span class="bu">YRUtils</span>, <span class="bu">CodecZlib</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>raw_fastq_dir <span class="op">=</span> <span class="st">"raw_fastq"</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>cre_bc_file <span class="op">=</span> <span class="st">"/data/users/dell/mpra/link_barcode_to_cre/final_result/redundant_cre_bc_pairs.tsv"</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>raw_bc_umi_dir <span class="op">=</span> <span class="st">"raw_bc_umi"</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>base_qual <span class="op">=</span> <span class="fl">20</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>seqkit_nthreads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(raw_bc_umi_dir)</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>cre_bc_df <span class="op">=</span> CSV.<span class="fu">read</span>(cre_bc_file, DataFrame; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>uniq_barcodes <span class="op">=</span> <span class="fu">Set</span>(cre_bc_df[<span class="op">:</span>, <span class="op">:</span>barcode])</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>raw_fastq_file_name_pattern <span class="op">=</span> <span class="st">r"</span><span class="sc">.+</span><span class="st">/</span><span class="ch">(?</span><span class="st">&lt;id&gt;</span><span class="ch">[</span><span class="st">a-zA-Z0-9</span><span class="ch">]</span><span class="sc">+</span><span class="ch">)</span><span class="st">_</span><span class="ch">(?</span><span class="st">&lt;type&gt;RNA</span><span class="ch">|</span><span class="st">DNA</span><span class="ch">)</span><span class="st">_</span><span class="ch">(?</span><span class="st">&lt;rep&gt;rep</span><span class="ch">[</span><span class="st">0-9</span><span class="ch">]</span><span class="sc">+</span><span class="ch">)(</span><span class="st">_</span><span class="ch">(?</span><span class="st">&lt;tech&gt;part</span><span class="ch">[</span><span class="st">0-9</span><span class="ch">]</span><span class="sc">+</span><span class="ch">))</span><span class="sc">?\.</span><span class="ch">(?</span><span class="st">&lt;read&gt;R</span><span class="ch">[</span><span class="st">123</span><span class="ch">])</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fq</span><span class="ch">|</span><span class="st">fastq</span><span class="ch">)</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">"</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>raw_fastq_files<span class="st"> = YRUtils</span><span class="sc">.</span><span class="st">BioUtils</span><span class="sc">.</span><span class="st">list_files</span><span class="ch">(</span><span class="st">raw_fastq_dir, raw_fastq_file_name_pattern; recursive=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>raw_fastq_nums <span class="op">=</span> YRUtils.BioUtils.<span class="fu">fq_num</span>(raw_fastq_files, seqkit_nthreads)</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>ms <span class="op">=</span> <span class="fu">match</span>.(raw_fastq_file_name_pattern, raw_fastq_files)</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>metadata_vec <span class="op">=</span> <span class="fu">Vector</span><span class="dt">{NTuple{7,String}}</span>(<span class="cn">undef</span>, <span class="fu">length</span>(ms))</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(ms)</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>    metadata_vec[i] <span class="op">=</span> (ms[i].match, ms[i][<span class="st">"id"</span>], ms[i][<span class="st">"type"</span>], ms[i][<span class="st">"rep"</span>], ms[i][<span class="st">"tech"</span>], ms[i][<span class="st">"read"</span>], <span class="fu">string</span>(raw_fastq_nums[ms[i].match]))</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> <span class="fu">unique</span>(<span class="fu">DataFrame</span>(metadata_vec, [<span class="op">:</span>file, <span class="op">:</span>id, <span class="op">:</span>type, <span class="op">:</span>rep, <span class="op">:</span>tech, <span class="op">:</span>read, <span class="op">:</span>num_seqs]))</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> <span class="fu">transform</span>(</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>    df,</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>    [<span class="op">:</span>id, <span class="op">:</span>type, <span class="op">:</span>rep, <span class="op">:</span>tech] <span class="op">=&gt;</span> <span class="fu">ByRow</span>((id, <span class="kw">type</span>, rep, tech) <span class="op">-&gt;</span> <span class="fu">join</span>([id, type, rep, tech], <span class="st">"_"</span>)) <span class="op">=&gt;</span> <span class="op">:</span>tech_sample,</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>    [<span class="op">:</span>id, <span class="op">:</span>type, <span class="op">:</span>rep] <span class="op">=&gt;</span> <span class="fu">ByRow</span>((id, <span class="kw">type</span>, rep) <span class="op">-&gt;</span> <span class="fu">join</span>([id, type, rep], <span class="st">"_"</span>)) <span class="op">=&gt;</span> <span class="op">:</span>rep_sample,</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>    [<span class="op">:</span>id, <span class="op">:</span>type] <span class="op">=&gt;</span> <span class="fu">ByRow</span>((id, <span class="kw">type</span>) <span class="op">-&gt;</span> <span class="fu">join</span>([id, type], <span class="st">"_"</span>)) <span class="op">=&gt;</span> <span class="op">:</span>type_sample</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a>tech_gdf <span class="op">=</span> <span class="fu">groupby</span>(df, <span class="op">:</span>tech_sample)</span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(<span class="fu">joinpath</span>(raw_bc_umi_dir, <span class="st">"fq_metadata.tsv"</span>), df; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a>df_dict <span class="op">=</span> <span class="fu">Dict</span>(<span class="fu">unique</span>(read_df[<span class="op">:</span>, <span class="op">:</span>tech_sample])[<span class="fl">1</span>] <span class="op">=&gt;</span> <span class="fu">DataFrame</span>() <span class="cf">for</span> read_df <span class="kw">in</span> tech_gdf)</span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> read_df <span class="kw">in</span> tech_gdf</span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Read in reads</span></span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a>    read_dict <span class="op">=</span> <span class="fu">Dict</span>(read_type <span class="op">=&gt;</span> <span class="fu">Vector</span><span class="dt">{Tuple{String,String,String}}</span>(<span class="cn">undef</span>, <span class="fu">parse</span>(<span class="dt">Int64</span>, num_seqs)) <span class="cf">for</span> (read_type, num_seqs) <span class="kw">in</span> <span class="fu">collect</span>(<span class="fu">zip</span>(read_df[<span class="op">:</span>, <span class="op">:</span>read], read_df[<span class="op">:</span>, <span class="op">:</span>num_seqs])))</span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true" tabindex="-1"></a>    <span class="bu">Threads</span>.<span class="pp">@threads</span> <span class="cf">for</span> (read_type, num_seqs, fq_file) <span class="kw">in</span> <span class="fu">collect</span>(<span class="fu">zip</span>(read_df[<span class="op">:</span>, <span class="op">:</span>read], read_df[<span class="op">:</span>, <span class="op">:</span>num_seqs], read_df[<span class="op">:</span>, <span class="op">:</span>file]))</span>
<span id="cb18-36"><a href="#cb18-36" aria-hidden="true" tabindex="-1"></a>        <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"start parsing "</span>, fq_file, <span class="st">" with read type "</span>, read_type, <span class="st">" and the number of sequences "</span>, num_seqs, <span class="st">" ..."</span>)</span>
<span id="cb18-37"><a href="#cb18-37" aria-hidden="true" tabindex="-1"></a>        <span class="fu">FASTQReader</span>(<span class="fu">GzipDecompressorStream</span>(<span class="fu">open</span>(fq_file))) <span class="cf">do</span> reader</span>
<span id="cb18-38"><a href="#cb18-38" aria-hidden="true" tabindex="-1"></a>            record <span class="op">=</span> FASTQ.<span class="fu">Record</span>()</span>
<span id="cb18-39"><a href="#cb18-39" aria-hidden="true" tabindex="-1"></a>            i <span class="op">=</span> <span class="fl">0</span></span>
<span id="cb18-40"><a href="#cb18-40" aria-hidden="true" tabindex="-1"></a>            <span class="cf">while</span> !<span class="fu">eof</span>(reader)</span>
<span id="cb18-41"><a href="#cb18-41" aria-hidden="true" tabindex="-1"></a>                i <span class="op">+=</span> <span class="fl">1</span></span>
<span id="cb18-42"><a href="#cb18-42" aria-hidden="true" tabindex="-1"></a>                <span class="fu">empty!</span>(record)</span>
<span id="cb18-43"><a href="#cb18-43" aria-hidden="true" tabindex="-1"></a>                <span class="fu">read!</span>(reader, record)</span>
<span id="cb18-44"><a href="#cb18-44" aria-hidden="true" tabindex="-1"></a>                read_dict[read_type][i] <span class="op">=</span> (<span class="fu">identifier</span>(record), <span class="fu">sequence</span>(record), <span class="fu">join</span>(<span class="fu">collect</span>(<span class="fu">quality_scores</span>(record)), <span class="st">"/"</span>))</span>
<span id="cb18-45"><a href="#cb18-45" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb18-46"><a href="#cb18-46" aria-hidden="true" tabindex="-1"></a>            <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"read in "</span>, i, <span class="st">" sequences in total for "</span>, fq_file)</span>
<span id="cb18-47"><a href="#cb18-47" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> i <span class="op">!=</span> <span class="fu">parse</span>(<span class="dt">Int64</span>, num_seqs)</span>
<span id="cb18-48"><a href="#cb18-48" aria-hidden="true" tabindex="-1"></a>                <span class="pp">@error</span> <span class="fu">string</span>(<span class="st">"parsing file "</span>, fq_file, <span class="st">" failed!"</span>)</span>
<span id="cb18-49"><a href="#cb18-49" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb18-50"><a href="#cb18-50" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb18-51"><a href="#cb18-51" aria-hidden="true" tabindex="-1"></a>        <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"parsing "</span>, fq_file, <span class="st">" with read type "</span>, read_type, <span class="st">" and the number of sequences "</span>, num_seqs, <span class="st">" done!"</span>)</span>
<span id="cb18-52"><a href="#cb18-52" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb18-53"><a href="#cb18-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-54"><a href="#cb18-54" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Count qualified barcodes and their UMIs</span></span>
<span id="cb18-55"><a href="#cb18-55" aria-hidden="true" tabindex="-1"></a>    len_vec <span class="op">=</span> <span class="fu">length</span>.(<span class="fu">values</span>(read_dict))</span>
<span id="cb18-56"><a href="#cb18-56" aria-hidden="true" tabindex="-1"></a>    <span class="co"># The three files should have the same number of lines</span></span>
<span id="cb18-57"><a href="#cb18-57" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="fu">length</span>(<span class="fu">unique</span>(len_vec)) <span class="op">==</span> <span class="fl">1</span></span>
<span id="cb18-58"><a href="#cb18-58" aria-hidden="true" tabindex="-1"></a>        bc_umi_vec <span class="op">=</span> <span class="fu">Vector</span><span class="dt">{Tuple{String,String,String,Vararg{Bool,9}}}</span>(<span class="cn">undef</span>, len_vec[<span class="fl">1</span>])</span>
<span id="cb18-59"><a href="#cb18-59" aria-hidden="true" tabindex="-1"></a>        <span class="bu">Threads</span>.<span class="pp">@threads</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>len_vec[<span class="fl">1</span>]</span>
<span id="cb18-60"><a href="#cb18-60" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Read IDs should be identical across R1, R2, and R3</span></span>
<span id="cb18-61"><a href="#cb18-61" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> read_dict[<span class="st">"R1"</span>][i][<span class="fl">1</span>] <span class="op">==</span> read_dict[<span class="st">"R2"</span>][i][<span class="fl">1</span>] <span class="op">==</span> read_dict[<span class="st">"R3"</span>][i][<span class="fl">1</span>]</span>
<span id="cb18-62"><a href="#cb18-62" aria-hidden="true" tabindex="-1"></a>                bc_umi_vec[i] <span class="op">=</span> (</span>
<span id="cb18-63"><a href="#cb18-63" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># Read 1</span></span>
<span id="cb18-64"><a href="#cb18-64" aria-hidden="true" tabindex="-1"></a>                    read_dict[<span class="st">"R1"</span>][i][<span class="fl">2</span>],</span>
<span id="cb18-65"><a href="#cb18-65" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># Read 2</span></span>
<span id="cb18-66"><a href="#cb18-66" aria-hidden="true" tabindex="-1"></a>                    read_dict[<span class="st">"R2"</span>][i][<span class="fl">2</span>],</span>
<span id="cb18-67"><a href="#cb18-67" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># UMI</span></span>
<span id="cb18-68"><a href="#cb18-68" aria-hidden="true" tabindex="-1"></a>                    read_dict[<span class="st">"R3"</span>][i][<span class="fl">2</span>],</span>
<span id="cb18-69"><a href="#cb18-69" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># Read sequences should only contain A, T, C, and G across R1, R2, and R3</span></span>
<span id="cb18-70"><a href="#cb18-70" aria-hidden="true" tabindex="-1"></a>                    !<span class="fu">occursin</span>(<span class="st">"N"</span>, <span class="fu">string</span>(read_dict[<span class="st">"R1"</span>][i][<span class="fl">2</span>], read_dict[<span class="st">"R2"</span>][i][<span class="fl">2</span>], read_dict[<span class="st">"R3"</span>][i][<span class="fl">2</span>])),</span>
<span id="cb18-71"><a href="#cb18-71" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># All base qualities &gt;= base_qual across R1, R2, and R3</span></span>
<span id="cb18-72"><a href="#cb18-72" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">all</span>(<span class="fu">parse</span>.(<span class="dt">Int</span>, <span class="fu">split</span>(<span class="fu">string</span>(read_dict[<span class="st">"R1"</span>][i][<span class="fl">3</span>], <span class="st">"/"</span>, read_dict[<span class="st">"R2"</span>][i][<span class="fl">3</span>], <span class="st">"/"</span>, read_dict[<span class="st">"R3"</span>][i][<span class="fl">3</span>]), <span class="st">"/"</span>)) <span class="op">.&gt;=</span> base_qual),</span>
<span id="cb18-73"><a href="#cb18-73" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># Read 1 and read 2 should be reverse and complementary</span></span>
<span id="cb18-74"><a href="#cb18-74" aria-hidden="true" tabindex="-1"></a>                    YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(read_dict[<span class="st">"R1"</span>][i][<span class="fl">2</span>]) <span class="op">==</span> read_dict[<span class="st">"R2"</span>][i][<span class="fl">2</span>],</span>
<span id="cb18-75"><a href="#cb18-75" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># Either read 1 or read 2 should be in the barcode library (not both in theory)</span></span>
<span id="cb18-76"><a href="#cb18-76" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># Read 1 in the barcode library?</span></span>
<span id="cb18-77"><a href="#cb18-77" aria-hidden="true" tabindex="-1"></a>                    read_dict[<span class="st">"R1"</span>][i][<span class="fl">2</span>] <span class="kw">in</span> uniq_barcodes,</span>
<span id="cb18-78"><a href="#cb18-78" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># Read 2 in the barcode library?</span></span>
<span id="cb18-79"><a href="#cb18-79" aria-hidden="true" tabindex="-1"></a>                    read_dict[<span class="st">"R2"</span>][i][<span class="fl">2</span>] <span class="kw">in</span> uniq_barcodes,</span>
<span id="cb18-80"><a href="#cb18-80" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># The reverse sequence of read 1 in the barcode library?</span></span>
<span id="cb18-81"><a href="#cb18-81" aria-hidden="true" tabindex="-1"></a>                    YRUtils.BioUtils.<span class="fu">rev_seq</span>(read_dict[<span class="st">"R1"</span>][i][<span class="fl">2</span>]) <span class="kw">in</span> uniq_barcodes,</span>
<span id="cb18-82"><a href="#cb18-82" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># The complementary sequence of read 1 in the barcode library?</span></span>
<span id="cb18-83"><a href="#cb18-83" aria-hidden="true" tabindex="-1"></a>                    YRUtils.BioUtils.<span class="fu">com_dna_seq</span>(read_dict[<span class="st">"R1"</span>][i][<span class="fl">2</span>]) <span class="kw">in</span> uniq_barcodes,</span>
<span id="cb18-84"><a href="#cb18-84" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># The reverse sequence of read 2 in the barcode library?</span></span>
<span id="cb18-85"><a href="#cb18-85" aria-hidden="true" tabindex="-1"></a>                    YRUtils.BioUtils.<span class="fu">rev_seq</span>(read_dict[<span class="st">"R2"</span>][i][<span class="fl">2</span>]) <span class="kw">in</span> uniq_barcodes,</span>
<span id="cb18-86"><a href="#cb18-86" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># The complementary sequence of read 2 in the barcode library?</span></span>
<span id="cb18-87"><a href="#cb18-87" aria-hidden="true" tabindex="-1"></a>                    YRUtils.BioUtils.<span class="fu">com_dna_seq</span>(read_dict[<span class="st">"R2"</span>][i][<span class="fl">2</span>]) <span class="kw">in</span> uniq_barcodes</span>
<span id="cb18-88"><a href="#cb18-88" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb18-89"><a href="#cb18-89" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span></span>
<span id="cb18-90"><a href="#cb18-90" aria-hidden="true" tabindex="-1"></a>                <span class="pp">@error</span> <span class="st">"read IDs are not identical across R1, R2, and R3"</span></span>
<span id="cb18-91"><a href="#cb18-91" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb18-92"><a href="#cb18-92" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb18-93"><a href="#cb18-93" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span></span>
<span id="cb18-94"><a href="#cb18-94" aria-hidden="true" tabindex="-1"></a>        <span class="pp">@error</span> <span class="st">"length(R1) == length(R2) == length(R3) is not true"</span></span>
<span id="cb18-95"><a href="#cb18-95" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb18-96"><a href="#cb18-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-97"><a href="#cb18-97" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Write statistics</span></span>
<span id="cb18-98"><a href="#cb18-98" aria-hidden="true" tabindex="-1"></a>    col4, col5, col6, col7, col8, col9, col10, col11, col12 <span class="op">=</span> (</span>
<span id="cb18-99"><a href="#cb18-99" aria-hidden="true" tabindex="-1"></a>        <span class="fu">getindex</span>.(bc_umi_vec, <span class="fl">4</span>),</span>
<span id="cb18-100"><a href="#cb18-100" aria-hidden="true" tabindex="-1"></a>        <span class="fu">getindex</span>.(bc_umi_vec, <span class="fl">5</span>),</span>
<span id="cb18-101"><a href="#cb18-101" aria-hidden="true" tabindex="-1"></a>        <span class="fu">getindex</span>.(bc_umi_vec, <span class="fl">6</span>),</span>
<span id="cb18-102"><a href="#cb18-102" aria-hidden="true" tabindex="-1"></a>        <span class="fu">getindex</span>.(bc_umi_vec, <span class="fl">7</span>),</span>
<span id="cb18-103"><a href="#cb18-103" aria-hidden="true" tabindex="-1"></a>        <span class="fu">getindex</span>.(bc_umi_vec, <span class="fl">8</span>),</span>
<span id="cb18-104"><a href="#cb18-104" aria-hidden="true" tabindex="-1"></a>        <span class="fu">getindex</span>.(bc_umi_vec, <span class="fl">9</span>),</span>
<span id="cb18-105"><a href="#cb18-105" aria-hidden="true" tabindex="-1"></a>        <span class="fu">getindex</span>.(bc_umi_vec, <span class="fl">10</span>),</span>
<span id="cb18-106"><a href="#cb18-106" aria-hidden="true" tabindex="-1"></a>        <span class="fu">getindex</span>.(bc_umi_vec, <span class="fl">11</span>),</span>
<span id="cb18-107"><a href="#cb18-107" aria-hidden="true" tabindex="-1"></a>        <span class="fu">getindex</span>.(bc_umi_vec, <span class="fl">12</span>)</span>
<span id="cb18-108"><a href="#cb18-108" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb18-109"><a href="#cb18-109" aria-hidden="true" tabindex="-1"></a>    <span class="fu">open</span>(<span class="fu">joinpath</span>(raw_bc_umi_dir, <span class="st">"fq_read_stats.txt"</span>), <span class="st">"a"</span>) <span class="cf">do</span> io</span>
<span id="cb18-110"><a href="#cb18-110" aria-hidden="true" tabindex="-1"></a>        stat_str <span class="op">=</span> <span class="fu">string</span>(</span>
<span id="cb18-111"><a href="#cb18-111" aria-hidden="true" tabindex="-1"></a>            <span class="st">"==&gt; "</span>, <span class="fu">unique</span>(read_df[<span class="op">:</span>, <span class="op">:</span>tech_sample])[<span class="fl">1</span>], <span class="st">" &lt;==</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb18-112"><a href="#cb18-112" aria-hidden="true" tabindex="-1"></a>            <span class="st">"1. The number of reads in total: "</span>, len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n\n</span><span class="st">"</span>,</span>
<span id="cb18-113"><a href="#cb18-113" aria-hidden="true" tabindex="-1"></a>            <span class="st">"2. % reads without Ns ("</span>, <span class="fu">sum</span>(col4), <span class="st">"): "</span>, <span class="fu">sum</span>(col4) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb18-114"><a href="#cb18-114" aria-hidden="true" tabindex="-1"></a>            <span class="st">"3. % reads with base qualities &gt;= "</span>, base_qual, <span class="st">" ("</span>, <span class="fu">sum</span>(col5), <span class="st">"): "</span>, <span class="fu">sum</span>(col5) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb18-115"><a href="#cb18-115" aria-hidden="true" tabindex="-1"></a>            <span class="st">"4. % reads passing 2 and 3: ("</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5), <span class="st">"): "</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n\n</span><span class="st">"</span>,</span>
<span id="cb18-116"><a href="#cb18-116" aria-hidden="true" tabindex="-1"></a>            <span class="st">"5. % reads (R1 and R2 are reverse and complementary) ("</span>, <span class="fu">sum</span>(col6), <span class="st">"): "</span>, <span class="fu">sum</span>(col6) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb18-117"><a href="#cb18-117" aria-hidden="true" tabindex="-1"></a>            <span class="st">"6. % reads passing 2, 3, and 5 ("</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6), <span class="st">"): "</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n\n</span><span class="st">"</span>,</span>
<span id="cb18-118"><a href="#cb18-118" aria-hidden="true" tabindex="-1"></a>            <span class="st">"7. % reads of R1 in the library ("</span>, <span class="fu">sum</span>(col7), <span class="st">"): "</span>, <span class="fu">sum</span>(col7) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb18-119"><a href="#cb18-119" aria-hidden="true" tabindex="-1"></a>            <span class="st">"8. % reads passing 2, 3, 5 and 7 ("</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col7), <span class="st">"): "</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col7) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n\n</span><span class="st">"</span>,</span>
<span id="cb18-120"><a href="#cb18-120" aria-hidden="true" tabindex="-1"></a>            <span class="st">"9. % reads of R2 in the library ("</span>, <span class="fu">sum</span>(col8), <span class="st">"): "</span>, <span class="fu">sum</span>(col8) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb18-121"><a href="#cb18-121" aria-hidden="true" tabindex="-1"></a>            <span class="st">"10. % reads passing 2, 3, 5 and 9 ("</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col8), <span class="st">"): "</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col8) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n\n</span><span class="st">"</span>,</span>
<span id="cb18-122"><a href="#cb18-122" aria-hidden="true" tabindex="-1"></a>            <span class="st">"11. % reads (both R1 and R2 are in the library) ("</span>, <span class="fu">sum</span>(col7 <span class="op">.&amp;&amp;</span> col8), <span class="st">"): "</span>, <span class="fu">sum</span>(col7 <span class="op">.&amp;&amp;</span> col8) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb18-123"><a href="#cb18-123" aria-hidden="true" tabindex="-1"></a>            <span class="st">"12. % reads passing 2, 3, 5 and 11 ("</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col7 <span class="op">.&amp;&amp;</span> col8), <span class="st">"): "</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col7 <span class="op">.&amp;&amp;</span> col8) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n\n</span><span class="st">"</span>,</span>
<span id="cb18-124"><a href="#cb18-124" aria-hidden="true" tabindex="-1"></a>            <span class="st">"13. % reads (one of R1 and R2 in the library, not both) ("</span>, <span class="fu">sum</span>(col7 <span class="op">.âŠ»</span> col8), <span class="st">"): "</span>, <span class="fu">sum</span>(col7 <span class="op">.âŠ»</span> col8) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb18-125"><a href="#cb18-125" aria-hidden="true" tabindex="-1"></a>            <span class="st">"14. % reads passing 2, 3, 5 and 13 ("</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> (col7 <span class="op">.âŠ»</span> col8)), <span class="st">"): "</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> (col7 <span class="op">.âŠ»</span> col8)) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n\n</span><span class="st">"</span>,</span>
<span id="cb18-126"><a href="#cb18-126" aria-hidden="true" tabindex="-1"></a>            <span class="st">"15. % reverse reads of R1 in the library ("</span>, <span class="fu">sum</span>(col9), <span class="st">"): "</span>, <span class="fu">sum</span>(col9) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb18-127"><a href="#cb18-127" aria-hidden="true" tabindex="-1"></a>            <span class="st">"16. % reads passing 2, 3, 5 and 15 ("</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col9), <span class="st">"): "</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col9) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n\n</span><span class="st">"</span>,</span>
<span id="cb18-128"><a href="#cb18-128" aria-hidden="true" tabindex="-1"></a>            <span class="st">"17. % complementary reads of R1 in the library ("</span>, <span class="fu">sum</span>(col10), <span class="st">"): "</span>, <span class="fu">sum</span>(col10) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb18-129"><a href="#cb18-129" aria-hidden="true" tabindex="-1"></a>            <span class="st">"18. % reads passing 2, 3, 5 and 17 ("</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col10), <span class="st">"): "</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col10) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n\n</span><span class="st">"</span>,</span>
<span id="cb18-130"><a href="#cb18-130" aria-hidden="true" tabindex="-1"></a>            <span class="st">"19. % reverse reads of R2 in the library ("</span>, <span class="fu">sum</span>(col11), <span class="st">"): "</span>, <span class="fu">sum</span>(col11) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb18-131"><a href="#cb18-131" aria-hidden="true" tabindex="-1"></a>            <span class="st">"20. % reads passing 2, 3, 5 and 19 ("</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col11), <span class="st">"): "</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col11) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n\n</span><span class="st">"</span>,</span>
<span id="cb18-132"><a href="#cb18-132" aria-hidden="true" tabindex="-1"></a>            <span class="st">"21. % complementary reads of R2 in the library ("</span>, <span class="fu">sum</span>(col12), <span class="st">"): "</span>, <span class="fu">sum</span>(col12) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb18-133"><a href="#cb18-133" aria-hidden="true" tabindex="-1"></a>            <span class="st">"22. % reads passing 2, 3, 5 and 21 ("</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col12), <span class="st">"): "</span>, <span class="fu">sum</span>(col4 <span class="op">.&amp;&amp;</span> col5 <span class="op">.&amp;&amp;</span> col6 <span class="op">.&amp;&amp;</span> col12) <span class="op">/</span> len_vec[<span class="fl">1</span>], <span class="st">"</span><span class="sc">\n\n\n</span><span class="st">"</span></span>
<span id="cb18-134"><a href="#cb18-134" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb18-135"><a href="#cb18-135" aria-hidden="true" tabindex="-1"></a>        <span class="fu">print</span>(io, stat_str)</span>
<span id="cb18-136"><a href="#cb18-136" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb18-137"><a href="#cb18-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-138"><a href="#cb18-138" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Filter out invalid Barcode-UMI pairs</span></span>
<span id="cb18-139"><a href="#cb18-139" aria-hidden="true" tabindex="-1"></a>    valid_bc_umi_vec <span class="op">=</span> bc_umi_vec[col4<span class="op">.&amp;&amp;</span>col5<span class="op">.&amp;&amp;</span>col6<span class="op">.&amp;&amp;</span>(col7<span class="op">.âŠ»</span>col8)]</span>
<span id="cb18-140"><a href="#cb18-140" aria-hidden="true" tabindex="-1"></a>    df_dict[<span class="fu">unique</span>(read_df[<span class="op">:</span>, <span class="op">:</span>tech_sample])[<span class="fl">1</span>]] <span class="op">=</span> <span class="fu">DataFrame</span>(</span>
<span id="cb18-141"><a href="#cb18-141" aria-hidden="true" tabindex="-1"></a>        [[(x[<span class="fl">1</span>], x[<span class="fl">3</span>]) for x <span class="kw">in</span> valid_bc_umi_vec[<span class="fu">getindex</span>.(valid_bc_umi_vec, <span class="fl">7</span>)]]; [(x[<span class="fl">2</span>], x[<span class="fl">3</span>]) for x <span class="kw">in</span> valid_bc_umi_vec[<span class="fu">getindex</span>.(valid_bc_umi_vec, <span class="fl">8</span>)]]],</span>
<span id="cb18-142"><a href="#cb18-142" aria-hidden="true" tabindex="-1"></a>        [<span class="op">:</span>barcode, <span class="op">:</span>umi]</span>
<span id="cb18-143"><a href="#cb18-143" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb18-144"><a href="#cb18-144" aria-hidden="true" tabindex="-1"></a>    df_dict[<span class="fu">unique</span>(read_df[<span class="op">:</span>, <span class="op">:</span>tech_sample])[<span class="fl">1</span>]][!, <span class="op">:</span>tech_sample] <span class="op">.=</span> <span class="fu">unique</span>(read_df[<span class="op">:</span>, <span class="op">:</span>tech_sample])[<span class="fl">1</span>]</span>
<span id="cb18-145"><a href="#cb18-145" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb18-146"><a href="#cb18-146" aria-hidden="true" tabindex="-1"></a>bc_umi_df <span class="op">=</span> <span class="fu">reduce</span>(vcat, <span class="fu">collect</span>(<span class="fu">values</span>(df_dict)); cols<span class="op">=:</span>setequal)</span>
<span id="cb18-147"><a href="#cb18-147" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(<span class="fu">joinpath</span>(raw_bc_umi_dir, <span class="st">"raw_bc_umi_pairs.tsv"</span>), bc_umi_df; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="solution-2" class="level4" data-number="2.3.2">
<h4 data-number="2.3.2" class="anchored" data-anchor-id="solution-2"><span class="header-section-number">2.3.2</span> Solution 2</h4>
<div id="6582e8d5" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">FASTX</span>, <span class="bu">DataFrames</span>, <span class="bu">CSV</span>, <span class="bu">YRUtils</span>, <span class="bu">CodecZlib</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>raw_fastq_dir <span class="op">=</span> <span class="st">"raw_fastq"</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>extract_valid_barcode_dir <span class="op">=</span> <span class="st">"extract_valid_barcode"</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>base_qual <span class="op">=</span> <span class="fl">20</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>seqkit_nthreads <span class="op">=</span> <span class="fl">40</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>parallel_block_size <span class="op">=</span> <span class="fl">500000</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(extract_valid_barcode_dir)</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>raw_fastq_file_name_pattern <span class="op">=</span> <span class="st">r"</span><span class="sc">.+</span><span class="st">/</span><span class="ch">(?</span><span class="st">&lt;id&gt;</span><span class="ch">[</span><span class="st">a-zA-Z0-9</span><span class="ch">]</span><span class="sc">+</span><span class="ch">)</span><span class="st">_</span><span class="ch">(?</span><span class="st">&lt;type&gt;RNA</span><span class="ch">|</span><span class="st">DNA</span><span class="ch">)</span><span class="st">_</span><span class="ch">(?</span><span class="st">&lt;rep&gt;rep</span><span class="ch">[</span><span class="st">0-9</span><span class="ch">]</span><span class="sc">+</span><span class="ch">)(</span><span class="st">_</span><span class="ch">(?</span><span class="st">&lt;tech&gt;part</span><span class="ch">[</span><span class="st">0-9</span><span class="ch">]</span><span class="sc">+</span><span class="ch">))</span><span class="sc">?\.</span><span class="ch">(?</span><span class="st">&lt;read&gt;R</span><span class="ch">[</span><span class="st">123</span><span class="ch">])</span><span class="sc">\.</span><span class="ch">(</span><span class="st">fq</span><span class="ch">|</span><span class="st">fastq</span><span class="ch">)</span><span class="sc">\.</span><span class="st">gz</span><span class="sc">$</span><span class="st">"</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>raw_fastq_files<span class="st"> = YRUtils</span><span class="sc">.</span><span class="st">BioUtils</span><span class="sc">.</span><span class="st">list_files</span><span class="ch">(</span><span class="st">raw_fastq_dir, raw_fastq_file_name_pattern; recursive=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>raw_fastq_nums <span class="op">=</span> YRUtils.BioUtils.<span class="fu">fq_num</span>(raw_fastq_files, seqkit_nthreads)</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>ms <span class="op">=</span> <span class="fu">match</span>.(raw_fastq_file_name_pattern, raw_fastq_files)</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>metadata_vec <span class="op">=</span> <span class="fu">Vector</span><span class="dt">{NTuple{7,String}}</span>(<span class="cn">undef</span>, <span class="fu">length</span>(ms))</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(ms)</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>    metadata_vec[i] <span class="op">=</span> (ms[i].match, ms[i][<span class="st">"id"</span>], ms[i][<span class="st">"type"</span>], ms[i][<span class="st">"rep"</span>], ms[i][<span class="st">"tech"</span>], ms[i][<span class="st">"read"</span>], <span class="fu">string</span>(raw_fastq_nums[ms[i].match]))</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> <span class="fu">unique</span>(<span class="fu">DataFrame</span>(metadata_vec, [<span class="op">:</span>file, <span class="op">:</span>id, <span class="op">:</span>type, <span class="op">:</span>rep, <span class="op">:</span>tech, <span class="op">:</span>read, <span class="op">:</span>num_seqs]))</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> <span class="fu">transform</span>(</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>    df,</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>    [<span class="op">:</span>id, <span class="op">:</span>type, <span class="op">:</span>rep, <span class="op">:</span>tech] <span class="op">=&gt;</span> <span class="fu">ByRow</span>((id, <span class="kw">type</span>, rep, tech) <span class="op">-&gt;</span> <span class="fu">join</span>([id, type, rep, tech], <span class="st">"_"</span>)) <span class="op">=&gt;</span> <span class="op">:</span>tech_sample,</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>    [<span class="op">:</span>id, <span class="op">:</span>type, <span class="op">:</span>rep] <span class="op">=&gt;</span> <span class="fu">ByRow</span>((id, <span class="kw">type</span>, rep) <span class="op">-&gt;</span> <span class="fu">join</span>([id, type, rep], <span class="st">"_"</span>)) <span class="op">=&gt;</span> <span class="op">:</span>rep_sample,</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>    [<span class="op">:</span>id, <span class="op">:</span>type] <span class="op">=&gt;</span> <span class="fu">ByRow</span>((id, <span class="kw">type</span>) <span class="op">-&gt;</span> <span class="fu">join</span>([id, type], <span class="st">"_"</span>)) <span class="op">=&gt;</span> <span class="op">:</span>type_sample</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(<span class="fu">joinpath</span>(extract_valid_barcode_dir, <span class="st">"fq_metadata.tsv"</span>), df; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a><span class="co"># global var</span></span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>top_dict <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{String,Set{String}}</span>()</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (<span class="kw">type</span>, type_subdf) <span class="kw">in</span> <span class="fu">pairs</span>(<span class="fu">groupby</span>(df, <span class="op">:</span><span class="kw">type</span>))</span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a>    <span class="kw">global</span> top_dict</span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a>    <span class="co"># local var</span></span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a>    type_dict <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{String,Set{String}}</span>()</span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (biorep, biorep_subdf) <span class="kw">in</span> <span class="fu">pairs</span>(<span class="fu">groupby</span>(type_subdf, <span class="op">:</span>rep))</span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a>        <span class="co"># local var</span></span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a>        biorep_df <span class="op">=</span> <span class="fu">DataFrame</span>(barcode<span class="op">=</span><span class="dt">String</span>[], count<span class="op">=</span><span class="dt">Int64</span>[])</span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a>        biorep_dict <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{String,Set{String}}</span>()</span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> (techrep, techrep_subdf) <span class="kw">in</span> <span class="fu">pairs</span>(<span class="fu">groupby</span>(biorep_subdf, <span class="op">:</span>tech))</span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true" tabindex="-1"></a>            <span class="co"># local var</span></span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true" tabindex="-1"></a>            techrep_dict <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{String,Set{String}}</span>()</span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Read in reads</span></span>
<span id="cb19-42"><a href="#cb19-42" aria-hidden="true" tabindex="-1"></a>            read_dict <span class="op">=</span> <span class="fu">Dict</span>(read_type <span class="op">=&gt;</span> <span class="fu">Vector</span><span class="dt">{Tuple{String,String,String}}</span>(<span class="cn">undef</span>, <span class="fu">parse</span>(<span class="dt">Int64</span>, num_seqs)) <span class="cf">for</span> (read_type, num_seqs) <span class="kw">in</span> <span class="fu">collect</span>(<span class="fu">zip</span>(techrep_subdf[<span class="op">:</span>, <span class="op">:</span>read], techrep_subdf[<span class="op">:</span>, <span class="op">:</span>num_seqs])))</span>
<span id="cb19-43"><a href="#cb19-43" aria-hidden="true" tabindex="-1"></a>            <span class="bu">Threads</span>.<span class="pp">@threads</span> <span class="cf">for</span> (read_type, num_seqs, fq_file) <span class="kw">in</span> <span class="fu">collect</span>(<span class="fu">zip</span>(techrep_subdf[<span class="op">:</span>, <span class="op">:</span>read], techrep_subdf[<span class="op">:</span>, <span class="op">:</span>num_seqs], techrep_subdf[<span class="op">:</span>, <span class="op">:</span>file]))</span>
<span id="cb19-44"><a href="#cb19-44" aria-hidden="true" tabindex="-1"></a>                <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"start parsing "</span>, fq_file, <span class="st">" with read type "</span>, read_type, <span class="st">" and the number of sequences "</span>, num_seqs, <span class="st">" ..."</span>)</span>
<span id="cb19-45"><a href="#cb19-45" aria-hidden="true" tabindex="-1"></a>                <span class="fu">FASTQReader</span>(<span class="fu">GzipDecompressorStream</span>(<span class="fu">open</span>(fq_file))) <span class="cf">do</span> reader</span>
<span id="cb19-46"><a href="#cb19-46" aria-hidden="true" tabindex="-1"></a>                    record <span class="op">=</span> FASTQ.<span class="fu">Record</span>()</span>
<span id="cb19-47"><a href="#cb19-47" aria-hidden="true" tabindex="-1"></a>                    i <span class="op">=</span> <span class="fl">0</span></span>
<span id="cb19-48"><a href="#cb19-48" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">while</span> !<span class="fu">eof</span>(reader)</span>
<span id="cb19-49"><a href="#cb19-49" aria-hidden="true" tabindex="-1"></a>                        i <span class="op">+=</span> <span class="fl">1</span></span>
<span id="cb19-50"><a href="#cb19-50" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">empty!</span>(record)</span>
<span id="cb19-51"><a href="#cb19-51" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">read!</span>(reader, record)</span>
<span id="cb19-52"><a href="#cb19-52" aria-hidden="true" tabindex="-1"></a>                        read_dict[read_type][i] <span class="op">=</span> (<span class="fu">identifier</span>(record), <span class="fu">sequence</span>(record), <span class="fu">join</span>(<span class="fu">collect</span>(<span class="fu">quality_scores</span>(record)), <span class="st">"/"</span>))</span>
<span id="cb19-53"><a href="#cb19-53" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">end</span></span>
<span id="cb19-54"><a href="#cb19-54" aria-hidden="true" tabindex="-1"></a>                    <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"read in "</span>, i, <span class="st">" sequences in total for "</span>, fq_file)</span>
<span id="cb19-55"><a href="#cb19-55" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span> i <span class="op">!=</span> <span class="fu">parse</span>(<span class="dt">Int64</span>, num_seqs)</span>
<span id="cb19-56"><a href="#cb19-56" aria-hidden="true" tabindex="-1"></a>                        <span class="pp">@error</span> <span class="fu">string</span>(<span class="st">"parsing file "</span>, fq_file, <span class="st">" failed!"</span>)</span>
<span id="cb19-57"><a href="#cb19-57" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">end</span></span>
<span id="cb19-58"><a href="#cb19-58" aria-hidden="true" tabindex="-1"></a>                <span class="cf">end</span></span>
<span id="cb19-59"><a href="#cb19-59" aria-hidden="true" tabindex="-1"></a>                <span class="pp">@info</span> <span class="fu">string</span>(<span class="st">"parsing "</span>, fq_file, <span class="st">" with read type "</span>, read_type, <span class="st">" and the number of sequences "</span>, num_seqs, <span class="st">" done!"</span>)</span>
<span id="cb19-60"><a href="#cb19-60" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb19-61"><a href="#cb19-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-62"><a href="#cb19-62" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Count qualified barcodes and their UMIs</span></span>
<span id="cb19-63"><a href="#cb19-63" aria-hidden="true" tabindex="-1"></a>            len_vec <span class="op">=</span> <span class="fu">length</span>.(<span class="fu">values</span>(read_dict))</span>
<span id="cb19-64"><a href="#cb19-64" aria-hidden="true" tabindex="-1"></a>            <span class="co"># The three files should have the same number of lines</span></span>
<span id="cb19-65"><a href="#cb19-65" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="fu">length</span>(<span class="fu">unique</span>(len_vec)) <span class="op">==</span> <span class="fl">1</span></span>
<span id="cb19-66"><a href="#cb19-66" aria-hidden="true" tabindex="-1"></a>                bc_umi_vec <span class="op">=</span> <span class="fu">Vector</span><span class="dt">{Tuple{String,String,String,Vararg{Bool,3}}}</span>(<span class="cn">undef</span>, len_vec[<span class="fl">1</span>])</span>
<span id="cb19-67"><a href="#cb19-67" aria-hidden="true" tabindex="-1"></a>                <span class="bu">Threads</span>.<span class="pp">@threads</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span>len_vec[<span class="fl">1</span>]</span>
<span id="cb19-68"><a href="#cb19-68" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># Read IDs should be identical across R1, R2, and R3</span></span>
<span id="cb19-69"><a href="#cb19-69" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span> read_dict[<span class="st">"R1"</span>][i][<span class="fl">1</span>] <span class="op">==</span> read_dict[<span class="st">"R2"</span>][i][<span class="fl">1</span>] <span class="op">==</span> read_dict[<span class="st">"R3"</span>][i][<span class="fl">1</span>]</span>
<span id="cb19-70"><a href="#cb19-70" aria-hidden="true" tabindex="-1"></a>                        bc_umi_vec[i] <span class="op">=</span> (</span>
<span id="cb19-71"><a href="#cb19-71" aria-hidden="true" tabindex="-1"></a>                            <span class="co"># Read 1</span></span>
<span id="cb19-72"><a href="#cb19-72" aria-hidden="true" tabindex="-1"></a>                            read_dict[<span class="st">"R1"</span>][i][<span class="fl">2</span>],</span>
<span id="cb19-73"><a href="#cb19-73" aria-hidden="true" tabindex="-1"></a>                            <span class="co"># Read 2</span></span>
<span id="cb19-74"><a href="#cb19-74" aria-hidden="true" tabindex="-1"></a>                            read_dict[<span class="st">"R2"</span>][i][<span class="fl">2</span>],</span>
<span id="cb19-75"><a href="#cb19-75" aria-hidden="true" tabindex="-1"></a>                            <span class="co"># UMI</span></span>
<span id="cb19-76"><a href="#cb19-76" aria-hidden="true" tabindex="-1"></a>                            read_dict[<span class="st">"R3"</span>][i][<span class="fl">2</span>],</span>
<span id="cb19-77"><a href="#cb19-77" aria-hidden="true" tabindex="-1"></a>                            <span class="co"># Read sequences should only contain A, T, C, and G across R1, R2, and R3</span></span>
<span id="cb19-78"><a href="#cb19-78" aria-hidden="true" tabindex="-1"></a>                            !<span class="fu">occursin</span>(<span class="st">"N"</span>, <span class="fu">string</span>(read_dict[<span class="st">"R1"</span>][i][<span class="fl">2</span>], read_dict[<span class="st">"R2"</span>][i][<span class="fl">2</span>], read_dict[<span class="st">"R3"</span>][i][<span class="fl">2</span>])),</span>
<span id="cb19-79"><a href="#cb19-79" aria-hidden="true" tabindex="-1"></a>                            <span class="co"># All base qualities &gt;= base_qual across R1, R2, and R3</span></span>
<span id="cb19-80"><a href="#cb19-80" aria-hidden="true" tabindex="-1"></a>                            <span class="fu">all</span>(<span class="fu">parse</span>.(<span class="dt">Int</span>, <span class="fu">split</span>(<span class="fu">string</span>(read_dict[<span class="st">"R1"</span>][i][<span class="fl">3</span>], <span class="st">"/"</span>, read_dict[<span class="st">"R2"</span>][i][<span class="fl">3</span>], <span class="st">"/"</span>, read_dict[<span class="st">"R3"</span>][i][<span class="fl">3</span>]), <span class="st">"/"</span>)) <span class="op">.&gt;=</span> base_qual),</span>
<span id="cb19-81"><a href="#cb19-81" aria-hidden="true" tabindex="-1"></a>                            <span class="co"># Read 1 and read 2 should be reverse and complementary</span></span>
<span id="cb19-82"><a href="#cb19-82" aria-hidden="true" tabindex="-1"></a>                            YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(read_dict[<span class="st">"R1"</span>][i][<span class="fl">2</span>]) <span class="op">==</span> read_dict[<span class="st">"R2"</span>][i][<span class="fl">2</span>]</span>
<span id="cb19-83"><a href="#cb19-83" aria-hidden="true" tabindex="-1"></a>                        )</span>
<span id="cb19-84"><a href="#cb19-84" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">else</span></span>
<span id="cb19-85"><a href="#cb19-85" aria-hidden="true" tabindex="-1"></a>                        <span class="pp">@error</span> <span class="st">"read IDs are not identical across R1, R2, and R3"</span></span>
<span id="cb19-86"><a href="#cb19-86" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">end</span></span>
<span id="cb19-87"><a href="#cb19-87" aria-hidden="true" tabindex="-1"></a>                <span class="cf">end</span></span>
<span id="cb19-88"><a href="#cb19-88" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span></span>
<span id="cb19-89"><a href="#cb19-89" aria-hidden="true" tabindex="-1"></a>                <span class="pp">@error</span> <span class="st">"length(R1) == length(R2) == length(R3) is not true"</span></span>
<span id="cb19-90"><a href="#cb19-90" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb19-91"><a href="#cb19-91" aria-hidden="true" tabindex="-1"></a>            read_dict <span class="op">=</span> <span class="cn">nothing</span></span>
<span id="cb19-92"><a href="#cb19-92" aria-hidden="true" tabindex="-1"></a>            GC.<span class="fu">gc</span>()</span>
<span id="cb19-93"><a href="#cb19-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-94"><a href="#cb19-94" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Filter out invalid Barcode-UMI pairs</span></span>
<span id="cb19-95"><a href="#cb19-95" aria-hidden="true" tabindex="-1"></a>            valid_bc_umi_vec <span class="op">=</span> bc_umi_vec[<span class="fu">getindex</span>.(bc_umi_vec, <span class="fl">4</span>)<span class="op">.&amp;&amp;</span><span class="fu">getindex</span>.(bc_umi_vec, <span class="fl">5</span>)<span class="op">.&amp;&amp;</span><span class="fu">getindex</span>.(bc_umi_vec, <span class="fl">6</span>)]</span>
<span id="cb19-96"><a href="#cb19-96" aria-hidden="true" tabindex="-1"></a>            bc_umi_vec <span class="op">=</span> <span class="cn">nothing</span></span>
<span id="cb19-97"><a href="#cb19-97" aria-hidden="true" tabindex="-1"></a>            GC.<span class="fu">gc</span>()</span>
<span id="cb19-98"><a href="#cb19-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-99"><a href="#cb19-99" aria-hidden="true" tabindex="-1"></a>            split_valid_bc_umi_vec <span class="op">=</span> <span class="fu">collect</span>(<span class="bu">Base</span>.<span class="bu">Iterators</span>.<span class="fu">partition</span>(valid_bc_umi_vec, parallel_block_size))</span>
<span id="cb19-100"><a href="#cb19-100" aria-hidden="true" tabindex="-1"></a>            techrep_dict_vec <span class="op">=</span> [<span class="fu">Dict</span><span class="dt">{String,Set{String}}</span>() for _ <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(split_valid_bc_umi_vec)]</span>
<span id="cb19-101"><a href="#cb19-101" aria-hidden="true" tabindex="-1"></a>            <span class="bu">Threads</span>.<span class="pp">@threads</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(split_valid_bc_umi_vec)</span>
<span id="cb19-102"><a href="#cb19-102" aria-hidden="true" tabindex="-1"></a>                <span class="cf">for</span> item <span class="kw">in</span> split_valid_bc_umi_vec[i]</span>
<span id="cb19-103"><a href="#cb19-103" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span> <span class="fu">haskey</span>(techrep_dict_vec[i], item[<span class="fl">1</span>])</span>
<span id="cb19-104"><a href="#cb19-104" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">push!</span>(techrep_dict_vec[i][item[<span class="fl">1</span>]], item[<span class="fl">3</span>])</span>
<span id="cb19-105"><a href="#cb19-105" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">elseif</span> <span class="fu">haskey</span>(techrep_dict_vec[i], item[<span class="fl">2</span>])</span>
<span id="cb19-106"><a href="#cb19-106" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">push!</span>(techrep_dict_vec[i][item[<span class="fl">2</span>]], item[<span class="fl">3</span>])</span>
<span id="cb19-107"><a href="#cb19-107" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">else</span></span>
<span id="cb19-108"><a href="#cb19-108" aria-hidden="true" tabindex="-1"></a>                        techrep_dict_vec[i][item[<span class="fl">1</span>]] <span class="op">=</span> <span class="fu">Set</span>([item[<span class="fl">3</span>]])</span>
<span id="cb19-109"><a href="#cb19-109" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">end</span></span>
<span id="cb19-110"><a href="#cb19-110" aria-hidden="true" tabindex="-1"></a>                <span class="cf">end</span></span>
<span id="cb19-111"><a href="#cb19-111" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb19-112"><a href="#cb19-112" aria-hidden="true" tabindex="-1"></a>            valid_bc_umi_vec <span class="op">=</span> <span class="cn">nothing</span></span>
<span id="cb19-113"><a href="#cb19-113" aria-hidden="true" tabindex="-1"></a>            split_valid_bc_umi_vec <span class="op">=</span> <span class="cn">nothing</span></span>
<span id="cb19-114"><a href="#cb19-114" aria-hidden="true" tabindex="-1"></a>            GC.<span class="fu">gc</span>()</span>
<span id="cb19-115"><a href="#cb19-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-116"><a href="#cb19-116" aria-hidden="true" tabindex="-1"></a>            <span class="cf">while</span> <span class="cn">true</span></span>
<span id="cb19-117"><a href="#cb19-117" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> <span class="fu">length</span>(techrep_dict_vec) <span class="op">&lt;=</span> <span class="fl">1</span></span>
<span id="cb19-118"><a href="#cb19-118" aria-hidden="true" tabindex="-1"></a>                    techrep_dict <span class="op">=</span> techrep_dict_vec[<span class="fl">1</span>]</span>
<span id="cb19-119"><a href="#cb19-119" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">break</span></span>
<span id="cb19-120"><a href="#cb19-120" aria-hidden="true" tabindex="-1"></a>                <span class="cf">else</span></span>
<span id="cb19-121"><a href="#cb19-121" aria-hidden="true" tabindex="-1"></a>                    split_techrep_dict_vec <span class="op">=</span> <span class="fu">convert</span>(<span class="dt">Vector</span>{<span class="dt">Vector</span>{<span class="dt">Dict</span>{<span class="dt">String</span>,<span class="dt">Set</span>{<span class="dt">String</span>}}}}, <span class="fu">collect</span>(<span class="bu">Base</span>.<span class="bu">Iterators</span>.<span class="fu">partition</span>(techrep_dict_vec, <span class="fl">2</span>)))</span>
<span id="cb19-122"><a href="#cb19-122" aria-hidden="true" tabindex="-1"></a>                <span class="cf">end</span></span>
<span id="cb19-123"><a href="#cb19-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-124"><a href="#cb19-124" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> <span class="fu">length</span>(split_techrep_dict_vec[<span class="kw">end</span>]) <span class="op">==</span> <span class="fl">1</span></span>
<span id="cb19-125"><a href="#cb19-125" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">push!</span>(split_techrep_dict_vec[<span class="kw">end</span><span class="op">-</span><span class="fl">1</span>], <span class="fu">pop!</span>(split_techrep_dict_vec[<span class="kw">end</span>]))</span>
<span id="cb19-126"><a href="#cb19-126" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">pop!</span>(split_techrep_dict_vec)</span>
<span id="cb19-127"><a href="#cb19-127" aria-hidden="true" tabindex="-1"></a>                <span class="cf">end</span></span>
<span id="cb19-128"><a href="#cb19-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-129"><a href="#cb19-129" aria-hidden="true" tabindex="-1"></a>                <span class="bu">Threads</span>.<span class="pp">@threads</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(split_techrep_dict_vec)</span>
<span id="cb19-130"><a href="#cb19-130" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">for</span> j <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="fu">length</span>(split_techrep_dict_vec[i])</span>
<span id="cb19-131"><a href="#cb19-131" aria-hidden="true" tabindex="-1"></a>                        <span class="cf">for</span> (k, v) <span class="kw">in</span> split_techrep_dict_vec[i][j]</span>
<span id="cb19-132"><a href="#cb19-132" aria-hidden="true" tabindex="-1"></a>                            <span class="cf">if</span> <span class="fu">haskey</span>(split_techrep_dict_vec[i][<span class="fl">1</span>], k)</span>
<span id="cb19-133"><a href="#cb19-133" aria-hidden="true" tabindex="-1"></a>                                split_techrep_dict_vec[i][<span class="fl">1</span>][k] <span class="op">=</span> <span class="fu">union</span>(split_techrep_dict_vec[i][<span class="fl">1</span>][k], v)</span>
<span id="cb19-134"><a href="#cb19-134" aria-hidden="true" tabindex="-1"></a>                            <span class="cf">elseif</span> <span class="fu">haskey</span>(split_techrep_dict_vec[i][<span class="fl">1</span>], YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k))</span>
<span id="cb19-135"><a href="#cb19-135" aria-hidden="true" tabindex="-1"></a>                                split_techrep_dict_vec[i][<span class="fl">1</span>][YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k)] <span class="op">=</span> <span class="fu">union</span>(split_techrep_dict_vec[i][<span class="fl">1</span>][YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k)], v)</span>
<span id="cb19-136"><a href="#cb19-136" aria-hidden="true" tabindex="-1"></a>                            <span class="cf">else</span></span>
<span id="cb19-137"><a href="#cb19-137" aria-hidden="true" tabindex="-1"></a>                                split_techrep_dict_vec[i][<span class="fl">1</span>][k] <span class="op">=</span> v</span>
<span id="cb19-138"><a href="#cb19-138" aria-hidden="true" tabindex="-1"></a>                            <span class="cf">end</span></span>
<span id="cb19-139"><a href="#cb19-139" aria-hidden="true" tabindex="-1"></a>                        <span class="cf">end</span></span>
<span id="cb19-140"><a href="#cb19-140" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">end</span></span>
<span id="cb19-141"><a href="#cb19-141" aria-hidden="true" tabindex="-1"></a>                <span class="cf">end</span></span>
<span id="cb19-142"><a href="#cb19-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-143"><a href="#cb19-143" aria-hidden="true" tabindex="-1"></a>                techrep_dict_vec <span class="op">=</span> <span class="fu">getindex</span>.(split_techrep_dict_vec, <span class="fl">1</span>)</span>
<span id="cb19-144"><a href="#cb19-144" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb19-145"><a href="#cb19-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-146"><a href="#cb19-146" aria-hidden="true" tabindex="-1"></a>            <span class="fu">open</span>(<span class="fu">joinpath</span>(extract_valid_barcode_dir, <span class="st">"fq_read_stats.txt"</span>), <span class="st">"a"</span>) <span class="cf">do</span> io</span>
<span id="cb19-147"><a href="#cb19-147" aria-hidden="true" tabindex="-1"></a>                stat_str <span class="op">=</span> <span class="fu">string</span>(</span>
<span id="cb19-148"><a href="#cb19-148" aria-hidden="true" tabindex="-1"></a>                    techrep_subdf.tech_sample[<span class="fl">1</span>], <span class="st">": "</span>, <span class="fu">length</span>(techrep_dict), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb19-149"><a href="#cb19-149" aria-hidden="true" tabindex="-1"></a>                <span class="fu">print</span>(io, stat_str)</span>
<span id="cb19-150"><a href="#cb19-150" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb19-151"><a href="#cb19-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-152"><a href="#cb19-152" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="fu">isempty</span>(biorep_dict)</span>
<span id="cb19-153"><a href="#cb19-153" aria-hidden="true" tabindex="-1"></a>                biorep_dict <span class="op">=</span> techrep_dict</span>
<span id="cb19-154"><a href="#cb19-154" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span></span>
<span id="cb19-155"><a href="#cb19-155" aria-hidden="true" tabindex="-1"></a>                <span class="cf">for</span> (k, v) <span class="kw">in</span> techrep_dict</span>
<span id="cb19-156"><a href="#cb19-156" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span> <span class="fu">haskey</span>(biorep_dict, k)</span>
<span id="cb19-157"><a href="#cb19-157" aria-hidden="true" tabindex="-1"></a>                        biorep_dict[k] <span class="op">=</span> <span class="fu">union</span>(biorep_dict[k], v)</span>
<span id="cb19-158"><a href="#cb19-158" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">elseif</span> <span class="fu">haskey</span>(biorep_dict, YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k))</span>
<span id="cb19-159"><a href="#cb19-159" aria-hidden="true" tabindex="-1"></a>                        biorep_dict[YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k)] <span class="op">=</span> <span class="fu">union</span>(biorep_dict[YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k)], v)</span>
<span id="cb19-160"><a href="#cb19-160" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">else</span></span>
<span id="cb19-161"><a href="#cb19-161" aria-hidden="true" tabindex="-1"></a>                        biorep_dict[k] <span class="op">=</span> v</span>
<span id="cb19-162"><a href="#cb19-162" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">end</span></span>
<span id="cb19-163"><a href="#cb19-163" aria-hidden="true" tabindex="-1"></a>                <span class="cf">end</span></span>
<span id="cb19-164"><a href="#cb19-164" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb19-165"><a href="#cb19-165" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb19-166"><a href="#cb19-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-167"><a href="#cb19-167" aria-hidden="true" tabindex="-1"></a>        biorep_df <span class="op">=</span> <span class="fu">sort</span>(<span class="fu">DataFrame</span>([(barcode<span class="op">=</span>k, count<span class="op">=</span><span class="fu">length</span>(v)) for (k, v) <span class="kw">in</span> biorep_dict]), <span class="op">:</span>count, rev<span class="op">=</span><span class="cn">true</span>)</span>
<span id="cb19-168"><a href="#cb19-168" aria-hidden="true" tabindex="-1"></a>        CSV.<span class="fu">write</span>(<span class="fu">joinpath</span>(extract_valid_barcode_dir, <span class="fu">string</span>(biorep_subdf.rep_sample[<span class="fl">1</span>], <span class="st">".barcode_count.tsv"</span>)), biorep_df; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb19-169"><a href="#cb19-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-170"><a href="#cb19-170" aria-hidden="true" tabindex="-1"></a>        <span class="fu">open</span>(<span class="fu">joinpath</span>(extract_valid_barcode_dir, <span class="st">"fq_read_stats.txt"</span>), <span class="st">"a"</span>) <span class="cf">do</span> io</span>
<span id="cb19-171"><a href="#cb19-171" aria-hidden="true" tabindex="-1"></a>            stat_str <span class="op">=</span> <span class="fu">string</span>(</span>
<span id="cb19-172"><a href="#cb19-172" aria-hidden="true" tabindex="-1"></a>                biorep_subdf.rep_sample[<span class="fl">1</span>], <span class="st">": "</span>, <span class="fu">length</span>(biorep_dict), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb19-173"><a href="#cb19-173" aria-hidden="true" tabindex="-1"></a>            <span class="fu">print</span>(io, stat_str)</span>
<span id="cb19-174"><a href="#cb19-174" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb19-175"><a href="#cb19-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-176"><a href="#cb19-176" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="fu">isempty</span>(type_dict)</span>
<span id="cb19-177"><a href="#cb19-177" aria-hidden="true" tabindex="-1"></a>            type_dict <span class="op">=</span> biorep_dict</span>
<span id="cb19-178"><a href="#cb19-178" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span></span>
<span id="cb19-179"><a href="#cb19-179" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> (k, v) <span class="kw">in</span> biorep_dict</span>
<span id="cb19-180"><a href="#cb19-180" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> <span class="fu">haskey</span>(type_dict, k)</span>
<span id="cb19-181"><a href="#cb19-181" aria-hidden="true" tabindex="-1"></a>                    type_dict[k] <span class="op">=</span> <span class="fu">union</span>(type_dict[k], v)</span>
<span id="cb19-182"><a href="#cb19-182" aria-hidden="true" tabindex="-1"></a>                <span class="cf">elseif</span> <span class="fu">haskey</span>(type_dict, YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k))</span>
<span id="cb19-183"><a href="#cb19-183" aria-hidden="true" tabindex="-1"></a>                    type_dict[YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k)] <span class="op">=</span> <span class="fu">union</span>(type_dict[YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k)], v)</span>
<span id="cb19-184"><a href="#cb19-184" aria-hidden="true" tabindex="-1"></a>                <span class="cf">else</span></span>
<span id="cb19-185"><a href="#cb19-185" aria-hidden="true" tabindex="-1"></a>                    type_dict[k] <span class="op">=</span> v</span>
<span id="cb19-186"><a href="#cb19-186" aria-hidden="true" tabindex="-1"></a>                <span class="cf">end</span></span>
<span id="cb19-187"><a href="#cb19-187" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb19-188"><a href="#cb19-188" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb19-189"><a href="#cb19-189" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb19-190"><a href="#cb19-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-191"><a href="#cb19-191" aria-hidden="true" tabindex="-1"></a>    <span class="fu">open</span>(<span class="fu">joinpath</span>(extract_valid_barcode_dir, <span class="st">"fq_read_stats.txt"</span>), <span class="st">"a"</span>) <span class="cf">do</span> io</span>
<span id="cb19-192"><a href="#cb19-192" aria-hidden="true" tabindex="-1"></a>        stat_str <span class="op">=</span> <span class="fu">string</span>(</span>
<span id="cb19-193"><a href="#cb19-193" aria-hidden="true" tabindex="-1"></a>            type_subdf.<span class="kw">type</span>[<span class="fl">1</span>], <span class="st">": "</span>, <span class="fu">length</span>(type_dict), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb19-194"><a href="#cb19-194" aria-hidden="true" tabindex="-1"></a>        <span class="fu">print</span>(io, stat_str)</span>
<span id="cb19-195"><a href="#cb19-195" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb19-196"><a href="#cb19-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-197"><a href="#cb19-197" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="fu">isempty</span>(top_dict)</span>
<span id="cb19-198"><a href="#cb19-198" aria-hidden="true" tabindex="-1"></a>        top_dict <span class="op">=</span> type_dict</span>
<span id="cb19-199"><a href="#cb19-199" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span></span>
<span id="cb19-200"><a href="#cb19-200" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> (k, v) <span class="kw">in</span> type_dict</span>
<span id="cb19-201"><a href="#cb19-201" aria-hidden="true" tabindex="-1"></a>            <span class="kw">global</span> top_dict</span>
<span id="cb19-202"><a href="#cb19-202" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="fu">haskey</span>(top_dict, k)</span>
<span id="cb19-203"><a href="#cb19-203" aria-hidden="true" tabindex="-1"></a>                top_dict[k] <span class="op">=</span> <span class="fu">union</span>(top_dict[k], v)</span>
<span id="cb19-204"><a href="#cb19-204" aria-hidden="true" tabindex="-1"></a>            <span class="cf">elseif</span> <span class="fu">haskey</span>(top_dict, YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k))</span>
<span id="cb19-205"><a href="#cb19-205" aria-hidden="true" tabindex="-1"></a>                top_dict[YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k)] <span class="op">=</span> <span class="fu">union</span>(top_dict[YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k)], v)</span>
<span id="cb19-206"><a href="#cb19-206" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span></span>
<span id="cb19-207"><a href="#cb19-207" aria-hidden="true" tabindex="-1"></a>                top_dict[k] <span class="op">=</span> v</span>
<span id="cb19-208"><a href="#cb19-208" aria-hidden="true" tabindex="-1"></a>            <span class="cf">end</span></span>
<span id="cb19-209"><a href="#cb19-209" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb19-210"><a href="#cb19-210" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb19-211"><a href="#cb19-211" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb19-212"><a href="#cb19-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-213"><a href="#cb19-213" aria-hidden="true" tabindex="-1"></a><span class="fu">open</span>(<span class="fu">joinpath</span>(extract_valid_barcode_dir, <span class="st">"fq_read_stats.txt"</span>), <span class="st">"a"</span>) <span class="cf">do</span> io</span>
<span id="cb19-214"><a href="#cb19-214" aria-hidden="true" tabindex="-1"></a>    stat_str <span class="op">=</span> <span class="fu">string</span>(</span>
<span id="cb19-215"><a href="#cb19-215" aria-hidden="true" tabindex="-1"></a>        <span class="st">"ALL: "</span>, <span class="fu">length</span>(top_dict), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb19-216"><a href="#cb19-216" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(io, stat_str)</span>
<span id="cb19-217"><a href="#cb19-217" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="attach-barcodes-to-cres" class="level3" data-number="2.4">
<h3 data-number="2.4" class="anchored" data-anchor-id="attach-barcodes-to-cres"><span class="header-section-number">2.4</span> Attach barcodes to CREs</h3>
<section id="solution-1-1" class="level4" data-number="2.4.1">
<h4 data-number="2.4.1" class="anchored" data-anchor-id="solution-1-1"><span class="header-section-number">2.4.1</span> Solution 1</h4>
<pre class="{r}"><code>library(vroom)
library(tidyverse)
library(ggprism)
library(YRUtils)

raw_bc_umi_pairs_file &lt;- "raw_bc_umi/raw_bc_umi_pairs.tsv"
# Here, we only keep those CRE-Barcode pairs, where each barcode is assigned to only one CRE
cre_bc_file &lt;- "/data/users/dell/mpra/link_barcode_to_cre/final_result/non_redundant_cre_bc_pairs.tsv"
output_dir &lt;- "cre_bc_count"

dir.create(output_dir)
cre_bc_df &lt;- vroom(cre_bc_file)
raw_bc_umi_pairs_df &lt;- vroom(raw_bc_umi_pairs_file)

# Count barcodes based on UMIs
biorep_bc_count_df &lt;- raw_bc_umi_pairs_df %&gt;%
    mutate(rep_sample = gsub("_part[0-9]+$", "", tech_sample)) %&gt;%
    select(-all_of(c("tech_sample"))) %&gt;%
    distinct() %&gt;%
    group_by(rep_sample, barcode) %&gt;%
    # Count the number of occurrences of each barcode in each biological replicate
    count(name = "barcode_count") %&gt;%
    ungroup() %&gt;%
    group_by(rep_sample) %&gt;%
    arrange(desc(barcode_count), .by_group = TRUE) %&gt;%
    ungroup()
vroom_write(biorep_bc_count_df, file = file.path(output_dir, "biorep_bc_count.tsv"))

# Attach CREs to barcodes
biorep_cre_bc_count_df &lt;- biorep_bc_count_df %&gt;%
    inner_join(cre_bc_df, by = "barcode", relationship = "many-to-many") %&gt;%
    group_by(rep_sample) %&gt;%
    arrange(desc(barcode_count), .by_group = TRUE) %&gt;%
    ungroup()
vroom_write(biorep_cre_bc_count_df, file = file.path(output_dir, "biorep_cre_bc_count.tsv"))

# Count the number of unique barcodes detected in each biological replicate
bc_num_each_biorep &lt;- biorep_bc_count_df %&gt;%
    group_by(rep_sample) %&gt;%
    count(name = "barcode_num") %&gt;%
    ungroup()
vroom_write(bc_num_each_biorep, file = file.path(output_dir, "bc_num_each_biorep.tsv"))

# The distribution of the number of barcodes belonging to each CRE
# The distribution of the number of CREs belonging to each barcode
left_cre_bc_df &lt;- biorep_cre_bc_count_df %&gt;%
    select(barcode, cre) %&gt;%
    distinct()
bc_count_df &lt;- left_cre_bc_df %&gt;%
    group_by(barcode) %&gt;%
    count() %&gt;%
    rename(cre_bc = barcode) %&gt;%
    mutate(type = "BC")
cre_count_df &lt;- left_cre_bc_df %&gt;%
    group_by(cre) %&gt;%
    count() %&gt;%
    rename(cre_bc = cre) %&gt;%
    mutate(type = if_else(str_detect(cre_bc, "^CTL"), "CTL", "CRE"))
count_df &lt;- bind_rows(cre_count_df, bc_count_df)

cre_quantiles &lt;- quantile(count_df$n[count_df$type %in% c("CTL", "CRE")], probs = seq(0, 1, 0.1))
cre_only_quantiles &lt;- quantile(count_df$n[count_df$type == "CRE"], probs = seq(0, 1, 0.1))
ctl_only_quantiles &lt;- quantile(count_df$n[count_df$type == "CTL"], probs = seq(0, 1, 0.1))
bc_quantiles &lt;- quantile(count_df$n[count_df$type == "BC"], probs = seq(0, 1, 0.1))
type_nums &lt;- table(count_df$type)

paste0(
    "1. CRE/CTL quantiles (", type_nums["CTL"] + type_nums["CRE"], "): \n",
    paste0(paste0(names(cre_quantiles), "\t", cre_quantiles), collapse = "\n"), "\n\n",
    "2. CRE only quantiles (", type_nums["CRE"], "): \n",
    paste0(paste0(names(cre_only_quantiles), "\t", cre_only_quantiles), collapse = "\n"), "\n\n",
    "3. CTL only quantiles (", type_nums["CTL"], "): \n",
    paste0(paste0(names(ctl_only_quantiles), "\t", ctl_only_quantiles), collapse = "\n"), "\n\n",
    "4. BC quantiles (", type_nums["BC"], "): \n",
    paste0(paste0(names(bc_quantiles), "\t", bc_quantiles), collapse = "\n")
) %&gt;% vroom_write_lines(file = file.path(output_dir, "quantiles.txt"))

p &lt;- ggplot(count_df, aes(type, log2(n), fill = type, color = type)) +
    geom_violin(scale = "width", alpha = 0.25, trim = TRUE) +
    geom_boxplot(width = 0.2, outliers = FALSE, alpha = 0.25) +
    scale_y_continuous(expand = expansion(mult = c(0.05, 0))) +
    labs(
        x = "Sequence Type",
        y = "log2(Count)"
    ) +
    theme_prism(base_size = 20, base_family = "Arial", border = FALSE) +
    theme(legend.position = "none")
ppreview(p, file = file.path(output_dir, "violin.pdf"))</code></pre>
</section>
<section id="solution-2-1" class="level4" data-number="2.4.2">
<h4 data-number="2.4.2" class="anchored" data-anchor-id="solution-2-1"><span class="header-section-number">2.4.2</span> Solution 2</h4>
<div id="702a972d" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">CSV</span>, <span class="bu">DataFrames</span>, <span class="bu">YRUtils</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>extract_valid_barcode_dir <span class="op">=</span> <span class="st">"extract_valid_barcode"</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>output_dir <span class="op">=</span> <span class="st">"cre_bc_count"</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>cre_bc_file <span class="op">=</span> <span class="st">"/data/users/dell/mpra/link_barcode_to_cre/final_result/non_redundant_cre_bc_pairs.tsv"</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(output_dir)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>cre_bc_df <span class="op">=</span> CSV.<span class="fu">read</span>(cre_bc_file, DataFrame; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>barcode_count_file_name_pattern <span class="op">=</span> <span class="st">r"</span><span class="sc">.+</span><span class="st">/</span><span class="ch">(?</span><span class="st">&lt;id&gt;</span><span class="ch">[</span><span class="st">a-zA-Z0-9</span><span class="ch">]</span><span class="sc">+</span><span class="ch">)</span><span class="st">_</span><span class="ch">(?</span><span class="st">&lt;type&gt;RNA</span><span class="ch">|</span><span class="st">DNA</span><span class="ch">)</span><span class="st">_</span><span class="ch">(?</span><span class="st">&lt;rep&gt;rep</span><span class="ch">[</span><span class="st">0-9</span><span class="ch">]</span><span class="sc">+</span><span class="ch">)</span><span class="sc">\.</span><span class="st">barcode_count</span><span class="sc">\.</span><span class="ch">(</span><span class="st">txt</span><span class="ch">|</span><span class="st">tsv</span><span class="ch">)</span><span class="sc">$</span><span class="st">"</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>barcode_count_files<span class="st"> = YRUtils</span><span class="sc">.</span><span class="st">BioUtils</span><span class="sc">.</span><span class="st">list_files</span><span class="ch">(</span><span class="st">extract_valid_barcode_dir, barcode_count_file_name_pattern; recursive=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>ms <span class="op">=</span> <span class="fu">match</span>.(barcode_count_file_name_pattern, barcode_count_files)</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>metadata_vec <span class="op">=</span> <span class="fu">Vector</span><span class="dt">{NTuple{4,String}}</span>(<span class="cn">undef</span>, <span class="fu">length</span>(ms))</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(ms)</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>    metadata_vec[i] <span class="op">=</span> (ms[i].match, ms[i][<span class="st">"id"</span>], ms[i][<span class="st">"type"</span>], ms[i][<span class="st">"rep"</span>])</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> <span class="fu">unique</span>(<span class="fu">DataFrame</span>(metadata_vec, [<span class="op">:</span>file, <span class="op">:</span>id, <span class="op">:</span>type, <span class="op">:</span>rep]))</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> <span class="fu">transform</span>(</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>    df,</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>    [<span class="op">:</span>id, <span class="op">:</span>type, <span class="op">:</span>rep] <span class="op">=&gt;</span> <span class="fu">ByRow</span>((id, <span class="kw">type</span>, rep) <span class="op">-&gt;</span> <span class="fu">join</span>([id, type, rep], <span class="st">"_"</span>)) <span class="op">=&gt;</span> <span class="op">:</span>rep_sample,</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>    [<span class="op">:</span>id, <span class="op">:</span>type] <span class="op">=&gt;</span> <span class="fu">ByRow</span>((id, <span class="kw">type</span>) <span class="op">-&gt;</span> <span class="fu">join</span>([id, type], <span class="st">"_"</span>)) <span class="op">=&gt;</span> <span class="op">:</span>type_sample</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(<span class="fu">joinpath</span>(output_dir, <span class="st">"barcode_count_file_metadata.tsv"</span>), df; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a>cre_barcode_count_df <span class="op">=</span> <span class="fu">crossjoin</span>(<span class="fu">DataFrame</span>(cre<span class="op">=</span><span class="dt">String</span>[], barcode<span class="op">=</span><span class="dt">String</span>[], value_barcode<span class="op">=</span><span class="dt">String</span>[], count<span class="op">=</span><span class="dt">Int64</span>[], CPM<span class="op">=</span><span class="dt">Float64</span>[]), <span class="fu">empty</span>(df))</span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> row_subdf <span class="kw">in</span> <span class="fu">eachrow</span>(df)</span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>    <span class="kw">global</span> cre_barcode_count_df</span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a>    barcode_count_df <span class="op">=</span> CSV.<span class="fu">read</span>(row_subdf.file, DataFrame; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>    barcode_count_df <span class="op">=</span> <span class="fu">crossjoin</span>(barcode_count_df, <span class="fu">DataFrame</span>(row_subdf))</span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a>    barcode_count_df.CPM <span class="op">=</span> barcode_count_df.count <span class="op">*</span> <span class="fl">1000000</span> <span class="op">/</span> <span class="fu">sum</span>(barcode_count_df.count)</span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a>    in_barcode_set <span class="op">=</span> <span class="fu">in</span>(<span class="fu">Set</span>(barcode_count_df.barcode))</span>
<span id="cb21-32"><a href="#cb21-32" aria-hidden="true" tabindex="-1"></a>    cre_barcode_dict <span class="op">=</span> <span class="fu">Dict</span>(k <span class="op">=&gt;</span> <span class="st">""</span> <span class="cf">for</span> k <span class="kw">in</span> cre_bc_df.barcode)</span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a>    <span class="bu">Threads</span>.<span class="pp">@threads</span> <span class="cf">for</span> k <span class="kw">in</span> cre_bc_df.barcode</span>
<span id="cb21-34"><a href="#cb21-34" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="fu">in_barcode_set</span>(k)</span>
<span id="cb21-35"><a href="#cb21-35" aria-hidden="true" tabindex="-1"></a>            cre_barcode_dict[k] <span class="op">=</span> k</span>
<span id="cb21-36"><a href="#cb21-36" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elseif</span> <span class="fu">in_barcode_set</span>(YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k))</span>
<span id="cb21-37"><a href="#cb21-37" aria-hidden="true" tabindex="-1"></a>            cre_barcode_dict[k] <span class="op">=</span> YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(k)</span>
<span id="cb21-38"><a href="#cb21-38" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb21-39"><a href="#cb21-39" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb21-40"><a href="#cb21-40" aria-hidden="true" tabindex="-1"></a>    bridge_barcode_df <span class="op">=</span> <span class="fu">DataFrame</span>([(key_barcode<span class="op">=</span>k, value_barcode<span class="op">=</span>v) for (k, v) <span class="kw">in</span> cre_barcode_dict])</span>
<span id="cb21-41"><a href="#cb21-41" aria-hidden="true" tabindex="-1"></a>    bridge_barcode_df <span class="op">=</span> <span class="fu">filter</span>(<span class="op">:</span>value_barcode <span class="op">=&gt;</span> x <span class="op">-&gt;</span> x <span class="op">!=</span> <span class="st">""</span>, bridge_barcode_df)</span>
<span id="cb21-42"><a href="#cb21-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-43"><a href="#cb21-43" aria-hidden="true" tabindex="-1"></a>    barcode_count_df <span class="op">=</span> <span class="fu">innerjoin</span>(bridge_barcode_df, barcode_count_df, on<span class="op">=:</span>value_barcode <span class="op">=&gt;</span> <span class="op">:</span>barcode)</span>
<span id="cb21-44"><a href="#cb21-44" aria-hidden="true" tabindex="-1"></a>    barcode_count_df <span class="op">=</span> <span class="fu">innerjoin</span>(cre_bc_df, barcode_count_df, on<span class="op">=:</span>barcode <span class="op">=&gt;</span> <span class="op">:</span>key_barcode)</span>
<span id="cb21-45"><a href="#cb21-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">append!</span>(cre_barcode_count_df, barcode_count_df; cols<span class="op">=:</span>setequal)</span>
<span id="cb21-46"><a href="#cb21-46" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb21-47"><a href="#cb21-47" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(<span class="fu">joinpath</span>(output_dir, <span class="st">"cre_barcode_count.tsv"</span>), cre_barcode_count_df; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
</section>
<section id="identify-active-cres" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="identify-active-cres"><span class="header-section-number">3</span> Identify active CREs</h2>
<section id="q95-threshold-method" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="q95-threshold-method"><span class="header-section-number">3.1</span> Q95 threshold method</h3>
<pre class="{r}"><code>library(vroom)
library(tidyverse)
library(YRUtils)

cre_bc_count_file &lt;- "cre_bc_count/cre_barcode_count.tsv"
output_dir &lt;- "identify_active.Q95"

dir.create(output_dir)
cre_bc_count_df &lt;- vroom(cre_bc_count_file)
cre_bc_cpm_df &lt;- pivot_wider(cre_bc_count_df, id_cols = c("rep", "cre", "barcode"), names_from = "type", values_from = "CPM", values_fill = 0) %&gt;%
    filter(DNA &gt; 0)

cre_bc_qs &lt;- cre_bc_cpm_df %&gt;%
    group_by(rep, cre) %&gt;%
    reframe(n = n()) %&gt;%
    pull(n) %&gt;%
    quantile(probs = seq(0, 1, 0.05))
sink(file = file.path(output_dir, "cre_bc_quantiles.txt"))
print(cre_bc_qs)
sink()

cre_bc_cpm_df &lt;- cre_bc_cpm_df %&gt;%
    group_by(rep, cre) %&gt;%
    reframe(DNA = sum(DNA), RNA = sum(RNA)) %&gt;%
    mutate(
        activity = RNA / DNA,
        cre_type = if_else(str_detect(cre, "^PSCE\\d+(_\\d+)?$"), "PSCE",
            if_else(str_detect(cre, "^CTL\\d+(_\\d+)?"), "CTL", "NA")
        ),
        cre_type = factor(cre_type, levels = c("CTL", "PSCE")),
        rep = factor(rep, levels = c("rep1", "rep2"))
    )
if (any(is.na(cre_bc_cpm_df$cre_type))) {
    stop("unexpected CRE type found")
}

p_density &lt;- ggplot(cre_bc_cpm_df, aes(log2(activity + 1), fill = cre_type, color = cre_type)) +
    geom_density(alpha = 0.1) +
    scale_y_continuous(expand = expansion(mult = 0.02)) +
    facet_grid(. ~ rep) +
    labs(color = "CRE Type", fill = "CRE Type") +
    theme_classic(base_size = 20, base_family = "Arial") +
    theme(strip.background = element_blank())
ppreview(p_density, file = file.path(output_dir, "density.pdf"))

cre_bc_cpm_ls &lt;- split(cre_bc_cpm_df, cre_bc_cpm_df$rep)
cre_bc_cpm_ls &lt;- lapply(cre_bc_cpm_ls, function(df) {
    ctl_activity_qs &lt;- df %&gt;%
        filter(cre_type == "CTL") %&gt;%
        pull(activity) %&gt;%
        quantile(probs = seq(0, 1, 0.05))
    psce_activity_qs &lt;- df %&gt;%
        filter(cre_type == "PSCE") %&gt;%
        pull(activity) %&gt;%
        quantile(probs = seq(0, 1, 0.05))
    psce_df &lt;- df %&gt;%
        filter(cre_type == "PSCE") %&gt;%
        mutate(is_active = if_else(activity &gt; ctl_activity_qs["95%"], TRUE, FALSE)) %&gt;%
        arrange(desc(is_active), desc(activity))
    p_violin &lt;- ggplot(df, aes(
        cre_type, activity,
        fill = cre_type, color = cre_type
    )) +
        geom_violin(trim = TRUE, scale = "width", alpha = 0.5) +
        scale_y_continuous(expand = expansion(mult = 0.02)) +
        labs(x = NULL, y = "Activity") +
        theme_classic(base_family = "Arial", base_size = 20) +
        theme(legend.position = "none")
    p_density &lt;- ggplot(df, aes(log2(activity + 1), fill = cre_type, color = cre_type)) +
        geom_density(alpha = 0.1) +
        geom_vline(xintercept = ctl_activity_qs["95%"], color = "grey", linetype = "dashed") +
        scale_y_continuous(expand = expansion(mult = 0.02)) +
        labs(color = "CRE Type", fill = "CRE Type") +
        theme_classic(base_size = 20, base_family = "Arial")

    list(
        df = df %&gt;% arrange(cre_type, desc(activity)),
        ctl_activity_qs = ctl_activity_qs,
        psce_activity_qs = psce_activity_qs,
        psce_df = psce_df,
        violin = p_violin,
        density = p_density
    )
})
for (rep in names(cre_bc_cpm_ls)) {
    vroom_write(cre_bc_cpm_ls[[rep]]$df, file = file.path(output_dir, paste0(rep, ".df.tsv")))
    vroom_write(cre_bc_cpm_ls[[rep]]$psce_df, file = file.path(output_dir, paste0(rep, ".psce_df.tsv")))
    sink(file = file.path(output_dir, paste0(rep, ".ctl_activity_quantiles.txt")))
    print(cre_bc_cpm_ls[[rep]]$ctl_activity_qs)
    sink()
    sink(file = file.path(output_dir, paste0(rep, ".psce_activity_quantiles.txt")))
    print(cre_bc_cpm_ls[[rep]]$psce_activity_qs)
    sink()
    ppreview(cre_bc_cpm_ls[[rep]]$violin, file = file.path(output_dir, paste0(rep, ".violin.pdf")))
    ppreview(cre_bc_cpm_ls[[rep]]$density, file = file.path(output_dir, paste0(rep, ".density.pdf")))
}

join_df &lt;- full_join(
    cre_bc_cpm_ls$rep1$psce_df %&gt;%
        filter(is_active) %&gt;%
        select(cre, activity, is_active),
    cre_bc_cpm_ls$rep2$psce_df %&gt;%
        filter(is_active) %&gt;%
        select(cre, activity, is_active),
    by = "cre", suffix = c(".rep1", ".rep2")
) %&gt;%
    mutate(both_are_active = if_else(is.na(is_active.rep1), "rep2",
        if_else(is.na(is_active.rep2), "rep1", "both")
    )) %&gt;%
    arrange(both_are_active)
vroom_write(join_df, file = file.path(output_dir, "rep1_rep2.psce_df.tsv"))</code></pre>
</section>
<section id="mpranalyze-method" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="mpranalyze-method"><span class="header-section-number">3.2</span> MPRAnalyze method</h3>
<pre class="{r}"><code>library(MPRAnalyze)
library(vroom)
library(tidyverse)
library(YRUtils)
library(magrittr)

cre_bc_count_file &lt;- "cre_bc_count/cre_barcode_count.tsv"
output_dir &lt;- "identify_active.MPRAnalyze"

dir.create(output_dir)
cre_bc_count_df &lt;- vroom(cre_bc_count_file)
cre_bc_count_df &lt;- pivot_wider(cre_bc_count_df, id_cols = c("rep", "cre", "barcode"), names_from = "type", values_from = "count", values_fill = 0) %&gt;%
    filter(DNA &gt; 0) %&gt;%
    group_by(rep, cre) %&gt;%
    reframe(DNA = sum(DNA), RNA = sum(RNA)) %&gt;%
    mutate(
        cre_type = if_else(str_detect(cre, "^PSCE\\d+(_\\d+)?$"), "PSCE",
            if_else(str_detect(cre, "^CTL\\d+(_\\d+)?"), "CTL", "NA")
        )
    )
dna_count_df &lt;- pivot_wider(cre_bc_count_df, id_cols = c("cre", "cre_type"), names_from = "rep", values_from = "DNA", values_fill = 0) %&gt;%
    arrange(cre) %&gt;%
    as.data.frame() %&gt;%
    set_rownames(.[["cre"]])
rna_count_df &lt;- pivot_wider(cre_bc_count_df, id_cols = c("cre", "cre_type"), names_from = "rep", values_from = "RNA", values_fill = 0) %&gt;%
    arrange(cre) %&gt;%
    as.data.frame() %&gt;%
    set_rownames(.[["cre"]])
if (!all(row.names(dna_count_df) == row.names(rna_count_df))) {
    stop("the row names of DNA are not identical to the row names of RNA")
}

col_names &lt;- c("rep1", "rep2")
obj &lt;- MpraObject(
    dnaCounts = as.matrix(dna_count_df[, col_names]),
    rnaCounts = as.matrix(rna_count_df[, col_names]),
    dnaAnnot = data.frame(replicate = col_names, row.names = col_names),
    rnaAnnot = data.frame(replicate = col_names, row.names = col_names),
    controls = dna_count_df$cre_type == "CTL"
)
obj &lt;- estimateDepthFactors(obj, lib.factor = "replicate", which.lib = "dna", depth.estimator = "uq")
obj &lt;- estimateDepthFactors(obj, lib.factor = "replicate", which.lib = "rna", depth.estimator = "uq")
obj &lt;- analyzeQuantification(obj, dnaDesign = ~replicate, rnaDesign = ~1)

alpha_df &lt;- getAlpha(obj) %&gt;%
    mutate(
        cre = row.names(.),
        cre_type = if_else(str_detect(cre, "^PSCE\\d+(_\\d+)?$"), "PSCE",
            if_else(str_detect(cre, "^CTL\\d+(_\\d+)?"), "CTL", "NA")
        ),
        cre_type = factor(cre_type, levels = c("CTL", "PSCE"))
    )
p_violin &lt;- ggplot(alpha_df, aes(
    cre_type, alpha,
    fill = cre_type, color = cre_type
)) +
    geom_violin(trim = TRUE, scale = "width", alpha = 0.5) +
    scale_y_continuous(expand = expansion(mult = 0.02)) +
    labs(x = NULL, y = "Activity") +
    theme_classic(base_family = "Arial", base_size = 20) +
    theme(legend.position = "none")
ppreview(p_violin, file = file.path(output_dir, "alpha.violin.pdf"))
p_density &lt;- ggplot(alpha_df, aes(alpha, fill = cre_type, color = cre_type)) +
    geom_density(alpha = 0.1) +
    scale_y_continuous(expand = expansion(mult = 0.02)) +
    labs(color = "CRE Type", fill = "CRE Type") +
    theme_classic(base_size = 20, base_family = "Arial")
ppreview(p_density, file = file.path(output_dir, "alpha.density.pdf"))

res_df &lt;- testEmpirical(obj, statistic = alpha_df$alpha)
res_df &lt;- bind_cols(alpha_df, res_df) %&gt;%
    mutate(
        pval.empirical.adj = p.adjust(pval.empirical, method = "fdr"),
        pval.mad.adj = p.adjust(pval.mad, method = "fdr"),
        pval.zscore.adj = p.adjust(pval.zscore, method = "fdr")
    ) %&gt;%
    filter(cre_type == "PSCE") %&gt;%
    arrange(pval.mad.adj) %&gt;%
    mutate(is_active = if_else(pval.mad.adj &lt; 0.05, TRUE, FALSE))
vroom_write(res_df, file = file.path(output_dir, "psce_df.tsv"))</code></pre>
</section>
</section>
<section id="peak-annotation" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="peak-annotation"><span class="header-section-number">4</span> Peak annotation</h2>
<pre class="{r}"><code>library(vroom)
library(tidyverse)
library(ggVennDiagram)
library(YRUtils)
library(ChIPseeker)
library(AnnotationDbi)
library(patchwork)

mpt_file &lt;- "/data/biodatabase/species/mm10/genome/anno/gencode.vM21.trna.ercc.phix.gtf.gz.gene_id_name_mapping_table.tsv"
txdb_file &lt;- "/data/biodatabase/species/mm10/genome/anno/gencode.vM21.primary_assembly.annotation_UCSC_names.TxDb.sqlite"
mpra_test_file &lt;- "/data/users/dell/mpra/link_barcode_to_cre/enzyme_v20250430/ref/2w_library.165bp.no_enzyme_cutting_sites.dealed.tsv"
rep1_file &lt;- "identify_active.Q95/rep1.psce_df.tsv"
rep2_file &lt;- "identify_active.Q95/rep2.psce_df.tsv"
mpranalyze_file &lt;- "identify_active.MPRAnalyze/psce_df.tsv"
output_dir &lt;- "identify_active.overlap"

dir.create(output_dir)
mpt_df &lt;- vroom(mpt_file) %&gt;%
    dplyr::select(gene_id, gene_name, gene_biotype, gene_version, gene_level)
txdb &lt;- loadDb(txdb_file)
mpra_test_df &lt;- vroom(mpra_test_file) %&gt;%
    dplyr::select(seqname, start, end, PSCE, PSCE_new_id) %&gt;%
    mutate(id = paste0(seqname, ":", start, "-", end))
rep1_df &lt;- vroom(rep1_file) %&gt;%
    filter(is_active) %&gt;%
    dplyr::select(-rep, -cre_type)
rep2_df &lt;- vroom(rep2_file) %&gt;%
    filter(is_active) %&gt;%
    dplyr::select(-rep, -cre_type)
mpranalyze_df &lt;- vroom(mpranalyze_file) %&gt;%
    filter(is_active) %&gt;%
    dplyr::select(-cre_type)
names(mpranalyze_df) &lt;- paste0(names(mpranalyze_df), ".mpranalyze")

join_df &lt;- full_join(rep1_df, rep2_df, by = "cre", suffix = c(".rep1", ".rep2")) %&gt;%
    full_join(mpranalyze_df, by = c("cre" = "cre.mpranalyze")) %&gt;%
    arrange(pval.mad.adj.mpranalyze) %&gt;%
    inner_join(mpra_test_df, by = c("cre" = "PSCE_new_id"))
vroom_write(join_df, file = file.path(output_dir, "overlap.df.tsv"))

ls &lt;- list(rep1 = rep1_df$cre, rep2 = rep2_df$cre, mpranalyze = mpranalyze_df$cre.mpranalyze)
p_venn &lt;- ggVennDiagram(ls, label_alpha = 0, label = "count", label_size = 12) +
    scale_fill_gradient(low = "grey90", high = "#FF7256") +
    xlim(c(-10, 10)) +
    theme(legend.position = "none")
ppreview(p_venn, file = file.path(output_dir, "overlap.venn.pdf"))

gr &lt;- join_df[, c("seqname", "start", "end", "cre")] %&gt;%
    as("GRanges")
peak_anno &lt;- annotatePeak(gr, tssRegion = c(-3000, 3000), TxDb = txdb)

p_bar &lt;- plotAnnoBar(peak_anno) +
    labs(title = NULL) +
    theme(text = element_text(family = "Arial", size = 20))
ppreview(p_bar, file = file.path(output_dir, "peak_anno.feature_distribution.pdf"))

p_tss &lt;- plotDistToTSS(peak_anno) +
    labs(title = NULL) +
    theme(text = element_text(family = "Arial", size = 20))
ppreview(p_tss, file = file.path(output_dir, "peak_anno.distance_to_tss.pdf"))

peak_anno_df &lt;- as.data.frame(peak_anno) %&gt;%
    dplyr::select(cre, annotation, geneId, distanceToTSS) %&gt;%
    inner_join(mpt_df, by = c("geneId" = "gene_id"))
final_df &lt;- inner_join(join_df, peak_anno_df, by = "cre")
vroom_write(final_df, file = file.path(output_dir, "overlap.peak_anno.df.tsv"))

gene_list_df &lt;- final_df %&gt;%
    dplyr::select(gene_name) %&gt;%
    distinct() %&gt;%
    mutate(cluster = "active_psce")
vroom_write(gene_list_df, file.path(output_dir, "overlap.peak_anno.gene_list.df.tsv"))</code></pre>
</section>
<section id="appendix" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="appendix"><span class="header-section-number">5</span> Appendix</h2>
<section id="identify-active-barcodes" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="identify-active-barcodes"><span class="header-section-number">5.1</span> Identify active barcodes</h3>
<section id="extract-barcode-sample-matrix" class="level4" data-number="5.1.1">
<h4 data-number="5.1.1" class="anchored" data-anchor-id="extract-barcode-sample-matrix"><span class="header-section-number">5.1.1</span> Extract barcode-sample matrix</h4>
<div id="4676a754" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">CSV</span>, <span class="bu">DataFrames</span>, <span class="bu">YRUtils</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>extract_valid_barcode_dir <span class="op">=</span> <span class="st">"extract_valid_barcode"</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>output_dir <span class="op">=</span> <span class="st">"bc_count"</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkpath</span>(output_dir)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>barcode_count_file_name_pattern <span class="op">=</span> <span class="st">r"</span><span class="sc">.+</span><span class="st">/</span><span class="ch">(?</span><span class="st">&lt;id&gt;</span><span class="ch">[</span><span class="st">a-zA-Z0-9</span><span class="ch">]</span><span class="sc">+</span><span class="ch">)</span><span class="st">_</span><span class="ch">(?</span><span class="st">&lt;type&gt;RNA</span><span class="ch">|</span><span class="st">DNA</span><span class="ch">)</span><span class="st">_</span><span class="ch">(?</span><span class="st">&lt;rep&gt;rep</span><span class="ch">[</span><span class="st">0-9</span><span class="ch">]</span><span class="sc">+</span><span class="ch">)</span><span class="sc">\.</span><span class="st">barcode_count</span><span class="sc">\.</span><span class="ch">(</span><span class="st">txt</span><span class="ch">|</span><span class="st">tsv</span><span class="ch">)</span><span class="sc">$</span><span class="st">"</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>barcode_count_files<span class="st"> = YRUtils</span><span class="sc">.</span><span class="st">BioUtils</span><span class="sc">.</span><span class="st">list_files</span><span class="ch">(</span><span class="st">extract_valid_barcode_dir, barcode_count_file_name_pattern; recursive=false, full_name=true</span><span class="ch">)</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>ms <span class="op">=</span> <span class="fu">match</span>.(barcode_count_file_name_pattern, barcode_count_files)</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>metadata_vec <span class="op">=</span> <span class="fu">Vector</span><span class="dt">{NTuple{4,String}}</span>(<span class="cn">undef</span>, <span class="fu">length</span>(ms))</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span>(ms)</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>    metadata_vec[i] <span class="op">=</span> (ms[i].match, ms[i][<span class="st">"id"</span>], ms[i][<span class="st">"type"</span>], ms[i][<span class="st">"rep"</span>])</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> <span class="fu">unique</span>(<span class="fu">DataFrame</span>(metadata_vec, [<span class="op">:</span>file, <span class="op">:</span>id, <span class="op">:</span>type, <span class="op">:</span>rep]))</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> <span class="fu">transform</span>(</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>    df,</span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>    [<span class="op">:</span>id, <span class="op">:</span>type, <span class="op">:</span>rep] <span class="op">=&gt;</span> <span class="fu">ByRow</span>((id, <span class="kw">type</span>, rep) <span class="op">-&gt;</span> <span class="fu">join</span>([id, type, rep], <span class="st">"_"</span>)) <span class="op">=&gt;</span> <span class="op">:</span>rep_sample,</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>    [<span class="op">:</span>id, <span class="op">:</span>type] <span class="op">=&gt;</span> <span class="fu">ByRow</span>((id, <span class="kw">type</span>) <span class="op">-&gt;</span> <span class="fu">join</span>([id, type], <span class="st">"_"</span>)) <span class="op">=&gt;</span> <span class="op">:</span>type_sample</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(<span class="fu">joinpath</span>(output_dir, <span class="st">"barcode_count_file_metadata.tsv"</span>), df; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>bridge_barcode_dict <span class="op">=</span> <span class="fu">Dict</span><span class="dt">{String,Dict{String,String}}</span>()</span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> row_subdf <span class="kw">in</span> <span class="fu">eachrow</span>(df)</span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a>    <span class="kw">global</span> bridge_barcode_dict</span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>    barcode_count_df <span class="op">=</span> CSV.<span class="fu">read</span>(row_subdf.file, DataFrame; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> barcode <span class="kw">in</span> barcode_count_df.barcode</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a>        <span class="kw">global</span> bridge_barcode_dict</span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="fu">haskey</span>(bridge_barcode_dict, barcode)</span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>            bridge_barcode_dict[barcode][row_subdf.rep_sample] <span class="op">=</span> barcode</span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elseif</span> <span class="fu">haskey</span>(bridge_barcode_dict, YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(barcode))</span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a>            bridge_barcode_dict[YRUtils.BioUtils.<span class="fu">rev_com_dna_seq</span>(barcode)][row_subdf.rep_sample] <span class="op">=</span> barcode</span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span></span>
<span id="cb25-33"><a href="#cb25-33" aria-hidden="true" tabindex="-1"></a>            bridge_barcode_dict[barcode] <span class="op">=</span> <span class="fu">Dict</span>(k <span class="op">=&gt;</span> <span class="st">""</span> <span class="cf">for</span> k <span class="kw">in</span> df.rep_sample)</span>
<span id="cb25-34"><a href="#cb25-34" aria-hidden="true" tabindex="-1"></a>            bridge_barcode_dict[barcode][row_subdf.rep_sample] <span class="op">=</span> barcode</span>
<span id="cb25-35"><a href="#cb25-35" aria-hidden="true" tabindex="-1"></a>        <span class="cf">end</span></span>
<span id="cb25-36"><a href="#cb25-36" aria-hidden="true" tabindex="-1"></a>    <span class="cf">end</span></span>
<span id="cb25-37"><a href="#cb25-37" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb25-38"><a href="#cb25-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-39"><a href="#cb25-39" aria-hidden="true" tabindex="-1"></a>all_barcode_count_df <span class="op">=</span> <span class="fu">crossjoin</span>(<span class="fu">DataFrame</span>(key_barcode<span class="op">=</span><span class="dt">String</span>[], value_barcode<span class="op">=</span><span class="dt">String</span>[], count<span class="op">=</span><span class="dt">Int64</span>[], CPM<span class="op">=</span><span class="dt">Float64</span>[]), <span class="fu">empty</span>(df))</span>
<span id="cb25-40"><a href="#cb25-40" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> row_subdf <span class="kw">in</span> <span class="fu">eachrow</span>(df)</span>
<span id="cb25-41"><a href="#cb25-41" aria-hidden="true" tabindex="-1"></a>    <span class="kw">global</span> bridge_barcode_dict</span>
<span id="cb25-42"><a href="#cb25-42" aria-hidden="true" tabindex="-1"></a>    barcode_count_df <span class="op">=</span> CSV.<span class="fu">read</span>(row_subdf.file, DataFrame; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb25-43"><a href="#cb25-43" aria-hidden="true" tabindex="-1"></a>    barcode_count_df.CPM <span class="op">=</span> barcode_count_df.count <span class="op">*</span> <span class="fl">1000000</span> <span class="op">/</span> <span class="fu">sum</span>(barcode_count_df.count)</span>
<span id="cb25-44"><a href="#cb25-44" aria-hidden="true" tabindex="-1"></a>    barcode_count_df <span class="op">=</span> <span class="fu">crossjoin</span>(barcode_count_df, <span class="fu">DataFrame</span>(row_subdf))</span>
<span id="cb25-45"><a href="#cb25-45" aria-hidden="true" tabindex="-1"></a>    bridge_barcode_df <span class="op">=</span> <span class="fu">DataFrame</span>([(key_barcode<span class="op">=</span>k, value_barcode<span class="op">=</span>v[row_subdf.rep_sample]) for (k, v) <span class="kw">in</span> bridge_barcode_dict if v[row_subdf.rep_sample] <span class="op">!=</span> <span class="st">""</span>])</span>
<span id="cb25-46"><a href="#cb25-46" aria-hidden="true" tabindex="-1"></a>    barcode_count_df <span class="op">=</span> <span class="fu">innerjoin</span>(bridge_barcode_df, barcode_count_df, on<span class="op">=:</span>value_barcode <span class="op">=&gt;</span> <span class="op">:</span>barcode)</span>
<span id="cb25-47"><a href="#cb25-47" aria-hidden="true" tabindex="-1"></a>    <span class="fu">append!</span>(all_barcode_count_df, barcode_count_df; cols<span class="op">=:</span>setequal)</span>
<span id="cb25-48"><a href="#cb25-48" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb25-49"><a href="#cb25-49" aria-hidden="true" tabindex="-1"></a>wide_barcode_count_df <span class="op">=</span> <span class="fu">unstack</span>(all_barcode_count_df, <span class="op">:</span>key_barcode, <span class="op">:</span>rep_sample, <span class="op">:</span>count; fill<span class="op">=</span><span class="fl">0</span>)</span>
<span id="cb25-50"><a href="#cb25-50" aria-hidden="true" tabindex="-1"></a>filtered_wide_barcode_count_df <span class="op">=</span> <span class="fu">filter</span>([<span class="op">:</span>HMC3MPRA_DNA_rep1, <span class="op">:</span>HMC3MPRA_DNA_rep2] <span class="op">=&gt;</span> (x, y) <span class="op">-&gt;</span> x <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;&amp;</span> y <span class="op">&gt;</span> <span class="fl">0</span>, wide_barcode_count_df)</span>
<span id="cb25-51"><a href="#cb25-51" aria-hidden="true" tabindex="-1"></a>wide_barcode_cpm_df <span class="op">=</span> <span class="fu">unstack</span>(all_barcode_count_df, <span class="op">:</span>key_barcode, <span class="op">:</span>rep_sample, <span class="op">:</span>CPM; fill<span class="op">=</span><span class="fl">0</span>)</span>
<span id="cb25-52"><a href="#cb25-52" aria-hidden="true" tabindex="-1"></a>filtered_wide_barcode_cpm_df <span class="op">=</span> <span class="fu">filter</span>([<span class="op">:</span>HMC3MPRA_DNA_rep1, <span class="op">:</span>HMC3MPRA_DNA_rep2] <span class="op">=&gt;</span> (x, y) <span class="op">-&gt;</span> x <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;&amp;</span> y <span class="op">&gt;</span> <span class="fl">0</span>, wide_barcode_cpm_df)</span>
<span id="cb25-53"><a href="#cb25-53" aria-hidden="true" tabindex="-1"></a>filtered_wide_barcode_cpm_df.rep1_ratio <span class="op">=</span> filtered_wide_barcode_cpm_df.HMC3MPRA_RNA_rep1 <span class="op">./</span> filtered_wide_barcode_cpm_df.HMC3MPRA_DNA_rep1</span>
<span id="cb25-54"><a href="#cb25-54" aria-hidden="true" tabindex="-1"></a>filtered_wide_barcode_cpm_df.rep2_ratio <span class="op">=</span> filtered_wide_barcode_cpm_df.HMC3MPRA_RNA_rep2 <span class="op">./</span> filtered_wide_barcode_cpm_df.HMC3MPRA_DNA_rep2</span>
<span id="cb25-55"><a href="#cb25-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-56"><a href="#cb25-56" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(<span class="fu">joinpath</span>(output_dir, <span class="st">"barcode_count.tsv"</span>), all_barcode_count_df; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb25-57"><a href="#cb25-57" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(<span class="fu">joinpath</span>(output_dir, <span class="st">"wide_barcode_count.tsv"</span>), wide_barcode_count_df; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb25-58"><a href="#cb25-58" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(<span class="fu">joinpath</span>(output_dir, <span class="st">"filtered_wide_barcode_count.tsv"</span>), filtered_wide_barcode_count_df; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb25-59"><a href="#cb25-59" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(<span class="fu">joinpath</span>(output_dir, <span class="st">"wide_barcode_cpm.tsv"</span>), wide_barcode_cpm_df; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb25-60"><a href="#cb25-60" aria-hidden="true" tabindex="-1"></a>CSV.<span class="fu">write</span>(<span class="fu">joinpath</span>(output_dir, <span class="st">"filtered_wide_barcode_cpm.tsv"</span>), filtered_wide_barcode_cpm_df; header<span class="op">=</span><span class="cn">true</span>, delim<span class="op">=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, append<span class="op">=</span><span class="cn">false</span>)</span>
<span id="cb25-61"><a href="#cb25-61" aria-hidden="true" tabindex="-1"></a>YRUtils.ShellUtils.<span class="fu">recur_pigz</span>(output_dir, <span class="st">r"</span><span class="sc">.+\.</span><span class="st">tsv</span><span class="sc">$</span><span class="st">"; </span>recursive<span class="st">=false, pigz_options="")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="simple-threshold-method" class="level4" data-number="5.1.2">
<h4 data-number="5.1.2" class="anchored" data-anchor-id="simple-threshold-method"><span class="header-section-number">5.1.2</span> Simple threshold method</h4>
<pre class="{r}"><code>library(vroom)
library(tidyverse)
library(YRUtils)

cpm_file &lt;- "bc_count/filtered_wide_barcode_cpm.tsv.gz"
output_dir &lt;- "identify_bc_active.2-fold"

dir.create(output_dir)
cpm_df &lt;- vroom(cpm_file) %&gt;%
    select(key_barcode, rep1_ratio, rep2_ratio) %&gt;%
    rename(Rep1 = rep1_ratio, Rep2 = rep2_ratio) %&gt;%
    pivot_longer(cols = c("Rep1", "Rep2"), names_to = "Replicate", values_to = "Activity") %&gt;%
    mutate(log2Activity = log2(Activity + 1))

p_density &lt;- ggplot(cpm_df, aes(log2Activity, fill = Replicate, color = Replicate)) +
    geom_density(alpha = 0.1) +
    geom_vline(xintercept = log2(3), linetype = "dashed", color = "grey") +
    scale_y_continuous(expand = expansion(mult = 0.02)) +
    scale_x_continuous(expand = expansion(mult = 0.02)) +
    theme_classic(base_size = 20, base_family = "Arial")
ppreview(p_density, file = file.path(output_dir, "density.pdf"))

stat_df &lt;- cpm_df %&gt;%
    mutate(is_active = log2Activity &gt; log2(3)) %&gt;%
    group_by(Replicate) %&gt;%
    reframe(n = n(), n_is_active = sum(is_active)) %&gt;%
    mutate(ratio = n_is_active / n)
vroom_write(stat_df, file = file.path(output_dir, "stat_df.tsv"))</code></pre>
</section>
<section id="mpranalyze-method-1" class="level4" data-number="5.1.3">
<h4 data-number="5.1.3" class="anchored" data-anchor-id="mpranalyze-method-1"><span class="header-section-number">5.1.3</span> MPRAnalyze method</h4>
<pre class="{r}"><code>library(MPRAnalyze)
library(vroom)
library(tidyverse)
library(YRUtils)
library(magrittr)
library(BiocParallel)

count_file &lt;- "bc_count/filtered_wide_barcode_count.tsv.gz"
output_dir &lt;- "identify_bc_active.MPRAnalyze"

dir.create(output_dir)
count_df &lt;- vroom(count_file)
dna_count_df &lt;- count_df %&gt;%
    as.data.frame() %&gt;%
    set_rownames(.[["key_barcode"]]) %&gt;%
    select(all_of(c("HMC3MPRA_DNA_rep1", "HMC3MPRA_DNA_rep2")))
rna_count_df &lt;- count_df %&gt;%
    as.data.frame() %&gt;%
    set_rownames(.[["key_barcode"]]) %&gt;%
    select(all_of(c("HMC3MPRA_RNA_rep1", "HMC3MPRA_RNA_rep2")))
if (!all(row.names(dna_count_df) == row.names(rna_count_df))) {
    stop("the row names of DNA are not identical to the row names of RNA")
}

col_names &lt;- c("rep1", "rep2")
obj &lt;- MpraObject(
    dnaCounts = as.matrix(dna_count_df),
    rnaCounts = as.matrix(rna_count_df),
    dnaAnnot = data.frame(replicate = col_names, row.names = colnames(dna_count_df)),
    rnaAnnot = data.frame(replicate = col_names, row.names = colnames(rna_count_df)),
    BPPARAM = MulticoreParam()
)
obj &lt;- estimateDepthFactors(obj, lib.factor = "replicate", which.lib = "dna", depth.estimator = "uq")
obj &lt;- estimateDepthFactors(obj, lib.factor = "replicate", which.lib = "rna", depth.estimator = "uq")
obj &lt;- analyzeQuantification(obj, dnaDesign = ~replicate, rnaDesign = ~1)

alpha_df &lt;- getAlpha(obj) %&gt;%
    mutate(
        barcode = row.names(.),
        log2Alpha = log2(alpha + 1)
    )
p_density &lt;- ggplot(alpha_df, aes(log2Alpha)) +
    geom_density(alpha = 0.1, fill = "magenta", color = "magenta") +
    scale_y_continuous(expand = expansion(mult = 0.02)) +
    scale_x_continuous(expand = expansion(mult = 0.02)) +
    theme_classic(base_size = 20, base_family = "Arial")
ppreview(p_density, file = file.path(output_dir, "alpha.density.pdf"))

res_df &lt;- testEmpirical(obj, statistic = alpha_df$alpha)
res_df &lt;- bind_cols(alpha_df, res_df) %&gt;%
    mutate(
        pval.mad.adj = p.adjust(pval.mad, method = "fdr"),
        pval.zscore.adj = p.adjust(pval.zscore, method = "fdr")
    ) %&gt;%
    arrange(pval.mad.adj) %&gt;%
    mutate(is_active = if_else(pval.mad.adj &lt; 0.05, TRUE, FALSE))
vroom_write(res_df, file = file.path(output_dir, "df.tsv"))</code></pre>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>Powered by <a href="https://quarto.org">Quarto</a></p>
<div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/edit/main/Blogs/Bioinformatics/posts/MPRA/mpra_analysis_pipeline/index.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/blob/main/Blogs/Bioinformatics/posts/MPRA/mpra_analysis_pipeline/index.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/yr-neurospace/NeuroBorder-WebPageSource/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>